{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "import re\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_folder = '../Data/QC_dev_data/counts' #Path to data with counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_data(data): #Reads the data\n",
    "    data_path = Path(data) #Creates path object\n",
    "    all_data = list(data_path.glob('*tsv')) #Gets list of all files in folder\n",
    "\n",
    "    dfs = [] #Empty list to hold dataframes for each gene\n",
    "\n",
    "    #Iterates through all_data list and reads .TSV\n",
    "    for data in all_data:\n",
    "        df = pd.read_csv(data, sep = '\\t')\n",
    "        dfs.append(df)\n",
    "\n",
    "    return dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_dfs(dfs): #Processes dataframes - gets mininum read counts\n",
    "\n",
    "    clean_dfs = [] #Empty list to hold processed dataframes\n",
    "\n",
    "    #Iterates through provided list of dataframes and does processing\n",
    "    for df in dfs:\n",
    "        new_columns = [('D5_min_count', 'D13_min_count'), ('D5_median_count','D13_median_count')] #List of tuples for new columns \n",
    "        count_cols = [['D05_R1_lib1','D05_R1_lib2','D05_R2_lib1','D05_R2_lib2', 'D05_R3_lib1', 'D05_R3_lib2'], ['D13_R1_lib1', 'D13_R1_lib2', 'D13_R2_lib1', 'D13_R2_lib2', 'D13_R3_lib1', 'D13_R3_lib2']] #Column names of columns wiht counts\n",
    "        all_count_cols = count_cols[0] + count_cols[1] #All count columns\n",
    "\n",
    "        #Iterates through day 5 and day 13 columns and determines min. and median read count\n",
    "        for d5, d13 in new_columns:\n",
    "            if 'min' in d5:\n",
    "                df[d5] = df[count_cols[0]].min(axis = 1)\n",
    "                df[d13] = df[count_cols[1]].min(axis = 1)\n",
    "            elif 'median' in d5:\n",
    "                df[d5] = df[count_cols[0]].median(axis = 1)\n",
    "                df[d13] = df[count_cols[1]].median(axis = 1)\n",
    "                \n",
    "        df['min_count'] = df[all_count_cols].min(axis = 1) #Gets global minimum read count across both timepoints\n",
    "        df['median_count'] = df[all_count_cols].median(axis = 1) #Gets global median read count across both timepoints\n",
    "\n",
    "        gene = df['exon'][0].split('_')[0] #Gets gene name\n",
    "        df['Gene'] = gene #Makes gene name column\n",
    "        df = df[['chrom', 'pos', 'ref', 'allele', 'exon', 'simplified_consequence', 'score', 'functional_consequence', 'standard_error', 'D5_min_count', 'D13_min_count', 'D5_median_count', 'D13_median_count']] #Keeps useful columns\n",
    "        \n",
    "        clean_dfs.append(df) #Appends to empty list\n",
    "    \n",
    "    return dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analysis_plots(dfs): #Creates analysis plots\n",
    "    \n",
    "    concat_df = pd.concat(dfs) #Concatenates all dataframes\n",
    "\n",
    "    concat_df = concat_df[concat_df['Gene'].isin(['BARD1'])] #Line used to extract one or a select few genes only\n",
    "\n",
    "    concat_df['exon_sort'] = concat_df['exon'].apply(lambda x: int(re.search(r'X(\\d+)', x).group(1)) if re.search(r'X(\\d+)', x) else 0) #Creates exon number column for facet sorting\n",
    "\n",
    "    #Creates std. error vs. minimum D5 read count plot\n",
    "    min_plot = alt.Chart(concat_df).mark_point().encode(\n",
    "        x = alt.X('D5_min_count',title = 'Min. D5 Count'),\n",
    "        y = alt.Y('standard_error', title = 'Std. Error'),\n",
    "        color = alt.Color('simplified_consequence', legend = alt.Legend(title = 'Consequence')),\n",
    "        tooltip = [alt.Tooltip('exon', title = 'Exon: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Func. Consequence: '),\n",
    "                   alt.Tooltip('standard_error', title = 'Std. Error: '),\n",
    "                   alt.Tooltip('D5_min_count', title = 'Min D5 Count: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        height = 400, \n",
    "        width = 600\n",
    "    ).facet('Gene'\n",
    "        ).resolve_scale(\n",
    "            x = 'independent'\n",
    "    ).interactive()\n",
    "\n",
    "\n",
    "    #Creates std. error vs. median D5 read count plot\n",
    "    med_plot = alt.Chart(concat_df).mark_point().encode(\n",
    "        x = alt.X('D5_median_count',title = 'Med. D5 Count'),\n",
    "        y = alt.Y('standard_error', title = 'Std. Error'),\n",
    "        color = alt.Color('simplified_consequence', legend = alt.Legend(title = 'Consequence')),\n",
    "        tooltip = [alt.Tooltip('exon', title = 'Exon: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Func. Consequence: '),\n",
    "                   alt.Tooltip('standard_error', title = 'Std. Error: '),\n",
    "                   alt.Tooltip('D5_median_count', title = 'Med. D5 Count: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        height = 400, \n",
    "        width = 600\n",
    "    ).facet('Gene'\n",
    "        ).resolve_scale(\n",
    "            x = 'independent'\n",
    "    ).interactive()\n",
    "\n",
    "    #Concatenates and displays min. and median plots\n",
    "    plots = alt.vconcat(min_plot, med_plot)\n",
    "    plots.display() \n",
    "\n",
    "    #Creates std. error vs. min. D5 read count faceted by variant consequence\n",
    "    consequences = alt.Chart(concat_df).mark_point().encode(\n",
    "        x = 'D5_min_count',\n",
    "        y = 'standard_error',\n",
    "        color = 'simplified_consequence',\n",
    "        tooltip = [alt.Tooltip('exon', title = 'Exon: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Func. Consequence: '),\n",
    "                   alt.Tooltip('standard_error', title = 'Std. Error: '),\n",
    "                   alt.Tooltip('min_count', title = 'Min D5 Count: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        height = 400, \n",
    "        width = 600\n",
    "    ).facet(column = 'simplified_consequence',\n",
    "            row = 'Gene'\n",
    "           )\n",
    "\n",
    "    consequences.display()\n",
    "\n",
    "    #Creates std. error vs. min. D5 read count faceted by functional consequence\n",
    "    function = alt.Chart(concat_df).mark_point().encode(\n",
    "        x = 'D5_min_count', \n",
    "        y = 'standard_error',\n",
    "        color = 'simplified_consequence',\n",
    "        tooltip = [alt.Tooltip('exon', title = 'Exon: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Func. Consequence: '),\n",
    "                   alt.Tooltip('standard_error', title = 'Std. Error: '),\n",
    "                   alt.Tooltip('D5_min_count', title = 'Min D5 Count: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        height = 400,\n",
    "        width = 600\n",
    "    ).facet(row = 'Gene', \n",
    "            column = 'functional_consequence'\n",
    "           )\n",
    "    function.display()\n",
    "\n",
    "    #Creates std. error vs. SGE score\n",
    "    sge_plot = alt.Chart(concat_df).mark_point().encode(\n",
    "        x = alt.X('score',title = 'SGE Score'),\n",
    "        y = alt.Y('standard_error', title = 'Std. Error'),\n",
    "        color = alt.Color('simplified_consequence', legend = alt.Legend(title = 'Consequence')),\n",
    "        tooltip = [alt.Tooltip('exon', title = 'Exon: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Func. Consequence: '),\n",
    "                   alt.Tooltip('standard_error', title = 'Std. Error: '),\n",
    "                   alt.Tooltip('D5_min_count', title = 'Min D5 Count: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        height = 400, \n",
    "        width = 600\n",
    "    ).facet('Gene'\n",
    "        ).resolve_scale(\n",
    "            x = 'independent'\n",
    "    ).interactive()\n",
    "\n",
    "    sge_plot.display()\n",
    "\n",
    "    #Creates std. error vs. min. D5 read count faceted by exon\n",
    "    byexon_plot = alt.Chart(concat_df).mark_point().encode(\n",
    "        x = alt.X('D5_min_count',title = 'D5 Min. Count'),\n",
    "        y = alt.Y('standard_error', title = 'Std. Error'),\n",
    "        color = alt.Color('simplified_consequence', legend = alt.Legend(title = 'Consequence')),\n",
    "        tooltip = [alt.Tooltip('exon', title = 'Exon: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Func. Consequence: '),\n",
    "                   alt.Tooltip('standard_error', title = 'Std. Error: '),\n",
    "                   alt.Tooltip('D5_min_count', title = 'Min D5 Count: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        height = 400, \n",
    "        width = 600\n",
    "    ).facet(facet = alt.Facet('exon', sort = {'field': 'exon_sort', 'op': 'min', 'order': 'ascending'})\n",
    "        ).resolve_scale(\n",
    "            x = 'independent'\n",
    "    ).interactive()\n",
    "\n",
    "    byexon_plot.display()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    dfs = read_data(data_folder)\n",
    "    final_dfs = process_dfs(dfs)\n",
    "    analysis_plots(final_dfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
