{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sge_genes = [('BARD1', '../Data/filtered_ppj_data/SGE/BARD1.tsv'),\n",
    "             ('BRCA2','../Data/filtered_ppj_data/SGE/BRCA2.tsv'), \n",
    "             ('CTCF', '../Data/filtered_ppj_data/SGE/CTCF.tsv'), \n",
    "             ('PALB2','../Data/filtered_ppj_data/SGE/PALB2.tsv'), \n",
    "             ('RAD51D','../Data/filtered_ppj_data/SGE/RAD51D.tsv'), \n",
    "             ('SFPQ','../Data/filtered_ppj_data/SGE/SFPQ.tsv'), \n",
    "             ('XRCC2', '../Data/filtered_ppj_data/SGE/XRCC2.tsv')\n",
    "            ]\n",
    "\n",
    "vampseq = [('F9 Ab001','../Data/filtered_ppj_data/VAMPseq/F9_ab001.csv'),\n",
    "           ('F9 Ab102','../Data/filtered_ppj_data/VAMPseq/F9_ab102.csv'),\n",
    "           ('F9 Ab124','../Data/filtered_ppj_data/VAMPseq/F9_ab124.csv'),\n",
    "           ('F9 Ab3570','../Data/filtered_ppj_data/VAMPseq/F9_ab3570.csv'),\n",
    "           ('F9 Strep','../Data/filtered_ppj_data/VAMPseq/F9_strep.csv' ),\n",
    "           ('G6PD','../Data/filtered_ppj_data/VAMPseq/G6PD_scores_consequence.csv'),\n",
    "           ('TSC2', '../Data/filtered_ppj_data/VAMPseq/TSC2_scores_consequences.csv')\n",
    "          ]\n",
    "\n",
    "alt.data_transformers.disable_max_rows() #gets rid of max data length problem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gmm_threshold(df):\n",
    "    \n",
    "    # find the GMM thresholds\n",
    "    target_value = 0.950\n",
    "\n",
    "    ab_df = df.loc[df['functional_consequence'].isin(['functionally_abnormal'])]\n",
    "    norm_df = df.loc[df['functional_consequence'].isin(['functionally_normal'])]\n",
    "    \n",
    "    # Calculate the absolute difference for the Normal (N) density\n",
    "    diffN = (norm_df['gmm_density_normal'] - target_value).abs()\n",
    "    # Find the index of the minimum difference\n",
    "    closest_index = diffN.idxmin()\n",
    "    # Retrieve the row with the closest value\n",
    "    closest_row_n = norm_df.loc[closest_index]\n",
    "    \n",
    "    # now repeat that for the abnormal density\n",
    "    # Calculate the absolute difference\n",
    "    diffA = (ab_df['gmm_density_abnormal'] - target_value).abs()\n",
    "    # Find the index of the minimum difference\n",
    "    closest_index = diffA.idxmin()\n",
    "    # Retrieve the row with the closest value\n",
    "    closest_row_a = ab_df.loc[closest_index]\n",
    "    \n",
    "    # now we get the scores that are the closest to the (n)ormal and (a)bnormal thresholds\n",
    "    upprthresh = closest_row_n['score']\n",
    "    lwrthresh = closest_row_a['score']\n",
    "\n",
    "    thresholds = [lwrthresh, upprthresh]\n",
    "\n",
    "    return thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_data(sge, vampseq):\n",
    "\n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained',\n",
    "        'Intron', \n",
    "        'UTR Variant',\n",
    "        'Stop Lost',\n",
    "        'Start Lost',\n",
    "        'Canonical Splice', \n",
    "        'Splice Region',\n",
    "    ]\n",
    "    \n",
    "    sge_dict = {}\n",
    "    for gene in sge:\n",
    "        gene_name, data_path = gene\n",
    "\n",
    "        df = pd.read_csv(data_path, sep = '\\t')\n",
    "        df = df.loc[df['variant_qc_flag'].isin(['PASS'])]\n",
    "\n",
    "        \n",
    "        df = df.rename(columns = {'consequence': 'Consequence'})\n",
    "        df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "        df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "        df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "        df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "        df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "        df.loc[df['Consequence'].str.contains('site'), 'Consequence'] = 'Canonical Splice'\n",
    "        df.loc[df['Consequence'].str.contains('ing_var'), 'Consequence'] = 'Splice Region'\n",
    "        df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "        df.loc[df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "        df.loc[df['ref'].str.len() > 1, 'Consequence'] = '3bp Deletion'\n",
    "        \n",
    "        sge_dict[gene_name] = df\n",
    "\n",
    "\n",
    "    vampseq_dict = {}\n",
    "    for gene in vampseq:\n",
    "        gene_name, data_path = gene\n",
    "\n",
    "        gene_df = pd.read_csv(data_path)\n",
    "        gene_df = gene_df.rename(columns = {'average_score': 'score', 'type': 'Consequence'})\n",
    "        gene_df.loc[gene_df['Consequence'] == 'Nonsense', 'Consequence'] = 'Stop Gained'\n",
    "        gene_df = gene_df.loc[~(gene_df['Consequence'].isin(['Deletion']))]\n",
    "        vampseq_dict[gene_name] = gene_df\n",
    "\n",
    "\n",
    "    return sge_dict, vampseq_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sge_histogram(df, gene_name, thresholds, thresholding = True): #makes histogram\n",
    "\n",
    "\n",
    "    length = str(len(df)) #gets length of data for title\n",
    "\n",
    "    #final_tital = gene_name + ' SGE Scores ' + '(n = ' + length + ')' #used to build title\n",
    "    final_tital = f' {gene_name} (n = {length})'\n",
    "    bins = 50 #number of bins\n",
    "\n",
    "    sorted = [\"Intron\", \"Missense\", \"Synonymous\", \"Stop Gained\", \"Splice\", \"Start Lost\", 'Stop Lost', 'UTR', '3bp Deletion'] #order for the legend\n",
    "    selection = alt.selection_point(fields=['Consequence'], bind='legend')\n",
    "\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a', # yellow\n",
    "    '#6AA84F', # med green\n",
    "    '#93C47D', # light green\n",
    "    '#888888', # med gray\n",
    "    '#000000', # black\n",
    "    '#1170AA', # darker blue\n",
    "    '#CFCFCF', # light gray\n",
    "    '#FF9A00'   #orange\n",
    "        \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained',\n",
    "        'Intron', \n",
    "        'UTR Variant',\n",
    "        'Stop Lost',\n",
    "        'Start Lost',\n",
    "        'Canonical Splice', \n",
    "        'Splice Region',\n",
    "        '3bp Deletion'\n",
    "    ]\n",
    "\n",
    "    df = df.loc[df['Consequence'].isin(variant_types)]\n",
    "    # Builds histogram with interactive legend\n",
    "    histogram = alt.Chart(df).mark_bar().encode(\n",
    "            alt.X('score', \n",
    "                  axis = alt.Axis(title = 'Functional Score', \n",
    "                                  labelFontSize = 16, \n",
    "                                  titleFontSize = 20, \n",
    "                                  values = [-0.6, -0.5, -0.4, -0.3, -0.2, -0.1, 0, 0.1, 0.2]\n",
    "                                 ), \n",
    "                  bin = alt.Bin(maxbins = bins)),\n",
    "            alt.Y('count()', \n",
    "                  axis = alt.Axis(title = 'Number of Variants', \n",
    "                                  labelFontSize = 16, \n",
    "                                  titleFontSize = 20)),\n",
    "            color = alt.Color('Consequence:N', \n",
    "                             scale = alt.Scale(range = palette,\n",
    "                                              domain = variant_types), \n",
    "                             legend = alt.Legend(titleFontSize = 16, \n",
    "                                                 labelFontSize = 14,\n",
    "                                                 orient = 'top-left',\n",
    "                                                offset = 10)),\n",
    "            opacity=alt.condition(selection, alt.value(1), alt.value(0.2))  # Highlight selected categories\n",
    "    ).add_params(\n",
    "        selection\n",
    "    ).properties(\n",
    "        width = 400,\n",
    "        height = 300,\n",
    "        title = alt.TitleParams(text = final_tital, fontSize = 22)\n",
    "    ).interactive()\n",
    "\n",
    "    if thresholding:\n",
    "\n",
    "        nf_line = alt.Chart(pd.DataFrame({'Functional Score': [thresholds[0]]})).mark_rule(color = 'red').encode(\n",
    "            x = 'Functional Score')\n",
    "    \n",
    "        func_line = alt.Chart(pd.DataFrame({'Functional Score': [thresholds[1]]})).mark_rule(color = 'blue').encode(\n",
    "            x = 'Functional Score')\n",
    "    \n",
    "    \n",
    "        histogram = alt.layer(histogram, nf_line, func_line)\n",
    "\n",
    "    else:\n",
    "        histogram = histogram \n",
    "        \n",
    "    return histogram\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def vamp_histogram(df, gene_name, thresholds, thresholding = True):\n",
    "\n",
    "    length = str(len(df)) #gets length of data for title\n",
    "\n",
    "    final_tital = gene_name +  ' (n = ' + length + ')' #used to build title\n",
    "    bins = 50 #number of bins\n",
    "\n",
    "    sorted = [\"Intron\", \"Missense\", \"Synonymous\", \"Stop Gained\", \"Splice\", \"Start Lost\", 'Stop Lost', 'UTR'] #order for the legend\n",
    "    selection = alt.selection_point(fields=['Consequence'], bind='legend')\n",
    "\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a', # yellow    \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained'\n",
    "    ]\n",
    "    \n",
    "    # Builds histogram with interactive legend\n",
    "    histogram = alt.Chart(df).mark_bar().encode(\n",
    "            alt.X('score', \n",
    "                  axis = alt.Axis(title = 'Functional Score', \n",
    "                                  labelFontSize = 16, \n",
    "                                  titleFontSize = 20,\n",
    "                                  values = [-0.2, 0, 0.2, 0.4, 0.6, 0.8, 1, 1.2, 1.4, 1.6]\n",
    "                                 ), \n",
    "                  bin = alt.Bin(maxbins = bins)),\n",
    "            alt.Y('count()', \n",
    "                  axis = alt.Axis(title = 'Number of Variants', \n",
    "                                  labelFontSize = 16, \n",
    "                                  titleFontSize = 20)),\n",
    "            color = alt.Color('Consequence:N', \n",
    "                             scale = alt.Scale(range = palette,\n",
    "                                              domain = variant_types), \n",
    "                             legend = alt.Legend(titleFontSize = 16, \n",
    "                                                 labelFontSize = 14,\n",
    "                                                 orient = 'top-left',\n",
    "                                                offset = 10)),\n",
    "            opacity=alt.condition(selection, alt.value(1), alt.value(0.2))  # Highlight selected categories\n",
    "    ).add_params(\n",
    "        selection\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400,\n",
    "        title = alt.TitleParams(text = final_tital, fontSize = 22)\n",
    "    ).interactive()\n",
    "\n",
    "    if thresholding:\n",
    "\n",
    "        nf_line = alt.Chart(pd.DataFrame({'Functional Score': [thresholds[0]]})).mark_rule(color = 'red').encode(\n",
    "            x = 'Functional Score')\n",
    "    \n",
    "        func_line = alt.Chart(pd.DataFrame({'Functional Score': [thresholds[1]]})).mark_rule(color = 'blue').encode(\n",
    "            x = 'Functional Score')\n",
    "    \n",
    "    \n",
    "        histogram = alt.layer(histogram, nf_line, func_line)\n",
    "\n",
    "    else:\n",
    "        histogram = histogram \n",
    "        \n",
    "    return histogram\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sge_strip_plot(df, thresholds, thresholding = True):\n",
    "\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a', # yellow\n",
    "    '#6AA84F', # med green\n",
    "    '#93C47D', # light green\n",
    "    '#888888', # med gray\n",
    "    '#000000', # black\n",
    "    '#1170AA', # darker blue\n",
    "    '#CFCFCF', # light gray\n",
    "    '#FF9A00'   #orange\n",
    "        \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained',\n",
    "        'Intron', \n",
    "        'UTR Variant',\n",
    "        'Stop Lost',\n",
    "        'Start Lost',\n",
    "        'Canonical Splice', \n",
    "        'Splice Region',\n",
    "        '3bp Deletion'\n",
    "    ]\n",
    "\n",
    "    df = df.loc[df['Consequence'].isin(variant_types)]\n",
    "    df['Consequence'] = pd.Categorical(df['Consequence'], \n",
    "                                   categories=variant_types, \n",
    "                                   ordered=True)\n",
    "\n",
    "    nf_line = alt.Chart(pd.DataFrame({'SGE Score': [thresholds[0]]})).mark_rule(color = 'red').encode(\n",
    "        x = 'SGE Score')\n",
    "\n",
    "    func_line = alt.Chart(pd.DataFrame({'SGE Score': [thresholds[1]]})).mark_rule(color = 'blue').encode(\n",
    "        x = 'SGE Score')\n",
    "\n",
    "    sort_order = variant_types\n",
    "            \n",
    "    plot = alt.Chart(df).mark_tick(opacity = 1).encode(\n",
    "        x = alt.X('score:Q',\n",
    "                  axis = alt.Axis(title = '', \n",
    "                                  titleFontSize = 18,\n",
    "                                 labelFontSize = 16, \n",
    "                                 values = [-0.6, -0.5, -0.4, -0.3, -0.2, -0.1, 0, 0.1, 0.2])\n",
    "                 ),\n",
    "        y = alt.Y('Consequence:N', \n",
    "                  sort = sort_order,\n",
    "                  axis = alt.Axis(title = '',\n",
    "                                 labelFontSize = 24)\n",
    "                 ),\n",
    "        color = alt.Color('Consequence:N',\n",
    "                legend=None,\n",
    "                scale = alt.Scale(range = palette,\n",
    "                                  domain = variant_types)\n",
    "                         )\n",
    "        ).properties(\n",
    "            width = 800,\n",
    "            height = 400\n",
    "        ).interactive()\n",
    "\n",
    "    if thresholding:\n",
    "        plot = alt.layer(plot, nf_line, func_line).resolve_scale(y = 'shared')\n",
    "\n",
    "\n",
    "    else:\n",
    "        plot = plot\n",
    "\n",
    "    return plot\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def vamp_strip(df, thresholds, thresholding = True):\n",
    "\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a' # yellow   \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained'\n",
    "    ]\n",
    "\n",
    "    df['Consequence'] = pd.Categorical(df['Consequence'], \n",
    "                                   categories=variant_types, \n",
    "                                   ordered=True)\n",
    "\n",
    "    nf_line = alt.Chart(pd.DataFrame({'VAMP-seq Score': [thresholds[0]]})).mark_rule(color = 'red').encode(\n",
    "        x = 'VAMP-seq Score')\n",
    "\n",
    "    func_line = alt.Chart(pd.DataFrame({'VAMP-seq Score': [thresholds[1]]})).mark_rule(color = 'green').encode(\n",
    "        x = 'VAMP-seq Score')\n",
    "\n",
    "    sort_order = variant_types\n",
    "            \n",
    "    plot = alt.Chart(df).mark_tick(opacity = 1).encode(\n",
    "        x = alt.X('score:Q',\n",
    "                  axis = alt.Axis(title = '', \n",
    "                                  titleFontSize = 18,\n",
    "                                 labelFontSize = 16, \n",
    "                                 values = [-0.2, 0, 0.2, 0.4, 0.6, 0.8, 1, 1.2, 1.4, 1.6])\n",
    "                 ),\n",
    "        y = alt.Y('Consequence:N', \n",
    "                  sort = sort_order,\n",
    "                  axis = alt.Axis(title = '',\n",
    "                                 labelFontSize = 24)\n",
    "                 ),\n",
    "        color = alt.Color('Consequence:N',\n",
    "                legend=None,\n",
    "                scale = alt.Scale(range = palette,\n",
    "                                  domain = variant_types)\n",
    "                         )\n",
    "        ).properties(\n",
    "            width = 800,\n",
    "            height = 400\n",
    "        ).interactive()\n",
    "\n",
    "    if thresholding:\n",
    "        plot = alt.layer(plot, nf_line, func_line).resolve_scale(y = 'shared')\n",
    "\n",
    "\n",
    "    else:\n",
    "        plot = plot\n",
    "\n",
    "    return plot\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sge_plots(sge_data):\n",
    "    genes = list(sge_data.keys())\n",
    "\n",
    "    histograms = {}\n",
    "    strip_plots = {}\n",
    "    for gene in genes:\n",
    "        df = sge_data[gene]\n",
    "        if gene == 'BRCA2':\n",
    "            gene_thresholds = [0,0]\n",
    "            histogram = sge_histogram(df, gene, gene_threshold, thresholding = False)\n",
    "            stripplot = sge_strip_plot(df, gene_thresholds, thresholding = False)\n",
    "            histograms[gene] = histogram\n",
    "            strip_plots[gene] = stripplot\n",
    "        else:\n",
    "            threshold_df = df.loc[~df['Consequence'].isin(['3bp Deletion'])]\n",
    "            gene_threshold = get_gmm_threshold(threshold_df)\n",
    "\n",
    "            print(f'{gene} Thresholds are: {gene_threshold}')\n",
    "            histogram = sge_histogram(df, gene, gene_threshold)\n",
    "            stripplot = sge_strip_plot(df, gene_threshold)\n",
    "            histograms[gene] = histogram\n",
    "            strip_plots[gene] = stripplot\n",
    "\n",
    "\n",
    "\n",
    "    for gene in genes:\n",
    "        histogram = histograms[gene]\n",
    "        stripplot = strip_plots[gene]\n",
    "        final = histogram\n",
    "        #final = histogram  & stripplot\n",
    "\n",
    "        final = final.configure_axis(\n",
    "            grid = False\n",
    "        ).configure_view(\n",
    "            stroke = None\n",
    "        ).resolve_scale(\n",
    "            x = 'shared'\n",
    "        )\n",
    "\n",
    "        to_save = '/Users/ivan/Desktop/pillar_project_figs/Histogram_wStripplot/' + gene + '_Histogram.svg'\n",
    "        #final.save(to_save)\n",
    "        #histogram.display()\n",
    "        final.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def vampseq_figs(vampseq_data):\n",
    "    genes = list(vampseq_data.keys())\n",
    "\n",
    "    histograms = {}\n",
    "    strip_plots = {}\n",
    "\n",
    "    for gene in genes:\n",
    "        df = vampseq_data[gene]\n",
    "        gene_threshold = [0,0]\n",
    "        histogram = vamp_histogram(df, gene, gene_threshold, thresholding = False)\n",
    "        strip = vamp_strip(df, gene_threshold, thresholding = False)\n",
    "        histograms[gene] = histogram\n",
    "        strip_plots[gene] = strip\n",
    "\n",
    "\n",
    "    for gene in genes:\n",
    "        histogram = histograms[gene]\n",
    "        stripplot = strip_plots[gene]\n",
    "\n",
    "        #final = histogram  & stripplot\n",
    "        final = histogram\n",
    "\n",
    "        final = final.configure_axis(\n",
    "            grid = False\n",
    "        ).configure_view(\n",
    "            stroke = None\n",
    "        ).resolve_scale(\n",
    "            x = 'shared'\n",
    "        )\n",
    "\n",
    "        to_save = '/Users/ivan/Desktop/pillar_project_figs/Histogram_wStripplot/' + gene + '_Histogram.svg'\n",
    "        #final.save(to_save)\n",
    "        \n",
    "        final.display()\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    sge_data, vampseq_data = read_data(sge_genes, vampseq)\n",
    "    sge_plots(sge_data)\n",
    "    vampseq_figs(vampseq_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
