{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "from natsort import natsorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "bard1_data = '../Data/BARD1_SGE_final_table.xlsx'\n",
    "brca1_data = '../Data/BRCA1_SGE_data.xlsx'\n",
    "\n",
    "#Estimated thresholds for data from Findlay et. al 2018\n",
    "brca1_int_cutoff = -0.748\n",
    "brca1_path_cutoff = -1.328\n",
    "\n",
    "domain = 'BRCT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_to_one(hgvs_str):\n",
    "    \n",
    "    aa_3to1 = {\n",
    "    'Ala': 'A', 'Arg': 'R', 'Asn': 'N', 'Asp': 'D', 'Cys': 'C',\n",
    "    'Gln': 'Q', 'Glu': 'E', 'Gly': 'G', 'His': 'H', 'Ile': 'I',\n",
    "    'Leu': 'L', 'Lys': 'K', 'Met': 'M', 'Phe': 'F', 'Pro': 'P',\n",
    "    'Ser': 'S', 'Thr': 'T', 'Trp': 'W', 'Tyr': 'Y', 'Val': 'V'\n",
    "    } #Dictionary for converting 3-letter amino acid codes to 1-letter codes\n",
    "\n",
    "    sub = hgvs_str.split('.')[1]\n",
    "\n",
    "    ogAA = sub[0:3]\n",
    "    newAA = sub[-3:]\n",
    "    pos = sub[3:-3]\n",
    "    \n",
    "    remap_ogAA = aa_3to1[ogAA]\n",
    "    remap_newAA = aa_3to1[newAA]\n",
    "\n",
    "    new_hgvs = 'p.' + remap_ogAA + pos + remap_newAA\n",
    "\n",
    "    return new_hgvs  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_data(bard1, brca1, brca1_int, brca1_path_max, domain):\n",
    "    bard1_df = pd.read_excel(bard1, sheet_name = 'scores')\n",
    "    bard1_df = bard1_df.loc[~bard1_df['variant_qc_flag'].isin(['WARN'])] #Filters out WARN variants in same codon as fixed edit\n",
    "    brca1_df = pd.read_excel(brca1, sheet_name = 'findlay_2018')\n",
    "\n",
    "    if domain == 'RING': #Sets domain for RING\n",
    "        brca1_region = list(range(1, 302))\n",
    "        bard1_region = list(range(26, 123))\n",
    "\n",
    "    elif domain == 'RING_loop': #Sets domain for RING domains without the large helices\n",
    "        brca1_region = list(range(67, 238)) #Residues 23 - 79\n",
    "        bard1_region = list(range(48, 97)) #Residues 48 - 96\n",
    "\n",
    "    elif domain == 'BRCT': #Sets domain for BRCT\n",
    "        brca1_region = list(range(4936, 5566))\n",
    "        bard1_region = list(range(568, 778))\n",
    "\n",
    "    brca1_df = brca1_df.loc[brca1_df['pos'].isin(brca1_region)] #Gets all data from Findlay 2018 dataset\n",
    "    brca1_df['functional_consequence'] = 'indeterminate' #Creates functional consequence column\n",
    "\n",
    "    #Puts variants into functional classes\n",
    "    brca1_df.loc[brca1_df['snv_score_minmax'] <= brca1_path_max, 'functional_consequence'] = 'functionally_abnormal'\n",
    "    brca1_df.loc[brca1_df['snv_score_minmax'] >= brca1_int, 'functional_consequence'] = 'functionally_normal'\n",
    "    brca1_df = brca1_df.loc[brca1_df['Consequence'].isin(['missense_variant'])] #Filters for missense variants\n",
    "    brca1_df['AApos'] = brca1_df['hgvs_pro'].transform(lambda x: int(x.split(':')[1].split('.')[1][3:-3]))\n",
    "    brca1_df['hgvs_pro'] = brca1_df['hgvs_pro'].transform(lambda x: x.split(':')[1])\n",
    "    brca1_df['hgvs_pro'] = brca1_df['hgvs_pro'].apply(convert_to_one)\n",
    "    \n",
    "    new_brca1 = pd.read_excel(brca1, sheet_name = 'dace_2025') #Reads Dace 2025 BRCA1 data\n",
    "    new_brca1 = new_brca1.loc[new_brca1['Consequence'].isin(['Missense'])] #Filters for missense variants\n",
    "    new_brca1['CDSpos'] = new_brca1['CDSpos'].astype(int)\n",
    "    new_brca1 = new_brca1.loc[new_brca1['CDSpos'].isin(brca1_region)] #Filters for variants in specified domain only\n",
    "\n",
    "    new_brca1['target'] = 'BRCA1_X2' #Placeholder column for downstream concatenation \n",
    "    new_brca1['functional_consequence'] = 'indeterminate' #Sets functional consequence collumn\n",
    "    new_brca1.loc[new_brca1['function_class'] == 'LoF', 'functional_consequence'] = 'functionally_abnormal' #LoF varinats are set as functionaly abnormal\n",
    "    new_brca1.loc[new_brca1['function_class'] == 'Neutral', 'functional_consequence'] = 'functionally_normal' #Neutral variants set to functionally normal\n",
    "    new_brca1_pos = list(set(new_brca1['CDSpos'].tolist())) #Gets positions in new data set\n",
    "    new_brca1 = new_brca1.rename(columns = {'CDSpos': 'pos', 'final_function_score': 'snv_score_minmax', 'protPos': 'AApos', 'pHGVS': 'hgvs_pro'}) #Renames columns for concatenation\n",
    "\n",
    "    new_brca1['AApos'] = new_brca1['AApos'].astype(int)\n",
    "    brca1_df = brca1_df.loc[~(brca1_df['pos'].isin(new_brca1_pos))] #Variants in 2018 dataset found in 2025 dataset are removed\n",
    "    brca1_df = pd.concat([brca1_df, new_brca1[['target','pos','Consequence','snv_score_minmax', 'functional_consequence', 'AApos', 'hgvs_pro']]]) #2018 and 2025 BRCA1 datasets are concatenated\n",
    "    brca1_df = brca1_df.rename(columns = {'snv_score_minmax': 'score'})\n",
    "    \n",
    "    #Processing BARD1 data\n",
    "    bard1_df = bard1_df.loc[bard1_df['consequence'].isin(['missense_variant'])] #Missense variants only\n",
    "    bard1_df['AApos'] = bard1_df['amino_acid_change'].transform(lambda x: x[1:-1]) #Gets amino acid position\n",
    "    bard1_df['AApos'] = bard1_df['AApos'].astype(int) #Sets data type as integer\n",
    "    bard1_df = bard1_df.loc[bard1_df['AApos'].isin(bard1_region)] #Gets variants in region of interst only\n",
    "    \n",
    "    return bard1_df, brca1_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def missense_sensitivity(bard1, brca1): #Tabulates missense sensitivity between BRCA1 and BARD1\n",
    "\n",
    "    #Gets counts and percent of missense variants in functionally normal and abnormal camps for BARD1\n",
    "    bard1_sensitive = bard1['functional_consequence'].value_counts().get('functionally_abnormal',0)\n",
    "    bard1_normal = bard1['functional_consequence'].value_counts().get('functionally_normal',0)\n",
    "    total_bard1_ring_missense = len(bard1)\n",
    "    bard1_sensitivity = bard1_sensitive / total_bard1_ring_missense\n",
    "    bard1_normal_percent = bard1_normal / total_bard1_ring_missense\n",
    "\n",
    "    #Analagous code for BRCA1\n",
    "    brca1_sensitive = brca1['functional_consequence'].value_counts().get('functionally_abnormal',0)\n",
    "    brca1_normal = brca1['functional_consequence'].value_counts().get('functionally_normal',0)\n",
    "    total_brca1_ring_missense = len(brca1)\n",
    "\n",
    "    brca1_sensitivity = brca1_sensitive / total_brca1_ring_missense\n",
    "    brca1_normal_percent = brca1_normal / total_brca1_ring_missense\n",
    "\n",
    "    #Sets up contingency table for Fischer exact testing\n",
    "    contingency_table = np.array([[brca1_sensitive, brca1_normal],\n",
    "                                  [bard1_sensitive, bard1_normal]])\n",
    "\n",
    "    odds_ratio, p_value = stats.fisher_exact(contingency_table)\n",
    "\n",
    "    print(f\"Fisher's exact test p-value: {p_value}\")\n",
    "    print(f\"Odds ratio: {odds_ratio}\")\n",
    "\n",
    "    #Dataframe passed on for plotting\n",
    "    to_plot = pd.DataFrame({\n",
    "        'Gene': ['BRCA1', 'BARD1','BRCA1', 'BARD1'],\n",
    "        'Percent of Variants': [brca1_sensitivity,bard1_sensitivity, brca1_normal_percent, bard1_normal_percent],\n",
    "        'Type': ['Abnormal Missense', 'Abnormal Missense', 'Normal Missense', 'Normal Missense']\n",
    "    })\n",
    "\n",
    "    display(to_plot)\n",
    "    return to_plot\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def missense_resi(bard1_df, brca1_df):\n",
    "\n",
    "    #brca1_df.loc[brca1_df['functional_consequence'] == 'indeterminate', 'functional_consequence'] = 'functionally_abnormal'\n",
    "\n",
    "    bard1_df['functional_consequence'] = pd.Categorical(bard1_df['functional_consequence'], categories = ['functionally_abnormal', 'indeterminate', 'functionally_normal'], ordered=True)\n",
    "    brca1_df['functional_consequence'] = pd.Categorical(brca1_df['functional_consequence'], categories = ['functionally_abnormal', 'indeterminate', 'functionally_normal'], ordered=True)\n",
    "    #brca1_df = brca1_df.loc[brca1_df['AApos'].isin([20, 26, 27, 28, 51, 52, 54, 55, 63, 64, 65, 66, 68, 104])]\n",
    "    \n",
    "    bard1_df = bard1_df[['functional_consequence', 'AApos']]\n",
    "    bard1_df = bard1_df.groupby('AApos').agg({\n",
    "        'functional_consequence': 'min'\n",
    "    }).reset_index()\n",
    "\n",
    "    brca1_df = brca1_df[['functional_consequence', 'AApos']]\n",
    "    brca1_df = brca1_df.groupby('AApos').agg({\n",
    "        'functional_consequence': 'min'\n",
    "    }).reset_index()\n",
    "\n",
    "    bard1_sensitive_count = bard1_df['functional_consequence'].value_counts().get('functionally_abnormal', 0)\n",
    "    brca1_sensitive_count = brca1_df['functional_consequence'].value_counts().get('functionally_abnormal',0)\n",
    "\n",
    "    \n",
    "\n",
    "    brca1_perc = brca1_sensitive_count / len(brca1_df)\n",
    "    bard1_perc = bard1_sensitive_count / len(bard1_df)\n",
    "\n",
    "    print('BARD1: ', str(bard1_sensitive_count), '\\n',\n",
    "            'BRCA1: ', str(brca1_sensitive_count), '\\n',\n",
    "          'BARD1 %: ', str(bard1_perc), '\\n',\n",
    "          'BRCA1 %: ', str(brca1_perc)\n",
    "         )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bar_plot(df): #Simple bar plot visualization\n",
    "\n",
    "    df['Percent of Variants'] = df['Percent of Variants'] * 100\n",
    "\n",
    "    plot = alt.Chart(df).mark_bar(color = 'white',\n",
    "                                 stroke = 'black').encode(\n",
    "        x = alt.X('Gene',\n",
    "                  axis = alt.Axis(title = '',\n",
    "                                 labelFontSize = 16,\n",
    "                                 titleFontSize = 18,\n",
    "                                  labelAngle = 0\n",
    "                                 )\n",
    "                 ),\n",
    "        y = alt.Y('Percent of Variants',\n",
    "                  axis = alt.Axis(title = '% LoF Missense Vars.',\n",
    "                                  labelFontSize = 16,\n",
    "                                  titleFontSize = 18\n",
    "                                 )\n",
    "                 ),\n",
    "        column = alt.Column('Type', \n",
    "                            title = '',\n",
    "                            header = alt.Header(\n",
    "                                titleFontSize = 18\n",
    "                            )\n",
    "                           )\n",
    "    ).properties(\n",
    "        width = 125,\n",
    "        height = 330\n",
    "    ).configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "\n",
    "    \n",
    "    #plot.save('/Users/ivan/Desktop/BARD1_draft_figs/supp_figs/suppfig_BRCTcomparison_barplot.png', ppi = 500)\n",
    "    plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def heatmap(df, gene):\n",
    "\n",
    "    if gene == 'BRCA1':\n",
    "        min = -1.5\n",
    "        max = 0\n",
    "        sort_order = ['S1655', 'G1656', 'T1700', 'K1702']\n",
    "\n",
    "    elif gene == 'BARD1':\n",
    "        min = -0.2\n",
    "        max = 0\n",
    "        sort_order = ['S575', 'G576', 'T617', 'K619']\n",
    "\n",
    "    color_domain = [min, max]\n",
    "    \n",
    "    plot = alt.Chart(df).mark_rect().encode(\n",
    "        x = alt.X('aa_plot:O',\n",
    "                  title = '',\n",
    "                  sort = sort_order,\n",
    "                  axis = alt.Axis(\n",
    "                      labelAngle = 0,\n",
    "                      ticks = False,\n",
    "                      labelFontSize = 16\n",
    "                  )\n",
    "                 ),\n",
    "        y = alt.Y('type_change:N',\n",
    "                  title = '',\n",
    "                  axis = alt.Axis(ticks = False,\n",
    "                                  labelFontSize = 16\n",
    "                                 )\n",
    "                 ),\n",
    "        color = alt.Color('score:Q',\n",
    "                         scale = alt.Scale(\n",
    "                             scheme = 'bluepurple',\n",
    "                             domain = color_domain,\n",
    "                             clamp = True,\n",
    "                             reverse = True\n",
    "                         ),\n",
    "                          legend = alt.Legend(\n",
    "                              title = 'SGE Score',\n",
    "                              titleFontSize = 18,\n",
    "                              labelFontSize = 14\n",
    "                          )\n",
    "                         )\n",
    "    ).properties(\n",
    "        width = 250,\n",
    "        height = 300,\n",
    "        title = alt.TitleParams(text = gene,\n",
    "                                fontSize = 20\n",
    "                               )\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "\n",
    "    plot.display()\n",
    "    return plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def phosphosite_heatmap(bard1_df, brca1_df):\n",
    "\n",
    "    bard1_phosphosite = [575, 576, 617, 619]\n",
    "    brca1_phosphosite = [1655, 1656, 1700, 1702]\n",
    "\n",
    "    bard1_df = bard1_df.copy()\n",
    "    brca1_df = brca1_df.copy()\n",
    "\n",
    "    bard1_df = bard1_df.loc[bard1_df['AApos'].isin(bard1_phosphosite)]\n",
    "\n",
    "    #WARN reannotation for G576 and T617\n",
    "    bard1_warn_reannotated = pd.DataFrame({'pos_id': ['214745805:A','214745805:G', '214745806:A', '214745806:G', '214745806:T' ,'214745805:T', '214745119:A', '214745119:C', '214745119:T', '214745121:A', '214745121:C', '214745121:G'],\n",
    "                                           'score': [-0.109977, -0.0175886, -0.0298891, -0.00470975, -0.0413407, 0.0208659, -0.00814722, -0.00228625, 0.00051614, -0.00129351, 0.018338, -0.0955636],\n",
    "                                           'amino_acid_change': ['G576V', 'G576A', 'G576C', 'G576R', 'G576S', 'G576D', 'T617T', 'T617T', 'T617T', 'T617S', 'T617A', 'T617P'],\n",
    "                                           'functional_consequence': ['functionally_abnormal', 'functionally_normal', 'functionally_normal', 'functionally_normal', 'indeterminate', 'functionally_normal', 'functionally_normal', 'functionally_normal', 'functionally_normal', 'functionally_normal', 'functionally_normal', 'functionally_abnormal'],\n",
    "                                           'AApos': [576, 576, 576, 576, 576, 576, 617, 617, 617, 617, 617, 617]\n",
    "                                          })\n",
    "\n",
    "    bard1_df = pd.concat([bard1_df, bard1_warn_reannotated]).reset_index()\n",
    "                                           \n",
    "    brca1_df = brca1_df.loc[brca1_df['AApos'].isin(brca1_phosphosite)]\n",
    "\n",
    "    bard1_df['nAA'] = bard1_df['amino_acid_change'].transform(lambda x: x[-1])\n",
    "    bard1_df['oAA'] = bard1_df['amino_acid_change'].transform(lambda x: x[0])\n",
    "    brca1_df['nAA'] = brca1_df['hgvs_pro'].transform(lambda x: x[-1])\n",
    "    brca1_df['oAA'] = brca1_df['hgvs_pro'].transform(lambda x: x.split('.')[1][0])\n",
    "\n",
    "    bard1_df['aa_plot'] = bard1_df['oAA'] + bard1_df['AApos'].astype(str)\n",
    "    brca1_df['aa_plot'] = brca1_df['oAA'] + brca1_df['AApos'].astype(str)\n",
    "\n",
    "    \n",
    "    \n",
    "    aa_properties = {\n",
    "        # Positive charge (basic)\n",
    "        'R': 'Positive',  # Arginine\n",
    "        'K': 'Positive',  # Lysine\n",
    "        'H': 'Positive',  # Histidine (can be positive at physiological pH)\n",
    "        \n",
    "        # Negative charge (acidic)\n",
    "        'D': 'Negative',  # Aspartic acid\n",
    "        'E': 'Negative',  # Glutamic acid\n",
    "        \n",
    "        # Polar uncharged\n",
    "        'S': 'Polar',     # Serine\n",
    "        'T': 'Polar',     # Threonine\n",
    "        'N': 'Polar',     # Asparagine\n",
    "        'Q': 'Polar',     # Glutamine\n",
    "        'Y': 'Polar',     # Tyrosine\n",
    "        'C': 'Polar',     # Cysteine (can form disulfide bonds)\n",
    "        \n",
    "        # Hydrophobic (nonpolar)\n",
    "        'A': 'Hydrophobic',  # Alanine\n",
    "        'V': 'Hydrophobic',  # Valine\n",
    "        'I': 'Hydrophobic',  # Isoleucine\n",
    "        'L': 'Hydrophobic',  # Leucine\n",
    "        'M': 'Hydrophobic',  # Methionine\n",
    "        'F': 'Hydrophobic',  # Phenylalanine\n",
    "        'W': 'Hydrophobic',  # Tryptophan\n",
    "        \n",
    "        # Special cases\n",
    "        'G': 'Special',   # Glycine (smallest, flexible)\n",
    "        'P': 'Special',   # Proline (helix breaker, rigid)\n",
    "    }\n",
    "\n",
    "    bard1_df['type_change'] = bard1_df['nAA'].map(aa_properties)\n",
    "    brca1_df['type_change'] = brca1_df['nAA'].map(aa_properties)\n",
    "\n",
    "    brca1_df = brca1_df.groupby(['aa_plot', 'type_change']).agg({\n",
    "        'score': 'median'}\n",
    "                                                               ).reset_index()\n",
    "\n",
    "    bard1_df = bard1_df.groupby(['aa_plot', 'type_change']).agg({\n",
    "        'score': 'median'}\n",
    "                                                               ).reset_index()\n",
    "\n",
    "    brca1_map = heatmap(brca1_df, 'BRCA1')\n",
    "    bard1_map = heatmap(bard1_df, 'BARD1')\n",
    "\n",
    "    return brca1_map, bard1_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    bard1, brca1 = read_data(bard1_data, brca1_data,brca1_int_cutoff, brca1_path_cutoff, domain)\n",
    "    #to_plot = missense_sensitivity(bard1,brca1)\n",
    "    #missense_resi(bard1,brca1)\n",
    "    #bar_plot(to_plot)\n",
    "\n",
    "    if domain == 'BRCT':\n",
    "        brca1_map, bard1_map = phosphosite_heatmap(bard1, brca1)\n",
    "\n",
    "        #brca1_map.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_5_brca1_phophomap.png', ppi = 400)\n",
    "        #bard1_map.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_5_bard1_phophomap.png', ppi = 400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
