{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = '/Users/ivan/Downloads/BARD1.snvscores.tsv' #path to SGE scores\n",
    "ref_path = '../Data/SNV_filtering_inputs/20240809_BARD1_SNVlib_ref_seqs_intron_annotated.xlsx' #path to reference sequence\n",
    "coord_file = '../Data/SNV_filtering_inputs/20250415_BARD1_filter_entry.xlsx'\n",
    "#coords = [(214809,214809500),(214797050,214797156)] #genomic coordinates for exon to make map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_region_coords(file): #Reads input file to get coordinates for each SGE target\n",
    "    df = pd.read_excel(file, sheet_name = 'targets')\n",
    "\n",
    "    coords = []\n",
    "\n",
    "    i = 0\n",
    "    while i < len(df):\n",
    "        target = df['target'][i]\n",
    "        target_start = df['start'][i]\n",
    "        target_end = df['end'][i]\n",
    "\n",
    "        start_end = (target_start, target_end)\n",
    "        full_tuple = (target, start_end)\n",
    "\n",
    "        coords.append(full_tuple)\n",
    "\n",
    "        i += 1\n",
    "        \n",
    "    return coords"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def read_scores(file,region): #reads scores\n",
    "    \n",
    "    data = pd.read_csv(file, sep = '\\t')\n",
    "    data = data.rename(columns = {'simplified_consequence': 'Consequence', 'score': 'snv_score_minmax'})\n",
    "    data['pos'] = data['pos'].astype(str)\n",
    "    data['pos_id'] = data['pos'] + ':' + data['allele']\n",
    "    \n",
    "    data = data[['exon','target','pos', 'pos_id', 'Consequence', 'snv_score_minmax', 'amino_acid_change', 'functional_consequence']]\n",
    "    data = data.loc[data['target'].isin([region])]\n",
    "    \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def get_reference(ref, coords,region): #pulls out reference sequence\n",
    "    start, end = coords\n",
    "\n",
    "    #Generates list of coordinates for each target\n",
    "    list_coords = []\n",
    "    for i in range(end, start + 1): #end and start flipped due to input having start coord > end coord on antisense strnad\n",
    "        list_coords.append(i)\n",
    "\n",
    "    ref = pd.read_excel(ref)\n",
    "    ref = ref.loc[ref['target'].isin([region])]\n",
    "    ref = ref[['target', 'Reference', 'pos']]\n",
    "    x_coord = ref.loc[ref['pos'].isin(list_coords)]\n",
    "    \n",
    "    return x_coord\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def reverse_complement(seq_string):\n",
    "    reverse_seq = seq_string[::-1]\n",
    "    reverse_comp_list = []\n",
    "    for char in reverse_seq:\n",
    "        if char == \"A\":\n",
    "            reverse_comp_list.append(\"T\")\n",
    "        elif char == \"G\":\n",
    "            reverse_comp_list.append(\"C\")\n",
    "        elif char == \"C\":\n",
    "            reverse_comp_list.append(\"G\")\n",
    "        else:\n",
    "            reverse_comp_list.append(\"A\")\n",
    "    reverse_compliment_str = \"\".join(reverse_comp_list)\n",
    "    return reverse_compliment_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def reverse_comp_ref(x_ref): #reverse complements reference for antisense gene\n",
    "\n",
    "    ref_list = x_ref['Reference'].tolist()\n",
    "    ref_string = ''.join(ref_list)\n",
    "    ref_string = ref_string.upper()\n",
    "    reversed = reverse_complement(ref_string)\n",
    "\n",
    "    reversed_ref = []\n",
    "    for char in reversed:\n",
    "        reversed_ref.append(char)\n",
    "\n",
    "    x_ref = x_ref[::-1].reset_index(drop = True)\n",
    "\n",
    "\n",
    "    x_ref['Reference'] = reversed_ref\n",
    "\n",
    "    x_ref_reversed = x_ref\n",
    "\n",
    "    return x_ref_reversed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def row_enumerate_ref(ref_df): #enumerates each row for heat map and each column for the base pair number\n",
    "    ref_df['Row'] = None\n",
    "\n",
    "    bases = ['A', 'C', 'G', 'T']\n",
    "\n",
    "    for base in bases:\n",
    "        ref_df.loc[ref_df['Reference'] == base, 'Row'] = base\n",
    "\n",
    "    bp_num = []\n",
    "    i = 0\n",
    "    while i < len(ref_df):\n",
    "        bp_num.append(i + 1)\n",
    "\n",
    "        i += 1\n",
    "    ref_df['Column'] = bp_num\n",
    "\n",
    "    return ref_df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_posid(string): #to reverse complement pos_id for antisense gene to get base change on coding strand\n",
    "    split = string.split(':')\n",
    "    reversed = reverse_complement(split[1])\n",
    "\n",
    "    split[1] = reversed\n",
    "\n",
    "    reversed_id = ':'.join(split)\n",
    "    \n",
    "    return reversed_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_data(df): #groups consequence of SNVs, enumerates row for base change\n",
    "    df = df.reset_index(drop = True)\n",
    "  \n",
    "    df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "    df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "    df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "    df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "    df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "    df.loc[df['Consequence'].str.contains('splic'), 'Consequence'] = 'Splice'\n",
    "    df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "    df.loc[df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "    \n",
    "\n",
    "    df['rev_pos_id'] = df['pos_id'].transform(lambda x: reverse_posid(x))\n",
    "    df['Row'] = df['rev_pos_id'].transform(lambda x: x[-1])\n",
    "\n",
    "    df = df[::-1].reset_index(drop = True)\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def column_enumerate_data(df,refdf): #enumerates column for basepair that was changed\n",
    "    column_dict = {} #dictionary to store the column number for each genomic coordinate\n",
    "\n",
    "    i = 0\n",
    "    while i < len(refdf): #makes the dictionary\n",
    "        coord = refdf['pos'][i]\n",
    "        col = refdf['Column'][i]\n",
    "\n",
    "        column_dict[coord] = col\n",
    "\n",
    "        i += 1\n",
    "\n",
    "    df['Column'] = np.nan #empty column to hold column values\n",
    "\n",
    "    j = 0\n",
    "    while j < len(df): #assigns the column values\n",
    "        id = df['pos_id'][j]\n",
    "        split = id.split(':')\n",
    "        coord = int(split[0])\n",
    "        col = column_dict[coord]\n",
    "\n",
    "        df.loc[df['pos_id'] == id, 'Column'] = col\n",
    "\n",
    "        j += 1\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def heatmap(data, letters,region):\n",
    "    \n",
    "    # Filter out the cells that will display letters from the heatmap dataset\n",
    "    heatmap_data = data.merge(letters, on=['Row', 'Column'], how='left', indicator=True)\n",
    "    heatmap_data = heatmap_data[heatmap_data['_merge'] == 'left_only'].drop(columns=['Reference', '_merge'])\n",
    "\n",
    "    # Define the rectangle size and spacing\n",
    "    rect_size = 15\n",
    "    spacing = 7.5\n",
    "\n",
    "    total_width = (rect_size + spacing) * len(letters) - spacing\n",
    "    total_height = (rect_size + spacing) * len(data['Row'].unique()) - spacing\n",
    "\n",
    "    target = region.split('_')\n",
    "    title_s = 'Exon ' + target[1]\n",
    "\n",
    "    # Create the background heatmap with borders\n",
    "    background = alt.Chart(heatmap_data).mark_rect(\n",
    "        width=rect_size,\n",
    "        height=rect_size,\n",
    "        strokeWidth=2\n",
    "    ).encode(\n",
    "        x=alt.X('pos_x:O', sort = 'descending', title='Basepair', axis=alt.Axis(labelAngle=270)),\n",
    "        y=alt.Y('Row:N', title='SNV'),\n",
    "        color=alt.condition(\n",
    "        'datum.is_wt == true',\n",
    "        alt.value('#000000'),\n",
    "        alt.condition(\n",
    "            alt.datum.score <= -0.5,\n",
    "            alt.value('#ff0000'),\n",
    "            alt.condition(\n",
    "                alt.datum.score >= 0,\n",
    "                alt.value('#0000ff'),\n",
    "                alt.Color('snv_score_minmax:Q', title = 'SGE Score',\n",
    "                          scale = alt.Scale(\n",
    "                              domain = [-0.5, 0],\n",
    "                              range = ['#ff0000', '#a6a6a6', '#0000ff']\n",
    "                          )\n",
    "                         )\n",
    "            )\n",
    "        )\n",
    "    ),\n",
    "        stroke=alt.Stroke('Consequence:N', title='Consequence', legend = alt.Legend(symbolFillColor = 'white')),\n",
    "        tooltip = [alt.Tooltip('amino_acid_change', title = 'Amino Acid Change: '),\n",
    "                   alt.Tooltip('Consequence', title = 'Consequence: '),\n",
    "                   alt.Tooltip('snv_score_minmax', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Functional Consequence: '),\n",
    "                   alt.Tooltip('pos_id', title = 'Position ID: ')\n",
    "                  ]\n",
    "                   \n",
    "    ).properties(\n",
    "        width= total_width,\n",
    "        height= total_height\n",
    "    )\n",
    "    \n",
    "    # Create the text overlay\n",
    "    letters['pos'] = letters['pos'].astype(str) #Sets data type of pos column to string\n",
    "    letters = letters.rename(columns = {'pos': 'pos_x'}) #Renames to have same x encoding as heat map\n",
    "\n",
    "    #Builds text overlay\n",
    "    text = alt.Chart(letters).mark_text(\n",
    "        align='center',\n",
    "        baseline='middle',\n",
    "        fontSize=14\n",
    "    ).encode(\n",
    "        x=alt.X('pos_x:O', sort = 'descending', title='Basepair'),\n",
    "        y=alt.Y('Row:N', title='SNV'),\n",
    "        text='Reference:N',\n",
    "        color=alt.value('black')\n",
    "    )\n",
    "    \n",
    "    # Combine the background and text\n",
    "    heatmap = alt.layer(\n",
    "        background, text\n",
    "    ).properties(\n",
    "        title = title_s\n",
    "    )\n",
    "\n",
    "    # Display the chart\n",
    "    heatmap.show()\n",
    "\n",
    "    return heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    all_maps = []\n",
    "    coords = get_region_coords(coord_file)\n",
    "    \n",
    "    for elem in coords:\n",
    "        region, ref_coords = elem\n",
    "        data = read_scores(file,region)\n",
    "        ref = get_reference(ref_path, ref_coords, region)\n",
    "        ref_reversed = reverse_comp_ref(ref)\n",
    "        ref_enumerated = row_enumerate_ref(ref_reversed)\n",
    "        sge_row_enumerated = process_data(data)\n",
    "        sge_ready = column_enumerate_data(sge_row_enumerated, ref_enumerated)\n",
    "        map = heatmap(sge_ready, ref_enumerated,region)\n",
    "        all_maps.append(map)\n",
    "\n",
    "    #i = 1\n",
    "    #map_1 = all_maps[0]\n",
    "    #while i < len(all_maps):\n",
    "        #if i == 1:\n",
    "            #joined = alt.vconcat(map_1,all_maps[1])\n",
    "\n",
    "            #i += 1\n",
    "        #elif 1 < i < len(all_maps):\n",
    "            #joined = alt.vconcat(joined,all_maps[i])\n",
    "\n",
    "            #i += 1\n",
    "        #else:\n",
    "            #joined = joined.configure_view(\n",
    "                    #stroke = None\n",
    "            #)\n",
    "\n",
    "            #i += 1\n",
    "\n",
    "        #i += 1\n",
    "    \n",
    "    #joined.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
