{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "from pathlib import Path\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '../Data/QC_dev_data/' #Path to data\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_data(data): #reads RNA/DNA count files and merges them\n",
    "               \n",
    "    data_path = Path(data) #Creates path variable\n",
    "    all_data = list(data_path.glob('*.tsv')) #Gets list of all TSV files in folder\n",
    "\n",
    "    all_dfs = [] #Empty list to hold dataframes for each gene dataset\n",
    "\n",
    "    #Iterates through all_data list and gets dataframe for each gene\n",
    "    for elem in all_data:\n",
    "        df = pd.read_csv(elem, sep = '\\t') #Reads TSV\n",
    "        df = df[['pos', 'simplified_consequence', 'standard_error', '95_ci_upper', '95_ci_lower','exon', 'score']] #Keeps useful columns\n",
    "        df = df.dropna(subset = ['simplified_consequence']) #Drops NA values from simplified consequence\n",
    "        all_dfs.append(df)\n",
    "\n",
    "    return all_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def std_error_vs_coord(list_df): #Creates plot of std. error vs. genomic coordinate\n",
    "\n",
    "    path = '/Users/ivan/Desktop/QC_plots/' #Path to plot output folder\n",
    "\n",
    "    #Iterates through each provided dataframe and generates a std. error vs. genomic coordinate plot for each gene\n",
    "    for gene in list_df:\n",
    "        \n",
    "        gene['exon_sort'] = gene['exon'].apply(lambda x: int(re.search(r'X(\\d+)', x).group(1)) if re.search(r'X(\\d+)', x) else 0) #Regex search to get exon number\n",
    "        gene_name = gene['exon'][0].split('_')[0] #Gets gene name for final plot saving\n",
    "  \n",
    "        #Builds std. error vs. genomic coordinate plot faceted by exon number\n",
    "        plot = alt.Chart(gene).mark_circle().encode(\n",
    "            x = alt.X('pos:O', axis = alt.Axis(labels = False, title = 'Genomic Coordinate')),\n",
    "            y = alt.Y('standard_error:Q', axis = alt.Axis(title = 'Standard Error')),\n",
    "            color = alt.Color('simplified_consequence', legend = alt.Legend(title = \n",
    "                                                                            'Consequence'\n",
    "                                                                           )),\n",
    "            tooltip = [alt.Tooltip('pos', title = 'Coordinate: '),\n",
    "                       alt.Tooltip('standard_error', title = 'Standard Error: '),\n",
    "                       alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                       alt.Tooltip('simplified_consequence', title = 'Consequence: ')\n",
    "                      ]\n",
    "        ).properties(\n",
    "            height = 400,\n",
    "            width = 600\n",
    "        ).facet(facet = alt.Facet('exon', title = 'Std. Error vs. Genomic Coordinate', sort = {'field': 'exon_sort', 'op': 'min', 'order': 'ascending'}),\n",
    "                columns = 3\n",
    "        ).resolve_scale(\n",
    "            x = 'independent'\n",
    "        )\n",
    "\n",
    "        full_path = path + gene_name + '.png' # Builds full path for final plot to be saved\n",
    "\n",
    "        #plot.save(full_path) #Saves plot\n",
    "        plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    df_list = read_data(data_path)\n",
    "    std_error_vs_coord(df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
