{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import altair as alt\n",
    "from natsort import natsorted\n",
    "from scipy.stats import gaussian_kde\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = '../Data/20250423_BARD1_snvscores_IGVFupload.tsv'\n",
    "scores_path = '/Users/ivan/Downloads/all_sge'\n",
    "output_folder = '/Users/ivan/Desktop/batch_sge_figs/'\n",
    "output_format = '.png'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_files(path):\n",
    "    scores = Path(path)\n",
    "\n",
    "    files = sorted(list(scores.glob('*tsv')))\n",
    "\n",
    "    return files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def histogram(file, output_folder, output_format):\n",
    "    \n",
    "    df = pd.read_csv(file, sep = '\\t')\n",
    "    df = df[['exon','consequence','score']]\n",
    "    first_exon = df['exon'][0]\n",
    "\n",
    "    gene_name = first_exon.split('_')[0]\n",
    "\n",
    "    df = df.rename(columns = {'score': 'snv_score', 'consequence': 'Consequence'})\n",
    "\n",
    "    df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "    df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "    df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "    df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "    df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "    df.loc[df['Consequence'].str.contains('splic'), 'Consequence'] = 'Splice'\n",
    "    df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "    df.loc[df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "\n",
    "    alt.data_transformers.disable_max_rows() #gets rid of max data length problem\n",
    "    \n",
    "    length = str(len(df)) #gets length of data for title\n",
    "\n",
    "    final_tital = 'Distribution of ' + gene_name + ' SGE Scores ' + '(n = ' + length + ')' #used to build title\n",
    "    bins = 50 #number of bins\n",
    "\n",
    "    sorted = [\"Intron\", \"Missense\", \"Synonymous\", \"Stop Gained\", \"Splice\", \"Start Lost\", 'Stop Lost', 'UTR'] #order for the legend\n",
    "\n",
    "    #Builds histogram\n",
    "    histogram = alt.Chart(df).mark_bar().encode(\n",
    "            alt.X('snv_score', axis = alt.Axis(title = 'SGE Score', labelFontSize = 16, titleFontSize = 20), bin = alt.Bin(maxbins = bins)),\n",
    "            alt.Y('count()', axis = alt.Axis(title = 'Number of Variants', labelFontSize = 16, titleFontSize = 20)),\n",
    "            color = alt.Color('Consequence:N', scale = alt.Scale(scheme = 'category10'), sort = sorted, legend = alt.Legend(titleFontSize = 16, labelFontSize = 14))\n",
    "    ).properties(\n",
    "        width = 800,\n",
    "        height = 400,\n",
    "        title = alt.TitleParams(text = final_tital, fontSize = 22)\n",
    "    ).interactive()\n",
    "\n",
    "    file_name = output_folder + gene_name + '_SGE_histogram' + output_format\n",
    "    \n",
    "    histogram.save(file_name)\n",
    "    histogram.display()\n",
    "\n",
    "    return gene_name\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def ridgeline_plot(file, gene_name, output_folder, output_format):\n",
    "    \n",
    "    def read_scores(file): #reads SGE scores\n",
    "    \n",
    "        df = pd.read_csv(file, sep = '\\t')\n",
    "        df = df.rename(columns = {'score': 'snv_score', 'consequence': 'Consequence'})\n",
    "        df = df[['exon','Consequence','snv_score']]\n",
    "        df = df.loc[df['Consequence'].isin(['missense_variant', 'synonymous_variant', 'intron_variant', 'stop_gained', 'start_lost', 'splicing_variant', 'splice_site_variant', 'UTR_variant'])]\n",
    "\n",
    "        return df\n",
    "\n",
    "    def prep_data(df): #Renames categories to be nicer\n",
    "    \n",
    "        df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "        df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Syn.'\n",
    "        df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "        df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop'\n",
    "        df.loc[df['Consequence'].str.contains('splic'), 'Consequence'] = 'Splice'\n",
    "        df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR'\n",
    "        df.loc[df['Consequence'].str.contains('start'), 'Consequence'] = 'Start Lost'\n",
    "   \n",
    "        return df\n",
    "\n",
    "    def compute_density(df, category): #used to calculate the density - mostly GPT code\n",
    "        values = df['snv_score'].values\n",
    "        density = gaussian_kde(values)\n",
    "        xs = np.linspace(values.min(), values.max(), 200)\n",
    "        ys = density(xs)\n",
    "    \n",
    "        return pd.DataFrame({'SGE_score': xs, 'density': ys, 'Consequence': category})\n",
    "\n",
    "    def make_plot(df,rawdf): #creates plot\n",
    "\n",
    "        df = pd.concat([df,rawdf], ignore_index = True) #concatenates the density data and raw data so that the ticks and density plots share a dataframe (required for graphing)\n",
    "    \n",
    "        #Base creates the density plots using the density data \n",
    "        base = alt.Chart(df).mark_area(\n",
    "            interpolate = 'monotone',\n",
    "            line = True\n",
    "        ).encode(\n",
    "            x = alt.X('SGE_score:Q', axis = alt.Axis(title = 'SGE Score', titleFontSize = 20, labelFontSize = 16)),\n",
    "            y = alt.Y('density:Q', axis = None),\n",
    "            color = alt.Color('Consequence:N', legend = None)\n",
    "        )\n",
    "    \n",
    "        #ticks creates the tick marks using the raw data\n",
    "        ticks = alt.Chart(df).mark_tick(\n",
    "            color = 'black',\n",
    "            thickness = 0.5,\n",
    "            size = 5\n",
    "        ).encode(\n",
    "            x = alt.X('snv_score:Q', title = ''),\n",
    "            y = alt.value(77.5)\n",
    "        )\n",
    "    \n",
    "        #Plots are layered\n",
    "        combined_plot = alt.layer(ticks, base).properties(\n",
    "            width = 1000,\n",
    "            height = 75\n",
    "        )\n",
    "\n",
    "        sorted = [\"Intron\", \"Missense\", \"Syn.\", \"Stop\", \"Splice\", \"Start Lost\", 'UTR']\n",
    "        \n",
    "        #plots are faceted by consequence of the variant to yield the final plot (faceting requires that all data is in the same dataframe\n",
    "        plot = combined_plot.facet(facet = alt.Facet(\n",
    "            row = alt.Row('Consequence:N', title = 'Consequence', sort = sorted)),\n",
    "            spacing = 2, \n",
    "        ).properties(\n",
    "            title = '',\n",
    "            bounds = 'flush'\n",
    "        ).configure_facet(\n",
    "            spacing= 1\n",
    "        ).configure_header(\n",
    "            titleFontSize = 20,\n",
    "            labelFontSize = 16\n",
    "        ).configure_title(\n",
    "            anchor='start'\n",
    "        ).configure_axis(\n",
    "            grid=False\n",
    "        ).configure_view(\n",
    "            stroke = None\n",
    "        )\n",
    "\n",
    "        file_name = output_folder + gene_name + '_SGE_RidgelinePlot' + output_format\n",
    "        plot.save(file_name)\n",
    "        plot.show()\n",
    "\n",
    "    def main():\n",
    "        alt.data_transformers.disable_max_rows()\n",
    "        data = read_scores(file)\n",
    "        relabeled_data = prep_data(data)\n",
    "        density_data = pd.concat([compute_density(relabeled_data[relabeled_data['Consequence'] == category], category)\n",
    "                              for category in relabeled_data['Consequence'].unique()])\n",
    "        \n",
    "        make_plot(density_data,relabeled_data)\n",
    "\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def corr_matrix(file, gene_name, output_folder, output_format):\n",
    "\n",
    "    def read_scores(file): #Reads SGE scores\n",
    "        df = pd.read_csv(file, sep = '\\t')\n",
    "    \n",
    "        all_columns = ['target','D05_R1_lib1','D05_R2_lib1','D05_R3_lib1','D13_R1_lib1','D13_R2_lib1','D13_R3_lib1'] #List of columns to retain\n",
    "        \n",
    "        df = df[all_columns] #Gets necessary columns\n",
    "        \n",
    "        df = df.rename(\n",
    "            columns = {\n",
    "                'D05_R1_lib1': 'D05 R1', \n",
    "                'D05_R2_lib1': 'D05 R2', \n",
    "                'D05_R3_lib1': 'D05 R3', \n",
    "                'D13_R1_lib1': 'D13 R1',\n",
    "                'D13_R2_lib1': 'D13 R2', \n",
    "                'D13_R3_lib1': 'D13 R3',\n",
    "            }\n",
    "        ) #Renames columns to be consistent with downstream code\n",
    "\n",
    "            \n",
    "        df['target'] = df['target'].str.extract(r'_X(.*)$') #Rewrites targets column to contain target name only\n",
    "    \n",
    "        return df\n",
    "\n",
    "    \n",
    "    def group_targets(df): #Groups targets for correlation testing\n",
    "        \n",
    "        grouped = df.groupby('target') #Creates groupby objects for each SGE region\n",
    "        r_lists = [] #list that will hold lists that have correlation\n",
    "        combos = [('D05 R1','D05 R2'),('D05 R1','D05 R3'),('D05 R2','D05 R3'),\n",
    "                  ('D13 R1','D13 R2'),('D13 R1','D13 R3'),('D13 R2','D13 R3')\n",
    "                 ] #pairwise arguments for correlation testing\n",
    "        \n",
    "        for group_name, group_df in grouped: #iterates through groupby objects\n",
    "            group_df = group_df[['D05 R1', 'D05 R2', 'D05 R3', \n",
    "               'D13 R1', 'D13 R2', 'D13 R3']] #pulls out count columns only\n",
    "            \n",
    "            for elem in combos: #iterates through each combination of replicates for r-testing\n",
    "                output = [] #output list\n",
    "                col1, col2 = elem #unpacks tuple that is combination\n",
    "                output.append(group_name) #adds SGE target name\n",
    "                output.append(col1 + ' vs ' + col2) #adds which replicates were tested\n",
    "                output.append(compute_r(group_df,col1,col2)) #appends r\n",
    "                r_lists.append(output) #appends final list to output list\n",
    "                \n",
    "        return r_lists\n",
    "    \n",
    "    \n",
    "    def compute_r(group,col1,col2): #does the correlation math\n",
    "        return group[col1].corr(group[col2])\n",
    "    \n",
    "    \n",
    "    def parse_r_lists(r_list): #Parses through grouped lists and create Pearson R values\n",
    "        \n",
    "        target = [] #list to hold SGE target names\n",
    "        test_type = [] #list to hold the replicates compared\n",
    "        r = [] #list to hold r values\n",
    "        \n",
    "        for elem in r_list: #iterates through each item in list and appends respective information\n",
    "            target.append(elem[0])\n",
    "            test_type.append(elem[1])\n",
    "            r.append(round(elem[2],3))\n",
    "    \n",
    "        data = {\n",
    "                'Targets' : target,\n",
    "                'Tests' : test_type,\n",
    "                'r_correlation': r\n",
    "        } #Dictionary to build dataframe\n",
    "        \n",
    "        to_map = pd.DataFrame(data) #Final dataframe with all R-values\n",
    "        \n",
    "        return to_map\n",
    "\n",
    "    def heatmap(pivoted):\n",
    "        \n",
    "        targets = set(pivoted['Targets'].tolist())\n",
    "        targets = natsorted(targets)\n",
    "        title = 'Correlation of Replicates' + ' (' + gene_name + ')'\n",
    "        \n",
    "        graph = alt.Chart(pivoted, title = alt.TitleParams(text = title, fontSize = 32)).mark_rect().encode(\n",
    "                    x = alt.X('Tests:N', axis = alt.Axis(title = '', titleFontSize = 28, labelFontSize = 22, labelLimit = 300, labelAngle = 45)),\n",
    "                    y = alt.Y('Targets', axis = alt.Axis(title = 'SGE Target Region', titleFontSize = 28, labelFontSize = 22), sort = targets),\n",
    "                    color = alt.Color('r_correlation:Q', scale = alt.Scale(domain = [.2, 1]), legend = alt.Legend(title = \"Pearson's r\", titleFontSize = 24,labelFontSize = 22)),\n",
    "                    tooltip = [alt.Tooltip('r_correlation', title = \"Pearson's r: \")]\n",
    "        ).properties(\n",
    "            width = 800,\n",
    "            height = 700\n",
    "        )\n",
    "\n",
    "        file_name = output_folder + gene_name + '_SGE_CorrelationMatrix' + output_format\n",
    "        graph.save(file_name)\n",
    "        graph.display()\n",
    "    \n",
    "    \n",
    "    def main():\n",
    "        data = read_scores(file) \n",
    "        r_lists = group_targets(data)\n",
    "        test = parse_r_lists(r_lists)\n",
    "        heatmap(test)\n",
    "\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def stacked_bars(file, gene_name, output_folder, output_format):\n",
    "\n",
    "    def read_data(path): #Reads data\n",
    "        df = pd.read_csv(file, sep = '\\t') #Reads SGE score file\n",
    "    \n",
    "        df = df.rename(columns = {'consequence': 'Consequence', 'score': 'snv_score'}) #Renames to harmonize old code with new column names\n",
    "        \n",
    "        df = df[['exon','Consequence','snv_score', 'functional_consequence']] #pulls out relevant columns in the dataframe\n",
    "    \n",
    "        filtered_consequences = ['missense_variant','synonymous_variant','stop_gained'] #Focusing on these variant types only \n",
    "    \n",
    "        df = df.loc[df['Consequence'].isin(filtered_consequences)] #Filters for desired variant types\n",
    "        \n",
    "        df = df.reset_index(drop = True) #Resets index\n",
    "        \n",
    "        return df\n",
    "\n",
    "    def prep_data(df): #renames VEP consequence categories to be more human friendly\n",
    "        \n",
    "        df.loc[df['Consequence'].str.contains('missense_variant'), 'Consequence'] = 'Missense'\n",
    "        df.loc[df['Consequence'].str.contains('synonymous_variant'), 'Consequence'] = 'Synonymous'\n",
    "        #df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "        df.loc[df['Consequence'].str.contains('stop_gained'), 'Consequence'] = 'Stop'\n",
    "        #df.loc[(df['Consequence'] == 'splice_polypyrimidine_tract_variant') |(df['Consequence'] == 'splice_region_variant') | (df['Consequence'] == 'splice_acceptor_variant') | (df['Consequence'] == 'splice_donor_region_variant') | (df['Consequence'] == 'splice_donor_5th_base_variant') | (df['Consequence'] == 'splice_donor_variant'),'Consequence'] = 'Splice'\n",
    "        #df.loc[df['Consequence'] == '3_prime_UTR_variant', 'Consequence'] = 'UTR'\n",
    "    \n",
    "        return df\n",
    "\n",
    "\n",
    "    def rewrite_targets(df): #Rewrites SGE target to get exon number only\n",
    "        \n",
    "        df['target'] = df['exon'].str.extract(r'_X(\\d+)')\n",
    "        \n",
    "        return df\n",
    "\n",
    "\n",
    "    def exon_stats(df): #Generates summary dataframe with % of variants in each functional class for each exon\n",
    "    \n",
    "        exon_df_list = [] #List to hold summary dataframes for each target\n",
    "        grouped_exons = df.groupby(['target','Consequence'])\n",
    "        \n",
    "        for group_name, group_df in grouped_exons:\n",
    "            exon, var_type = group_name\n",
    "            scores = group_df['functional_consequence'].tolist()\n",
    "    \n",
    "            non = 0 #counters for nonfunctional group\n",
    "    \n",
    "            inter = 0 #counter for indeterminate group\n",
    "            \n",
    "            for elem in scores: #Iterates through grouped scores and determines number of variants in each functional class\n",
    "                if elem == 'functionally_abnormal': #Boolean for non-functional \n",
    "                    non += 1\n",
    "                elif elem == 'indeterminate': #Boolean for indeterminate\n",
    "                    inter += 1\n",
    "    \n",
    "            non_per = (non / len(group_df)) * 100 #Gets % non-functional\n",
    "            inter_per = (inter / len(group_df)) * 100 #Gets % intermediate\n",
    "            same_per = (100 - (non_per + inter_per)) #Gets % functional\n",
    "    \n",
    "            #Creates 3 separate dataframes for each functional class\n",
    "            exon_non_df = pd.DataFrame({'Exon': exon, 'Consequence': var_type, 'Function Type': 'Nonfunctional', 'Percent': non_per}, index = [0])\n",
    "            exon_inter_df = pd.DataFrame({'Exon': exon, 'Consequence': var_type, 'Function Type': 'Indeterminate', 'Percent': inter_per}, index = [0])\n",
    "            exon_same_df = pd.DataFrame({'Exon': exon, 'Consequence': var_type, 'Function Type': 'Functional', 'Percent': same_per}, index = [0])\n",
    "                \n",
    "            exon_df = pd.concat([exon_non_df, exon_inter_df, exon_same_df]) #Concatnates all dataframes together\n",
    "    \n",
    "            exon_df_list.append(exon_df) #Appends to final dataframe list\n",
    "    \n",
    "        all_exons_df = pd.concat(exon_df_list) #Concatenates all summary dataframes\n",
    "        \n",
    "        all_exons_df = all_exons_df.reset_index(drop = True) #resets index\n",
    "        \n",
    "        return all_exons_df\n",
    "\n",
    "\n",
    "    def stacked_bars(df): #Creates the stacked bar chart\n",
    "        \n",
    "        exons = natsorted(set(list(df['Exon'].tolist()))) #Gets sorted list of exons\n",
    "        \n",
    "        #Builds stacked bar chart\n",
    "        chart = alt.Chart(df).mark_bar().encode(\n",
    "                x = alt.X('Exon:O', axis = alt.Axis(labelAngle = 0, labelFontSize = 16, titleFontSize = 20), sort = exons),\n",
    "                y = alt.Y('Percent',axis = alt.Axis(labelFontSize = 16, titleFontSize = 20)),\n",
    "                tooltip = [alt.Tooltip('Function Type', title = 'Functional Class: '), \n",
    "                            alt.Tooltip('Percent', title = 'Percent: ')],\n",
    "                color = alt.Color('Function Type', title = 'Functional Class', legend = alt.Legend(titleFontSize = 18, labelFontSize = 16))\n",
    "        ).properties(\n",
    "            width =400,\n",
    "            height = 500\n",
    "        ).facet(facet = alt.Facet('Consequence',\n",
    "            sort = ['Synonymous', 'Missense', 'Stop']\n",
    "                    )\n",
    "        )\n",
    "    \n",
    "        chart = chart.configure_header(\n",
    "            titleFontSize = 20, \n",
    "            labelFontSize = 16\n",
    "        )\n",
    "\n",
    "        file_name = output_folder + gene_name + '_SGE_StackedBars' + output_format\n",
    "        chart.save(file_name)\n",
    "        chart.show()\n",
    "\n",
    "\n",
    "    def main():\n",
    "        data = read_data(file)\n",
    "        reannotated = prep_data(data)\n",
    "        num_exons = rewrite_targets(data)\n",
    "        all_stats = exon_stats(num_exons)\n",
    "        stacked_bars(all_stats)\n",
    "\n",
    "\n",
    "\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def strip_plot(file, gene_name,output_folder,output_format):\n",
    "    \n",
    "    def read_data(file):\n",
    "        df = pd.read_csv(file, sep = '\\t')\n",
    "        df = df.rename(columns = {'consequence': 'Consequence'})\n",
    "        df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "        df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "        df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "        df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "        df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "        df.loc[df['Consequence'].str.contains('site'), 'Consequence'] = 'Canonical Splice'\n",
    "        df.loc[df['Consequence'].str.contains('ing_var'), 'Consequence'] = 'Splice Region'\n",
    "        df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "        df.loc[df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "        \n",
    "        return df\n",
    "\n",
    "    def strip_plot(df):\n",
    "    \n",
    "        \n",
    "        nf_line = alt.Chart(pd.DataFrame({'x': [-0.089]})).mark_rule(color = 'red').encode(\n",
    "            x = 'x')\n",
    "    \n",
    "        func_lin = alt.Chart(pd.DataFrame({'x': [-0.077]})).mark_rule(color = 'blue').encode(\n",
    "            x = 'x')\n",
    "    \n",
    "        sorted = [\"Intron\", \"Missense\", \"Synonymous\", \"Stop Gained\", \"Splice\", \"Start Lost\", 'Stop Lost', 'Splice Region','Canonical Splice', 'UTR']\n",
    "        controls_sorted = [\"Intron\",  \"Synonymous\", \"Stop Gained\", \"Missense\", \"Splice\", \"Start Lost\", 'Stop Lost', 'UTR']\n",
    "        df = df.loc[df['Consequence'].isin(sorted)]\n",
    "\n",
    "        final_tital = 'Distribution of ' + gene_name + ' SGE Scores' #used to build title\n",
    "    \n",
    "        # Get the category10 colors\n",
    "        category10_colors = [\n",
    "            '#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd', \n",
    "            '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf'\n",
    "        ]\n",
    "        \n",
    "        # Map each category to a specific color\n",
    "        color_mapping = {\n",
    "            \"Intron\": category10_colors[0],\n",
    "            \"Synonymous\": category10_colors[1],\n",
    "            \"Stop Gained\": category10_colors[2],\n",
    "            \"Missense\": category10_colors[3],\n",
    "            \"Splice\": category10_colors[4],\n",
    "            \"Start Lost\": category10_colors[5],\n",
    "            \"Stop Lost\": category10_colors[6],\n",
    "            \"UTR\": category10_colors[7]\n",
    "        }\n",
    "        \n",
    "        plot = alt.Chart(df).mark_tick(opacity = 1).encode(\n",
    "            x = alt.X('score:Q',\n",
    "                      axis = alt.Axis(title = '', \n",
    "                                      titleFontSize = 20,\n",
    "                                     labelFontSize = 24)\n",
    "                     ),\n",
    "            y = alt.Y('Consequence:N', \n",
    "                      sort = sorted,\n",
    "                      axis = alt.Axis(title = '',\n",
    "                                     labelFontSize = 24)\n",
    "                     ),\n",
    "            color = alt.Color('Consequence:N',\n",
    "                    legend=None,\n",
    "                    sort = sorted,\n",
    "                    scale = alt.Scale(scheme = 'category10')\n",
    "                             )\n",
    "            ).properties(\n",
    "                width = 800,\n",
    "                height = 400,\n",
    "                title = alt.TitleParams(final_tital, fontSize = 22)\n",
    "            )\n",
    "        \n",
    "        #plot = plot + nf_line + func_lin\n",
    "    \n",
    "        plot = plot.configure_axis(\n",
    "            grid = False\n",
    "        )\n",
    "\n",
    "        file_name = output_folder + gene_name + '_SGE_StripPlot' + output_format\n",
    "        plot.save(file_name)\n",
    "        plot.display()\n",
    "        \n",
    "\n",
    "    def main():\n",
    "        data = read_data(file)\n",
    "        strip_plot(data)\n",
    "\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def main():\n",
    "    files = get_files(scores_path)\n",
    "\n",
    "    for file in files:\n",
    "        gene_name = histogram(file, output_folder, output_format)\n",
    "        ridgeline_plot(file, gene_name, output_folder, output_format)\n",
    "        corr_matrix(file, gene_name, output_folder, output_format)\n",
    "        stacked_bars(file, gene_name, output_folder, output_format)\n",
    "        strip_plot(file, gene_name,output_folder,output_format)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
