{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import numpy as np\n",
    "from natsort import natsorted\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_output = '/Users/ivan/Desktop/test_excel_outputs/20250710_BARD1_202505_data_RNAclassified_beta_collapsed.xlsx'\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_output(file):\n",
    "    df = pd.read_excel(file)\n",
    "    #df = df.drop(columns = ['L2RNA/DNA'])\n",
    "    \n",
    "    df['AApos'] = df['AAsub'].transform(lambda x: x[1: -1])\n",
    "    df = df.loc[~(df['AApos'].isin(['-']))]\n",
    "    df['AApos'] = df['AApos'].astype(int)\n",
    "    df['exon'] = df['target'].str.extract(r'(BARD1_X\\d+)')\n",
    "    \n",
    "    abnormal_df = df[df['RNA_classification'].isin(['high', 'low'])]\n",
    "    normal_df = df[df['RNA_classification'].isin(['normal'])]\n",
    "\n",
    "    return abnormal_df, normal_df, df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_rna_performance(df):\n",
    "\n",
    "    plot = alt.Chart(df).mark_point().encode(\n",
    "        x = 'AApos',\n",
    "        y = 'RNA/DNA',\n",
    "        color = alt.Color('functional_consequence', \n",
    "                          sort = ['functionally_normal', 'indeterminate', 'functionally_abnormal']\n",
    "                         ),\n",
    "        shape = 'RNA_classification',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Region: '),\n",
    "                   alt.Tooltip('AAsub', title = 'Amino Acid Substituion: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 1200,\n",
    "        height = 400, \n",
    "        title = 'Distribution of RNA/DNA Across BARD1'\n",
    "    ).interactive()\n",
    "\n",
    "\n",
    "    chart = alt.Chart(df).mark_point().encode(\n",
    "        x = alt.X('RNA/DNA',\n",
    "                  scale = alt.Scale(domain = [-1.5, 2.5])\n",
    "                 ),\n",
    "        y = alt.Y('snv_score'),\n",
    "        color = 'Consequence',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Region: '),\n",
    "                   alt.Tooltip('AAsub', title = 'Amino Acid Substituion: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400, \n",
    "        title = 'SGE Score vs. RNA Score'\n",
    "    ).interactive()\n",
    "\n",
    "    nf_line = alt.Chart(pd.DataFrame({'y': [-0.089]})).mark_rule(color = 'red').encode(\n",
    "        y = 'y')\n",
    "\n",
    "    func_line = alt.Chart(pd.DataFrame({'y': [-0.077]})).mark_rule(color = 'blue').encode(\n",
    "        y = 'y')\n",
    "\n",
    "    rna_func_line = alt.Chart(pd.DataFrame({'x': [1]})).mark_rule(color = 'red').encode(\n",
    "        x = 'x')\n",
    "    rna_nf_line = alt.Chart(pd.DataFrame({'x': [0]})).mark_rule(color = 'red').encode(\n",
    "        x = 'x')\n",
    "    chart = chart + nf_line + func_line + rna_func_line + rna_nf_line\n",
    "    \n",
    "    plot.display()\n",
    "    chart.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def stem_plot(df):\n",
    "\n",
    "    df = df.loc[df['RNA/DNA'] < 10]\n",
    "    df = df.copy()\n",
    "\n",
    "    abnormal_df = df.loc[(df['functional_consequence'].isin(['functionally_abnormal'])) & (df['RNA_classification'].isin(['low']))]\n",
    "\n",
    "    df['rule_value'] = np.nan\n",
    "    df.loc[(df['functional_consequence'] == 'functionally_abnormal') & (df['RNA_classification'] == 'low'), 'rule_value'] = df.loc[(df['functional_consequence'] == 'functionally_abnormal') & (df['RNA_classification'] == 'low'), 'RNA/DNA']\n",
    "\n",
    "    exons = list(set(df['exon'].tolist()))\n",
    "    exons = natsorted(exons)\n",
    "\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a', # yellow\n",
    "    '#6AA84F', # med green\n",
    "    '#93C47D', # light green\n",
    "    '#888888', # med gray\n",
    "    '#000000', # black\n",
    "    '#1170AA', # darker blue\n",
    "    '#CFCFCF' # light gray\n",
    "        \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'synonymous_variant',\n",
    "        'missense_variant',  \n",
    "        'stop_gained',\n",
    "        'intron_variant', \n",
    "        'UTR_variant',\n",
    "        'stop_lost',\n",
    "        'start_lost',\n",
    "        'splice_site_variant', \n",
    "        'splicing_variant',\n",
    "    ]\n",
    "\n",
    "    \n",
    "    base = alt.Chart(df)\n",
    "    base_dot = base.mark_point(filled = True, size = 50).encode(\n",
    "        x = alt.X('CDSpos:Q', \n",
    "                  scale = alt.Scale(zero = False, \n",
    "                                    padding = 5\n",
    "                                   )\n",
    "                 ),\n",
    "        y = alt.Y('RNA/DNA:Q', \n",
    "                  title = 'RNA/DNA',\n",
    "                  scale = alt.Scale(domain = [-2.1, 3],\n",
    "                                    padding = 5)\n",
    "                 ),\n",
    "        color = alt.Color('Consequence',\n",
    "                          scale = alt.Scale(\n",
    "                              range = palette, \n",
    "                              domain = variant_types\n",
    "                          )\n",
    "                         ),\n",
    "        shape = alt.Shape('functional_consequence',\n",
    "                          sort = ['functionally_normal', \n",
    "                                  'functionally_abnormal', \n",
    "                                  'indeterminate'\n",
    "                                 ]\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '), \n",
    "                   alt.Tooltip('AAsub', title = 'Amino Acid Sub: '),\n",
    "                   alt.Tooltip('Consequence', title = 'Consequence: '),\n",
    "                   alt.Tooltip('snv_score', title = 'SGE Score: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 800, \n",
    "        height = 400\n",
    "    )\n",
    "\n",
    "    rule = base.mark_rule().encode(\n",
    "        x = alt.X('CDSpos:Q',\n",
    "                  scale = alt.Scale(zero = False,\n",
    "                                    padding = 5\n",
    "                                   )\n",
    "                 ),\n",
    "        y = alt.Y('rule_value:Q', \n",
    "                  scale = alt.Scale(padding = 5)\n",
    "                 )\n",
    "    )\n",
    "\n",
    "    chart = (rule + base_dot)\n",
    "    \n",
    "\n",
    "    chart = chart.facet(facet = alt.Facet('exon:N',\n",
    "                                         sort = exons\n",
    "                                         ),\n",
    "                        columns = 4\n",
    "                       ).resolve_scale(\n",
    "        x = 'independent', \n",
    "        y = 'independent'\n",
    "    ).resolve_axis(\n",
    "    x='independent',\n",
    "    y='independent'\n",
    "    ).interactive()\n",
    "    \n",
    "    chart.display()\n",
    "    #base_dot.display()\n",
    "    #rule.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nmd_across_gene(df):\n",
    "    \n",
    "    df = df.loc[df['Consequence'].isin(['stop_gained'])]\n",
    "\n",
    "    color_scale = alt.Scale(\n",
    "        domain =['functionally_normal', 'functionally_abnormal'],\n",
    "        range = ['#1f77b4', '#d62728']\n",
    "    )\n",
    "    \n",
    "    plot = alt.Chart(df).mark_point().encode(\n",
    "        x = 'AApos',\n",
    "        y = 'RNA/DNA',\n",
    "        color = alt.Color('functional_consequence', \n",
    "                          scale = color_scale,\n",
    "                         ),\n",
    "        shape = 'RNA_classification',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Region: '),\n",
    "                   alt.Tooltip('AApos', title = 'Amino Acid Position: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 800,\n",
    "        height = 400, \n",
    "        title = 'NMD Across BARD1'\n",
    "    ).interactive()\n",
    "\n",
    "    plot.display()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def missense_rna_performance(df):\n",
    "\n",
    "    df = df.loc[df['Consequence'].isin(['missense_variant'])]\n",
    "    \n",
    "    plot = alt.Chart(df).mark_point().encode(\n",
    "        x = 'AApos',\n",
    "        y = 'RNA/DNA',\n",
    "        color = alt.Color('functional_consequence', \n",
    "                          sort = ['functionally_normal', 'indeterminate', 'functionally_abnormal']\n",
    "                         ),\n",
    "        shape = 'RNA_classification',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Region: '),\n",
    "                   alt.Tooltip('AAsub', title = 'Amino Acid Substituion: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 800,\n",
    "        height = 400, \n",
    "        title = 'RNA Performance of Missense Variants'\n",
    "    ).interactive()\n",
    "\n",
    "    plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def synonymous_rna_performance(df):\n",
    "    \n",
    "    df = df.loc[df['Consequence'].isin(['synonymous_variant'])]\n",
    "\n",
    "    color_scale = alt.Scale(\n",
    "        domain =['functionally_normal', 'functionally_abnormal'],\n",
    "        range = ['#1f77b4', '#d62728']\n",
    "    )\n",
    "    \n",
    "    plot = alt.Chart(df).mark_point().encode(\n",
    "        x = 'AApos',\n",
    "        y = 'RNA/DNA',\n",
    "        color = alt.Color('functional_consequence', \n",
    "                          scale = color_scale,\n",
    "                         ),\n",
    "        shape = 'RNA_classification',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Region: '),\n",
    "                   alt.Tooltip('AApos', title = 'Amino Acid Position: '), \n",
    "                   alt.Tooltip('snv_score', title = 'SGE Score: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 800,\n",
    "        height = 400, \n",
    "        title = 'RNA Performance of Synonymous Variants'\n",
    "    ).interactive()\n",
    "\n",
    "    plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    ab_data, norm_data, all_data = read_output(rna_output)\n",
    "    all_rna_performance(all_data)\n",
    "    stem_plot(all_data)\n",
    "    nmd_across_gene(all_data)\n",
    "    missense_rna_performance(ab_data)\n",
    "    synonymous_rna_performance(ab_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
