{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sge = '../Data/20250813_BARD1scores_final_FILTERED.xlsx'\n",
    "gnomad = '../Data/20240905_BARD1_gnomADv4.1.0_SNVs.xlsx' #for Ambry\n",
    "carriers = '/Users/ivan/Downloads/7_genes_for_Starita_02282025_hg38.xlsx'\n",
    "ambry = '/Users/ivan/Downloads/Ambry_BARD1.xlsx'\n",
    "bridges_all = '../Data/BRIDGES_data/20250815_BRIDGES_missense_all.xlsx'\n",
    "bridges_population = '../Data/BRIDGES_data/20250815_BRIDGES_missense_population.xlsx'\n",
    "\n",
    "aa_dict = {\n",
    "    'Ala': 'A',\n",
    "    'Cys': 'C',\n",
    "    'Asp': 'D',\n",
    "    'Glu': 'E',\n",
    "    'Phe': 'F',\n",
    "    'Gly': 'G',\n",
    "    'His': 'H',\n",
    "    'Ile': 'I',\n",
    "    'Lys': 'K',\n",
    "    'Leu': 'L',\n",
    "    'Met': 'M',\n",
    "    'Asn': 'N',\n",
    "    'Pro': 'P',\n",
    "    'Gln': 'Q',\n",
    "    'Arg': 'R',\n",
    "    'Ser': 'S',\n",
    "    'Thr': 'T',\n",
    "    'Val': 'V',\n",
    "    'Trp': 'W',\n",
    "    'Tyr': 'Y',\n",
    "    'Ter': '*'   # Termination codon\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_sge(sge):\n",
    "    sge = pd.read_excel(sge) #Reads SGE data\n",
    "    sge = sge.loc[~(sge['functional_consequence'].isin(['indeterminate']))]\n",
    "\n",
    "    sge = sge.rename(columns = {'functional_consequence': 'Classification', \n",
    "                               'consequence': 'Consequence'})\n",
    "\n",
    "    sge.loc[sge['Classification'] == 'functionally_normal', 'Classification'] = 'F'\n",
    "    sge.loc[sge['Classification'] == 'functionally_abnormal', 'Classification'] = 'NF'\n",
    "    sge.loc[sge['Classification'] == 'indeterminate', 'Classification'] = 'I'\n",
    "\n",
    "    return sge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_ambry(ambry, gnomad):\n",
    "    ambry = pd.read_excel(ambry)\n",
    "    ambry = ambry[['Gene', 'c_variant', 'p_variant', 'Classification']]\n",
    "    ambry['amino_acid_change'] = ambry['p_variant'].str[2:]\n",
    "\n",
    "    ambry = ambry.dropna(subset = ['amino_acid_change']).copy()\n",
    "    ambry = ambry.reset_index(drop = True)\n",
    "\n",
    "    \n",
    "    gnomad = pd.read_excel(gnomad)\n",
    "    gnomad = gnomad[['Chromosome', 'Position', 'Reference', 'Alternate', 'Allele Frequency', 'Protein Consequence']]\n",
    "    gnomad['Position'] = gnomad['Position'].astype(str)\n",
    "    gnomad['pos_id'] = gnomad['Position'] + ':' + gnomad['Alternate']\n",
    "\n",
    "    return ambry, gnomad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_carriers_data(cc): #Reads all data\n",
    "    cc = pd.read_excel(cc) #Reads case-control data\n",
    "    cc = cc[cc['CAVA_GENE'].isin(['BARD1'])] #Filters only for BARD1\n",
    "    cc = cc[['Classification', '#CHROM', 'REF', 'ALT', 'CAVA_GENE', 'CAVA_CSN', 'CAVA_SO', 'Sample_AAF', 'Sample_ID', 'CaseControl','ER_status1', 'hg38_start']].copy() #Keeps necessary columns\n",
    "    cc = cc[cc['ER_status1'].isin([0, 777])] #ER negative cases only \n",
    "\n",
    "    return cc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_bridges(all, population):\n",
    "    bridges_all = pd.read_excel(all, sheet_name = 'BARD1')\n",
    "    bridges_pop = pd.read_excel(population, sheet_name = 'BARD1')\n",
    "\n",
    "    raw_dfs = [bridges_all, bridges_pop]\n",
    "    cleaned_dfs = []\n",
    "    \n",
    "    for df in raw_dfs:\n",
    "        df = df[['Cases', 'Controls', 'chr', 'ref', 'alt', 'hg38_pos']]\n",
    "        df = df.rename(columns = {'hg38_pos': 'pos'})\n",
    "\n",
    "        df['pos_id'] = df['pos'].astype(str) + ':' + df['alt']\n",
    "\n",
    "        df = df[['Cases', 'Controls', 'pos_id']]\n",
    "        cleaned_dfs.append(df)\n",
    "\n",
    "\n",
    "    return cleaned_dfs   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_vars(sge):\n",
    "\n",
    "    #under development - trying to figure out which parameters are best\n",
    "    sge = sge[sge['Consequence'].isin(['missense_variant'])]\n",
    "    sge = sge.copy()\n",
    "    sge['AApos'] = sge['amino_acid_change'].transform(lambda x: int(x[1:-1]))\n",
    "    \n",
    "    analysis_type = input('Analysis Region (RING/ARD/BRCT/Structured/All): ')\n",
    "    \n",
    "\n",
    "    if analysis_type == 'BRCT':\n",
    "        brct = list(range(565, 778))\n",
    "        sge = sge.loc[sge['AApos'].isin(brct)]\n",
    "    elif analysis_type == 'ARD': \n",
    "        ard = list(range(425, 546))\n",
    "        sge = sge.loc[sge['AApos'].isin(ard)]\n",
    "    elif analysis_type == 'RING':\n",
    "        ring = list(range(26, 123))\n",
    "        sge = sge.loc[sge['AApos'].isin(ring)]\n",
    "    elif analysis_type == 'Structured':\n",
    "        idr = list(range(123, 425)) + list(range(546, 565))\n",
    "        sge = sge.loc[~(sge['AApos'].isin(idr))]  \n",
    "    elif analysis_type == 'All': \n",
    "        pass\n",
    "        \n",
    "    sge_nf = sge[sge['Classification'].isin(['NF'])]\n",
    "    sge_func = sge[sge['Classification'].isin(['F'])]\n",
    "\n",
    "    return sge, sge_func, sge_nf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def process_carriers(cc): #Adds Position ID column to case-control data and splits into cases and controls\n",
    "    \n",
    "    cc['pos_id'] = None #Creates emtpy pos_id column\n",
    "    cc = cc[cc['ALT'].str.len() == 1].copy()\n",
    "    cc['hg38_start'] = cc['hg38_start'].astype(str) #Sets hg38 coordinates as str data type\n",
    "    cc['pos_id'] = cc['hg38_start'] + ':' + cc['ALT'] #Creates position ID\n",
    "\n",
    "    cc_control = cc[cc['CaseControl'].isin(['Control'])] #Creates control df\n",
    "    cc_case = cc[cc['CaseControl'].isin(['Case'])] #creates case df\n",
    "\n",
    "    return cc, cc_control, cc_case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_ambry(sge, gnomad, ambry):\n",
    "    \n",
    "    sge_gnomad = pd.merge(sge, gnomad, on = 'pos_id', how = 'inner')\n",
    "\n",
    "    gnomad_func = sge_gnomad['Classification'].value_counts()['F']\n",
    "    gnomad_nf = sge_gnomad['Classification'].value_counts()['NF']\n",
    "    \n",
    "    sge_ambry = pd.merge(sge,ambry, on = 'amino_acid_change', how = 'inner')\n",
    "\n",
    "    ambry_func = sge_ambry['Classification_x'].value_counts()['F']\n",
    "    ambry_nf = sge_ambry['Classification_x'].value_counts()['NF'] \n",
    "\n",
    "    contingency_table = np.array([[ambry_nf, ambry_func],\n",
    "                                    [gnomad_nf, gnomad_func]])\n",
    "\n",
    "    df = pd.DataFrame(contingency_table, columns = ['NF', 'F'], index = ['Ambry', 'gnomAD'])\n",
    "    display(df)\n",
    "\n",
    "\n",
    "    oddsratio, p_value = stats.fisher_exact(contingency_table) #Tabulates odds-ratio and p-value from Fischer's exact test\n",
    "    table = sm.stats.Table2x2(contingency_table) #Generates confidence intervals\n",
    "    \n",
    "    result = table.summary() #Summary table with Odds and Risk ratios\n",
    "    display(result) #Displays summary stats\n",
    "    \n",
    "    \n",
    "    #Prints out summary stats\n",
    "    print('Analysis for Ambry: ')\n",
    "    print(f\"Odds Ratio: {table.oddsratio:.2f}\")\n",
    "    print(f\"95% CI: ({table.oddsratio_confint()[0]:.2f}, {table.oddsratio_confint()[1]:.2f})\")\n",
    "    print(f\"P-value (Fisher's exact): {p_value:.4f}\")\n",
    "    oddsratio, p_value = stats.fisher_exact(contingency_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_carriers(cc, sge):\n",
    "    merged = pd.merge(cc, sge, on = 'pos_id', how = 'inner') #Merges case-control and SGE data\n",
    "    merged = merged.dropna(subset = ['Classification_y']) #drops any columsn without a classification\n",
    "    \n",
    "    contingency_tab = merged[['CaseControl', 'Classification_y']] #Creates dataframe for contingency table\n",
    "\n",
    "    contingency_tab = pd.crosstab(merged['CaseControl'], merged['Classification_y']) #Creates contingency table\n",
    "    contingency_tab = contingency_tab[contingency_tab.columns[::-1]] #Swaps NF and F columns in df\n",
    "    table_array = contingency_tab.values #Gets values from contingency table\n",
    "\n",
    "    print(table_array)\n",
    "    oddsratio, p_value = stats.fisher_exact(table_array) #Tabulates odds-ratio and p-value from Fischer's exact test\n",
    "    table = sm.stats.Table2x2(table_array) #Generates confidence intervals\n",
    "    \n",
    "    result = table.summary() #Summary table with Odds and Risk ratios\n",
    "    display(contingency_tab) #Displays contingency table \n",
    "    display(result) #Displays summary stats\n",
    "    \n",
    "\n",
    "    \n",
    "    #Prints out summary stats\n",
    "    print('Analysis for CARRIERS Data Set: ')\n",
    "    print(f\"Odds Ratio: {table.oddsratio:.2f}\")\n",
    "    print(f\"95% CI: ({table.oddsratio_confint()[0]:.2f}, {table.oddsratio_confint()[1]:.2f})\")\n",
    "    print(f\"P-value (Fisher's exact): {p_value:.4f}\")\n",
    "\n",
    "    df = pd.DataFrame({'data': ['CARRIERS'], 'OR': [table.oddsratio], 'LCB': [table.oddsratio_confint()[0]], 'UCB': [table.oddsratio_confint()[1]]\n",
    "                      })\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_bridges(sge, dfs):\n",
    "\n",
    "    i = 0\n",
    "    labels = ['BRIDGES All', 'BRIDGES Population']\n",
    "    for df in dfs:\n",
    "        merged = pd.merge(df, sge, on = 'pos_id', how = 'inner')\n",
    "\n",
    "        summary = merged.pivot_table(\n",
    "            values = ['Cases', 'Controls'],\n",
    "            index = 'Classification',\n",
    "            aggfunc = 'sum'\n",
    "        )\n",
    "\n",
    "        print(summary.transpose())\n",
    "        table_array = np.array([[summary['Cases']['NF'], summary['Cases']['F']],\n",
    "                                 [summary['Controls']['NF'], summary['Controls']['F']]])\n",
    "\n",
    "        \n",
    "        oddsratio, p_value = stats.fisher_exact(table_array) #Tabulates odds-ratio and p-value from Fischer's exact test\n",
    "        table = sm.stats.Table2x2(table_array) #Generates confidence intervals\n",
    "        \n",
    "        result = table.summary() #Summary table with Odds and Risk ratios\n",
    "        display(result) #Displays summary stats\n",
    "    \n",
    "\n",
    "    \n",
    "        #Prints out summary stats\n",
    "        print('Analysis for ', labels[i], ' Data Set: ')\n",
    "        print(f\"Odds Ratio: {table.oddsratio:.2f}\")\n",
    "        print(f\"95% CI: ({table.oddsratio_confint()[0]:.2f}, {table.oddsratio_confint()[1]:.2f})\")\n",
    "        print(f\"P-value (Fisher's exact): {p_value:.4f}\")        \n",
    "        \n",
    "        i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def cc_fig(df):\n",
    "\n",
    "    #Manual input for ambry data, will combine notebooks\n",
    "    ambry_df = pd.DataFrame({'data': ['Ambry & gnomAD'], 'OR': [1.378], 'LCB': [1.133], 'UCB': [1.677]\n",
    "                      })\n",
    "\n",
    "    df = pd.concat([df, ambry_df])\n",
    "\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7' # dusty blue\n",
    "     ]\n",
    "\n",
    "    datasets = [\n",
    "        'Ambry & gnomAD',\n",
    "        'CARRIERS'\n",
    "    ]\n",
    "    \n",
    "    base = alt.Chart(df)\n",
    "\n",
    "    points = base.mark_point(\n",
    "        filled = True,\n",
    "        size = 50\n",
    "    ).encode(\n",
    "        x = alt.X('data',\n",
    "                 axis = alt.Axis(title = '',\n",
    "                                 labels = False,\n",
    "                                 labelFontSize = 16\n",
    "                                )\n",
    "                 ),\n",
    "        y = alt.Y('OR',\n",
    "                 axis = alt.Axis(\n",
    "                     title = 'Odds Ratio',\n",
    "                     labelFontSize = 16,\n",
    "                     titleFontSize = 18\n",
    "                                 )\n",
    "                 ),\n",
    "        color = alt.Color('data',\n",
    "                          scale = alt.Scale(domain = datasets,\n",
    "                                            range = palette\n",
    "                                           ),\n",
    "                         legend = alt.Legend(title = '',\n",
    "                                             labelFontSize = 14, \n",
    "                                             titleFontSize = 26\n",
    "                                            )\n",
    "                         )\n",
    "    )\n",
    "\n",
    "    ci_bars = base.mark_errorbar().encode(\n",
    "        x = 'data',\n",
    "        y = alt.Y('LCB:Q', axis = alt.Axis(title = '')),\n",
    "        y2 = 'UCB:Q',\n",
    "        color = 'data'\n",
    "    )\n",
    "\n",
    "    plot = (points + ci_bars).properties(\n",
    "        height = 600, \n",
    "        width = 100\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    ).configure_axis(\n",
    "        grid = False\n",
    "    )\n",
    "\n",
    "    plot.display()\n",
    "\n",
    "    #plot.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_4d_ORplot.png', ppi = 500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    raw_sge_df = read_sge(sge)\n",
    "    ambry_df, gnomad_df = read_ambry(ambry, gnomad)\n",
    "    cc_df = read_carriers_data(carriers)\n",
    "    bridges_dfs = read_bridges(bridges_all, bridges_population)\n",
    "    \n",
    "    sge_all,sge_func, sge_nf = classify_vars(raw_sge_df)\n",
    "    \n",
    "    carriers_all, carriers_controls, carriers_cases = process_carriers(cc_df)\n",
    "    analyze_ambry(raw_sge_df, gnomad_df, ambry_df)\n",
    "    cc_df = analyze_carriers(carriers_all, sge_all)\n",
    "    analyze_bridges(sge_all, bridges_dfs)\n",
    "    #cc_fig(cc_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
