{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import altair as alt\n",
    "from scipy.stats import pearsonr\n",
    "import re\n",
    "\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.default_inference import DefaultInference\n",
    "from pydeseq2.ds import DeseqStats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_path = '../Data/ATG_lib_data/counts' #Path to counts for ATG library variants\n",
    "x1a_annotations = '../Data/ATG_lib_data/X1A_annotations.xlsx' #Path to file with annotations for X1A variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_counts(path): #Reads all counts\n",
    "    data_path = Path(counts_path) #Creates path object\n",
    "\n",
    "    all_files = list(data_path.glob('*tsv')) #Creates list of all count TSV files\n",
    "\n",
    "    counts_dict = {} #Empty dictionary to collect counts dataframes\n",
    "    \n",
    "    for path in all_files: #Iterates through all_files list and reads counts TSV files into dataframes and adds to counts_dict dictionary\n",
    "\n",
    "        if 'RNA' in str(path): #Reads RNA counts\n",
    "            df = pd.read_csv(path, sep = '\\t')\n",
    "\n",
    "            if 'RNA_R1R2R3' in str(path):\n",
    "                counts_dict['RNA_D05_rep1'] = df\n",
    "            elif 'RNA_R4R5R6' in str(path):\n",
    "                counts_dict['RNA_D05_rep2'] = df\n",
    "            elif 'RNA_R7R8R9' in str(path):\n",
    "                counts_dict['RNA_D05_rep3'] = df\n",
    "                \n",
    "        else: #Reads DNA counts\n",
    "            df = pd.read_csv(path, sep = '\\t')\n",
    "    \n",
    "            str_path = str(path)\n",
    "    \n",
    "            if 'lib_counts' in str_path:\n",
    "                counts_dict['lib'] = df\n",
    "            elif 'NC' in str_path:\n",
    "                counts_dict['NC'] = df\n",
    "            elif 'R1R2R3_D05' in str_path:\n",
    "                counts_dict['D05_rep1'] = df\n",
    "            elif 'R4R5R6_D05' in str_path:\n",
    "                counts_dict['D05_rep2'] = df\n",
    "            elif 'R7R8R9_D05' in str_path:\n",
    "                counts_dict['D05_rep3'] = df\n",
    "            elif 'R3_D13' in str_path:\n",
    "                counts_dict['D13_rep1'] = df\n",
    "            elif 'R6_D13' in str_path:\n",
    "                counts_dict['D13_rep2'] = df\n",
    "            elif 'R9_D13' in str_path:\n",
    "                counts_dict['D13_rep3'] = df\n",
    "\n",
    "    return counts_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_freq(dict): #Adds DNA/RNA frequency columns\n",
    "    \n",
    "    keys = list(dict.keys()) #Gets all keys from dictionary\n",
    "\n",
    "    freq_dicts = {} #New dictionary to hold dataframes with frequency columns\n",
    "\n",
    "    #Iterates through provided dict and adds frequency column\n",
    "    for key in keys:\n",
    "        if key != 'NC': #NC skipped\n",
    "            df = dict[key] #Gets dataframe from dictionary\n",
    "\n",
    "            total_count = df['Count'].sum() #Gets total number of variants\n",
    "            df['freq'] = df['Count'] / total_count #Creates frequency column\n",
    "\n",
    "            freq_dicts[key] = df #Adds to new dictionary\n",
    "\n",
    "    keys = list(freq_dicts.keys()) #Gets list of keys for future use in downstream functions\n",
    "    return freq_dicts,keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate_snvs(dna_sequence): #Mutates all possible SNVs of provided DNA sequence\n",
    "    snvs = []\n",
    "    i = 0\n",
    "    while i < len(dna_sequence):\n",
    "        if dna_sequence[i] == \"A\":\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        elif dna_sequence[i] == \"T\":\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        elif dna_sequence[i] == \"C\":\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        else:\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "        i += 1\n",
    "    return snvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement_string(seq_string): #Reverse complement and returns string\n",
    "    reverse_seq = seq_string[::-1]\n",
    "    reverse_comp_list = []\n",
    "    for char in reverse_seq:\n",
    "        if char == \"A\":\n",
    "            reverse_comp_list.append(\"T\")\n",
    "        elif char == \"G\":\n",
    "            reverse_comp_list.append(\"C\")\n",
    "        elif char == \"C\":\n",
    "            reverse_comp_list.append(\"G\")\n",
    "        else:\n",
    "            reverse_comp_list.append(\"A\")\n",
    "    reverse_compliment_str = \"\".join(reverse_comp_list)\n",
    "    return reverse_compliment_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_strings(str1, str2, coord_offset): #Special function to compare strings and return position with difference\n",
    "    \n",
    "    list_str1 = [] #List for string 1 characters\n",
    "    list_str2 = [] #List for string 2 charaacters\n",
    "\n",
    "    #Appends strings 1 and 2 to respective lists\n",
    "    for char in str1:\n",
    "        list_str1.append(char)\n",
    "\n",
    "    for char in str2:\n",
    "        list_str2.append(char)\n",
    "\n",
    "    #Iterates and compares each element in str1 and 2 lists. \n",
    "    i = 0 \n",
    "    while i < len(list_str1):\n",
    "        if list_str1[i] == list_str2[i]:\n",
    "            i += 1\n",
    "        else:\n",
    "            output_str = str(i + coord_offset) + ':' + list_str2[i] #If characters not the same, difference and position returned, coord_offset used for genomic coordinate\n",
    "            \n",
    "            return output_str\n",
    "            \n",
    "            i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_snv_lib(): #Annotates all SNVs in SNV lib with position IDs\n",
    "    seq = 'CCATGGA' #original WT 7bp library around Met26\n",
    "\n",
    "    coding_snvs = mutate_snvs(seq) #mutates all possible SNVs\n",
    "    rev_coding_snvs = {} #Dictionary to hold SNVs that have been reverse complemented (used for pos_id merging)\n",
    "\n",
    "    for snv in coding_snvs: #Iterates through all SNVs and reverse complements\n",
    "        rev_coding_snvs[snv] = reverse_complement_string(snv)\n",
    "\n",
    "    rev_seq = 'TCCATGG'\n",
    "\n",
    "    mapped_snvs = {}\n",
    "    for snv in coding_snvs: #iterates through SNVs and generates position IDs for merging\n",
    "        pos_id = compare_strings(rev_seq, rev_coding_snvs[snv], 214809490)\n",
    "        mapped_snvs[snv] = pos_id\n",
    "\n",
    "\n",
    "\n",
    "    return mapped_snvs\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_vars(dict,keys, snv_map, annotations): #Annotates variants with Consequences and AA substitution\n",
    "\n",
    "    annotation_df = pd.read_excel(annotations) #Reads annotations file\n",
    "    annotated_dfs = {} #Diciotnary to hold annotated data frames\n",
    "    \n",
    "    for key in keys: #Annotates all dataframes\n",
    "        df = dict[key] #Gets one dataframe\n",
    "        df['canonical_start'] = df['Sequence'].transform(lambda x: x[15:18]) #Gets canonical start codon\n",
    "        df['second_start'] = df['Sequence'].transform(lambda x: x[90:93]) #Gets second start codon\n",
    "        df['snv_lib'] = df['Sequence'].transform(lambda x: x[88:95]) #Adds SNV lib sequence\n",
    "\n",
    "        df['pos_id'] = df['snv_lib'].transform(lambda x: snv_map[x]) #Adds pos_id\n",
    "        df['start_pos_id'] = df['canonical_start'] + ':' + df['pos_id'] #Adds combined column with canonical start codon and pos_id\n",
    "        df = pd.merge(df, annotation_df, how = 'left', on = 'pos_id') #Merges\n",
    "\n",
    "        df = df[['Count', 'freq', 'canonical_start', 'second_start', 'pos_id', 'pos', 'allele', 'AAsub', 'Consequence','start_pos_id']] #Gets important columns\n",
    "        df = df.loc[:,['pos', 'allele', 'pos_id', 'Consequence', 'AAsub', 'canonical_start', 'second_start', 'Count', 'freq', 'start_pos_id']] #Reorders columns \n",
    "\n",
    "        new_count_name = key + '_count' #Name for replicate count column\n",
    "        new_freq_name = key + '_freq' #Name for replicate frequency column\n",
    "\n",
    "        df = df.rename(columns = {'Count': new_count_name, 'freq': new_freq_name}) #Columns for counts and frequency renamed\n",
    "        \n",
    "        annotated_dfs[key] = df #Annotated df added to dictionary\n",
    "\n",
    "    return annotated_dfs\n",
    "\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_dfs(dict, keys): #Merges all dataframes\n",
    "    \n",
    "    base_df = dict['lib'] #Base dataframe with column names and annotated variants\n",
    "    base_df = base_df[['pos', 'allele', 'pos_id', 'Consequence', 'AAsub', 'canonical_start', 'second_start', 'start_pos_id']] #Key columns pulled\n",
    "\n",
    "    for key in keys: #Merges all dataframes\n",
    "        to_merge = dict[key] #Gets one dataframe for merging\n",
    "\n",
    "        count_name = key + '_count' #Gets count column name\n",
    "        freq_name = key + '_freq' #Gets freq column name\n",
    "        to_merge = to_merge[['start_pos_id', count_name, freq_name]] #Gets columns for merging\n",
    "\n",
    "        base_df = pd.merge(base_df, to_merge, how = 'left', on = 'start_pos_id') #Merges\n",
    "\n",
    "    #RNA vs. DNA freq columsn used in RNA\n",
    "    base_df['rep1_RNAovDNA'] = base_df['RNA_D05_rep1_freq'] / base_df['D05_rep1_freq']\n",
    "    base_df['rep2_RNAovDNA'] = base_df['RNA_D05_rep2_freq'] / base_df['D05_rep2_freq']\n",
    "    base_df['rep3_RNAovDNA'] = base_df['RNA_D05_rep3_freq'] / base_df['D05_rep3_freq']\n",
    "    \n",
    "    return base_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def qc_scatter(pair_list, df): #Function for QC scatter plot\n",
    "\n",
    "    all_plots = []\n",
    "    for pair in pair_list:\n",
    "        rep1, rep2 = pair\n",
    "        r, p_value = pearsonr(df[rep1], df[rep2])\n",
    "        scatter = alt.Chart(df).mark_point().encode(\n",
    "            x = rep1,\n",
    "            y = rep2,\n",
    "            tooltip = [alt.Tooltip('canonical_start', title = 'Canonical Start: '),\n",
    "                       alt.Tooltip('second_start', title = 'Second Start: '),\n",
    "                       alt.Tooltip('AAsub', title = 'AA Substitution')\n",
    "                      ]\n",
    "        ).properties(\n",
    "            title = rep1 + ' vs. ' + rep2 + ' (r = ' + str(round(r,3)) + ')'\n",
    "        ).interactive()\n",
    "\n",
    "        trendline = scatter.transform_regression(rep1, rep2).mark_line(color = 'red')\n",
    "        scatter = trendline + scatter\n",
    "\n",
    "        all_plots.append(scatter)\n",
    "\n",
    "    final_plot = alt.hconcat(all_plots[0], all_plots[1], all_plots[2])\n",
    "    #final_plot.display()\n",
    "\n",
    "    return final_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def qc_plots(df): #Generates DNA replicate vs. replicate frequency correlation plots and RNA correlation plots\n",
    "    \n",
    "    d5_pairs = [('D05_rep1_freq', 'D05_rep2_freq'), ('D05_rep1_freq', 'D05_rep3_freq'), ('D05_rep2_freq', 'D05_rep3_freq')] #Pairs of D5 replicates to be tested\n",
    "    d13_pairs = [('D13_rep1_freq', 'D13_rep2_freq'), ('D13_rep1_freq', 'D13_rep3_freq'), ('D13_rep2_freq', 'D13_rep3_freq')] #Pairs of D13 replicats to be tested\n",
    "\n",
    "    d5_all = qc_scatter(d5_pairs, df) #Generates all D5 scatter plots\n",
    "    \n",
    "    d13_all = qc_scatter(d13_pairs, df) #Generates all D13 scatter plots\n",
    " \n",
    "    full_corr_plot = d5_all & d13_all #Concatenates all plots \n",
    "    full_corr_plot.display() #Dispalys plot\n",
    "    \n",
    "    #Library QC plot with double stacked bars\n",
    "    lib_plot = alt.Chart(df).mark_bar().encode(\n",
    "        x = 'pos:O',\n",
    "        y = 'lib_freq',\n",
    "        color = 'canonical_start'\n",
    "    ).properties(\n",
    "        title = 'Replicate vs. Replicate DNA Freq.',\n",
    "        width = 600,\n",
    "        height = 400\n",
    "    )\n",
    "\n",
    "    lib_plot.display()\n",
    "\n",
    "\n",
    "    #RNA QC plots\n",
    "    d5_rna_pairs = [('RNA_D05_rep1_freq', 'RNA_D05_rep2_freq'), ('RNA_D05_rep1_freq', 'RNA_D05_rep3_freq'), ('RNA_D05_rep2_freq', 'RNA_D05_rep3_freq')] #Rep vs. rep D5 RNA frequency\n",
    "    intra_rep_rna_dna_pairs = [('RNA_D05_rep1_freq', 'D05_rep1_freq'), ('RNA_D05_rep2_freq', 'D05_rep2_freq'), ('RNA_D05_rep3_freq', 'D05_rep3_freq')] #Rep vs. rep D5 RNA vs. DNA\n",
    "    rnadna_ratios = [('rep1_RNAovDNA', 'rep2_RNAovDNA'), ('rep1_RNAovDNA', 'rep3_RNAovDNA'), ('rep2_RNAovDNA', 'rep3_RNAovDNA')] #Rep vs. rep RNA/DNA frequency\n",
    "\n",
    "    #Builds all scatter plots\n",
    "    rna_vs_rna = qc_scatter(d5_rna_pairs, df)\n",
    "    intra_rep_rna_dna = qc_scatter(intra_rep_rna_dna_pairs, df)\n",
    "    rnadna_ratios_qc = qc_scatter(rnadna_ratios, df)\n",
    "\n",
    "    full_rna_plot = rna_vs_rna & intra_rep_rna_dna & rnadna_ratios_qc\n",
    "\n",
    "    full_rna_plot.display()\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def dna_score(df): #Adds DNA scores\n",
    "    original_df = df[['start_pos_id', 'lib_count', 'lib_freq',\n",
    "       'D05_rep2_count', 'D05_rep2_freq', 'RNA_D05_rep3_count',\n",
    "       'RNA_D05_rep3_freq', 'RNA_D05_rep1_count', 'RNA_D05_rep1_freq',\n",
    "       'D13_rep2_count', 'D13_rep2_freq', 'D05_rep3_count', 'D05_rep3_freq',\n",
    "       'RNA_D05_rep2_count', 'RNA_D05_rep2_freq', 'D13_rep1_count',\n",
    "       'D13_rep1_freq', 'D05_rep1_count', 'D05_rep1_freq', 'D13_rep3_count',\n",
    "       'D13_rep3_freq']]\n",
    "    \n",
    "    df = df.rename(columns = {'D05_rep1_count': 'D05_R1',\n",
    "                              'D05_rep2_count': 'D05_R2',\n",
    "                              'D05_rep3_count': 'D05_R3',\n",
    "                              'D13_rep1_count': 'D13_R1',\n",
    "                              'D13_rep2_count': 'D13_R2',\n",
    "                              'D13_rep3_count': 'D13_R3'\n",
    "                             }\n",
    "                  )\n",
    "\n",
    "\n",
    "    df = df.drop(columns = ['D05_rep1_freq', 'D05_rep2_freq', 'D05_rep3_freq',\n",
    "                            'D13_rep1_freq', 'D13_rep2_freq', 'D13_rep3_freq',\n",
    "                            'RNA_D05_rep1_count', 'RNA_D05_rep2_count', 'RNA_D05_rep3_count',\n",
    "                            'RNA_D05_rep1_freq', 'RNA_D05_rep2_freq', 'RNA_D05_rep3_freq'\n",
    "                           ]\n",
    "                )\n",
    "\n",
    "\n",
    "    value_cols = [col for col in df.columns if \"D13\"  in col or \"D05\"  in col or \"lib_count\" in col]\n",
    "    dfmelt = pd.melt(df, id_vars=['start_pos_id'], value_vars=value_cols)\n",
    "    dfpivot = pd.pivot(dfmelt, index='variable', columns=['start_pos_id'], values='value')\n",
    "    dfpivot = dfpivot.rename_axis(None, axis=1)\n",
    "    dfpivot = dfpivot.rename_axis(None, axis=0)\n",
    "    \n",
    "    metanames = ['D05_R1', 'D05_R2', 'D05_R3', 'D13_R1', 'D13_R2', 'D13_R3', 'lib_count']\n",
    "    metadays = [5, 5, 5, 13, 13, 13, 0]\n",
    "\n",
    "    metadf = pd.DataFrame(\n",
    "        {'sample_name': metanames,\n",
    "        'time': metadays,\n",
    "        }\n",
    "    )\n",
    "    metadf = metadf.set_index('sample_name').rename_axis(None, axis=0)\n",
    "\n",
    "    inference = DefaultInference(n_cpus=1)\n",
    "    dds = DeseqDataSet(\n",
    "        counts=dfpivot,\n",
    "        metadata=metadf,\n",
    "        design=\"~time\",\n",
    "        refit_cooks=True\n",
    "    )\n",
    "    \n",
    "    dds.deseq2()\n",
    "    contrast = [\"time\", 1, 0]\n",
    "    stat_res = DeseqStats(dds, contrast=contrast, inference=inference)\n",
    "    stat_res.summary()\n",
    "    resdf = stat_res.results_df\n",
    "\n",
    "    resdf = resdf.reset_index(names='start_pos_id')#.drop(columns=['index'])\n",
    "    resdf = resdf.merge(df[[\n",
    "        \"pos\", \"allele\", \"start_pos_id\", 'AAsub', 'Consequence', 'canonical_start', 'second_start'\n",
    "    ]])\n",
    "    resdf[\"target\"] = 'BARD1_X1A_ATG'\n",
    "\n",
    "    resdf = resdf[['start_pos_id', 'log2FoldChange', 'lfcSE', 'pos', 'allele', 'AAsub', 'Consequence', 'canonical_start', 'second_start', 'target']]\n",
    "    resdf = resdf.rename(columns = {'log2FoldChange': 'score', 'lfcSE': 'standard_error'})\n",
    "\n",
    "    resdf = resdf.loc[:, ['target', 'pos', 'allele', 'start_pos_id', 'Consequence', 'AAsub', 'canonical_start', 'second_start', 'score', 'standard_error']]\n",
    "\n",
    "    def starts(df):\n",
    "        df['comb_starts'] = df['canonical_start'] + df['second_start']\n",
    "\n",
    "        df['start_consequence'] = None\n",
    "\n",
    "        df.loc[df['comb_starts'].str.match(r'^.{3}ATG$'), 'start_consequence'] = 'canonical_lost'\n",
    "        df.loc[df['comb_starts'].str.match(r'^ATG.{3}$'), 'start_consequence'] = 'second_lost'\n",
    "        df.loc[df['comb_starts'] == 'ATGATG', 'start_consequence'] = 'WT'\n",
    "        df.loc[df['start_consequence'].isna(), 'start_consequence'] = 'both_lost'\n",
    "\n",
    "        #df = df.drop(columns = ['comb_starts'])\n",
    "        \n",
    "        return df\n",
    "\n",
    "    resdf = starts(resdf)\n",
    "\n",
    "    resdf = pd.merge(resdf, original_df, on = 'start_pos_id', how = 'left')\n",
    "    #resdf.to_excel('../Data/ATG_lib_data/20250409_BARD1_X1A_ATG_scored.xlsx', index = False)\n",
    "\n",
    "    return resdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def visualize_scores(df): #Visualizes DNA scores\n",
    "\n",
    "    ordered_var_list = ['synonymous_variant', 'missense_variant', 'stop_gained']\n",
    "    facet_sort = ['WT', 'canonical_lost', 'second_lost', 'both_lost']\n",
    "    histogram = alt.Chart(df).mark_bar().encode(\n",
    "        x = alt.X('score', title = 'SGE Score', bin = alt.Bin(maxbins = 40)),\n",
    "        y = alt.Y('count()', title = 'Number of Variants'),\n",
    "        color = alt.Color('Consequence', sort = ordered_var_list)\n",
    "    ).properties(\n",
    "        width = 600, \n",
    "        height = 400\n",
    "    ).interactive()\n",
    "\n",
    "    facetgram = histogram.facet(columns = 2, facet = alt.Facet('start_consequence', title = 'Start Codon Impact', sort = facet_sort))\n",
    "\n",
    "    facetgram.display()\n",
    "\n",
    "    scattergram = alt.Chart(df).mark_point().encode(\n",
    "        x = alt.X('pos:O', title = 'Genomic Coordinate', axis = alt.Axis(labels = False)),\n",
    "        y = alt.Y('score', title = 'SGE Score'),\n",
    "        color = alt.Color('Consequence',sort = ordered_var_list),\n",
    "        tooltip = [alt.Tooltip('AAsub', title = 'AA Substitution: ')]\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400\n",
    "    ).facet(\n",
    "        columns = 2,\n",
    "        facet = alt.Facet('start_consequence',\n",
    "                          title = 'Start Codon Impact',\n",
    "                          sort = facet_sort\n",
    "                         )\n",
    "    )\n",
    "\n",
    "    scattergram.display()\n",
    "\n",
    "    std_error = alt.Chart(df).mark_point().encode(\n",
    "        x = 'pos:O',\n",
    "        y = 'standard_error',\n",
    "        color = 'start_consequence'\n",
    "    )\n",
    "\n",
    "    std_error.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    counts_dict = read_counts(counts_path)\n",
    "    freq_dicts, keys = add_freq(counts_dict)\n",
    "    mapped_snvs = annotate_snv_lib()\n",
    "    annotated = annotate_vars(freq_dicts, keys, mapped_snvs, x1a_annotations)\n",
    "    print(mapped_snvs)\n",
    "    combined_df = combine_dfs(annotated, keys)\n",
    "    qc_plots(combined_df)\n",
    "    scored_df = dna_score(combined_df)\n",
    "    visualize_scores(scored_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
