{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.default_inference import DefaultInference\n",
    "from pydeseq2.ds import DeseqStats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "bard1_data = '../Data/20250508_BARD1scores_update_FILTERED.xlsx'\n",
    "depth_file = '/Users/ivan/Desktop/test_excel_outputs/20250717_BARD1_CollapsedMedianDepth.xlsx'\n",
    "\n",
    "cutoffs = [-3.438968 * 0.028675 + 0.009242,-3.018904 * 0.028675 + 0.009242]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_data(file):\n",
    "    df = pd.read_excel(file)\n",
    "    df = df.loc[(df['exon'].isin(['BARD1_X4'])) & (df['simplified_consequence'].isin(['missense_variant', 'synonymous_variant']))]\n",
    "    df['overlapping'] = 'N'\n",
    "    df.loc[df['snvlib_lib2'].notna(), 'overlapping'] = 'Y'\n",
    "\n",
    "    df['consequence_overlap'] = df['functional_consequence'] + '+' + df['overlapping']\n",
    "\n",
    "    not_normal = df.loc[(df['functional_consequence'].isin(['functionally_abnormal', 'indeterminate'])) & (df['snvlib_lib2'].notna())]\n",
    "    \n",
    "    return df, not_normal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def std_error(df, depth):\n",
    "    df['min_tc_count'] = df[['D05_R1_lib1', 'D05_R1_lib2', 'D05_R2_lib1', 'D05_R2_lib2', 'D05_R3_lib1', 'D05_R3_lib2',\n",
    "                   'D13_R1_lib1','D13_R1_lib2', 'D13_R2_lib1', 'D13_R2_lib2', 'D13_R3_lib1', 'D13_R3_lib2']].min(axis = 1)\n",
    "    \n",
    "    chart = alt.Chart(df).mark_point().encode(\n",
    "        x = alt.X('pos',\n",
    "                  axis = alt.Axis(labels = False),\n",
    "                  scale = alt.Scale(domain = [214780602,214781509],\n",
    "                                   reverse = True)\n",
    "                 ),\n",
    "        y = 'standard_error',\n",
    "        color = 'functional_consequence',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('amino_acid_change', title = 'Amino Acid Sub.: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Functional Class: ')\n",
    "                  ]\n",
    "                           \n",
    "    ).properties(\n",
    "        width = 2000,\n",
    "        height = 400,\n",
    "        title = 'Std. Error vs. Position'\n",
    "    )\n",
    "\n",
    "    count_chart = alt.Chart(df).mark_point().encode(\n",
    "        x = alt.X('pos',\n",
    "                  axis = alt.Axis(labels = False),\n",
    "                  scale = alt.Scale(domain = [214780602,214781509],\n",
    "                                   reverse = True)\n",
    "                 ),\n",
    "        y = alt.Y('min_tc_count',\n",
    "                  scale = alt.Scale(domain = [0, 1000])\n",
    "                 ),\n",
    "        color = 'functional_consequence',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('amino_acid_change', title = 'Amino Acid Sub.: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Functional Class: '),\n",
    "                   alt.Tooltip('min_tc_count', title = 'Min. Count in TC Library: ')\n",
    "                  ]\n",
    "                           \n",
    "    ).properties(\n",
    "        width = 2000,\n",
    "        height = 400,\n",
    "        title = 'Min. Count vs. Position'\n",
    "    )\n",
    "\n",
    "    score_chart = alt.Chart(df).mark_point().encode(\n",
    "        x = alt.X('pos',\n",
    "                  axis = alt.Axis(labels = False),\n",
    "                  scale = alt.Scale(domain = [214780602,214781509],\n",
    "                                   reverse = True)\n",
    "                 ),\n",
    "        y = 'score',\n",
    "        color = 'functional_consequence',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('amino_acid_change', title = 'Amino Acid Sub.: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Functional Class: ')\n",
    "                  ]\n",
    "                           \n",
    "    ).properties(\n",
    "        width = 2000,\n",
    "        height = 400,\n",
    "        title = 'SGE Score vs. Position'\n",
    "    )\n",
    "\n",
    "\n",
    "    depth_df = pd.read_excel(depth)\n",
    "    depth_df = depth_df.loc[depth_df['pos'].isin(list(range(214780602,214781509 + 1)))]\n",
    "    \n",
    "    depth_chart = alt.Chart(depth_df).mark_line().encode(\n",
    "        x = alt.X('pos',\n",
    "                  axis = alt.Axis(labels = False),\n",
    "                  scale = alt.Scale(domain = [214780602,214781509],\n",
    "                                   reverse = True)\n",
    "                 ),\n",
    "        y = 'median_depth',\n",
    "        color = 'target',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('amino_acid', title = 'Amino Acid: ')\n",
    "                  ]\n",
    "                           \n",
    "    ).properties(\n",
    "        width = 2000,\n",
    "        height = 400,\n",
    "        title = 'Median Normalized Depth vs. Position'        \n",
    "    )\n",
    "\n",
    "\n",
    "    final_plot = chart & depth_chart\n",
    "\n",
    "    final_plot = final_plot & score_chart & count_chart\n",
    "\n",
    "    final_plot = final_plot.interactive() \n",
    "    final_plot.display() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def counts_for_abnormal(ab_df):\n",
    "    df = ab_df[['snvlib_lib1', 'snvlib_lib2',\n",
    "                   'D05_R1_lib1', 'D05_R1_lib2', 'D05_R2_lib1', 'D05_R2_lib2', 'D05_R3_lib1', 'D05_R3_lib2',\n",
    "                   'D13_R1_lib1','D13_R1_lib2', 'D13_R2_lib1', 'D13_R2_lib2', 'D13_R3_lib1', 'D13_R3_lib2', 'amino_acid_change'\n",
    "                  ]]\n",
    "    df_long = df.melt(\n",
    "        id_vars=['amino_acid_change'], \n",
    "        value_vars=[\n",
    "                   'D05_R1_lib1', 'D05_R1_lib2', 'D05_R2_lib1', 'D05_R2_lib2', 'D05_R3_lib1', 'D05_R3_lib2',\n",
    "                   'D13_R1_lib1','D13_R1_lib2', 'D13_R2_lib1', 'D13_R2_lib2', 'D13_R3_lib1', 'D13_R3_lib2'],\n",
    "        var_name='library', \n",
    "        value_name='count'\n",
    "    )\n",
    "\n",
    "    chart = alt.Chart(df_long).mark_bar().encode(\n",
    "        x=alt.X('library:N', axis=alt.Axis(labelAngle=-45)),\n",
    "        y='count:Q',\n",
    "        color='library:N',\n",
    "        tooltip = [alt.Tooltip('library', title = 'Library: '),\n",
    "                   alt.Tooltip('count', title = 'Count: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        height = 150,\n",
    "        width = 200\n",
    "    ).facet(\n",
    "        facet='amino_acid_change:N',\n",
    "        columns=14  # Number of columns in the grid\n",
    "    )\n",
    "\n",
    "    chart.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlap_nonoverlap_barchart(df):\n",
    "\n",
    "    summary_df = df.groupby(['functional_consequence', 'overlapping']).size().reset_index(name='count')\n",
    "    summary_df['percentage_normal/ab'] = summary_df.groupby('functional_consequence')['count'].transform(lambda x: (x / x.sum()) * 100)\n",
    "    summary_df['percentage_overlapping'] = summary_df.groupby('overlapping')['count'].transform(lambda x: (x / x.sum()) * 100)\n",
    "    \n",
    "\n",
    "    totals = summary_df.groupby('overlapping')['count'].sum().reset_index()\n",
    "    totals['functional_consequence'] = 'All Variants'\n",
    "    totals['percentage_normal/ab'] = 100  # or 100.0 if you prefer\n",
    "    totals['percentage_overlapping'] = (totals['count'] / totals['count'].sum()) * 100\n",
    "\n",
    "    summary_df = pd.concat([summary_df, totals], ignore_index=True)\n",
    "    print(summary_df)\n",
    "\n",
    "    chart = alt.Chart(summary_df).mark_bar().encode(\n",
    "        x = alt.X('functional_consequence',\n",
    "                  sort = ['All Variants', 'functionally_normal', 'functionally_abnormal'],\n",
    "                  axis = alt.Axis(labelAngle = 0\n",
    "                                 )\n",
    "                 ),\n",
    "        y = alt.Y('count'\n",
    "                 ),\n",
    "        color = alt.Color('overlapping',\n",
    "                          sort = ['Y', 'N'],\n",
    "                          legend = alt.Legend(title = 'In Overlap?')\n",
    "                         )\n",
    "    ).properties(\n",
    "        width = 300,\n",
    "        height = 300,\n",
    "        title = 'X4 Missense Variants'\n",
    "    )\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def overlapping_scatter(df):\n",
    "\n",
    "    df = df.loc[(df['snvlib_lib2'].notna()) & (df['functional_consequence'].isin(['functionally_abnormal']))]\n",
    "   \n",
    "\n",
    "    d5_reps_pairs_lib1 = [('D13_R1_lib1', 'D13_R2_lib1'),\n",
    "                     ('D13_R1_lib1', 'D13_R3_lib1'),\n",
    "                     ('D13_R2_lib1', 'D13_R3_lib1')\n",
    "                    ]\n",
    "\n",
    "    d5_reps_pairs_lib2 = [('D13_R1_lib2', 'D13_R2_lib2'),\n",
    "                     ('D13_R1_lib2', 'D13_R3_lib2'),\n",
    "                     ('D13_R2_lib2', 'D13_R3_lib2')\n",
    "                    ]\n",
    "\n",
    "    print('Library 1: ')\n",
    "    for elem in d5_reps_pairs_lib1:\n",
    "        rep1, rep2 = elem\n",
    "        d5_chart_lib1 = alt.Chart(df).mark_point().encode(\n",
    "            x = rep1,\n",
    "            y = rep2\n",
    "        ).interactive()\n",
    "\n",
    "       \n",
    "\n",
    "        d5_chart_lib1 = d5_chart_lib1.facet('target')\n",
    "        '''\n",
    "        d5_chart_lib1 = d5_chart_lib1.resolve_scale(\n",
    "            x = 'independent',\n",
    "            y = 'independent'\n",
    "        )\n",
    "        '''\n",
    "        d5_chart_lib1.display()\n",
    "\n",
    "    \n",
    "        \n",
    "    print('Library 2: ')\n",
    "    for elem in d5_reps_pairs_lib2:\n",
    "        rep1, rep2 = elem\n",
    "        d5_chart_lib1 = alt.Chart(df).mark_point().encode(\n",
    "            x = rep1,\n",
    "            y = rep2\n",
    "        ).interactive()\n",
    "\n",
    "       \n",
    "\n",
    "        d5_chart_lib1 = d5_chart_lib1.facet('target')\n",
    "\n",
    "        '''\n",
    "        d5_chart_lib1 = d5_chart_lib1.resolve_scale(\n",
    "            x = 'independent',\n",
    "            y = 'independent'\n",
    "        )\n",
    "        '''\n",
    "        d5_chart_lib1.display()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rescore_pilot(df, cutoffs):\n",
    "\n",
    "    #df = df.loc[df['target'].isin(['BARD1_X4A'])]\n",
    "    '''\n",
    "\n",
    "    base = alt.Chart(df)\n",
    "    original_score_chart = base.mark_point().encode(\n",
    "        x = alt.X('pos',\n",
    "                  axis = alt.Axis(labels = False),\n",
    "                  scale = alt.Scale(domain = [214780602,214781509],\n",
    "                                   reverse = True)\n",
    "                 ), \n",
    "        y = 'score',\n",
    "        color = alt.Color('simplified_consequence',\n",
    "                          sort = ['synonymous_variant', 'missense_variant']\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('amino_acid_change', title = 'Amino Acid Sub.: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'Functional Class: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 2000,\n",
    "        height = 400,\n",
    "        title = 'Original Scores'\n",
    "    )\n",
    "    nf_line = base.mark_rule(color='red').encode(\n",
    "        y=alt.datum(cutoffs[0])\n",
    "    )\n",
    "    \n",
    "    func_line = base.mark_rule(color='blue').encode(\n",
    "        y=alt.datum(cutoffs[1])\n",
    "    )\n",
    "    \n",
    "    original_score_chart = original_score_chart + nf_line + func_line\n",
    "    \n",
    "\n",
    "    original_score_chart.display()\n",
    "    '''\n",
    "\n",
    "    lib1_df = df[['pos_id', 'snvlib_lib1', 'D05_R1_lib1', 'D05_R2_lib1', 'D05_R3_lib1', 'D13_R1_lib1', 'D13_R2_lib1' ,'D13_R3_lib1']]\n",
    "    lib1_df = lib1_df.rename(columns = {'D05_R1_lib1' : 'D05_R1',\n",
    "                                        'D05_R2_lib1' : 'D05_R2',\n",
    "                                        'D05_R3_lib1' : 'D05_R3', \n",
    "                                        'D13_R1_lib1' : 'D13_R1',\n",
    "                                        'D13_R2_lib1' : 'D13_R2',\n",
    "                                        'D13_R3_lib1' : 'D13_R3',\n",
    "                                        'snvlib_lib1' : 'lib_count'\n",
    "                                       }\n",
    "                            )\n",
    "\n",
    "    value_cols = [col for col in lib1_df.columns if \"D13\"  in col or \"D05\"  in col or \"lib_count\" in col]\n",
    "    dfmelt = pd.melt(lib1_df, id_vars=['pos_id'], value_vars=value_cols)\n",
    "    dfpivot = pd.pivot(dfmelt, index='variable', columns=['pos_id'], values='value')\n",
    "    dfpivot = dfpivot.rename_axis(None, axis=1)\n",
    "    dfpivot = dfpivot.rename_axis(None, axis=0)\n",
    "\n",
    "\n",
    "    metanames = ['D05_R1', 'D05_R2', 'D05_R3', 'D13_R1', 'D13_R2', 'D13_R3', 'lib_count']\n",
    "    metadays = [5, 5, 5, 13, 13, 13, 0]\n",
    "\n",
    "    metadf = pd.DataFrame(\n",
    "        {'sample_name': metanames,\n",
    "        'time': metadays,\n",
    "        }\n",
    "    )\n",
    "    metadf = metadf.set_index('sample_name').rename_axis(None, axis=0)\n",
    "\n",
    "    inference = DefaultInference(n_cpus=1)\n",
    "    dds = DeseqDataSet(\n",
    "        counts=dfpivot,\n",
    "        metadata=metadf,\n",
    "        design=\"~time\",\n",
    "        refit_cooks=True\n",
    "    )\n",
    "    \n",
    "    dds.deseq2()\n",
    "    contrast = [\"time\", 1, 0]\n",
    "    stat_res = DeseqStats(dds, contrast=contrast, inference=inference)\n",
    "    stat_res.summary()\n",
    "    resdf = stat_res.results_df\n",
    "\n",
    "    resdf = resdf.reset_index(names = 'pos_id')\n",
    "\n",
    "    resdf = resdf.merge(df[[\n",
    "        \"pos\", \"allele\", \"pos_id\", 'amino_acid_change', 'simplified_consequence', \n",
    "    ]])\n",
    "    resdf[\"target\"] = 'BARD1_X4A'\n",
    "    \n",
    "    resdf = resdf[['pos_id', 'log2FoldChange', 'lfcSE', 'pos', 'allele', 'amino_acid_change', 'simplified_consequence', 'target']]\n",
    "    lib1_scored_df = resdf.rename(columns = {'log2FoldChange': 'score', 'lfcSE': 'standard_error'})\n",
    "\n",
    "\n",
    "    lib2_df = df[['pos_id', 'snvlib_lib2', 'D05_R1_lib2', 'D05_R2_lib2', 'D05_R3_lib2', 'D13_R1_lib2', 'D13_R2_lib2' ,'D13_R3_lib2']]\n",
    "    lib2_df = lib2_df.rename(columns = {'D05_R1_lib2' : 'D05_R1',\n",
    "                                        'D05_R2_lib2' : 'D05_R2',\n",
    "                                        'D05_R3_lib2' : 'D05_R3', \n",
    "                                        'D13_R1_lib2' : 'D13_R1',\n",
    "                                        'D13_R2_lib2' : 'D13_R2',\n",
    "                                        'D13_R3_lib2' : 'D13_R3',\n",
    "                                        'snvlib_lib2' : 'lib_count'\n",
    "                                       }\n",
    "                            )\n",
    "\n",
    "    lib2_df.to_excel('/Users/ivan/Desktop/test_excel_outputs/20250718_X4Overlap_lib2Counts.xlsx', index = None)\n",
    "    \n",
    "    value_cols = [col for col in lib2_df.columns if \"D13\"  in col or \"D05\"  in col or \"lib_count\" in col]\n",
    "    dfmelt = pd.melt(lib2_df, id_vars=['pos_id'], value_vars=value_cols)\n",
    "    dfpivot = pd.pivot(dfmelt, index='variable', columns=['pos_id'], values='value')\n",
    "    dfpivot = dfpivot.rename_axis(None, axis=1)\n",
    "    dfpivot = dfpivot.rename_axis(None, axis=0)\n",
    "\n",
    "\n",
    "    metanames = ['D05_R1', 'D05_R2', 'D05_R3', 'D13_R1', 'D13_R2', 'D13_R3', 'lib_count']\n",
    "    metadays = [5, 5, 5, 13, 13, 13, 0]\n",
    "\n",
    "    metadf = pd.DataFrame(\n",
    "        {'sample_name': metanames,\n",
    "        'time': metadays,\n",
    "        }\n",
    "    )\n",
    "    metadf = metadf.set_index('sample_name').rename_axis(None, axis=0)\n",
    "\n",
    "    inference = DefaultInference(n_cpus=1)\n",
    "    dds = DeseqDataSet(\n",
    "        counts=dfpivot,\n",
    "        metadata=metadf,\n",
    "        design=\"~time\",\n",
    "        refit_cooks=True\n",
    "    )\n",
    "    \n",
    "    dds.deseq2()\n",
    "    contrast = [\"time\", 1, 0]\n",
    "    stat_res = DeseqStats(dds, contrast=contrast, inference=inference)\n",
    "    stat_res.summary()\n",
    "    resdf = stat_res.results_df\n",
    "\n",
    "    resdf = resdf.reset_index(names = 'pos_id')\n",
    "\n",
    "    resdf = resdf.merge(df[[\n",
    "        \"pos\", \"allele\", \"pos_id\", 'amino_acid_change', 'simplified_consequence', \n",
    "    ]])\n",
    "    resdf[\"target\"] = 'BARD1_X4A'\n",
    "    \n",
    "    resdf = resdf[['pos_id', 'log2FoldChange', 'lfcSE', 'pos', 'allele', 'amino_acid_change', 'simplified_consequence', 'target']]\n",
    "    lib2_scored_df = resdf.rename(columns = {'log2FoldChange': 'score', 'lfcSE': 'standard_error'})\n",
    "\n",
    "    rescored_dfs = [('Lib1 & Lib2 Counts', df), ('Lib1 Counts', lib1_scored_df), ('Lib2 Counts', lib2_scored_df)]\n",
    "\n",
    "    score_plots = []\n",
    "    se_plots = []\n",
    "    for elem in rescored_dfs: \n",
    "        title, df = elem\n",
    "        base = alt.Chart(df)\n",
    "        original_score_chart = base.mark_point().encode(\n",
    "            x = alt.X('pos',\n",
    "                      axis = alt.Axis(labels = False),\n",
    "                      scale = alt.Scale(domain = [214780602,214781509],\n",
    "                                       reverse = True)\n",
    "                     ), \n",
    "            y = alt.Y('score',\n",
    "                      scale = alt.Scale(\n",
    "                          domain = [-0.25, 0.1]\n",
    "                      \n",
    "                      )\n",
    "                     ),\n",
    "            color = alt.Color('simplified_consequence',\n",
    "                              sort = ['synonymous_variant', 'missense_variant']\n",
    "                             ),\n",
    "            tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                       alt.Tooltip('amino_acid_change', title = 'Amino Acid Sub.: '),\n",
    "                       alt.Tooltip('score', title = 'SGE Score: ')\n",
    "                      ]\n",
    "        ).properties(\n",
    "            width = 2000,\n",
    "            height = 400,\n",
    "            title = 'With ' + title + ' SGE Scores'\n",
    "        )\n",
    "        nf_line = base.mark_rule(color='red').encode(\n",
    "            y=alt.datum(cutoffs[0])\n",
    "        )\n",
    "        \n",
    "        func_line = base.mark_rule(color='blue').encode(\n",
    "            y=alt.datum(cutoffs[1])\n",
    "        )\n",
    "\n",
    "        original_score_chart = original_score_chart + nf_line + func_line\n",
    "        score_plots.append(original_score_chart)\n",
    "\n",
    "        #SE Chart\n",
    "        se_chart = base.mark_point().encode(\n",
    "            x = alt.X('pos',\n",
    "                      axis = alt.Axis(labels = False),\n",
    "                      scale = alt.Scale(domain = [214780602,214781509],\n",
    "                                       reverse = True)\n",
    "                     ), \n",
    "            y = alt.Y('standard_error',\n",
    "                      scale = alt.Scale(\n",
    "                          domain = [0, 0.1]\n",
    "                      \n",
    "                      )\n",
    "                     ),\n",
    "            color = alt.Color('simplified_consequence',\n",
    "                              sort = ['synonymous_variant', 'missense_variant']\n",
    "                             ),\n",
    "            tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                       alt.Tooltip('amino_acid_change', title = 'Amino Acid Sub.: '),\n",
    "                       alt.Tooltip('score', title = 'SGE Score: ')\n",
    "                      ]\n",
    "        ).properties(\n",
    "            width = 2000,\n",
    "            height = 400,\n",
    "            title = 'With ' + title + ' Stanard Error'\n",
    "        )\n",
    "\n",
    "        se_plots.append(se_chart)\n",
    "        \n",
    "    score_plot = score_plots[0] & score_plots[1] & score_plots[2]\n",
    "    se_plot = se_plots[0] & se_plots[1] & se_plots[2]\n",
    "    score_plot.display()\n",
    "    se_plot.display()\n",
    "        \n",
    "    return lib2_scored_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_rescored(data_path, rescored, cutoffs):\n",
    "\n",
    "    rescored['exon'] = 'BARD1_X4'\n",
    "    pos_id_to_remove = rescored['pos_id'].tolist()\n",
    "    original_df = pd.read_excel(data_path)\n",
    "\n",
    "    original_df = original_df[['exon', 'pos_id', 'score', 'standard_error', 'pos', 'allele', 'amino_acid_change', 'simplified_consequence', 'target']]\n",
    "    \n",
    "    df = original_df.loc[~(original_df['pos_id'].isin(pos_id_to_remove))]\n",
    "\n",
    "    df = pd.concat([df, rescored])\n",
    "    df['functional_consequence'] = 'indeterminate'\n",
    "    df.loc[df['score'] <= cutoffs[0], 'functional_consequence'] = 'functionally_abnormal'\n",
    "    df.loc[df['score'] >= cutoffs[1], 'functional_consequence'] = 'functionally_normal'\n",
    "\n",
    "    df = df.loc[:, ['exon', 'target', 'pos', 'allele', 'pos_id', 'score', 'standard_error', 'simplified_consequence', 'amino_acid_change', 'functional_consequence']]\n",
    "    \n",
    "    #df.to_excel('/Users/ivan/Desktop/test_excel_outputs/20250718_BARD1scores_X4Rescored_filtered.xlsx', index = False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    df,abnormal_df = process_data(bard1_data)\n",
    "    std_error(df,depth_file)\n",
    "    counts_for_abnormal(abnormal_df)\n",
    "    overlap_nonoverlap_barchart(df)\n",
    "    #overlapping_scatter(df)\n",
    "    rescored_df = rescore_pilot(abnormal_df, cutoffs)\n",
    "    merge_rescored(bard1_data, rescored_df, cutoffs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "11",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
