{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "gnom_path = '../Data/20240905_BARD1_gnomADv4.1.0_SNVs.xlsx'\n",
    "scores = '../Data/20250813_BARD1scores_final_FILTERED.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gnomAD(gnomAD_path): #Reads gnomAD file\n",
    "    \n",
    "    unfiltered = pd.read_excel(gnomAD_path) #Reads gnomAD file\n",
    "    filtered = unfiltered[['gnomAD ID', 'Allele Frequency']] #Gets necessary columns \n",
    "    \n",
    "    return filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rewrite(df): #Adds a pos_id column for merging\n",
    "\n",
    "    df['pos_id'] = df['gnomAD ID'].transform(lambda x: x[2:11] + ':' + x[14])\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_scores(df, scores): #Merges gnomAD data frame and SGE scores\n",
    "    \n",
    "    scores = pd.read_excel(scores) #Reads SGE scores\n",
    "    scores = scores.rename(columns = {'consequence': 'Consequence', 'score': 'snv_score_minmax'}) #Renames columns to harmonize column names\n",
    "    scores['pos'] = scores['pos'].astype(str) #Sets data type of position column to string\n",
    "    scores['pos_id'] = scores['pos'] + ':' + scores['alt'] #adds pos_id column to SGE data\n",
    "    sge_scores = scores[['exon', 'pos_id','Consequence','snv_score_minmax', 'amino_acid_change']] #pulls necessary columns\n",
    "    \n",
    "    df = pd.merge(sge_scores, df, on = 'pos_id', how = 'inner') #Merges both data frames\n",
    "\n",
    "    df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "    df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "    df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "    df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "    df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "    df.loc[df['Consequence'].str.contains('site'), 'Consequence'] = 'Canonical Splice'\n",
    "    df.loc[df['Consequence'].str.contains('ing_var'), 'Consequence'] = 'Splice Region'\n",
    "    df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "    df.loc[df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def graph(df): #Generates scatter plot of gnomAD allele frequency vs. SGE score\n",
    "    \n",
    "\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a', # yellow\n",
    "    '#6AA84F', # med green\n",
    "    '#93C47D', # light green\n",
    "    '#888888', # med gray\n",
    "    '#000000', # black\n",
    "    '#1170AA', # darker blue\n",
    "    '#CFCFCF' # light gray\n",
    "        \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained',\n",
    "        'Intron', \n",
    "        'UTR Variant',\n",
    "        'Stop Lost',\n",
    "        'Start Lost',\n",
    "        'Canonical Splice', \n",
    "        'Splice Region',\n",
    "    ]\n",
    "\n",
    "    #Builds scatter plot\n",
    "    graph = alt.Chart(df).mark_point().encode(\n",
    "        x=alt.X('snv_score_minmax', \n",
    "                axis=alt.Axis(\n",
    "                    title='SGE Score', \n",
    "                    labelFontSize = 18, \n",
    "                    titleFontSize = 20\n",
    "                )\n",
    "               ),\n",
    "        y = alt.Y('Allele Frequency', \n",
    "                  axis = alt.Axis(\n",
    "                      grid = False, \n",
    "                      labelFontSize = 18, \n",
    "                      titleFontSize = 20,\n",
    "                      format = 'e'\n",
    "                  ), \n",
    "                  scale = alt.Scale(type = 'log'\n",
    "                                   )\n",
    "                 ),\n",
    "        color = alt.Color('Consequence', \n",
    "                          scale = alt.Scale(range = palette,\n",
    "                                  domain = variant_types), \n",
    "                          legend = alt.Legend(\n",
    "                              titleFontSize = 16, \n",
    "                              labelFontSize = 14,\n",
    "                              labelLimit = 500\n",
    "                          )\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('exon', title = 'Exon: '),\n",
    "                   alt.Tooltip('pos_id', title = 'Pos. ID: '),\n",
    "                   alt.Tooltip('Consequence', title = 'Consequence: '),\n",
    "                   alt.Tooltip('amino_acid_change', title = 'AA Sub.: '),\n",
    "                   alt.Tooltip('snv_score_minmax', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('Allele Frequency', title = 'MAF: ')\n",
    "                  ]\n",
    "    ).configure_axis(\n",
    "        grid = False\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400,\n",
    "        title = alt.TitleParams( text = 'gnomAD Allele Frequency vs. SGE Score' + ' (n = ' + str(len(df)) + ')', fontSize = 22)\n",
    "    ).interactive()\n",
    "    \n",
    "    #graph.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_4a_gnomAD.png', ppi = 500)\n",
    "    graph.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    filt = read_gnomAD(gnom_path)\n",
    "    rewritten = rewrite(filt)\n",
    "    merged = add_scores(rewritten, scores)\n",
    "    #merged.to_excel('/Users/ivan/Desktop/20250703_BARD1_gnomAD_vars.xlsx', index = False)\n",
    "    graph(merged)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "main() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
