{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "bard1_data = '../Data/BARD1_SGE_final_table.xlsx'\n",
    "brca1_data = '../Data/BRCA1_SGE_data.xlsx'\n",
    "\n",
    "#Gets thresholds for BARD1 data\n",
    "threshold_df = pd.read_excel(bard1_data, sheet_name = 'thresholds')\n",
    "\n",
    "cutoffs = {'BARD1': [threshold_df['min'][0], threshold_df['max'][0]],\n",
    "           'BRCA1': [-1.328,-0.748] #BRCA1 thresholds from Findlay et al. 2018\n",
    "          }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_zn_residues(bard1, brca1): #Pulls out data for missense variants impacting Zn2+ binding residues\n",
    "\n",
    "    aa_3to1 = {\n",
    "    'Ala': 'A', 'Arg': 'R', 'Asn': 'N', 'Asp': 'D', 'Cys': 'C',\n",
    "    'Gln': 'Q', 'Glu': 'E', 'Gly': 'G', 'His': 'H', 'Ile': 'I',\n",
    "    'Leu': 'L', 'Lys': 'K', 'Met': 'M', 'Phe': 'F', 'Pro': 'P',\n",
    "    'Ser': 'S', 'Thr': 'T', 'Trp': 'W', 'Tyr': 'Y', 'Val': 'V'\n",
    "    } #Amino acid three letter code to one letter code map for BRCA1 data\n",
    "\n",
    "    #Reads data\n",
    "    bard1_df = pd.read_excel(bard1, sheet_name = 'scores')\n",
    "    brca1_df = pd.read_excel(brca1, sheet_name = 'findlay_2018')\n",
    "    \n",
    "    bard1_df = bard1_df.rename(columns = {'consequence': 'Consequence'}) #Renaming columns for consistency\n",
    "    brca1_df = brca1_df.rename(columns = {'snv_score_minmax': 'score'}) #Renaming columns for consistency\n",
    "    \n",
    "    df_list = [bard1_df, brca1_df]\n",
    "\n",
    "    #Re-labels functional consequences for each consequence type\n",
    "    for df in df_list:\n",
    "        df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "        df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "        df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "        df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "        df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "        df.loc[df['Consequence'].str.contains('site'), 'Consequence'] = 'Canonical Splice'\n",
    "        df.loc[df['Consequence'].str.contains('ing_var'), 'Consequence'] = 'Splice Region'\n",
    "        df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "        df.loc[df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "\n",
    "    \n",
    "    #Gets BARD1 amino acids that coordinate Zn2+\n",
    "    bard1_df['amino_acid'] = bard1_df['amino_acid_change'].transform(lambda x: x[:-1])\n",
    "    bard1_df = bard1_df.loc[bard1_df['amino_acid'].isin(['C50','C53','C66', 'H68', 'C71', 'C74', 'C83', 'C86'])] #Zn2+ binding residues from Klevit lab structure paper\n",
    "    bard1_df = bard1_df[['amino_acid', 'amino_acid_change', 'Consequence', 'functional_consequence', 'score']]\n",
    "\n",
    "    #Gets BRCA1 amino acids that coordinate Zn2+\n",
    "    brca1_df['AApos'] = brca1_df['hgvs_pro'].transform(lambda x: x.split(':')[1].split('.')[1][3:-3]) #Gets amino acid position from the HGVS protein notation\n",
    "    brca1_df['amino_acid'] = brca1_df['hgvs_pro'].transform(lambda x: x.split(':')[1].split('.')[1][0:3]) #Gets original amino acid from the HGVS protein notation\n",
    "    brca1_df['amino_acid'] = brca1_df['amino_acid'].map(aa_3to1) #Remaps the 3-letter amino acid code to 1-letter code\n",
    "    brca1_df['amino_acid'] = brca1_df['amino_acid'] + brca1_df['AApos'] #Gets the full amino acid with position\n",
    "    brca1_df['amino_acid_change'] = brca1_df['hgvs_pro'].transform(lambda x: x.split(':')[1].split('.')[1][-3:]) #Gets amino acid substitution\n",
    "    \n",
    "    brca1_df['amino_acid_change'] = brca1_df['amino_acid_change'].map(aa_3to1) #Maps BRCA1 3 letter amino acid codes to one letter codes\n",
    " \n",
    "    brca1_df['amino_acid_change'] = brca1_df['amino_acid'] + brca1_df['amino_acid_change']\n",
    "    brca1_df = brca1_df.loc[brca1_df['amino_acid'].isin(['C24', 'C27', 'C44', 'C47', 'C39', 'H41', 'C61', 'C64'])] #Zn2+ binding residues from Klevit lab structure paper\n",
    "    brca1_df = brca1_df[['score', 'Consequence', 'amino_acid', 'amino_acid_change']]\n",
    "\n",
    "    to_return = {'BARD1': bard1_df,\n",
    "                 'BRCA1': brca1_df\n",
    "                }\n",
    "    \n",
    "    return to_return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def quick_missense_stats(data_dfs): #Quick numbers on number of Zn2+ binding missense variants that are functionally abnormal\n",
    "    \n",
    "    df = data_dfs['BARD1']\n",
    "    df = df.loc[df['Consequence'].isin(['Missense'])]\n",
    "    zn_missense_vars = len(df)\n",
    "    ab_missense_vars = df['functional_consequence'].value_counts().get('functionally_abnormal',0)\n",
    "\n",
    "    print('Total Missense Vars: ', str(zn_missense_vars), '\\n',\n",
    "            'Abnormal Missense Vars: ', str(ab_missense_vars)\n",
    "         )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def strip_plot(data_dfs, cutoffs): #Builds side-by-side strip plots to compare/contrast distribution of Zn2+ binding missense variants for BRCA1 and BARD1\n",
    "    \n",
    "    keys = ['BARD1', 'BRCA1'] #Gene names\n",
    "    plots = [] #List to hold plots\n",
    "\n",
    "    #Iterates through each gene and builds the strip plot\n",
    "    for key in keys:\n",
    "        df = data_dfs[key]\n",
    "\n",
    "        nf_line = alt.Chart(pd.DataFrame({'x': [cutoffs[key][0]]})).mark_rule(color = 'red').encode(\n",
    "            x = 'x')\n",
    "    \n",
    "        func_lin = alt.Chart(pd.DataFrame({'x': [cutoffs[key][1]]})).mark_rule(color = 'blue').encode(\n",
    "            x = 'x')\n",
    "    \n",
    "        sort_order = [\"Intron\", \"Missense\", \"Synonymous\", \"Stop Gained\", \"Splice\", \"Start Lost\", 'Stop Lost', 'UTR']\n",
    "        controls_sorted = [\"Intron\",  \"Synonymous\", \"Stop Gained\", \"Missense\", \"Splice\", \"Start Lost\", 'Stop Lost', 'UTR']\n",
    "    \n",
    "    \n",
    "        # Get the category10 colors\n",
    "        category10_colors = [\n",
    "            '#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd', \n",
    "            '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf'\n",
    "        ]\n",
    "        \n",
    "        # Map each category to a specific color\n",
    "        color_mapping = {\n",
    "            \"Intron\": category10_colors[0],\n",
    "            \"Synonymous\": category10_colors[2],\n",
    "            \"Stop Gained\": category10_colors[3],\n",
    "            \"Missense\": category10_colors[1],\n",
    "            \"Splice\": category10_colors[4],\n",
    "            \"Start Lost\": category10_colors[5],\n",
    "            \"Stop Lost\": category10_colors[6],\n",
    "            \"UTR\": category10_colors[7]\n",
    "        }\n",
    "        \n",
    "        plot = alt.Chart(df).mark_tick(opacity = 1).encode(\n",
    "            x = alt.X('score:Q',\n",
    "                      axis = alt.Axis(title = '', \n",
    "                                      titleFontSize = 20,\n",
    "                                     labelFontSize = 24)\n",
    "                     ),\n",
    "            y = alt.Y('amino_acid:N', \n",
    "                      sort = 'ascending',\n",
    "                      axis = alt.Axis(title = '',\n",
    "                                     labelFontSize = 24)\n",
    "                     ),\n",
    "            color = alt.Color('Consequence:N',\n",
    "                    sort = sort_order,\n",
    "                    scale = alt.Scale(domain = list(color_mapping.keys()),\n",
    "                                      range = list(color_mapping.values())\n",
    "                             )\n",
    "                             ),\n",
    "            tooltip = [alt.Tooltip('Consequence', title = 'Consequence: '),\n",
    "                        alt.Tooltip('amino_acid_change', title = 'Amino Acid Change: ')]\n",
    "            ).properties(\n",
    "                width = 800,\n",
    "                height = 400,\n",
    "                title = 'Variants at ' + key + ' Zn-Binding Residues'\n",
    "            )\n",
    "        \n",
    "        plot = plot + nf_line + func_lin\n",
    "\n",
    "        plots.append(plot)\n",
    "\n",
    "    final_plot = plots[0] | plots[1] #Concatenates plots\n",
    "    \n",
    "    final_plot = final_plot.configure_tick(\n",
    "        thickness = 2\n",
    "    )\n",
    "    final_plot.display()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def bard1_plot_only(df, cutoffs): #Builds clean strip plot showing distribution of BARD1 Zn2+ binding variants only\n",
    "    \n",
    "    nf_line = alt.Chart(pd.DataFrame({'x': [cutoffs['BARD1'][0]]})).mark_rule(color = 'red').encode(\n",
    "            x = 'x')\n",
    "    \n",
    "    func_lin = alt.Chart(pd.DataFrame({'x': [cutoffs['BARD1'][1]]})).mark_rule(color = 'blue').encode(\n",
    "        x = 'x')\n",
    "\n",
    "    sort_order = [\"Intron\", \"Missense\", \"Synonymous\", \"Stop Gained\", \"Splice\", \"Start Lost\", 'Stop Lost', 'UTR']\n",
    "    controls_sorted = [\"Intron\",  \"Synonymous\", \"Stop Gained\", \"Missense\", \"Splice\", \"Start Lost\", 'Stop Lost', 'UTR']\n",
    "\n",
    "\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a', # yellow\n",
    "    '#CFCFCF' # light gray\n",
    "        \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained', \n",
    "        'Splice Region',\n",
    "    ]\n",
    "    \n",
    "    plot = alt.Chart(df).mark_tick(opacity = 1, thickness = 2).encode(\n",
    "        x = alt.X('score:Q',\n",
    "                  axis = alt.Axis(title = 'SGE Score', \n",
    "                                  values = [-0.6, -0.4,-0.2,  0,  0.2],\n",
    "                                  titleFontSize = 20,\n",
    "                                 labelFontSize = 22)\n",
    "                 ),\n",
    "        y = alt.Y('amino_acid:N', \n",
    "                  sort = ['C50', 'C53', 'C66', 'H68', 'C71', 'C74','C83', 'C86'],\n",
    "                  axis = alt.Axis(title = 'Residue',\n",
    "                                  titleFontSize = 20,\n",
    "                                 labelFontSize = 22)\n",
    "                 ),\n",
    "        color = alt.Color('Consequence:N',\n",
    "                sort = variant_types,\n",
    "                scale = alt.Scale(domain = variant_types,\n",
    "                                  range = palette\n",
    "                         ),\n",
    "                legend = alt.Legend(titleFontSize = 18, \n",
    "                                     labelFontSize = 16,\n",
    "                         )\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('Consequence', title = 'Consequence: '),\n",
    "                    alt.Tooltip('amino_acid_change', title = 'Amino Acid Change: ')]\n",
    "        ).properties(\n",
    "            width = 400,\n",
    "            height = 300,\n",
    "            title = alt.TitleParams(text = 'Variants at ' + 'BARD1' + ' Zn-Binding Residues',\n",
    "                                    fontSize = 22\n",
    "                                   )\n",
    "        )\n",
    "    \n",
    "    plot = (plot + nf_line + func_lin).configure_axis(\n",
    "        grid = False).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "\n",
    "    plot.display()\n",
    "\n",
    "    return plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    dfs = get_zn_residues(bard1_data, brca1_data)\n",
    "    quick_missense_stats(dfs)\n",
    "    strip_plot(dfs, cutoffs)\n",
    "    bard1_plot = bard1_plot_only(dfs['BARD1'], cutoffs)\n",
    "\n",
    "    #bard1_plot.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_5c_ZnStrip.png', ppi = 500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
