{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "This notebook builds Extended Data Fig. 5a(orthogonal assay histogram and strip plots). The output is manually annotated for orthogonal assay used and variants called abnormal in orthogonal assays but normal in SGE."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import altair as alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sge = '../Data/final_tables/supplementary_file_1_BARD1_SGE_final_table.xlsx' #sge data tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_scores(scores): #Reads SGE scores\n",
    "    df = pd.read_excel(scores, sheet_name = 'scores')#reads TSV\n",
    "\n",
    "    threshold_df = pd.read_excel(scores, sheet_name = 'thresholds')\n",
    "\n",
    "    thresholds = [threshold_df['min'][0], threshold_df['max'][0]]\n",
    "\n",
    "    df = df.loc[~df['amino_acid_change'].isin(['-'])] #pulls out missense variants only\n",
    "    df = df.loc[~df['variant_qc_flag'].isin(['WARN'])]\n",
    "    df = df.rename(columns = {'score' : 'snv_score_minmax', 'amino_acid_change': 'AAsub'}) #renames score column and amino acid change column\n",
    "\n",
    "   \n",
    "    \n",
    "    return df, thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_orthogonal(file): #Reads orthogonal assay file\n",
    "    \n",
    "    df = pd.read_excel(file, sheet_name = 'orthogonal_data') #reads excel\n",
    "\n",
    "    df['Orthogonal Assay'] = 'ok' #Adds column to hold orthogonal assay characterization\n",
    "\n",
    "    df = df.reset_index(drop = True)\n",
    "                       \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge(scores,orthogonal): #Merges dataframes\n",
    "    \n",
    "    df = pd.merge(scores, orthogonal, how = 'inner', on = 'AAsub') #merged on intersection of AA substitutions\n",
    "    df = df.drop_duplicates(subset = 'pos_id') #drops duplicate SNVs that can appear\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def characterize_orthogonal(df): #Determines how variant performed in orthogonal assay(s)\n",
    "\n",
    "    df = df.dropna(subset = ['pos_id']) #drops any variants without score\n",
    "\n",
    "    #This block iterates through the dataframe and assigns an orthogonal assay function to each variant\n",
    "    for i in range(len(df)):\n",
    "        row_raw = df.iloc[i].tolist() #pulls out the row of data and stores as list\n",
    "        row = [] #empty list to hold each element of the row\n",
    "\n",
    "        #Iterates through row_raw and converts to string and removes extra spaces\n",
    "        for elem in row_raw:\n",
    "            elem = str(elem)\n",
    "            row.append(elem.strip())\n",
    "\n",
    "        #Boolean tests for how that variant performed in orthogonal assay\n",
    "        if 'Abnormal' in row:\n",
    "            df.iloc[i, len(row) - 1] = 'Abnormal'\n",
    "        elif 'Intermediate' in row:\n",
    "            df.iloc[i, len(row) - 1] = 'Intermediate'\n",
    "        else:\n",
    "            df.iloc[i, len(row) - 1] = 'Normal'\n",
    "\n",
    "    #This chunk assigns what assay(s) were used to characterize each variant\n",
    "    df['Assay Type'] = np.nan #establishes an empty column\n",
    "    df['Assay Type'] = np.where((df['Adamovich2019_HDR'].notna()) | (df['Lee2015_HDR'].notna()),'HDR', df['Assay Type']) #Finds variants done by HDR \n",
    "    df['Assay Type'] = np.where((pd.isna(df['Adamovich2019_HDR'])) & (pd.notna(df['Adamovich2019_western'])),'Western Only', df['Assay Type']) #Finds variants done only by Western\n",
    "    df['Assay Type'] = np.where((pd.notna(df['Dai2021_brca1foci'])) & (pd.notna(df['Dai2021_bard1foci'])) & (pd.notna(df['Dai2021_Ubbinding'])), #finds variants done by nuclear foci and Ub binding\n",
    "                                'Ub Binding', df['Assay Type'])\n",
    "    df['Assay Type'] = np.where((pd.notna(df['Dai2021_Ubbinding'])) & (pd.isna(df['Dai2021_brca1foci'])) & (pd.isna(df['Dai2021_bard1foci'])), #Finds Variants only done by Ub binding\n",
    "                                'Ub Binding', df['Assay Type'])\n",
    "    df['Assay Type'] = np.where(df['Witus2021_Ubactivity'].notna(),'Ubiquitylation Activity', df['Assay Type']) #Finds variants done by Ub activity\n",
    "    df['Assay Type'] = np.where(df['Becker2021_PARPiSensitivity'].notna(),'PARPi Sensitivity', df['Assay Type']) #Finds variants done by Ub activity\n",
    "    \n",
    "\n",
    "    df = df.loc[~df['Assay Type'].isin(['Western Only'])]\n",
    "    return df\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_histogram(df, thresholds): #Makes histogram that displays distribution of SGE scores for variants asayed by orthogonal assays\n",
    "    \n",
    "    nf_line = alt.Chart(pd.DataFrame({'x': [thresholds[0]]})).mark_rule(color = 'red').encode(\n",
    "        x = 'x')\n",
    "\n",
    "    func_lin = alt.Chart(pd.DataFrame({'x': [thresholds[1]]})).mark_rule(color = 'blue').encode(\n",
    "        x = 'x')\n",
    "    \n",
    "    bins = 50 #number of bins\n",
    "    scale = [-2,2] #scale\n",
    "    ticks = list(range(-2,2)) #tick marks\n",
    "    sorted = ['Normal','Abnormal','Intermediate'] #order for the legend\n",
    "\n",
    "    #df = df.loc[df['Assay Type'].isin(['HDR'])]\n",
    "    histogram = alt.Chart(df).mark_bar().encode(\n",
    "            alt.X('snv_score_minmax', axis = alt.Axis(title = '', labels = False, ticks = False), bin = alt.Bin(maxbins = bins), scale = alt.Scale(domain = [-0.45, 0.1])),\n",
    "            alt.Y('count()', axis = alt.Axis(title = 'Number of Variants', labelFontSize = 16, titleFontSize = 20)),\n",
    "            color = alt.Color('Orthogonal Assay:N', scale = alt.Scale(scheme = 'category10'), sort = sorted, legend = alt.Legend(titleFontSize = 16, labelFontSize = 14))\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400,\n",
    "        title = alt.TitleParams(text = f' Functional Scores for Variants with Orthogonal Assay Data (n = {len(df)})', fontSize = 22)\n",
    "    ).interactive()\n",
    "\n",
    "    #histogram = histogram + nf_line + func_lin\n",
    "    return histogram"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_dotplot(df, thresholds): #Makes dotplot to show how variants in orthogonal assays scored in SGE\n",
    "\n",
    "    nf_line = alt.Chart(pd.DataFrame({'snv_score_minmax': [thresholds[0]]})).mark_rule(\n",
    "        color='red',\n",
    "        strokeWidth=2,  # Make line more visible\n",
    "        opacity=0.7     # Slight transparency\n",
    "    ).encode(\n",
    "        x='snv_score_minmax:Q'  # Specify quantitative type\n",
    "    )\n",
    "\n",
    "    func_line = alt.Chart(pd.DataFrame({'snv_score_minmax': [thresholds[1]]})).mark_rule(\n",
    "            color='blue',\n",
    "            strokeWidth=2,\n",
    "            opacity=0.7\n",
    "        ).encode(\n",
    "            x='snv_score_minmax:Q'\n",
    "        )\n",
    "\n",
    "    #df = df.loc[df['Assay Type'].isin(['HDR'])]\n",
    "    sorted = ['Normal','Abnormal','Intermediate'] #order for the legend\n",
    "\n",
    "    dotplot = alt.Chart(df).mark_tick(opacity = 1).encode(\n",
    "        x = alt.X('snv_score_minmax', \n",
    "                  axis = alt.Axis(title = 'Functional Score', \n",
    "                                  values = [-0.4, -0.3, -0.2, -0.1, 0, 0.1], \n",
    "                                  labelFontSize = 16, \n",
    "                                  titleFontSize = 20\n",
    "                                 ), \n",
    "                  scale = alt.Scale(domain = [-0.45, 0.1])\n",
    "                 ),\n",
    "        y = alt.Y('Orthogonal Assay:N', axis = alt.Axis(title = '', labelFontSize = 16, titleFontSize = 20)),\n",
    "        color = alt.Color('Orthogonal Assay:N', scale = alt.Scale(scheme = 'category10'),sort = sorted, legend = alt.Legend(titleFontSize = 16, labelFontSize = 14)),\n",
    "        tooltip = [alt.Tooltip('snv_score_minmax', title = 'Functional Score: '),\n",
    "                   alt.Tooltip('exon', title = 'BARD1 Exon: '),\n",
    "                   alt.Tooltip('AAsub', title = 'Amino Acid Sub.: '),\n",
    "                   alt.Tooltip('functional_consequence', title = 'SGE Functional Consequence: '),\n",
    "                   alt.Tooltip('Assay Type', title = 'Orthogonal Assay: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 100,\n",
    "        title = alt.TitleParams(text = '')\n",
    "    ).facet('Assay Type:N', columns = 1, spacing = 20).resolve_scale(\n",
    "        y = 'independent'\n",
    "    ).interactive()\n",
    "\n",
    "    \n",
    "    return dotplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    data = read_orthogonal(sge)\n",
    "    sge_scores, thresholds = read_scores(sge)\n",
    "    merged = merge(sge_scores,data)\n",
    "    ready_df = characterize_orthogonal(merged)\n",
    "    histogram = make_histogram(ready_df, thresholds)\n",
    "    dotplot = make_dotplot(ready_df, thresholds)\n",
    "    \n",
    "    final_plot = alt.vconcat(histogram, dotplot)\n",
    "\n",
    "    final_plot = final_plot.configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    ).resolve_scale(\n",
    "        x = 'shared'\n",
    "    ).interactive()\n",
    "    \n",
    "    final_plot.display()\n",
    "\n",
    "\n",
    "    #final_plot.save('/Users/ivan/Desktop/BARD1_draft_figs/supp_figs/suppfig_bard1_orthogonal.svg')\n",
    "    #ready_df.to_excel('/Users/ivan/Desktop/20241205_BARD1_SGEvsOrthogonalAssays.xlsx', index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
