{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = '../Data/BARD1_SGE_final_table.xlsx' #SGE datafile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_data(file):\n",
    "    df = pd.read_excel(file, sheet_name = 'scores')\n",
    "    df = df.loc[~df['variant_qc_flag'].isin(['WARN'])]\n",
    "    gnomad_df = df.dropna(subset = ['gnomad_af'])\n",
    "    gnomad_df['dataset'] = 'gnomAD'\n",
    "    gnomad_df = gnomad_df.drop(columns = ['regeneron_maf'])\n",
    "    \n",
    "    reg_df = df.dropna(subset = ['regeneron_maf'])\n",
    "    reg_df = reg_df.drop(columns = ['gnomad_af'])\n",
    "    reg_df['dataset'] = 'Regeneron'\n",
    "    \n",
    "    gnomad_df = gnomad_df.rename(columns = {'gnomad_af': 'Allele Frequency'})\n",
    "    reg_df = reg_df.rename(columns = {'regeneron_maf': 'Allele Frequency'})\n",
    "\n",
    "\n",
    "    df = pd.concat([gnomad_df, reg_df])\n",
    "\n",
    "    df = df.rename(columns = {'consequence': 'Consequence', 'score': 'snv_score_minmax'}) #Renames columns to harmonize column names\n",
    "\n",
    "    #Renames variant consequences to be cleaner\n",
    "    df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "    df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "    df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "    df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "    df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "    df.loc[df['Consequence'].str.contains('site'), 'Consequence'] = 'Canonical Splice'\n",
    "    df.loc[df['Consequence'].str.contains('ing_var'), 'Consequence'] = 'Splice Region'\n",
    "    df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "    df.loc[df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "\n",
    "    df['log_AF'] = np.log10(df['Allele Frequency'])\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def graph(df): #Generates scatter plot of allele frequencies vs. SGE score\n",
    "    \n",
    "\n",
    "    #Color palettes\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a', # yellow\n",
    "    '#6AA84F', # med green\n",
    "    '#93C47D', # light green\n",
    "    '#888888', # med gray\n",
    "    '#000000', # black\n",
    "    '#1170AA', # darker blue\n",
    "    '#CFCFCF' # light gray\n",
    "        \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained',\n",
    "        'Intron', \n",
    "        'UTR Variant',\n",
    "        'Stop Lost',\n",
    "        'Start Lost',\n",
    "        'Canonical Splice', \n",
    "        'Splice Region',\n",
    "    ]\n",
    "\n",
    "    #Builds scatter plot\n",
    "    graph = alt.Chart(df).mark_point().encode(\n",
    "        x=alt.X('snv_score_minmax', \n",
    "                axis=alt.Axis(\n",
    "                    title='SGE Score', \n",
    "                    labelFontSize = 18, \n",
    "                    titleFontSize = 20\n",
    "                )\n",
    "               ),\n",
    "        y = alt.Y('Allele Frequency', \n",
    "                  axis = alt.Axis(\n",
    "                      grid = False, \n",
    "                      labelFontSize = 18, \n",
    "                      titleFontSize = 20,\n",
    "                      format = 'e'\n",
    "                  ), \n",
    "                  scale = alt.Scale(type = 'log'\n",
    "                                   )\n",
    "                 ),\n",
    "        color = alt.Color('Consequence', \n",
    "                          scale = alt.Scale(range = palette,\n",
    "                                  domain = variant_types), \n",
    "                          legend = alt.Legend(\n",
    "                              titleFontSize = 16, \n",
    "                              labelFontSize = 14,\n",
    "                              labelLimit = 500\n",
    "                          )\n",
    "                         ),\n",
    "        shape = alt.Shape('dataset',\n",
    "                         legend = alt.Legend(\n",
    "                             title = 'Dataset',\n",
    "                             titleFontSize = 16,\n",
    "                             labelFontSize = 14\n",
    "                         )\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('exon', title = 'Exon: '),\n",
    "                   alt.Tooltip('pos_id', title = 'Pos. ID: '),\n",
    "                   alt.Tooltip('Consequence', title = 'Consequence: '),\n",
    "                   alt.Tooltip('amino_acid_change', title = 'AA Sub.: '),\n",
    "                   alt.Tooltip('snv_score_minmax', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('Allele Frequency', title = 'MAF: ')\n",
    "                  ]\n",
    "    ).configure_axis(\n",
    "        grid = False\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400,\n",
    "        title = alt.TitleParams( text = 'Allele Frequency vs. SGE Score' + ' (n = ' + str(len(df)) + ')', fontSize = 22)\n",
    "    ).interactive()\n",
    "    \n",
    "    #graph.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_4a_MAF.png', ppi = 500)\n",
    "    graph.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def heatmap(df):\n",
    "\n",
    "    \n",
    "    plot = alt.Chart(df).mark_rect().encode(\n",
    "        x = alt.X('snv_score_minmax:Q',\n",
    "                  axis = alt.Axis(\n",
    "                      title = 'SGE Score',\n",
    "                      titleFontSize = 20,\n",
    "                      labelFontSize = 18,\n",
    "                      values = [-0.5, -0.4, -0.3, -0.2, -0.1, 0, 0.1]\n",
    "                  ),\n",
    "                  bin = alt.Bin(maxbins = 50)\n",
    "                 ),\n",
    "        y = alt.Y('log_AF:Q',\n",
    "                  bin = alt.Bin(maxbins = 20),\n",
    "                  axis = alt.Axis(\n",
    "                      title = 'log10(Allele Frequency)',\n",
    "                      titleFontSize = 20,\n",
    "                      labelFontSize = 18,\n",
    "                      values = [0, -1, -2, -3, -4, -5, -6]\n",
    "                  )\n",
    "                  ),\n",
    "        color = alt.Color('count():Q',\n",
    "                          scale = alt.Scale(scheme = 'lighttealblue'),\n",
    "                          legend = alt.Legend(title = '# of Variants',\n",
    "                                              titleFontSize = 16,\n",
    "                                              labelFontSize = 14\n",
    "                                             )\n",
    "                         )\n",
    "    ).properties(\n",
    "        width = 500,\n",
    "        height = 300,\n",
    "        title = alt.TitleParams( text = 'Allele Frequency vs. SGE Score' + ' (n = ' + str(len(df)) + ')', fontSize = 22)\n",
    "    ).configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "\n",
    "    plot.display()\n",
    "\n",
    "    return plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def density(df):\n",
    "    print(df)\n",
    "\n",
    "    sns.set_style(\"whitegrid\")\n",
    "    plt.figure(figsize=(6, 4))\n",
    "\n",
    "    sns.kdeplot(\n",
    "        data = df,\n",
    "        x = 'snv_score_minmax',\n",
    "        y = 'log_AF',\n",
    "        fill = True,\n",
    "        cmap = 'viridis',\n",
    "        levels = 10,\n",
    "        thresh = 0.05,\n",
    "        bw_adjust = 0.5\n",
    "    )\n",
    "\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    merged = read_data(file)\n",
    "    graph(merged)\n",
    "    binned_map = heatmap(merged)\n",
    "    density(merged)\n",
    "\n",
    "    #binned_map.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_4a_MAF.png', ppi = 500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "main() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
