{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import altair as alt\n",
    "from scipy.stats import pearsonr\n",
    "import re\n",
    "\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.default_inference import DefaultInference\n",
    "from pydeseq2.ds import DeseqStats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_path = '../Data/ATG_lib_data/counts'\n",
    "x1a_annotations = '../Data/ATG_lib_data/X1A_annotations.xlsx'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_counts(path):\n",
    "    data_path = Path(counts_path)\n",
    "\n",
    "    all_files = list(data_path.glob('*tsv'))\n",
    "\n",
    "    counts_dict = {}\n",
    "\n",
    "    for path in all_files:\n",
    "        df = pd.read_csv(path, sep = '\\t')\n",
    "\n",
    "        str_path = str(path)\n",
    "\n",
    "        if 'lib_counts' in str_path:\n",
    "            counts_dict['lib'] = df\n",
    "        elif 'NC' in str_path:\n",
    "            counts_dict['NC'] = df\n",
    "        elif 'R1R2R3_D05' in str_path:\n",
    "            counts_dict['D05_rep1'] = df\n",
    "        elif 'R4R5R6_D05' in str_path:\n",
    "            counts_dict['D05_rep2'] = df\n",
    "        elif 'R7R8R9_D05' in str_path:\n",
    "            counts_dict['D05_rep3'] = df\n",
    "        elif 'R3_D13' in str_path:\n",
    "            counts_dict['D13_rep1'] = df\n",
    "        elif 'R6_D13' in str_path:\n",
    "            counts_dict['D13_rep2'] = df\n",
    "        elif 'R9_D13' in str_path:\n",
    "            counts_dict['D13_rep3'] = df\n",
    "\n",
    "    return counts_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def add_freq(dict):\n",
    "    keys = list(dict.keys())\n",
    "\n",
    "    freq_dicts = {}\n",
    "    for key in keys:\n",
    "        if key != 'NC':\n",
    "            df = dict[key]\n",
    "\n",
    "            total_count = df['Count'].sum()\n",
    "            df['freq'] = df['Count'] / total_count\n",
    "\n",
    "            freq_dicts[key] = df\n",
    "\n",
    "    keys = list(freq_dicts.keys())\n",
    "    return freq_dicts,keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def mutate_snvs(dna_sequence): #Mutates all possible SNVs of provided DNA sequence\n",
    "    snvs = []\n",
    "    i = 0\n",
    "    while i < len(dna_sequence):\n",
    "        if dna_sequence[i] == \"A\":\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        elif dna_sequence[i] == \"T\":\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        elif dna_sequence[i] == \"C\":\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        else:\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "        i += 1\n",
    "    return snvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def reverse_complement_string(seq_string): #Reverse complement and returns string\n",
    "    reverse_seq = seq_string[::-1]\n",
    "    reverse_comp_list = []\n",
    "    for char in reverse_seq:\n",
    "        if char == \"A\":\n",
    "            reverse_comp_list.append(\"T\")\n",
    "        elif char == \"G\":\n",
    "            reverse_comp_list.append(\"C\")\n",
    "        elif char == \"C\":\n",
    "            reverse_comp_list.append(\"G\")\n",
    "        else:\n",
    "            reverse_comp_list.append(\"A\")\n",
    "    reverse_compliment_str = \"\".join(reverse_comp_list)\n",
    "    return reverse_compliment_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def compare_strings(str1, str2, coord_offset):\n",
    "    list_str1 = []\n",
    "    list_str2 = []\n",
    "\n",
    "    for char in str1:\n",
    "        list_str1.append(char)\n",
    "\n",
    "    for char in str2:\n",
    "        list_str2.append(char)\n",
    "\n",
    "    i = 0 \n",
    "    while i < len(list_str1):\n",
    "        if list_str1[i] == list_str2[i]:\n",
    "            i += 1\n",
    "        else:\n",
    "            output_str = str(i + coord_offset) + ':' + list_str2[i]\n",
    "            \n",
    "            return output_str\n",
    "            \n",
    "            i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def annotate_snv_lib():\n",
    "    seq = 'CCATGGA'\n",
    "\n",
    "    coding_snvs = mutate_snvs(seq)\n",
    "    rev_coding_snvs = {}\n",
    "\n",
    "    for snv in coding_snvs:\n",
    "        rev_coding_snvs[snv] = reverse_complement_string(snv)\n",
    "\n",
    "    rev_seq = 'TCCATGG'\n",
    "\n",
    "    mapped_snvs = {}\n",
    "    for snv in coding_snvs:\n",
    "        pos_id = compare_strings(rev_seq, rev_coding_snvs[snv], 214809490)\n",
    "        mapped_snvs[snv] = pos_id\n",
    "\n",
    "\n",
    "\n",
    "    return mapped_snvs\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_vars(dict,keys, snv_map, annotations):\n",
    "\n",
    "    annotation_df = pd.read_excel(annotations)\n",
    "    annotated_dfs = {}\n",
    "    for key in keys:\n",
    "        df = dict[key]\n",
    "        df['canonical_start'] = df['Sequence'].transform(lambda x: x[15:18])\n",
    "        df['second_start'] = df['Sequence'].transform(lambda x: x[90:93])\n",
    "        df['snv_lib'] = df['Sequence'].transform(lambda x: x[88:95])\n",
    "\n",
    "        df['pos_id'] = df['snv_lib'].transform(lambda x: snv_map[x])\n",
    "        df['start_pos_id'] = df['canonical_start'] + ':' + df['pos_id']\n",
    "        df = pd.merge(df, annotation_df, how = 'left', on = 'pos_id')\n",
    "\n",
    "        df = df[['Count', 'freq', 'canonical_start', 'second_start', 'pos_id', 'pos', 'allele', 'AAsub', 'Consequence','start_pos_id']]\n",
    "        df = df.loc[:,['pos', 'allele', 'pos_id', 'Consequence', 'AAsub', 'canonical_start', 'second_start', 'Count', 'freq', 'start_pos_id']]\n",
    "\n",
    "        new_count_name = key + '_count'\n",
    "        new_freq_name = key + '_freq'\n",
    "\n",
    "        df = df.rename(columns = {'Count': new_count_name, 'freq': new_freq_name})\n",
    "        annotated_dfs[key] = df\n",
    "\n",
    "    return annotated_dfs\n",
    "\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_dfs(dict, keys):\n",
    "    base_df = dict['lib']\n",
    "    base_df = base_df[['pos', 'allele', 'pos_id', 'Consequence', 'AAsub', 'canonical_start', 'second_start', 'start_pos_id']]\n",
    "\n",
    "    for key in keys:\n",
    "        to_merge = dict[key]\n",
    "\n",
    "        count_name = key + '_count'\n",
    "        freq_name = key + '_freq'\n",
    "        to_merge = to_merge[['start_pos_id', count_name, freq_name]]\n",
    "\n",
    "        base_df = pd.merge(base_df, to_merge, how = 'left', on = 'start_pos_id')\n",
    "        \n",
    "\n",
    "    return base_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def qc_plots(df):\n",
    "    d5_pairs = [('D05_rep1_freq', 'D05_rep2_freq'), ('D05_rep1_freq', 'D05_rep3_freq'), ('D05_rep2_freq', 'D05_rep3_freq')]\n",
    "    d13_pairs = [('D13_rep1_freq', 'D13_rep2_freq'), ('D13_rep1_freq', 'D13_rep3_freq'), ('D13_rep2_freq', 'D13_rep3_freq')]\n",
    "\n",
    "    d5_plots = []\n",
    "    for pair in d5_pairs:\n",
    "        rep1, rep2 = pair\n",
    "        r, p_value = pearsonr(df[rep1], df[rep2])\n",
    "        scatter = alt.Chart(df).mark_point().encode(\n",
    "            x = rep1,\n",
    "            y = rep2,\n",
    "            tooltip = [alt.Tooltip('canonical_start', title = 'Canonical Start: '),\n",
    "                       alt.Tooltip('second_start', title = 'Second Start: '),\n",
    "                       alt.Tooltip('AAsub', title = 'AA Substitution')\n",
    "                      ]\n",
    "        ).properties(\n",
    "            title = rep1 + ' vs. ' + rep2 + ' (r = ' + str(round(r,3)) + ')'\n",
    "        ).interactive()\n",
    "\n",
    "        trendline = scatter.transform_regression(rep1, rep2).mark_line(color = 'red')\n",
    "        scatter = trendline + scatter\n",
    "        d5_plots.append(scatter)\n",
    "\n",
    "    d5_all = alt.hconcat(d5_plots[0], d5_plots[1], d5_plots[2])\n",
    "    \n",
    "    d13_plots = []\n",
    "    for pair in d13_pairs:\n",
    "        rep1, rep2 = pair\n",
    "        r, p_value = pearsonr(df[rep1], df[rep2])\n",
    "        \n",
    "        scatter = alt.Chart(df).mark_point().encode(\n",
    "            x = rep1,\n",
    "            y = rep2,\n",
    "            tooltip = [alt.Tooltip('canonical_start', title = 'Canonical Start: '),\n",
    "                       alt.Tooltip('second_start', title = 'Second Start: '),\n",
    "                       alt.Tooltip('AAsub', title = 'AA Substitution')\n",
    "                      ]\n",
    "        ).properties(\n",
    "            title = rep1 + ' vs. ' + rep2 + ' (r = ' + str(round(r,3)) + ')'\n",
    "        ).interactive()\n",
    "        trendline = scatter.transform_regression(rep1, rep2).mark_line(color = 'red')\n",
    "\n",
    "        scatter = trendline + scatter\n",
    "        d13_plots.append(scatter)\n",
    "        \n",
    "    d13_all = alt.hconcat(d13_plots[0], d13_plots[1], d13_plots[2])\n",
    "\n",
    "    full_corr_plot = d5_all & d13_all\n",
    "    full_corr_plot.display()\n",
    "\n",
    "\n",
    "    lib_plot = alt.Chart(df).mark_bar().encode(\n",
    "        x = 'pos:O',\n",
    "        y = 'lib_freq',\n",
    "        color = 'canonical_start'\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400\n",
    "    )\n",
    "\n",
    "    lib_plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def score(df):\n",
    "\n",
    "    df = df.rename(columns = {'D05_rep1_count': 'D05_R1',\n",
    "                              'D05_rep2_count': 'D05_R2',\n",
    "                              'D05_rep3_count': 'D05_R3',\n",
    "                              'D13_rep1_count': 'D13_R1',\n",
    "                              'D13_rep2_count': 'D13_R2',\n",
    "                              'D13_rep3_count': 'D13_R3'\n",
    "                             }\n",
    "                  )\n",
    "\n",
    "\n",
    "    df = df.drop(columns = ['D05_rep1_freq', 'D05_rep2_freq', 'D05_rep3_freq',\n",
    "                            'D13_rep1_freq', 'D13_rep2_freq', 'D13_rep3_freq'\n",
    "                           ]\n",
    "                )\n",
    "\n",
    "\n",
    "    value_cols = [col for col in df.columns if \"D13\"  in col or \"D05\"  in col or \"lib_count\" in col]\n",
    "    dfmelt = pd.melt(df, id_vars=['start_pos_id'], value_vars=value_cols)\n",
    "    dfpivot = pd.pivot(dfmelt, index='variable', columns=['start_pos_id'], values='value')\n",
    "    dfpivot = dfpivot.rename_axis(None, axis=1)\n",
    "    dfpivot = dfpivot.rename_axis(None, axis=0)\n",
    "    \n",
    "    metanames = ['D05_R1', 'D05_R2', 'D05_R3', 'D13_R1', 'D13_R2', 'D13_R3', 'lib_count']\n",
    "    metadays = [5, 5, 5, 13, 13, 13, 0]\n",
    "\n",
    "    metadf = pd.DataFrame(\n",
    "        {'sample_name': metanames,\n",
    "        'time': metadays,\n",
    "        }\n",
    "    )\n",
    "    metadf = metadf.set_index('sample_name').rename_axis(None, axis=0)\n",
    "\n",
    "    inference = DefaultInference(n_cpus=1)\n",
    "    dds = DeseqDataSet(\n",
    "        counts=dfpivot,\n",
    "        metadata=metadf,\n",
    "        design=\"~time\",\n",
    "        refit_cooks=True\n",
    "    )\n",
    "    \n",
    "    dds.deseq2()\n",
    "    contrast = [\"time\", 1, 0]\n",
    "    stat_res = DeseqStats(dds, contrast=contrast, inference=inference)\n",
    "    stat_res.summary()\n",
    "    resdf = stat_res.results_df\n",
    "\n",
    "    resdf = resdf.reset_index(names='start_pos_id')#.drop(columns=['index'])\n",
    "    resdf = resdf.merge(df[[\n",
    "        \"pos\", \"allele\", \"start_pos_id\", 'AAsub', 'Consequence', 'canonical_start', 'second_start'\n",
    "    ]])\n",
    "    resdf[\"target\"] = 'BARD1_X1A_ATG'\n",
    "\n",
    "    resdf = resdf[['start_pos_id', 'log2FoldChange', 'lfcSE', 'pos', 'allele', 'AAsub', 'Consequence', 'canonical_start', 'second_start', 'target']]\n",
    "    resdf = resdf.rename(columns = {'log2FoldChange': 'score', 'lfcSE': 'standard_error'})\n",
    "\n",
    "    resdf = resdf.loc[:, ['target', 'pos', 'allele', 'start_pos_id', 'Consequence', 'AAsub', 'canonical_start', 'second_start', 'score', 'standard_error']]\n",
    "\n",
    "    def starts(df):\n",
    "        df['comb_starts'] = df['canonical_start'] + df['second_start']\n",
    "\n",
    "        df['start_consequence'] = None\n",
    "\n",
    "        df.loc[df['comb_starts'].str.match(r'^.{3}ATG$'), 'start_consequence'] = 'canonical_lost'\n",
    "        df.loc[df['comb_starts'].str.match(r'^ATG.{3}$'), 'start_consequence'] = 'second_lost'\n",
    "        df.loc[df['comb_starts'] == 'ATGATG', 'start_consequence'] = 'WT'\n",
    "        df.loc[df['start_consequence'].isna(), 'start_consequence'] = 'both_lost'\n",
    "\n",
    "        #df = df.drop(columns = ['comb_starts'])\n",
    "        \n",
    "        return df\n",
    "\n",
    "    resdf = starts(resdf)\n",
    "\n",
    "    resdf.to_excel('/Users/ivan/Desktop/ATG_QC.xlsx')\n",
    "    return resdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def visualize_scores(df):\n",
    "\n",
    "    ordered_var_list = ['synonymous_variant', 'missense_variant', 'stop_gained']\n",
    "    facet_sort = ['WT', 'canonical_lost', 'second_lost', 'both_lost']\n",
    "    histogram = alt.Chart(df).mark_bar().encode(\n",
    "        x = alt.X('score', title = 'SGE Score', bin = alt.Bin(maxbins = 40)),\n",
    "        y = alt.Y('count()', title = 'Number of Variants'),\n",
    "        color = alt.Color('Consequence', sort = ordered_var_list)\n",
    "    ).properties(\n",
    "        width = 600, \n",
    "        height = 400\n",
    "    ).interactive()\n",
    "\n",
    "    facetgram = histogram.facet(columns = 2, facet = alt.Facet('start_consequence', title = 'Start Codon Impact', sort = facet_sort))\n",
    "\n",
    "    facetgram.display()\n",
    "\n",
    "    scattergram = alt.Chart(df).mark_point().encode(\n",
    "        x = alt.X('pos:O', title = 'Genomic Coordinate', axis = alt.Axis(labels = False)),\n",
    "        y = alt.Y('score', title = 'SGE Score'),\n",
    "        color = alt.Color('Consequence',sort = ordered_var_list),\n",
    "        tooltip = [alt.Tooltip('AAsub', title = 'AA Substitution: ')]\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400\n",
    "    ).facet(\n",
    "        columns = 2,\n",
    "        facet = alt.Facet('start_consequence',\n",
    "                          title = 'Start Codon Impact',\n",
    "                          sort = facet_sort\n",
    "                         )\n",
    "    )\n",
    "\n",
    "    scattergram.display()\n",
    "\n",
    "    std_error = alt.Chart(df).mark_point().encode(\n",
    "        x = 'pos:O',\n",
    "        y = 'standard_error',\n",
    "        color = 'start_consequence'\n",
    "    )\n",
    "\n",
    "    std_error.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    counts_dict = read_counts(counts_path)\n",
    "    freq_dicts, keys = add_freq(counts_dict)\n",
    "    mapped_snvs = annotate_snv_lib()\n",
    "    annotated = annotate_vars(freq_dicts, keys, mapped_snvs, x1a_annotations)\n",
    "    combined_df = combine_dfs(annotated, keys)\n",
    "    qc_plots(combined_df)\n",
    "    scored_df = score(combined_df)\n",
    "    visualize_scores(scored_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
