{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "vep_file = '/Users/ivan/Downloads/20250707_BARD1_All_MisSynSpliceRegion_VEPrun.xlsx'\n",
    "sge_file = '../Data/20250508_BARD1scores_update_FILTERED.xlsx'\n",
    "rna_file = '/Users/ivan/Desktop/20250708_BARD1_202505_data_RNAclassified_beta.xlsx'\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_vep(vep_file, sge_file):\n",
    "\n",
    "    vep_df = pd.read_excel(vep_file)\n",
    "    sge_df = pd.read_excel(sge_file)\n",
    "\n",
    "    vep_df = vep_df[['Location', 'Allele', 'BLOSUM62', 'SpliceAI_pred_DS_AG', 'SpliceAI_pred_DS_AL', 'SpliceAI_pred_DS_DG', 'SpliceAI_pred_DS_DL']]\n",
    "\n",
    "    vep_df['pos'] = vep_df['Location'].transform(lambda x: x[-9:])\n",
    "    vep_df['pos_id'] = vep_df['pos'] + ':' + vep_df['Allele']\n",
    "    vep_df = vep_df.drop(columns = ['Location', 'Allele', 'pos'])\n",
    "    vep_df['max_SpliceAI'] = vep_df[['SpliceAI_pred_DS_AG', 'SpliceAI_pred_DS_AL', 'SpliceAI_pred_DS_DG', 'SpliceAI_pred_DS_DL']].max(axis = 1)\n",
    "    \n",
    "    merged_df = pd.merge(vep_df, sge_df, on = 'pos_id', how = 'left')\n",
    "    \n",
    "    return merged_df, vep_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def scatterplot(df):\n",
    "\n",
    "    scatter = alt.Chart(df).mark_circle().encode(\n",
    "        x = 'score',\n",
    "        y = 'max_SpliceAI',\n",
    "        color = alt.Color('simplified_consequence',\n",
    "                          sort = ['synonymous_variant', 'missense_variant', 'splicing_variant']\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('pos_id', title = 'Variant: '),\n",
    "                   alt.Tooltip('amino_acid_change', title = 'Amino Acid Sub.: '),\n",
    "                   alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('max_SpliceAI', title = 'Max SpliceAI DS: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400\n",
    "    ).interactive()\n",
    "\n",
    "    high_prec = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        y = alt.Y(datum = (0.8))\n",
    "    )\n",
    "    recommended =  alt.Chart(df).mark_rule(color = 'yellow').encode(\n",
    "        y = alt.Y(datum = (0.5))\n",
    "    )\n",
    "    high_recall = alt.Chart(df).mark_rule(color = 'green').encode(\n",
    "        y = alt.Y(datum = (0.2))\n",
    "    )\n",
    "\n",
    "    LFCupper = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        x = alt.X(datum = (-0.077))\n",
    "    )\n",
    "\n",
    "    LFClower = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        x = alt.X(datum = (-0.089))\n",
    "    )\n",
    "    scatter = scatter + high_prec + recommended + high_recall + LFCupper + LFClower\n",
    "    scatter.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_wRNA(rna_file, vep_df):\n",
    "    rna_df = pd.read_excel(rna_file)\n",
    "\n",
    "    rna_merged_df = pd.merge(vep_df, rna_df, on = 'pos_id', how = 'inner') \n",
    "    rna_abnormal = rna_merged_df.loc[rna_merged_df['RNA_classification'].isin(['low'])]\n",
    "    rna_dna_abnormal = rna_abnormal.loc[rna_abnormal['functional_consequence'].isin(['functionally_abnormal'])]\n",
    "\n",
    "    to_plot = [('All Variants', rna_merged_df),\n",
    "               ('RNA Abnormal Variants', rna_abnormal),\n",
    "               ('Functionally Abnormal and Abnormal RNA', rna_dna_abnormal)\n",
    "              ]\n",
    "\n",
    "    all_scatters = []\n",
    "    for plot_title, data in to_plot:\n",
    "        scatter = alt.Chart(data).mark_circle().encode(\n",
    "            x = 'RNA/DNA',\n",
    "            y = 'max_SpliceAI',\n",
    "            color = alt.Color('Consequence',\n",
    "                              sort = ['synonymous_variant', 'missense_variant', 'splicing_variant']\n",
    "                             ),\n",
    "            shape = 'functional_consequence',\n",
    "            tooltip = [alt.Tooltip('pos_id', title = 'Variant: '),\n",
    "                       alt.Tooltip('AAsub', title = 'Amino Acid Sub.: '),\n",
    "                       alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                       alt.Tooltip('RNA/DNA', title = 'RNA Score: '),\n",
    "                       alt.Tooltip('max_SpliceAI', title = 'Max SpliceAI DS: ')\n",
    "                      ]\n",
    "            ).properties(\n",
    "                width = 600,\n",
    "                height = 400,\n",
    "                title = plot_title\n",
    "            ).interactive()\n",
    "\n",
    "        high_prec = alt.Chart(rna_merged_df).mark_rule(color = 'red').encode(\n",
    "            y = alt.Y(datum = (0.8))\n",
    "        )\n",
    "        recommended =  alt.Chart(rna_merged_df).mark_rule(color = 'yellow').encode(\n",
    "            y = alt.Y(datum = (0.5))\n",
    "        )\n",
    "        high_recall = alt.Chart(rna_merged_df).mark_rule(color = 'green').encode(\n",
    "            y = alt.Y(datum = (0.2))\n",
    "        )\n",
    "\n",
    "        scatter = scatter + high_prec + recommended + high_recall\n",
    "        all_scatters.append(scatter)\n",
    "\n",
    "    final_scatter = all_scatters[0] | all_scatters[1] | all_scatters[2]\n",
    "    final_scatter.display()\n",
    "\n",
    "\n",
    "    scatter = alt.Chart(rna_merged_df).mark_circle().encode(\n",
    "        x = 'RNA/DNA',\n",
    "        y = 'max_SpliceAI',\n",
    "        color = alt.Color('Consequence',\n",
    "                          sort = ['synonymous_variant', 'missense_variant', 'splicing_variant']\n",
    "                         ),\n",
    "        shape = 'functional_consequence',\n",
    "        tooltip = [alt.Tooltip('pos_id', title = 'Variant: '),\n",
    "                   alt.Tooltip('AAsub', title = 'Amino Acid Sub.: '),\n",
    "                   alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('RNA/DNA', title = 'RNA Score: '),\n",
    "                   alt.Tooltip('max_SpliceAI', title = 'Max SpliceAI DS: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400\n",
    "    ).interactive()\n",
    "\n",
    "\n",
    "\n",
    "    LFCupper = alt.Chart(rna_merged_df).mark_rule(color = 'red').encode(\n",
    "        x = alt.X(datum = (0))\n",
    "    )\n",
    "\n",
    "    LFClower = alt.Chart(rna_merged_df).mark_rule(color = 'red').encode(\n",
    "        x = alt.X(datum = (1))\n",
    "    )\n",
    "\n",
    "    scatter_faceted = scatter.facet('functional_consequence')\n",
    "    scatter = scatter + high_prec + recommended + high_recall + LFCupper + LFClower\n",
    "    #scatter.display()\n",
    "\n",
    "    scatter_faceted.display()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    sge_merged_df, vep_df = process_vep(vep_file, sge_file)\n",
    "    scatterplot(sge_merged_df)\n",
    "    merge_wRNA(rna_file, vep_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
