{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "from sklearn.metrics import roc_curve, auc, confusion_matrix\n",
    "from typing import Optional, Tuple, Dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ClinVar data file\n",
    "file = '../Data/20250912_BARD1_ClinVarSNVs_1StarPlus.txt'\n",
    "clinvar_dels = '../Data/20250912_BARD1_ClinVarDels_1StarPlus.txt'\n",
    "\n",
    "#SGE SNV and deletion data\n",
    "sge = '../Data/20250825_BARD1snvscores_filtered.xlsx'\n",
    "sge_dels = '../Data/20250829_BARD1delscores.tsv'\n",
    "\n",
    "#Gets thresholds for BARD1\n",
    "bard1df = pd.read_excel(sge)\n",
    "# find the GMM thresholds\n",
    "target_value = 0.950\n",
    "# Calculate the absolute difference for the Normal (N) density\n",
    "diffN = (bard1df['gmm_density_normal'] - target_value).abs()\n",
    "# Find the index of the minimum difference\n",
    "closest_index = diffN.idxmin()\n",
    "# Retrieve the row with the closest value\n",
    "closest_row_n = bard1df.loc[closest_index]\n",
    "\n",
    "# now repeat that for the abnormal density\n",
    "# Calculate the absolute difference\n",
    "diffA = (bard1df['gmm_density_abnormal'] - target_value).abs()\n",
    "# Find the index of the minimum difference\n",
    "closest_index = diffA.idxmin()\n",
    "# Retrieve the row with the closest value\n",
    "closest_row_a = bard1df.loc[closest_index]\n",
    "\n",
    "# now we get the scores that are the closest to the (n)ormal and (a)bnormal thresholds\n",
    "uppr = closest_row_n['score']\n",
    "lwr = closest_row_a['score']\n",
    "\n",
    "path_max = lwr\n",
    "benign_min = uppr\n",
    "thresholds = [lwr,uppr]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_data(file, del_file): #Reads ClinVar data\n",
    "    \n",
    "    df = pd.read_csv(file, delimiter='\\t') #reads ClinVar SNV tabular .txt \n",
    "    df = df[['Name','Protein change','GRCh38Chromosome','GRCh38Location','Germline classification']] #pulls useful columns\n",
    "    df = df.dropna(subset = ['GRCh38Location']) #Drops variants without genomic coordinate\n",
    "    df.GRCh38Location = df.GRCh38Location.astype(int) #Sets coordinates to integer data type\n",
    "    df['Base Change'] = None #preps for next function\n",
    "\n",
    "    del_df = pd.read_csv(del_file, sep = '\\t') #Reads ClinVar deletions\n",
    "    del_df = del_df.loc[del_df['GRCh38Location'].str.contains('-')] #Splits coordinates\n",
    "    del_df['start'] = del_df['GRCh38Location'].transform(lambda x: x.split(' - ')[0]) #Gets deletion start coordinate\n",
    "    del_df['end'] = del_df['GRCh38Location'].transform(lambda x: x.split(' - ')[1]) #Gets deletion end coordinate\n",
    "\n",
    "    #Sets coordinate data types to integer\n",
    "    del_df['start'] = del_df['start'].astype(int) \n",
    "    del_df['end'] = del_df['end'].astype(int)\n",
    "\n",
    "    del_df['del_length'] = del_df['end'] - del_df['start'] #Calculates deletion length \n",
    "\n",
    "    del_df = del_df.loc[del_df['del_length'].isin([2])] #Pulls out 3bp deletions\n",
    "    del_df['Base Change'] = del_df['start'].astype(str) + '-' + del_df['end'].astype(str) #Sets base change column to coordinate spanned by deletion\n",
    "    del_df = del_df[['Base Change', 'Germline classification']] #Pulls out necessary columns\n",
    "\n",
    "\n",
    "    return df, del_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pair(base): #ClinVar gives base changes on negative sense strand, SGE pos_id on positive sense\n",
    "    if base == 'A':\n",
    "        return 'T'\n",
    "    elif base == 'T':\n",
    "        return 'A'\n",
    "    elif base == 'C':\n",
    "        return 'G'\n",
    "    else:\n",
    "        return 'C'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prep_sge(sge, sge_dels, thresholds): #reads SGE data\n",
    "\n",
    "    #Reads SNV data and renames consequences\n",
    "    df = pd.read_excel(sge)\n",
    "    df = df.rename(columns = {'consequence': 'Consequence', 'score': 'snv_score'})\n",
    "    df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "    df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "    df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "    df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "    df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "    df.loc[df['Consequence'].str.contains('site'), 'Consequence'] = 'Canonical Splice'\n",
    "    df.loc[df['Consequence'].str.contains('ing_var'), 'Consequence'] = 'Splice Region'\n",
    "    df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "    df.loc[df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "\n",
    "    df['Function Type'] = 'Indeterminate'\n",
    "    df.loc[df['functional_consequence'] == 'functionally_normal', 'Function Type'] = 'Benign'\n",
    "    df.loc[df['functional_consequence'] == 'functionally_abnormal', 'Function Type'] = 'Pathogenic'\n",
    "    df = df[['target', 'Consequence', 'pos_id', 'snv_score', 'Function Type']] #Pulls these columns \n",
    "\n",
    "    #Reads 3bp del data and renames consequences\n",
    "    del_df = pd.read_csv(sge_dels, sep = '\\t')\n",
    "    del_df = del_df.rename(columns = {'consequence': 'Consequence', 'score': 'snv_score'})\n",
    "    del_df.loc[del_df['Consequence'].str.contains('site'), 'Consequence'] = 'Canonical Splice'\n",
    "    del_df.loc[del_df['Consequence'].str.contains('ing_var'), 'Consequence'] = 'Splice Region'\n",
    "    del_df.loc[del_df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "    del_df.loc[del_df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "    del_df.loc[del_df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "    del_df.loc[del_df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "    del_df.loc[del_df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "    del_df.loc[del_df['Consequence'] == 'inframe_indel', 'Consequence'] = 'Inframe Indel'\n",
    "    \n",
    "    del_df['pos_id'] = del_df['start'].astype(str) + '-' + del_df['end'].astype(str) #Sets position ID type to string\n",
    "\n",
    "    #Functional classification of dels\n",
    "    del_df['Function Type'] = 'Indeterminate'\n",
    "    del_df.loc[del_df['snv_score'] <= thresholds[0], 'Function Type'] = 'Pathogenic'\n",
    "    del_df.loc[del_df['snv_score'] >= thresholds[1], 'Function Type'] = 'Benign'\n",
    "\n",
    "    df = pd.concat([df, del_df]) #Dataframes concatenated\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_base_changes(df): #Creates pos_id column in format of SGE datafile for ClinVar data    \n",
    "    k = 0\n",
    "    while k < len(df):\n",
    "        var = df['Name'][k]\n",
    "        coord = str(df['GRCh38Location'][k])\n",
    "        k += 1\n",
    "        i = 0\n",
    "        j = 3\n",
    "        while j < (len(var) + 1):\n",
    "            test_str = var[i:j]\n",
    "            j += 1\n",
    "            i += 1\n",
    "            sense_base = get_pair(test_str[2])\n",
    "            if test_str[1] == '>':\n",
    "                change = coord + \":\" + sense_base\n",
    "                df.loc[df['Name'] == var, 'Base Change'] = change\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge(clin,clin_del, sge, nf_cutoff, func_cutoff):\n",
    "    #merges ClinVar dataframe and SGE dataframe based on shared pos_id\n",
    "    clin_data = pd.concat([clin, clin_del])\n",
    "\n",
    "    sge_data = sge\n",
    "    df = pd.merge(clin_data, sge_data, left_on = 'Base Change', right_on = 'pos_id', how = 'inner')\n",
    "\n",
    "    #makes dataframe nicer if output needed    \n",
    "    df = df[['Name', 'Protein change', 'Germline classification', 'Base Change', 'Consequence', 'target', 'snv_score', 'Function Type']]\n",
    "    reordered = ['target', 'Name', 'Protein change', 'Base Change', 'Consequence', 'snv_score', 'Germline classification', 'Function Type']\n",
    "    df = df[reordered]\n",
    "\n",
    "    df.rename(columns = {'target': 'target', 'snv_score': 'SGE Score'}, inplace = True)\n",
    "    #df.to_excel('/Users/ivan/Desktop/test_excel_outputs/20250718_BARD1SGE_ClinVar_vars.xlsx', index = None)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rename_germline(df): #Renames ClinVar germline classifications to fit on plot better\n",
    "\n",
    "    df = df.copy() #so the original df isn't overwritten\n",
    "    \n",
    "    #renames germline classification categories\n",
    "    df.loc[(df['Germline classification'] == 'Pathogenic') | (df['Germline classification'] == 'Likely pathogenic'), 'Germline classification'] = 'P/LP'\n",
    "    df.loc[(df['Germline classification'] == 'Benign') | (df['Germline classification'] == 'Likely benign'), 'Germline classification' ] = 'B/LB'\n",
    "    df.loc[df['Germline classification'] == 'Uncertain significance', 'Germline classification' ] = 'VUS'\n",
    "    df.loc[df['Germline classification'] == 'Conflicting classifications of pathogenicity', 'Germline classification' ] = 'Conflicting'\n",
    "    df.loc[df['Germline classification'] == 'Benign/Likely benign', 'Germline classification'] = 'B/LB'\n",
    "    df.loc[df['Germline classification'] == 'Pathogenic/Likely pathogenic', 'Germline classification'] = 'P/LP'\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def histogram(df, nf_cutoff, func_cutoff): #creates histogram and strip plot showing distribution of scores for variants in ClinVar\n",
    "\n",
    "    df.loc[df['Base Change'].str.contains('-'), 'Consequence'] = '3bp Deletion' #Sets consequence of variants for deletion variants\n",
    "\n",
    "    #Color palette\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a', # yellow\n",
    "    '#6AA84F', # med green\n",
    "    '#93C47D', # light green\n",
    "    '#888888', # med gray\n",
    "    '#000000', # black\n",
    "    '#1170AA', # darker blue\n",
    "    '#CFCFCF', # light gray\n",
    "    '#FF9A00'   #orange    \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained',\n",
    "        'Intron', \n",
    "        'UTR Variant',\n",
    "        'Stop Lost',\n",
    "        'Start Lost',\n",
    "        'Canonical Splice', \n",
    "        'Splice Region', \n",
    "        '3bp Deletion'\n",
    "    ]\n",
    "    \n",
    "    #Lines for functional and non-functional cutoffs\n",
    "    nf_line = alt.Chart(pd.DataFrame({'x': [nf_cutoff]})).mark_rule(color = 'red').encode(\n",
    "        x = 'x')\n",
    "\n",
    "    func_lin = alt.Chart(pd.DataFrame({'x': [func_cutoff]})).mark_rule(color = 'blue').encode(\n",
    "        x = 'x')\n",
    "    \n",
    "    #sets bins and domain of scale for histograms\n",
    "    bins = 50\n",
    "    scale = [-3,2]\n",
    "    ticks = list(range(-3,3))\n",
    "\n",
    "    \n",
    "    classifications = ['Benign', 'Benign/Likely benign', 'Likely benign', 'Pathogenic', 'Likely pathogenic', 'Pathogenic/Likely pathogenic', 'Uncertain significance', 'Conflicting classifications of pathogenicity']\n",
    "    p_df = df.loc[df['Germline classification'].isin(classifications)]\n",
    "    p_df.loc[p_df['Germline classification'] == 'Benign/Likely benign','Germline classification'] = 'Likely benign' \n",
    "    p_df.loc[p_df['Germline classification'] == 'Pathogenic/Likely pathogenic', 'Germline classification'] = 'Likely pathogenic'\n",
    "    p_df.loc[(p_df['Germline classification'] == 'Uncertain significance') | (p_df['Germline classification'] == 'Conflicting classifications of pathogenicity'), 'Germline classification'] = 'Uncertain significance'\n",
    "    \n",
    "    #p_df.to_excel('/Users/ivan/Desktop/test_excel_outputs/20250912_BARD1SGE_SNVsDels_inClinVar.xlsx', index = False)\n",
    "    \n",
    "    #Creates histogram showing distribution of variants in ClinVar vs. SGE score\n",
    "    plp_hist = alt.Chart(p_df).mark_bar().encode(\n",
    "        alt.X('SGE Score', axis = alt.Axis(title = 'SGE Score', labelFontSize = 18, titleFontSize = 20), bin = alt.Bin(maxbins = bins)),\n",
    "        alt.Y('count()', axis = alt.Axis(title = 'Number of Variants', labelFontSize = 18, titleFontSize = 20)),\n",
    "        color = alt.Color('Germline classification:N',\n",
    "                          legend = alt.Legend(titleFontSize = 16, \n",
    "                                              labelFontSize = 14\n",
    "                                             ),\n",
    "                          scale = alt.Scale(\n",
    "                              domain = ['Benign', 'Likely benign', 'Pathogenic', 'Likely pathogenic', 'Uncertain significance'],\n",
    "                              range = ['#1D7AAB', '#63A1C4', '#CA7682','#E6B1B8', '#A0A0A0']\n",
    "                          )\n",
    "                         )\n",
    "    ).properties(\n",
    "        width = 800,\n",
    "        height = 400,\n",
    "        title = alt.TitleParams(text = 'BARD1 ClinVar Variants in SGE ' + ' (n= ' + str(len(p_df)) + ')', fontSize = 22)\n",
    "    )\n",
    "    plp_hist = plp_hist + nf_line + func_lin\n",
    "    plp_hist = plp_hist.configure_axis(\n",
    "        grid = False\n",
    "    )\n",
    "    plp_hist.show()\n",
    "\n",
    "    #Creates strip plot showing distribution of variants in ClinVar vs. SGE score\n",
    "    plp_strip = alt.Chart(p_df).mark_tick().encode(\n",
    "        x = alt.X('SGE Score',\n",
    "                  title = 'SGE Score',\n",
    "                  axis = alt.Axis(labelFontSize = 16, \n",
    "                                  titleFontSize = 20\n",
    "                                 )\n",
    "                 ),\n",
    "        y = alt.Y('Germline classification',\n",
    "                 sort = ['Benign', 'Likely benign', 'Uncertain significance', 'Likely pathogenic', 'Pathogenic'],\n",
    "                 axis = alt.Axis(\n",
    "                     labelFontSize = 16,\n",
    "                     titleFontSize = 20,\n",
    "                     labelLimit = 1000\n",
    "                 )\n",
    "                 ),\n",
    "        color = alt.Color('Consequence',\n",
    "                         scale = alt.Scale(\n",
    "                             range = palette,\n",
    "                             domain = variant_types\n",
    "                         ),\n",
    "                          legend = alt.Legend(titleFontSize = 16,\n",
    "                                              labelFontSize = 14\n",
    "                                             )\n",
    "                         ),\n",
    "        tooltip = ['Base Change']\n",
    "    ).properties(\n",
    "        width = 800,\n",
    "        height = 400,\n",
    "        title = alt.TitleParams(text = 'BARD1 ClinVar Variants in SGE ' + ' (n= ' + str(len(p_df)) + ')', fontSize = 22)\n",
    "    )\n",
    "\n",
    "    #Builds summary df with variant counts for each germline classification category\n",
    "    summary_df = p_df['Germline classification'].value_counts().reset_index()\n",
    "    summary_df['count'] = summary_df['count'].astype(str)\n",
    "    summary_df['text'] = '(n = ' + summary_df['count'] + ')'\n",
    "\n",
    "    #Builds y-axis labels with variant counts\n",
    "    counts = alt.Chart(summary_df).mark_text(\n",
    "        align='right',\n",
    "        dx=-10,  # Slight offset to the left of the y-axis\n",
    "        dy=20,  # Offset below the category label\n",
    "        fontSize=16,\n",
    "        color='black'\n",
    "    ).encode(\n",
    "        y=alt.Y('Germline classification:N', sort = ['Benign', 'Likely benign', 'Uncertain significance', 'Likely pathogenic', 'Pathogenic']),\n",
    "        text='text',\n",
    "        x=alt.value(0)  # Position at the left edge\n",
    "    )\n",
    "\n",
    "    #Builds final strip plot\n",
    "    plp_strip = (plp_strip + counts + nf_line + func_lin).configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "    \n",
    "    plp_strip.display()\n",
    "\n",
    "\n",
    "    blb_df = p_df.loc[p_df['Germline classification'].isin(['Benign', 'Likely Benign'])]\n",
    "    plp_df = p_df.loc[p_df['Germline classification'].isin(['Pathogenic', 'Likely pathogenic'])]\n",
    "    blb_scores = blb_df['SGE Score']\n",
    "    plp_scores = plp_df['SGE Score']\n",
    "\n",
    "    plp_mean = plp_scores.mean()\n",
    "    blb_mean = blb_scores.mean()\n",
    "    t_statistic, p_value = stats.ttest_1samp(a = blb_scores, popmean = plp_mean)\n",
    "\n",
    "    print(\"Mean BLB Score: \", blb_mean, '\\n',\n",
    "          'Mean PLP Score: ', plp_mean, '\\n',\n",
    "          'One Sample T-Test P-value: ', p_value)\n",
    "    \n",
    "    #extracts VUS and conflicting data from ClinVar\n",
    "    vus = ['Uncertain significance', 'Conflicting classifications of pathogenicity']\n",
    "    v_df = df.loc[df['Germline classification'].isin(vus)]\n",
    "    new_path = v_df.loc[v_df['SGE Score'] <= path_max]\n",
    "    new_benign = v_df.loc[v_df['SGE Score'] >= benign_min]\n",
    "\n",
    "    v_df.loc[v_df['Germline classification'] == 'Conflicting classifications of pathogenicity', 'Germline classification'] = 'Conflicting'\n",
    "    #v_df.to_excel('output.xlsx')\n",
    "    \n",
    "    return plp_hist, plp_strip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pr_df(df): #makes dataframe needed for ROC\n",
    "    #filter out non P/LP and B/LB variants\n",
    "    nv_df = df[['Base Change','Germline classification', 'Function Type']]\n",
    "    non_vus_list = ['Benign', 'Benign/Likely benign', 'Likely benign', 'Pathogenic', 'Likely pathogenic', 'Pathogenic/Likely pathogenic']\n",
    "    nv_df = nv_df.loc[nv_df['Germline classification'].isin(non_vus_list)]\n",
    "\n",
    "    #creates columns in dataframe needed for ROC analysis\n",
    "    nv_df = nv_df.copy()\n",
    "    nv_df['Germline Num'] = np.nan\n",
    "    nv_df['SGE Num'] = np.nan\n",
    "    nv_df['target'] = np.nan\n",
    "\n",
    "    #reindexes new df\n",
    "    new_index = []\n",
    "    for i in range(len(nv_df)):\n",
    "            new_index.append(i)\n",
    "    nv_df = nv_df.reset_index(drop = True)\n",
    "    nv_df.index = new_index\n",
    "\n",
    "    #assigns 1s and 0s to each variant type - both for ClinVar classification and SGE classification\n",
    "    nv_df.loc[(nv_df['Germline classification'] == 'Benign') | (nv_df['Germline classification'] == 'Likely benign') | (nv_df['Germline classification'] == 'Benign/Likely benign') , 'Germline Num'] = 1\n",
    "    nv_df.loc[(nv_df['Germline classification'] == 'Pathogenic') | (nv_df['Germline classification'] == 'Likely pathogenic') | (nv_df['Germline classification'] == 'Pathogenic/Likely pathogenic') , 'Germline Num'] = 0\n",
    "\n",
    "    nv_df.loc[nv_df['Function Type'] == 'Pathogenic', 'SGE Num'] = 0\n",
    "    nv_df.loc[(nv_df['Function Type'] == 'Benign') | (nv_df['Function Type'] == 'Intermediate'), 'SGE Num'] = 1\n",
    "\n",
    "    #determines if ClinVar an SGE agree (1 - yes, 0 - no)\n",
    "    i = 0\n",
    "    while i < len(nv_df):\n",
    "        clin = nv_df['Germline Num'][i]\n",
    "        sge = nv_df['SGE Num'][i]\n",
    "        id = nv_df['Base Change'][i]\n",
    "\n",
    "        if clin == sge:\n",
    "            nv_df.loc[nv_df['Base Change'] == id, 'target'] = 1\n",
    "        else:\n",
    "            nv_df.loc[nv_df['Base Change'] == id, 'target'] = 0\n",
    "\n",
    "        i += 1\n",
    "    \n",
    "    \n",
    "    return nv_df\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def roc_qc(df):\n",
    "    i = 0\n",
    "    concor = 0\n",
    "    total = 0\n",
    "    while i < len(df):\n",
    "        test = df['target'][i]\n",
    "        if test == 1:\n",
    "            concor += 1\n",
    "            total += 1\n",
    "        else:\n",
    "            total += 1\n",
    "        i += 1\n",
    "    print('Cocordant: ' + str(concor))\n",
    "    print('Discordant: ' + str(total-concor))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def concor_stats(df):\n",
    "    i = 0\n",
    "    plp_total = 0\n",
    "    plp_concor = 0\n",
    "    blb_total = 0\n",
    "    blb_concor = 0\n",
    "    while i < len(df):\n",
    "        test = df['target'][i]\n",
    "        type = df['Germline classification'][i]\n",
    "        if type == 'Pathogenic' or type == 'Likely pathogenic' or type == 'Pathogenic/Likely pathogenic':\n",
    "            plp_total += 1\n",
    "            if test == 1:\n",
    "                plp_concor += 1\n",
    "                i += 1\n",
    "            else:\n",
    "                i += 1\n",
    "        elif type == 'Benign' or type == 'Likely benign' or type == 'Benign/Likely benign':\n",
    "            blb_total += 1\n",
    "            if test == 1:\n",
    "                blb_concor += 1\n",
    "                i += 1\n",
    "            else: \n",
    "                i += 1\n",
    "    total = plp_total + blb_total\n",
    "    total_concor = plp_concor + blb_concor\n",
    "    \n",
    "    print(str(total_concor),' of ', str(total), ' variants concordant')\n",
    "    print(str(plp_concor), ' of ', str(plp_total), ' P/LP variants concordant')\n",
    "    print(str(blb_concor), ' of ', str(blb_total), ' B/LB variants concordant')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_roc_curve(df: pd.DataFrame,\n",
    "                      prediction_column: str = 'SGE Num',\n",
    "                      true_label_column: str = 'Germline Num',\n",
    "                      model_name: str = 'SGE Classifier') -> Tuple[pd.DataFrame, float, Dict]:\n",
    "    \"\"\"\n",
    "    Generate ROC curve data comparing SGE predictions to Germline truth.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : DataFrame\n",
    "        Input dataframe with classifications\n",
    "    prediction_column : str\n",
    "        Column name containing predictions (default: 'SGE Num')\n",
    "    true_label_column : str\n",
    "        Column name containing true labels (default: 'Germline Num')\n",
    "    model_name : str\n",
    "        Name for this model/classification\n",
    "    \n",
    "    Returns:\n",
    "    --------\n",
    "    Tuple of (ROC DataFrame, AUC value, metrics dictionary)\n",
    "    \"\"\"\n",
    "    # Remove any rows with NaN values\n",
    "    mask = ~(df[prediction_column].isna() | df[true_label_column].isna())\n",
    "    clean_df = df[mask].copy()\n",
    "    \n",
    "    y_true = clean_df[true_label_column].values\n",
    "    y_pred = clean_df[prediction_column].values\n",
    "    \n",
    "    # Calculate ROC curve\n",
    "    fpr, tpr, thresholds = roc_curve(y_true, y_pred)\n",
    "    roc_auc = auc(fpr, tpr)\n",
    "    \n",
    "    # Calculate additional metrics\n",
    "    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()\n",
    "    \n",
    "    metrics = {\n",
    "        'AUC': roc_auc,\n",
    "        'Sensitivity (TPR)': tp / (tp + fn) if (tp + fn) > 0 else 0,\n",
    "        'Specificity (TNR)': tn / (tn + fp) if (tn + fp) > 0 else 0,\n",
    "        'PPV (Precision)': tp / (tp + fp) if (tp + fp) > 0 else 0,\n",
    "        'NPV': tn / (tn + fn) if (tn + fn) > 0 else 0,\n",
    "        'Accuracy': (tp + tn) / (tp + tn + fp + fn),\n",
    "        'True Positives': tp,\n",
    "        'True Negatives': tn,\n",
    "        'False Positives': fp,\n",
    "        'False Negatives': fn,\n",
    "        'Total Samples': len(y_true)\n",
    "    }\n",
    "    \n",
    "    # Create DataFrame with ROC data\n",
    "    roc_df = pd.DataFrame({\n",
    "        'FPR': fpr,\n",
    "        'TPR': tpr,\n",
    "        'Threshold': thresholds,\n",
    "        'Model': model_name,\n",
    "        'AUC': roc_auc\n",
    "    })\n",
    "    \n",
    "    return roc_df, roc_auc, metrics\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_roc_altair(df: pd.DataFrame,\n",
    "                    prediction_column: str = 'SGE Num',\n",
    "                    true_label_column: str = 'Germline Num',\n",
    "                    model_name: str = 'SGE Classifier',\n",
    "                    width: int = 500,\n",
    "                    height: int = 500,\n",
    "                    title: Optional[str] = None,\n",
    "                    include_diagonal: bool = True,\n",
    "                    show_metrics: bool = True) -> alt.Chart:\n",
    "    \"\"\"\n",
    "    Create an Altair ROC curve visualization comparing SGE to Germline.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : DataFrame\n",
    "        Input dataframe\n",
    "    prediction_column : str\n",
    "        Column with SGE predictions (0 or 1)\n",
    "    true_label_column : str\n",
    "        Column with Germline truth labels (0 or 1)\n",
    "    model_name : str\n",
    "        Name for the model in the visualization\n",
    "    width : int\n",
    "        Chart width in pixels\n",
    "    height : int\n",
    "        Chart height in pixels\n",
    "    title : str, optional\n",
    "        Chart title (auto-generated if None)\n",
    "    include_diagonal : bool\n",
    "        Whether to include diagonal reference line\n",
    "    show_metrics : bool\n",
    "        Whether to print performance metrics\n",
    "    \n",
    "    Returns:\n",
    "    --------\n",
    "    Altair Chart object\n",
    "    \"\"\"\n",
    "    # Generate ROC data and metrics\n",
    "    roc_df, auc_value, metrics = generate_roc_curve(\n",
    "        df, prediction_column, true_label_column, model_name\n",
    "    )\n",
    "    \n",
    "    # Print metrics if requested\n",
    "    if show_metrics:\n",
    "        print(f\"\\n{'='*50}\")\n",
    "        print(f\"Performance Metrics: {model_name}\")\n",
    "        print(f\"{'='*50}\")\n",
    "        print(f\"AUC: {metrics['AUC']:.4f}\")\n",
    "        print(f\"Accuracy: {metrics['Accuracy']:.4f}\")\n",
    "        print(f\"Sensitivity (TPR): {metrics['Sensitivity (TPR)']:.4f}\")\n",
    "        print(f\"Specificity (TNR): {metrics['Specificity (TNR)']:.4f}\")\n",
    "        print(f\"PPV (Precision): {metrics['PPV (Precision)']:.4f}\")\n",
    "        print(f\"NPV: {metrics['NPV']:.4f}\")\n",
    "        print(f\"\\nConfusion Matrix:\")\n",
    "        print(f\"  True Positives: {metrics['True Positives']}\")\n",
    "        print(f\"  True Negatives: {metrics['True Negatives']}\")\n",
    "        print(f\"  False Positives: {metrics['False Positives']}\")\n",
    "        print(f\"  False Negatives: {metrics['False Negatives']}\")\n",
    "        print(f\"  Total Samples: {metrics['Total Samples']}\")\n",
    "        print(f\"{'='*50}\\n\")\n",
    "    \n",
    "    # Add AUC to model name for legend\n",
    "    roc_df['Model_Label'] = f\"{model_name} (AUC = {auc_value:.3f})\"\n",
    "    \n",
    "    # Create the ROC curve\n",
    "    roc_line = alt.Chart(roc_df).mark_line(\n",
    "        strokeWidth=3,\n",
    "        color='darkorange'\n",
    "    ).encode(\n",
    "        x=alt.X('FPR:Q',\n",
    "                title='False Positive Rate (1 - Specificity)',\n",
    "                scale=alt.Scale(domain=[0, 1]),\n",
    "                axis=alt.Axis(format='.1f', grid=True, tickCount=6, labelFontSize = 18, titleFontSize = 20, titleFontWeight = 'bold')),\n",
    "        y=alt.Y('TPR:Q',\n",
    "                title='True Positive Rate (Sensitivity)',\n",
    "                scale=alt.Scale(domain=[0, 1]),\n",
    "                axis=alt.Axis(format='.1f', grid=True, tickCount=6, labelFontSize = 18, titleFontSize = 20, titleFontWeight = 'bold')),\n",
    "        tooltip=[\n",
    "            alt.Tooltip('FPR:Q', format='.3f', title='False Positive Rate'),\n",
    "            alt.Tooltip('TPR:Q', format='.3f', title='True Positive Rate'),\n",
    "            alt.Tooltip('Threshold:Q', format='.3f', title='Threshold'),\n",
    "            alt.Tooltip('AUC:Q', format='.4f', title='AUC')\n",
    "        ]\n",
    "    )\n",
    "    \n",
    "    # Add interactive points for detailed hover\n",
    "    points = alt.Chart(roc_df).mark_circle(\n",
    "        size=30,\n",
    "        color='steelblue',\n",
    "        opacity=0\n",
    "    ).encode(\n",
    "        x='FPR:Q',\n",
    "        y='TPR:Q',\n",
    "        tooltip=[\n",
    "            alt.Tooltip('FPR:Q', format='.3f', title='FPR'),\n",
    "            alt.Tooltip('TPR:Q', format='.3f', title='TPR (Sensitivity)'),\n",
    "            alt.Tooltip('Threshold:Q', format='.3f', title='Threshold')\n",
    "        ]\n",
    "    ).add_params(\n",
    "        alt.selection_point(on='mouseover', nearest=True, empty=False)\n",
    "    )\n",
    "    \n",
    "    # Create optimal point (closest to top-left corner)\n",
    "    optimal_idx = np.argmax(roc_df['TPR'] - roc_df['FPR'])\n",
    "    optimal_point = alt.Chart(pd.DataFrame({\n",
    "        'FPR': [roc_df.iloc[optimal_idx]['FPR']],\n",
    "        'TPR': [roc_df.iloc[optimal_idx]['TPR']],\n",
    "        'Label': [f'Optimal (FPR={roc_df.iloc[optimal_idx][\"FPR\"]:.3f}, TPR={roc_df.iloc[optimal_idx][\"TPR\"]:.3f})']\n",
    "    })).mark_point(\n",
    "        size=150,\n",
    "        color='red',\n",
    "        filled=True\n",
    "    ).encode(\n",
    "        x='FPR:Q',\n",
    "        y='TPR:Q',\n",
    "        tooltip=['Label:N']\n",
    "    )\n",
    "    \n",
    "    # Combine line, points, and optimal point\n",
    "    roc_chart = roc_line + points\n",
    "    \n",
    "    # Add diagonal reference line if requested\n",
    "    if include_diagonal:\n",
    "        diagonal_df = pd.DataFrame({'x': [0, 1], 'y': [0, 1]})\n",
    "        \n",
    "        diagonal = alt.Chart(diagonal_df).mark_line(\n",
    "            strokeDash=[5, 5],\n",
    "            color='gray',\n",
    "            opacity=0.5,\n",
    "            strokeWidth=1.5\n",
    "        ).encode(\n",
    "            x=alt.X('x:Q', scale=alt.Scale(domain=[0, 1])),\n",
    "            y=alt.Y('y:Q', scale=alt.Scale(domain=[0, 1]))\n",
    "        )\n",
    "        \n",
    "        # Add annotation for random classifier\n",
    "        random_text = alt.Chart(pd.DataFrame({\n",
    "            'x': [0.5],\n",
    "            'y': [0.48],\n",
    "            'text': ['']\n",
    "        })).mark_text(\n",
    "            angle=315,\n",
    "            fontSize=11,\n",
    "            color='gray',\n",
    "            opacity=0.7\n",
    "        ).encode(\n",
    "            x='x:Q',\n",
    "            y='y:Q',\n",
    "            text='text:N'\n",
    "        )\n",
    "        \n",
    "        final_chart = diagonal + random_text + roc_chart\n",
    "    else:\n",
    "        final_chart = roc_chart\n",
    "    \n",
    "    # Add AUC text annotation\n",
    "    auc_text = alt.Chart(pd.DataFrame({\n",
    "        'x': [0.95],\n",
    "        'y': [0.05],\n",
    "        'text': [f'AUC = {auc_value:.4f}']\n",
    "    })).mark_text(\n",
    "        align='right',\n",
    "        baseline='top',\n",
    "        fontSize=24,\n",
    "        fontWeight='bold',\n",
    "        color='black'\n",
    "    ).encode(\n",
    "        x=alt.X('x:Q', scale=alt.Scale(domain=[0, 1])),\n",
    "        y=alt.Y('y:Q', scale=alt.Scale(domain=[0, 1])),\n",
    "        text='text:N'\n",
    "    )\n",
    "    \n",
    "    final_chart = final_chart + auc_text\n",
    "    \n",
    "    # Set title\n",
    "    if title is None:\n",
    "        title = f'ROC Curve: {prediction_column} vs {true_label_column} (Ground Truth)'\n",
    "    \n",
    "    # Configure and finalize the chart\n",
    "    final_chart = final_chart.properties(\n",
    "        width=width,\n",
    "        height=height,\n",
    "        title={\n",
    "            'text': title,\n",
    "            'fontSize': 16,\n",
    "            'fontWeight': 'bold'\n",
    "        }\n",
    "    ).configure_axis(\n",
    "        gridOpacity=0.25,\n",
    "        labelFontSize=12,\n",
    "        titleFontSize=13,\n",
    "        titleFontWeight='normal'\n",
    "    ).configure_view(\n",
    "        strokeWidth=1,\n",
    "        stroke='lightgray'\n",
    "    )\n",
    "    \n",
    "    return final_chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_roc_chart(df: pd.DataFrame,\n",
    "                    show_metrics: bool = True) -> alt.Chart:\n",
    "    \"\"\"\n",
    "    Simple function to create ROC curve for SGE vs Germline.\n",
    "    \n",
    "    Parameters:\n",
    "    -----------\n",
    "    df : DataFrame\n",
    "        Your data with SGE Num and Germline Num columns\n",
    "    show_metrics : bool\n",
    "        Whether to print performance metrics\n",
    "    \n",
    "    Returns:\n",
    "    --------\n",
    "    Altair chart\n",
    "    \"\"\"\n",
    "    return plot_roc_altair(\n",
    "        df=df,\n",
    "        prediction_column='SGE Num',\n",
    "        true_label_column='Germline Num',\n",
    "        model_name='SGE Classifier',\n",
    "        show_metrics=show_metrics\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    alt.data_transformers.disable_max_rows()\n",
    "    clin_data, clin_dels_data = read_data(file, clinvar_dels)\n",
    "    print('ClinVar Variants: ', len(clin_data))\n",
    "    with_base = get_base_changes(clin_data)\n",
    "    sge_data = prep_sge(sge, sge_dels, thresholds)\n",
    "    clinvar_data = merge(with_base, clin_dels_data, sge_data,path_max, benign_min)\n",
    "    germ_relabeled = rename_germline(clinvar_data)\n",
    "    df_pr= pr_df(clinvar_data)\n",
    "\n",
    "    #df_pr.to_excel('/Users/ivan/Desktop/test_excel_outputs/20250912_BARD1vClinVar.xlsx', index = False)\n",
    "    concor_stats(df_pr)\n",
    "    final_histogram, final_stripplot = histogram(clinvar_data, path_max, benign_min)\n",
    "\n",
    "    roc = create_roc_chart(df_pr, show_metrics = True)\n",
    "\n",
    "\n",
    "    roc.display()\n",
    "    #roc.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_4b_ClinVar_ROC.png', ppi = 500)\n",
    "    #final_stripplot.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_4b_clinvar_stripplot_.png', ppi = 500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
