{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "from scipy import stats\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "scores = '../Data/20250825_BARD1snvscores_filtered.xlsx'\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_scores(file): #reads scores from excel file\n",
    "\n",
    "    df = pd.read_excel(file) #reads scores\n",
    "\n",
    "    df = df.rename(columns = {'consequence': 'Consequence'}) #Renames consequence column\n",
    "\n",
    "    # find the GMM thresholds\n",
    "    target_value = 0.950\n",
    "    # Calculate the absolute difference for the Normal (N) density\n",
    "    diffN = (df['gmm_density_normal'] - target_value).abs()\n",
    "    # Find the index of the minimum difference\n",
    "    closest_index = diffN.idxmin()\n",
    "    # Retrieve the row with the closest value\n",
    "    closest_row_n = df.loc[closest_index]\n",
    "    \n",
    "    # now repeat that for the abnormal density\n",
    "    # Calculate the absolute difference\n",
    "    diffA = (df['gmm_density_abnormal'] - target_value).abs()\n",
    "    # Find the index of the minimum difference\n",
    "    closest_index = diffA.idxmin()\n",
    "    # Retrieve the row with the closest value\n",
    "    closest_row_a = df.loc[closest_index]\n",
    "    \n",
    "    # now we get the scores that are the closest to the (n)ormal and (a)bnormal thresholds\n",
    "    score_n_95 = closest_row_n['score']\n",
    "    score_a_95 = closest_row_a['score']\n",
    "    \n",
    "    thresholds = [score_a_95, score_n_95]\n",
    "    \n",
    "    return df, thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prep_data(df): #Adds annotation for which domain (or not) a particular variant falls into\n",
    "\n",
    "    #lists of BARD1 protein domains and respective protein domains\n",
    "    ring_coords = [(214809412,214809494),(214797061, 214797117), (214792297,214792445), (214781508, 214781509)] #Residues 26 to 122\n",
    "    ard_coords = [(214780560,214780601),(214769232,214769312),(214767482,214767654),(214752486,214752555)] #Residues 425 to 545\n",
    "    brct_coords = [(214745722,214745830),(214745067,214745159),(214730411,214730508),(214728685,214729008)] #Residues 568 to 777\n",
    "    idr_coords = [(214809495,214809569), (214780602,214781507),(214745831,214745854)] #Residues 1 to 25, 123 to 424 and 546 to 567\n",
    "\n",
    "    all_coords = {'ring': ring_coords, 'idr': idr_coords, 'ard': ard_coords, 'brct': brct_coords}\n",
    "    domains = list(all_coords.keys())\n",
    "\n",
    "    #Generates dictionaries that contain {domain name: [list of coordinates in that domain]}\n",
    "    domain_coords = {}\n",
    "    for domain in domains:\n",
    "        domain_list = []\n",
    "        all_domain_coords = all_coords[domain]\n",
    "\n",
    "        for pair in all_domain_coords:\n",
    "            start, end = pair\n",
    "\n",
    "            for i in range(start, end + 1):\n",
    "                domain_list.append(i)\n",
    "\n",
    "        domain_coords[domain] = domain_list\n",
    "\n",
    "    df['Domain'] = None #creates empty column for domain\n",
    "\n",
    "\n",
    "    #assigns domains for each datapoint\n",
    "    df.loc[df['pos'].isin(domain_coords['ring']), 'Domain'] = 'RING'\n",
    "    df.loc[df['pos'].isin(domain_coords['brct']), 'Domain'] = 'BRCT'\n",
    "    df.loc[df['pos'].isin(domain_coords['ard']), 'Domain'] = 'ARD'\n",
    "    df.loc[df['pos'].isin(domain_coords['idr']), 'Domain'] = 'IDRs'\n",
    "\n",
    "    df = df.loc[df['Domain'].isin(['RING','BRCT','ARD', 'IDRs'])] #gets data just from domains\n",
    "    df = df.reset_index(drop = True) #cleans up indices\n",
    "\n",
    "    \n",
    "    #Renames variant consequences to be more human readable\n",
    "    df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "    df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "    df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "    df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "    df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "    df.loc[df['Consequence'].str.contains('splic'), 'Consequence'] = 'Splice'\n",
    "\n",
    "    df = df.loc[df['Consequence'].isin(['Missense'])] #Gets missense variants only\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def percent_missense(df): #Quick numerical statistics to compare and contrast structured domains to disordered domains\n",
    "    \n",
    "    \n",
    "    ring_df = df.loc[df['Domain'].isin(['RING'])]\n",
    "    ard_df = df.loc[df['Domain'].isin(['ARD'])]\n",
    "    brct_df = df.loc[df['Domain'].isin(['BRCT'])]\n",
    "    idr_df = df.loc[df['Domain'].isin(['IDRs'])]\n",
    "\n",
    "    #Counts number of variants that are LoF vs. normal in each domain\n",
    "    total_sensitive = df['functional_consequence'].value_counts().get('functionally_abnormal',0)\n",
    "    ring_sensitive = ring_df['functional_consequence'].value_counts().get('functionally_abnormal',0)\n",
    "    ring_normal = ring_df['functional_consequence'].value_counts().get('functionally_normal', 0)\n",
    "    ard_sensitive = ard_df['functional_consequence'].value_counts().get('functionally_abnormal',0)\n",
    "    ard_normal = ard_df['functional_consequence'].value_counts().get('functionally_normal',0)\n",
    "    brct_sensitive = brct_df['functional_consequence'].value_counts().get('functionally_abnormal',0)\n",
    "    brct_normal = brct_df['functional_consequence'].value_counts().get('functionally_normal',0)\n",
    "    idr_sensitive = idr_df['functional_consequence'].value_counts().get('functionally_abnormal',0)\n",
    "    idr_normal = idr_df['functional_consequence'].value_counts().get('functionally_normal',0)\n",
    "    idr_int = idr_df['functional_consequence'].value_counts().get('indeterminate',0)\n",
    "\n",
    "    domain_sensitive = ring_sensitive + ard_sensitive + brct_sensitive #Gets total count of LoF missense variants in domains\n",
    "\n",
    "    #Gets total number of variants in each group\n",
    "    ring_vars = len(ring_df)\n",
    "    ard_vars = len(ard_df)\n",
    "    brct_vars = len(brct_df)\n",
    "    structured_total = ring_vars + ard_vars + brct_vars\n",
    "    idr_vars = len(idr_df)\n",
    "\n",
    "    #Tabulates percentages\n",
    "    ring_perc = ring_sensitive / ring_vars\n",
    "    ard_perc = ard_sensitive / ard_vars\n",
    "    brct_perc = brct_sensitive / brct_vars\n",
    "    idr_perc = idr_sensitive / idr_vars\n",
    "    structured_perc = domain_sensitive / total_sensitive\n",
    "\n",
    "    structured_enrichment = ((ring_sensitive + ard_sensitive + brct_sensitive) / structured_total) / idr_perc #Calculates how many times more LoF missense variants are in the structured domains vs. IDR\n",
    "\n",
    "    \n",
    "    print('RING % Sensitive Misense: ', str(100 * ring_perc), '\\n',\n",
    "         'ARD % Sensitive Missense: ', str(100 * ard_perc), '\\n',\n",
    "          'BRCT % Sensitive Missense: ', str(100 * brct_perc), '\\n',\n",
    "          'IDR % Sensitive Missense: ', str(100 * idr_perc), '\\n',\n",
    "          'IDR Missense Vars: ', str(idr_vars), '\\n',\n",
    "          'IDR Abnormal: ', str(idr_sensitive), '\\n',\n",
    "          'IDR Indeterminate: ', str(idr_int), '\\n', \n",
    "          'Totla Structure Missense Vars: ', str(structured_total), '\\n',\n",
    "          'Total LOF Missense: ', str(total_sensitive), '\\n', \n",
    "          'LOF in Domains: ', str(domain_sensitive), '\\n', \n",
    "          '%LOF in Domains: ', str(structured_perc * 100), '\\n',\n",
    "          'Domain LOF Enrichment', str(structured_enrichment)\n",
    "         )\n",
    "\n",
    "\n",
    "    #Fischer exact testing. Gives odds of finding LoF missense variant in the structured domains vs. IDRs with p-value\n",
    "    contingency_table = np.array([[domain_sensitive, (structured_total - domain_sensitive)],\n",
    "                                  [idr_sensitive, (idr_vars - idr_sensitive)]])\n",
    "\n",
    "    odds_ratio, p_value = stats.fisher_exact(contingency_table)\n",
    "\n",
    "    print(f\"Fisher's exact test p-value: {p_value}\")\n",
    "    print(f\"Odds ratio: {odds_ratio}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def strip_plot(df, thresholds): #Builds strip plot showing distribution of missense variants across the differet domains\n",
    "    \n",
    "    nf_line = alt.Chart(pd.DataFrame({'x': [thresholds[0]]})).mark_rule(color = 'red').encode(\n",
    "        x = 'x')\n",
    "\n",
    "    func_lin = alt.Chart(pd.DataFrame({'x': [thresholds[1]]})).mark_rule(color = 'blue').encode(\n",
    "        x = 'x')\n",
    "\n",
    "    sorted = ['RING', 'ARD', 'BRCT', 'IDRS']\n",
    "    controls_sorted = [\"Intron\",  \"Synonymous\", \"Stop Gained\", \"Missense\", \"Splice\", \"Start Lost\", 'Stop Lost', 'UTR']\n",
    "\n",
    "    palette = [\n",
    "    '#8dc4ed', # RING\n",
    "    '#b1b1b1', # IDR\n",
    "    '#9abd9a', # ARD\n",
    "    '#F6BF93', # BRCT\n",
    "    ]\n",
    "\n",
    "\n",
    "    domains = [\n",
    "        'RING',\n",
    "        'IDRs',\n",
    "        'ARD',\n",
    "        'BRCT'\n",
    "    ]\n",
    "    \n",
    "    plot = alt.Chart(df).mark_tick(opacity = 1, color = 'gray').encode(\n",
    "        x = alt.X('score:Q',\n",
    "                  axis = alt.Axis(title = 'SGE Score (Missense Vars. Only)', \n",
    "                                  titleFontSize = 20,\n",
    "                                 labelFontSize = 16,\n",
    "                                 values = [-0.8, -0.6, -0.4, -0.2, 0, 0.2]),\n",
    "                  scale = alt.Scale(domain = [-0.8, 0.3])\n",
    "                 ),\n",
    "        y = alt.Y('Domain:N', \n",
    "                  sort = sorted,\n",
    "                  axis = alt.Axis(title = '',\n",
    "                                 labelFontSize = 20)\n",
    "                 )\n",
    "        ).properties(\n",
    "            width = 800,\n",
    "            height = 400\n",
    "        )\n",
    "    \n",
    "    plot = plot + nf_line + func_lin\n",
    "\n",
    "    plot = plot.configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "    \n",
    "    #plot.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_2c_structure_faceted_stripplot.png', ppi = 500)\n",
    "    plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cumulator_plot(df,thresholds): #Cumulator plot showing cumulative % of SNVs vs. SGE score\n",
    "    \n",
    "    missub = df.sort_values(by=\"score\")\n",
    "    gbobj = missub.groupby('Domain')['score']\n",
    "    missub[\"cumulative_percentage\"] = (gbobj.cumcount()  + 1)/ gbobj.transform('count') * 100.0\n",
    "\n",
    "    lthresh = thresholds[0]\n",
    "    uthresh = thresholds[1]\n",
    "\n",
    "    palette = [\n",
    "    '#B9DBF4', # RING\n",
    "    '#b1b1b1', # IDR\n",
    "    '#9abd9a', # ARD\n",
    "    '#F6BF93', # BRCT\n",
    "    ]\n",
    "\n",
    "\n",
    "    domains = [\n",
    "        'RING',\n",
    "        'IDRs',\n",
    "        'ARD',\n",
    "        'BRCT'\n",
    "    ]\n",
    "    \n",
    "    \n",
    "    # draw the plot\n",
    "    chart = alt.Chart(missub, height=300, width=300).mark_line(size=5).encode(\n",
    "        x=alt.X('score:Q', title='SGE score').axis(labelFlush=False),\n",
    "        y=alt.Y('cumulative_percentage:Q', title='Cumulative percentage of missense SNVs'),\n",
    "        color = alt.Color('Domain:N', \n",
    "                          title = 'Domain',\n",
    "                          sort = ['RING', 'ARD', 'BRCT', 'IDRs'],\n",
    "                          scale = alt.Scale(range = palette,\n",
    "                                            domain = domains\n",
    "                                           ),\n",
    "                          legend = alt.Legend(symbolStrokeWidth = 6,\n",
    "                                              symbolSize = 250\n",
    "                                             )\n",
    "                         )\n",
    "    )\n",
    "    \n",
    "    threshl = alt.Chart().mark_rule(color='darkred', strokeDash=[8,8], strokeWidth=1).encode(\n",
    "        x=alt.datum(lthresh)\n",
    "    )\n",
    "    threshu = alt.Chart().mark_rule(color='darkgreen', strokeDash=[8,8], strokeWidth=1).encode(\n",
    "        x=alt.datum(uthresh)\n",
    "    )\n",
    "    \n",
    "    rectlow = alt.Chart().mark_rect(color='red').encode(\n",
    "        x=alt.datum(lthresh),\n",
    "        opacity=alt.value(0.05),\n",
    "    )\n",
    "    recthigh = alt.Chart().mark_rect(color='green').encode(\n",
    "        x2=alt.value(300),\n",
    "        x=alt.datum(uthresh),\n",
    "        opacity=alt.value(0.05),\n",
    "    )\n",
    "    \n",
    "    \n",
    "    final_plot = (chart + threshl + threshu + rectlow + recthigh)\n",
    "\n",
    "    final_plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    data, thresholds = read_scores(scores)\n",
    "    domained = prep_data(data)\n",
    "    percent_missense(domained)\n",
    "    strip_plot(domained, thresholds)\n",
    "    cumulator_plot(domained, thresholds)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
