{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import statistics\n",
    "import altair as alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sge_input_files = '../Data/filtered_ppj_data/pillar_project_data_inputs.xlsx'\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_inputs(inputs):\n",
    "    df = pd.read_excel(inputs)\n",
    "\n",
    "    genes = df['gene'].tolist()\n",
    "    df.set_index('gene', inplace = True)\n",
    "\n",
    "    return df, genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_thresholds(df):\n",
    "    \n",
    "    # find the GMM thresholds\n",
    "    target_value = 0.950\n",
    "    # Calculate the absolute difference for the Normal (N) density\n",
    "    diffN = (df['gmm_density_normal'] - target_value).abs()\n",
    "    # Find the index of the minimum difference\n",
    "    closest_index = diffN.idxmin()\n",
    "    # Retrieve the row with the closest value\n",
    "    closest_row_n = df.loc[closest_index]\n",
    "    \n",
    "    # now repeat that for the abnormal density\n",
    "    # Calculate the absolute difference\n",
    "    diffA = (df['gmm_density_abnormal'] - target_value).abs()\n",
    "    # Find the index of the minimum difference\n",
    "    closest_index = diffA.idxmin()\n",
    "    # Retrieve the row with the closest value\n",
    "    closest_row_a = df.loc[closest_index]\n",
    "    \n",
    "    # now we get the scores that are the closest to the (n)ormal and (a)bnormal thresholds\n",
    "    score_n_95 = closest_row_n['score']\n",
    "    score_a_95 = closest_row_a['score']\n",
    "\n",
    "    thresholds = [score_a_95, score_n_95]\n",
    "\n",
    "    return thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_gsp_data(input_df, gene):\n",
    "    input_path = input_df['sge_file'][gene]\n",
    "    \n",
    "    df = pd.read_excel(input_path)\n",
    "    df = df.drop(columns = ['functional_consequence'])\n",
    "    df = df.rename(columns = {'gmm_consequence_0.95': 'functional_consequence'})\n",
    "    \n",
    "    df = df.loc[~(df['amino_acid_change'].isin(['---']))]\n",
    "    df = df.loc[~(df['consequence'].isin(['synonymous_variant']))]\n",
    "    \n",
    "    df['og_AA'] = df['amino_acid_change'].transform(lambda x: x[0]) #Makes column with the original amino acid\n",
    "    df['AA_change'] = df['amino_acid_change'].transform(lambda x: x[-1]) #makes column with amino acid change\n",
    "    df['AApos'] = df['amino_acid_change'].transform(lambda x: x[1: len(x)-1]) #makes column with residue position\n",
    "    df['AApos'] = df['AApos'].astype(int)\n",
    "\n",
    "    df.loc[df['functional_consequence'] == 'functionally_normal', 'functional_consequence'] = 'Normal'\n",
    "    df.loc[df['functional_consequence'] == 'functionally_abnormal', 'functional_consequence'] = 'Abnormal'\n",
    "    df.loc[df['functional_consequence'] == 'indeterminate', 'functional_consequence'] = 'Indeterminate'\n",
    "\n",
    "    df = df[['AA_change', 'og_AA', 'AApos', 'functional_consequence', 'consequence', 'score', 'gmm_density_normal', 'gmm_density_abnormal']]\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_min_mean_scores(df, thresholds):\n",
    "\n",
    "    lwr = thresholds[0]\n",
    "    uppr = thresholds[1]\n",
    "    \n",
    "    mis_df = df.loc[df['consequence'].isin(['missense_variant'])]\n",
    "\n",
    "    min_df = mis_df.groupby('AApos')['score'].min().reset_index()\n",
    "    mean_df = mis_df.groupby('AApos')['score'].mean().reset_index()\n",
    "\n",
    "\n",
    "    min_df['AA_change'] = 'Mis. Min.'\n",
    "    mean_df['AA_change'] = 'Mis. Mean'\n",
    "\n",
    "    min_df['og_AA'] = 'Mis. Min.'\n",
    "    mean_df['og_AA'] = 'Mis. Mean'\n",
    "\n",
    "    min_df['full_sub'] = 'Mis. Min.'\n",
    "    mean_df['full_sub'] = 'Mis. Mean'\n",
    "\n",
    "\n",
    "    min_df['functional_consequence'] = 'Indeterminate'\n",
    "    mean_df['functional_consequence'] = 'Indeterminate'\n",
    "\n",
    "    min_df.loc[min_df['score'] >= uppr, 'functional_consequence'] = 'Normal'\n",
    "    min_df.loc[min_df['score'] <= lwr, 'functional_consequence'] = 'Abnormal'\n",
    "    mean_df.loc[mean_df['score'] >= uppr, 'functional_consequence'] = 'Normal'\n",
    "    mean_df.loc[mean_df['score'] <= lwr, 'functional_consequence'] = 'Abnormal'\n",
    "    \n",
    "    min_df.AApos = min_df.AApos.astype(int) #set datatype as int\n",
    "    mean_df.AApos = mean_df.AApos.astype(int) #set datatype as int\n",
    "\n",
    "    df = pd.concat([df,min_df, mean_df]) #concatenates dataframes\n",
    "    df.loc[(df['AA_change'] == '*', 'AA_change')] = 'Stop' #Renames stop-gained variants\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def heatmap(df, gene):\n",
    "\n",
    "    order = ['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y', 'Stop', 'Mis. Min.', 'Mis. Mean']\n",
    "\n",
    "    df = df.loc[~(df['AA_change'].isin(['Mis. Min.', 'Mis. Mean']))]\n",
    "    bins = len(df)\n",
    "    sge_map = alt.Chart(df).mark_rect().encode(\n",
    "        x = alt.X('AApos:Q',\n",
    "                  title = 'Amino Acid Position',\n",
    "                  bin = alt.Bin(maxbins = bins, minstep = 1),\n",
    "                  axis = alt.Axis(values = list(range(0, bins, 50))\n",
    "                                 )\n",
    "                 ),\n",
    "        y = alt.Y('AA_change:O',\n",
    "                  title = 'Amino Acid Substitution',\n",
    "                  sort = order\n",
    "                 ),\n",
    "        color = alt.Color('score',\n",
    "                          scale = alt.Scale(scheme = 'bluepurple',\n",
    "                                            domain = [-0.2, 0],\n",
    "                                            reverse = True\n",
    "                                           )\n",
    "                         ),\n",
    "        tooltip = ['score']\n",
    "    ).properties(\n",
    "        width = 1750,\n",
    "        height = 300,\n",
    "        title = alt.TitleParams(gene, fontSize = 20)\n",
    "    )\n",
    "\n",
    "    sge_map.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    sge_inputs, sge_genes = read_inputs(sge_input_files)\n",
    "\n",
    "    #sge_genes = ['BARD1']\n",
    "\n",
    "    for gene in sge_genes:\n",
    "        if gene == 'BRCA2':\n",
    "            continue\n",
    "        df = process_gsp_data(sge_inputs, gene)\n",
    "        thresholds = get_thresholds(df)\n",
    "        final_df = get_min_mean_scores(df, thresholds)\n",
    "        heatmap(final_df, gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
