{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_file = '../Data/20250825_BARD1.snvscores.tsv'\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_data(file):\n",
    "    df = pd.read_csv(file, sep = '\\t')\n",
    "\n",
    "    # find the GMM thresholds\n",
    "    target_value = 0.950\n",
    "    # Calculate the absolute difference for the Normal (N) density\n",
    "    diffN = (df['gmm_density_normal'] - target_value).abs()\n",
    "    # Find the index of the minimum difference\n",
    "    closest_index = diffN.idxmin()\n",
    "    # Retrieve the row with the closest value\n",
    "    closest_row_n = df.loc[closest_index]\n",
    "    \n",
    "    # now repeat that for the abnormal density\n",
    "    # Calculate the absolute difference\n",
    "    diffA = (df['gmm_density_abnormal'] - target_value).abs()\n",
    "    # Find the index of the minimum difference\n",
    "    closest_index = diffA.idxmin()\n",
    "    # Retrieve the row with the closest value\n",
    "    closest_row_a = df.loc[closest_index]\n",
    "    \n",
    "    # now we get the scores that are the closest to the (n)ormal and (a)bnormal thresholds\n",
    "    uppr = closest_row_n['score']\n",
    "    lwr = closest_row_a['score']\n",
    "\n",
    "    thresholds = [lwr, uppr]\n",
    "\n",
    "    #Filter for variants used in gmm threshold drawing\n",
    "    neutral_df = df.loc[df['consequence'].isin(['synonymous_variant', 'intron_variant'])]\n",
    "    lower, upper = neutral_df['score'].quantile([0.025, 0.975])\n",
    "\n",
    "    neutral_df = neutral_df[neutral_df['score'].between(lower, upper)]\n",
    "    \n",
    "    neutral_df.loc[(neutral_df['consequence'] == 'synonymous_variant') | (neutral_df['consequence'] == 'intron_variant'), 'consequence'] = 'Intron/Syn. Variant'\n",
    "    \n",
    "    abnormal_df = df.loc[~(df['exon'].isin(['BARD1_X1', 'BARD1_X4', 'BARD1_X11']))]\n",
    "    abnormal_df = abnormal_df.copy()\n",
    "    abnormal_df = abnormal_df.loc[abnormal_df['consequence'].isin(['stop_gained'])]\n",
    "    abnormal_df.loc[abnormal_df['consequence'] == 'stop_gained', 'consequence'] = 'Stop Gained'\n",
    "\n",
    "    dfs = [abnormal_df, neutral_df]\n",
    "    \n",
    "    return dfs, thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def density_plot(dfs, thresholds):\n",
    "\n",
    "    plots = []\n",
    "\n",
    "    nf_data = pd.DataFrame({'score': [thresholds[0]]})\n",
    "    func_data = pd.DataFrame({'score': [thresholds[1]]})\n",
    "\n",
    "\n",
    "    nf_line = alt.Chart(nf_data).mark_rule(color = 'red').encode(\n",
    "        x = 'score')\n",
    "\n",
    "    func_lin = alt.Chart(func_data).mark_rule(color = 'blue').encode(\n",
    "        x = 'score')\n",
    "\n",
    "    nf_text = alt.Chart(nf_data).mark_text(\n",
    "        align = 'left',\n",
    "        dx = -55,\n",
    "        dy = -190,\n",
    "        color = 'black',\n",
    "        fontSize = 16\n",
    "    ).encode(\n",
    "        x = 'score',\n",
    "        text = alt.Text('score:Q', format = '.3f')\n",
    "    )\n",
    "\n",
    "    func_text = alt.Chart(func_data).mark_text(\n",
    "        align = 'right',\n",
    "        dx = 55,\n",
    "        dy = -190,\n",
    "        color = 'black',\n",
    "        fontSize = 16\n",
    "    ).encode(\n",
    "        x = 'score',\n",
    "        text = alt.Text('score:Q', format = '.3f')\n",
    "    )\n",
    "        \n",
    "    \n",
    "    for df in dfs:\n",
    "        plot = alt.Chart(df).transform_density(\n",
    "            'score',\n",
    "            as_ = ['score', 'density'],\n",
    "            groupby = ['consequence']\n",
    "        ).mark_area(fillOpacity = 0.7, \n",
    "                   interpolate = 'monotone').encode(\n",
    "            x = alt.X('score:Q',\n",
    "                      title = 'SGE Score',\n",
    "                      axis = alt.Axis(titleFontSize = 20,\n",
    "                                      labelFontSize = 16\n",
    "                                     )\n",
    "                     ),\n",
    "            y = alt.Y('density:Q',\n",
    "                      title = 'Density',\n",
    "                      axis = alt.Axis(titleFontSize = 20,\n",
    "                                      labelFontSize = 16\n",
    "                                     )\n",
    "                     ),\n",
    "            color = alt.Color('consequence',\n",
    "                              legend = alt.Legend(title = 'Consequence',\n",
    "                                                 titleFontSize = 20,\n",
    "                                                 labelFontSize = 16)\n",
    "                             )\n",
    "        ).properties(\n",
    "            width = 600,\n",
    "            height = 400\n",
    "        )\n",
    "\n",
    "        plots.append(plot)\n",
    "\n",
    "\n",
    "    final_plot = alt.layer(plots[0], plots[1], nf_line, func_lin, nf_text, func_text).configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "    final_plot.display()\n",
    "\n",
    "    #final_plot.save('/Users/ivan/Desktop/BARD1_draft_figs/supp_figs/gmm_densities.pdf', dpi = 500)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    data, thresholds = read_data(score_file)\n",
    "    density_plot(data, thresholds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
