{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_path = '../Data/ATG_lib_data/counts/for_rna' #Path to counts for ATG library variants\n",
    "x1a_annotations = '../Data/ATG_lib_data/X1A_annotations.xlsx' #Path to file with annotations for X1A variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_counts(path): #Reads all counts\n",
    "    data_path = Path(counts_path) #Creates path object\n",
    "\n",
    "    all_files = list(data_path.glob('*tsv')) #Creates list of all count TSV files\n",
    "\n",
    "    counts_dict = {} #Empty dictionary to collect counts dataframes\n",
    "    \n",
    "    for path in all_files: #Iterates through all_files list and reads counts TSV files into dataframes and adds to counts_dict dictionary\n",
    "\n",
    "        if 'RNA' in str(path): #Reads RNA counts\n",
    "            df = pd.read_csv(path, sep = '\\t')\n",
    "\n",
    "            if 'RNA_R1R2R3' in str(path):\n",
    "                counts_dict['RNA_D05_rep1'] = df\n",
    "            elif 'RNA_R4R5R6' in str(path):\n",
    "                counts_dict['RNA_D05_rep2'] = df\n",
    "            elif 'RNA_R7R8R9' in str(path):\n",
    "                counts_dict['RNA_D05_rep3'] = df\n",
    "                \n",
    "        else: #Reads DNA counts\n",
    "            df = pd.read_csv(path, sep = '\\t')\n",
    "    \n",
    "            str_path = str(path)\n",
    "    \n",
    "            if 'lib_counts' in str_path:\n",
    "                counts_dict['lib'] = df\n",
    "            elif 'NC' in str_path:\n",
    "                counts_dict['NC'] = df\n",
    "            elif 'R1R2R3_D05' in str_path:\n",
    "                counts_dict['D05_rep1'] = df\n",
    "            elif 'R4R5R6_D05' in str_path:\n",
    "                counts_dict['D05_rep2'] = df\n",
    "            elif 'R7R8R9_D05' in str_path:\n",
    "                counts_dict['D05_rep3'] = df\n",
    "            elif 'R3_D13' in str_path:\n",
    "                counts_dict['D13_rep1'] = df\n",
    "            elif 'R6_D13' in str_path:\n",
    "                counts_dict['D13_rep2'] = df\n",
    "            elif 'R9_D13' in str_path:\n",
    "                counts_dict['D13_rep3'] = df\n",
    "\n",
    "    keys = list(counts_dict.keys())\n",
    "    \n",
    "    return counts_dict, keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_freq(dict): #Adds DNA/RNA frequency columns\n",
    "    \n",
    "    keys = list(dict.keys()) #Gets all keys from dictionary\n",
    "\n",
    "    freq_dicts = {} #New dictionary to hold dataframes with frequency columns\n",
    "\n",
    "    #Iterates through provided dict and adds frequency column\n",
    "    for key in keys:\n",
    "        if key != 'NC': #NC skipped\n",
    "            df = dict[key] #Gets dataframe from dictionary\n",
    "\n",
    "            total_count = df['Count'].sum() #Gets total number of variants\n",
    "            df['freq'] = df['Count'] / total_count #Creates frequency column\n",
    "\n",
    "            freq_dicts[key] = df #Adds to new dictionary\n",
    "\n",
    "    keys = list(freq_dicts.keys()) #Gets list of keys for future use in downstream functions\n",
    "    return freq_dicts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def mutate_snvs(dna_sequence): #Mutates all possible SNVs of provided DNA sequence\n",
    "    snvs = []\n",
    "    i = 0\n",
    "    while i < len(dna_sequence):\n",
    "        if dna_sequence[i] == \"A\":\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        elif dna_sequence[i] == \"T\":\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        elif dna_sequence[i] == \"C\":\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        else:\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "        i += 1\n",
    "    return snvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare_strings(str1, str2, coord_offset): #Special function to compare strings and return position with difference\n",
    "    \n",
    "    list_str1 = [] #List for string 1 characters\n",
    "    list_str2 = [] #List for string 2 charaacters\n",
    "\n",
    "    #Appends strings 1 and 2 to respective lists\n",
    "    for char in str1:\n",
    "        list_str1.append(char)\n",
    "\n",
    "    for char in str2:\n",
    "        list_str2.append(char)\n",
    "\n",
    "    #Iterates and compares each element in str1 and 2 lists. \n",
    "    i = 0 \n",
    "    while i < len(list_str1):\n",
    "        if list_str1[i] == list_str2[i]:\n",
    "            i += 1\n",
    "        else:\n",
    "            output_str = str(i + coord_offset) + ':' + list_str2[i] #If characters not the same, difference and position returned, coord_offset used for genomic coordinate\n",
    "            \n",
    "            return output_str\n",
    "            \n",
    "            i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def reverse_complement_string(seq_string): #Reverse complement and returns string\n",
    "    reverse_seq = seq_string[::-1]\n",
    "    reverse_comp_list = []\n",
    "    for char in reverse_seq:\n",
    "        if char == \"A\":\n",
    "            reverse_comp_list.append(\"T\")\n",
    "        elif char == \"G\":\n",
    "            reverse_comp_list.append(\"C\")\n",
    "        elif char == \"C\":\n",
    "            reverse_comp_list.append(\"G\")\n",
    "        else:\n",
    "            reverse_comp_list.append(\"A\")\n",
    "    reverse_compliment_str = \"\".join(reverse_comp_list)\n",
    "    return reverse_compliment_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_snv_lib(): #Annotates all SNVs in SNV lib with position IDs\n",
    "    seq = 'CCATGGA' #original WT 7bp library around Met26\n",
    "\n",
    "    coding_snvs = mutate_snvs(seq) #mutates all possible SNVs\n",
    "    rev_coding_snvs = {} #Dictionary to hold SNVs that have been reverse complemented (used for pos_id merging)\n",
    "\n",
    "    for snv in coding_snvs: #Iterates through all SNVs and reverse complements\n",
    "        rev_coding_snvs[snv] = reverse_complement_string(snv)\n",
    "\n",
    "    rev_seq = 'TCCATGG'\n",
    "\n",
    "    mapped_snvs = {}\n",
    "    for snv in coding_snvs: #iterates through SNVs and generates position IDs for merging\n",
    "        pos_id = compare_strings(rev_seq, rev_coding_snvs[snv], 214809490)\n",
    "        mapped_snvs[snv] = pos_id\n",
    "\n",
    "\n",
    "\n",
    "    return mapped_snvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_vars(dict,keys, snv_map, annotations): #Annotates variants with Consequences and AA substitution\n",
    "\n",
    "    annotation_df = pd.read_excel(annotations) #Reads annotations file\n",
    "    annotated_dfs = {} #Diciotnary to hold annotated data frames\n",
    "    \n",
    "    for key in keys: #Annotates all dataframes\n",
    "        df = dict[key] #Gets one dataframe\n",
    "        df['canonical_start'] = df['Sequence'].transform(lambda x: x[15:18]) #Gets canonical start codon\n",
    "        df['second_start'] = df['Sequence'].transform(lambda x: x[90:93]) #Gets second start codon\n",
    "        df['snv_lib'] = df['Sequence'].transform(lambda x: x[88:95]) #Adds SNV lib sequence\n",
    "\n",
    "        df['pos_id'] = df['snv_lib'].transform(lambda x: snv_map[x]) #Adds pos_id\n",
    "        df['start_pos_id'] = df['canonical_start'] + ':' + df['pos_id'] #Adds combined column with canonical start codon and pos_id\n",
    "        df = pd.merge(df, annotation_df, how = 'left', on = 'pos_id') #Merges\n",
    "\n",
    "        df = df[['Count', 'freq', 'canonical_start', 'second_start', 'pos_id', 'pos', 'allele', 'AAsub', 'Consequence','start_pos_id']] #Gets important columns\n",
    "        df = df.loc[:,['pos', 'allele', 'pos_id', 'Consequence', 'AAsub', 'canonical_start', 'second_start', 'Count', 'start_pos_id', 'freq']] #Reorders columns \n",
    "\n",
    "        new_count_name = key + '_count' #Name for replicate count column\n",
    "        new_freq_name = key + '_freq' #Name for replicate frequency column\n",
    "\n",
    "        df = df.rename(columns = {'Count': new_count_name, 'freq': new_freq_name}) #Columns for counts and frequency renamed\n",
    "        \n",
    "        annotated_dfs[key] = df #Annotated df added to dictionary\n",
    "\n",
    "    return annotated_dfs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_annotated(dfs,keys):\n",
    "\n",
    "    file_names = {'RNA_D05_rep1': 'BARD1_X1A_RNA_R1R2R3_ATG_SGE049_counts.tsv',\n",
    "                  'RNA_D05_rep2': 'BARD1_X1A_RNA_R4R5R6_ATG_SGE049_counts.tsv',\n",
    "                  'RNA_D05_rep3': 'BARD1_X1A_RNA_R7R8R9_ATG_SGE049_counts.tsv',\n",
    "                  'D05_rep1': 'BARD1_X1A_ATG_SGE049_R1R2R3_D05_counts.tsv',\n",
    "                  'D05_rep2': 'BARD1_X1A_ATG_SGE049_R4R5R6_D05_counts.tsv',\n",
    "                  'D05_rep3': 'BARD1_X1A_ATG_SGE049_R7R8R9_D05_counts.tsv'\n",
    "                 }\n",
    "    \n",
    "    for elem in keys:\n",
    "        if 'RNA' in elem:\n",
    "            path = '../Data/ATG_lib_data/counts/for_rna/RNA_counts/'\n",
    "\n",
    "            full_name = path + file_names[elem] \n",
    "            df = dfs[elem]\n",
    "            count_name = elem + '_count'\n",
    "            df = df.rename(columns = {count_name: 'count'})\n",
    "\n",
    "            df.to_csv(full_name, sep = '\\t', index = False)\n",
    "            \n",
    "        else:\n",
    "            path = '../Data/ATG_lib_data/counts/for_rna/DNA_counts/'\n",
    "\n",
    "            full_name = path + file_names[elem]\n",
    "\n",
    "            df = dfs[elem]\n",
    "            count_name = elem + '_count'\n",
    "            df = df.rename(columns = {count_name: 'count'})\n",
    "\n",
    "            df.to_csv(full_name, sep = '\\t', index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    counts_dict, keys = read_counts(counts_path)\n",
    "    counts_dict = add_freq(counts_dict)\n",
    "    mapped_snvs = annotate_snv_lib()\n",
    "    annotated = annotate_vars(counts_dict, keys, mapped_snvs, x1a_annotations)\n",
    "    save_annotated(annotated, keys)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
