{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "from pathlib import Path\n",
    "import re\n",
    "from natsort import natsorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "read_depth_path = '../Data/depth_data/depth_files'\n",
    "gsp_input_file = '../Data/depth_data/deletion_inputs.xlsx'\n",
    "sge_file = '../Data/20250508_BARD1scores_update.tsv'\n",
    "target_coords = '../Data/SNV_filtering_inputs/20250415_BARD1_filter_entry.xlsx'\n",
    "cut_sites = '../Data/20241217_BARD1_sgRNA_cutsites.xlsx'\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_input(file, cut_sites): #Reads input file containing coordinates for all exons\n",
    "    \n",
    "    input_params = pd.read_excel(file) #Reads input file\n",
    "\n",
    "    #Loop that creates list of genomic coordinates for coding sequence\n",
    "    i = 0\n",
    "    cds_coords = [] #List to hold coding coordinates\n",
    "    while i < len(input_params):\n",
    "        start = input_params['start'][i] #Gets starting coordinate\n",
    "        end = input_params['end'][i] #Gets end coordinate\n",
    "\n",
    "        #Makes coding coordinates\n",
    "        for j in range(start, end + 1):\n",
    "            cds_coords.append(j)\n",
    "        \n",
    "        i += 1\n",
    "\n",
    "    cut_sites = pd.read_excel(cut_sites)\n",
    "    cut_sites.set_index('target', inplace = True)\n",
    "\n",
    "    return cds_coords, cut_sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_depth(depth_path, sge,coding_coords, target_coords, cut_df): #Processes all depth files in directory and annotates them\n",
    "    \n",
    "    file_path = Path(depth_path) #Creates path object for depth files \n",
    "\n",
    "    depth_files = sorted(list(file_path.glob('*.tsv'))) #Gets all depth files\n",
    "    \n",
    "    columns = ['region', 'offset', 'depth'] #Column names for renaming depth files\n",
    "    \n",
    "    target_coordinates = pd.read_excel(target_coords, sheet_name = 'targets') #Reads input with SGE target coordinates\n",
    "    target_coordinates.set_index('target', inplace = True) #Sets target name to index\n",
    "\n",
    "    all_dfs = [] #Empty list to hold processed dataframes\n",
    "\n",
    "    #For loop iterates through all depth files and processes them\n",
    "    for file in depth_files:\n",
    "        df = pd.read_csv(file, sep = '\\t') #Reads depth file\n",
    "        df = df.set_axis(columns, axis = 1) #Renames columns \n",
    "\n",
    "        min_depth = df['depth'].min() #Gets minimum read count in file\n",
    "        max_depth = df['depth'].max() #Gets maximum read count in file\n",
    "\n",
    "        df['normdepth'] = df['depth'] / max_depth #Calculates normalized depth based on proportion of maximum read counts \n",
    "\n",
    "        full_region = df['region'][0] #Gets full SGE target\n",
    "        region_start = int(re.findall(r':(\\d+)-', full_region)[0]) #Gets starting coordinate for sequencing amplicon \n",
    "\n",
    "        df['pos'] = region_start + df['offset'] - 1 #Generates genomic coordinates for all regions based on offset column and starting coordinate\n",
    "\n",
    "        file_str = str(file) #Sets file name to string data type\n",
    "        region_rep = re.findall(r'/([^/]+)_D13\\.depth\\.tsv$', file_str)[0] #Gets region and replicate string\n",
    "\n",
    "        region_rep_split = region_rep.split('_') #Splits string on '_'\n",
    "        target = region_rep_split[0] + '_' + region_rep_split[1] #Gets target name\n",
    "        exon_test = region_rep_split[1][1:-1] #Gets exon \n",
    "\n",
    "        region_start = target_coordinates.loc[target, 'end'] #Gets SGE target starting coordinate\n",
    "        region_end = target_coordinates.loc[target, 'start'] #Gets SGE target end coordinate (end/start flipped due to antisense gene)\n",
    "    \n",
    "        region_coords = [] #List to hold coordinates in SGE target\n",
    "\n",
    "        for k in range(region_start, region_end + 1): #Loop creates coordinates for SGE target\n",
    "            region_coords.append(k)\n",
    "\n",
    "        #Booleans to get name of exon\n",
    "        if len(exon_test) > 0: #tests for all regions but X2\n",
    "            exon = exon_test\n",
    "        else: #Exception for X2\n",
    "            exon = '2'\n",
    "            \n",
    "        full_rep = region_rep_split[2] #Gets replicate value\n",
    "\n",
    "        #Boolean tests to assign replicate number\n",
    "        if full_rep == 'R1R4' or full_rep == 'R1R2R3':\n",
    "            rep = 'R1'\n",
    "        elif full_rep == 'R2R5' or full_rep == 'R4R5R6':\n",
    "            rep = 'R2'\n",
    "        elif full_rep == 'R3R6' or full_rep == 'R7R8R9':\n",
    "            rep = 'R3'\n",
    "\n",
    "        cut_site = cut_df.loc[target, 'pos']\n",
    "        \n",
    "        #Sets columns with identifying information\n",
    "        df['target'] = target\n",
    "        df['exon'] = exon\n",
    "        df['repl'] = rep\n",
    "        df['day'] = 'D13'\n",
    "        \n",
    "        df = df.loc[df['pos'].isin(region_coords)] #Dataframe filtered for coordinates in SGE target edited region only\n",
    "\n",
    "        df['cut_site_distance'] = -(df['pos'] - cut_site)\n",
    "         \n",
    "        all_dfs.append(df) #Dataframe appended to list\n",
    "    \n",
    "    final_df = pd.concat(all_dfs) #All dataframes concatenated\n",
    "    final_df = final_df.loc[final_df['pos'].isin(coding_coords)] #Dataframes filtered for coding sequencing only \n",
    "\n",
    "\n",
    "    raw_sge = pd.read_csv(sge, sep = '\\t') #Reads SGE score file\n",
    "    raw_sge['amino_acid'] = raw_sge['amino_acid_change'].transform(lambda x: x[0:-1]) #Creates amino acid column \n",
    "    raw_sge = raw_sge.loc[~(raw_sge['amino_acid'].isin(['--']))] #Drops columns without amino acid \n",
    "    annotation_df = raw_sge[['pos', 'amino_acid']] #Keeps position column for mergining and amino acid column for annotation \n",
    "\n",
    "    final_df = pd.merge(final_df, annotation_df, on = 'pos', how = 'left') #Depth and annotation_df merged to annotate with amino acids\n",
    "   \n",
    "    \n",
    "\n",
    "    final_df['id'] = final_df['pos']  + final_df['depth']  + final_df['normdepth'] #Unique ID created for each datapoint \n",
    "\n",
    "    final_df = final_df.drop_duplicates(subset = 'id', keep = 'first') #Any duplicates dropped\n",
    "    final_df = final_df.drop(columns = ['id']) #ID column dropped \n",
    "\n",
    "    return final_df    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_read_depth(df): #Depth dataframe processed for visualization\n",
    "\n",
    "    df['pos'] = df['pos'].astype(str) #Sets 'pos' column to string datatype\n",
    "    df['target_id'] = df['target'] + ':' + df['pos'] #Builds a target ID column for target-based collapsing to median\n",
    "    \n",
    "    grouped = df.groupby('target_id') #Groups dataframe by target ID\n",
    "\n",
    "    #Creates dataframe with annotated CDS positions\n",
    "    cds_annotated = df.groupby('pos').agg({ #Grouping by position allows for accurate CDS pos. to be assigned\n",
    "    'normdepth': 'median', \n",
    "    'target': 'first',\n",
    "    'exon': 'first',\n",
    "    'amino_acid': 'last'\n",
    "    }).reset_index()\n",
    "\n",
    "    \n",
    "    cds_pos = [] #List to hold CDS position values\n",
    "\n",
    "    for i in range(len(cds_annotated)): #Builds CDS position values\n",
    "        cds_pos.append(i+1)\n",
    "\n",
    "    cds_pos = cds_pos[::-1] #Reverses values due to negative sense gene\n",
    "\n",
    "    \n",
    "    \n",
    "    cds_annotated['CDSpos'] = cds_pos #Adds CDS position column\n",
    "    cds_annotated = cds_annotated[['pos', 'CDSpos']] #Drops unncessary column s\n",
    "\n",
    "    #cds_annotated.to_excel('/Users/ivan/Desktop/BARD1_GenomicCoords_wCDS.xlsx', index = False)\n",
    "    \n",
    "    #Collapses depth calculations to median based on shared target and position for all 3 replicates\n",
    "    median_depth = grouped.agg({\n",
    "        'normdepth': 'median',\n",
    "        'pos': 'first',\n",
    "        'target': 'first',\n",
    "        'exon': 'first',\n",
    "        'amino_acid': 'first',\n",
    "        'cut_site_distance': 'first'\n",
    "    }\n",
    "                              )\n",
    "\n",
    "    median_depth = pd.merge(median_depth, cds_annotated, on = 'pos', how = 'left') #Merges with CDS annotated dataframe with add CDS position\n",
    "    median_depth = median_depth.rename(columns = {'normdepth': 'median_depth'}) #depth column renamed \n",
    "    \n",
    "    median_depth['AApos'] = median_depth['amino_acid'].str[1:]\n",
    "    #median_depth.to_excel('/Users/ivan/Desktop/test_excel_outputs/20250806_DepthOutput.xlsx', index = None)\n",
    "    print(median_depth)\n",
    "\n",
    "    #cutsite_depth = df.groupby(\n",
    "    return median_depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def graph_rep_depth(df,median_depth): #Generates line plots pre-median faceted by replicate\n",
    "\n",
    "    median_depth = median_depth[['pos', 'CDSpos']]\n",
    "    df = pd.merge(df, median_depth, on = 'pos', how = 'left')\n",
    "    \n",
    "    \n",
    "    sorted = natsorted(set(df['target'].tolist()))\n",
    "    plot =  alt.Chart(df).mark_line(point = True).encode(\n",
    "        x = alt.X('CDSpos',\n",
    "                  axis = alt.Axis(title = 'CDS Position'),\n",
    "                  scale = alt.Scale(domain = [-50, 2384]\n",
    "                                   )\n",
    "                 ),\n",
    "        y = alt.Y('normdepth', \n",
    "                  axis = alt.Axis(title = 'Normalized Depth'),\n",
    "                  scale = alt.Scale(domain = [0, 1.03]\n",
    "                                   )\n",
    "                 ),\n",
    "        color = alt.Color('target', sort = sorted, legend = alt.Legend(title = 'SGE Target',\n",
    "                                                                      symbolLimit = 0)),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('amino_acid', title = 'Amino Acid: '),\n",
    "                   alt.Tooltip('normdepth', title = 'Normalized Depth: ')]\n",
    "    ).properties(\n",
    "        width = 1000, \n",
    "        height = 400\n",
    "    ).interactive()\n",
    "    \n",
    "    plot = plot.facet('repl', columns = 1)\n",
    "    plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def graph_median_depth(df, cut_coords): #Generates final line plot based on median collapsed depth\n",
    "\n",
    "    df['pos'] = df['pos'].astype(int)\n",
    "    annotate_cutsite_df = df[['pos', 'CDSpos']]\n",
    "\n",
    "    cds_cut_df = pd.merge(cut_coords, annotate_cutsite_df, on = 'pos', how = 'inner')\n",
    "    cds_cut_df = cds_cut_df.drop_duplicates()\n",
    "    print(len(cds_cut_df))\n",
    "    sorted = natsorted(set(df['target'].tolist()))\n",
    "\n",
    "    tableau20 = ['#4E79A7', '#F28E2C', '#E15759', '#76B7B2', '#59A14F', \n",
    "             '#EDC949', '#AF7AA1', '#FF9DA7', '#9C755F', '#BAB0AB',\n",
    "             '#4E79A7', '#F28E2C', '#E15759', '#76B7B2', '#59A14F',\n",
    "             '#EDC949', '#AF7AA1', '#FF9DA7', '#9C755F', '#BAB0AB']\n",
    "\n",
    "    category20 = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd',\n",
    "                  '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf',\n",
    "                  '#aec7e8', '#ffbb78', '#98df8a', '#ff9896', '#c5b0d5',\n",
    "                  '#c49c94', '#f7b6d2', '#c7c7c7', '#dbdb8d', '#9edae5']\n",
    "    \n",
    "    # Combine them\n",
    "    combined_colors = tableau20 + category20  # 40 colors total\n",
    "    \n",
    "    # Or remove similar colors\n",
    "    plot_domain = [-25, 2359]\n",
    "    combined_colors = list(dict.fromkeys(tableau20 + category20))\n",
    "\n",
    "    #Builds median read depth plot\n",
    "    plot =  alt.Chart(df).mark_line(point = True).encode(\n",
    "        x = alt.X('CDSpos',\n",
    "                  axis = alt.Axis(title = 'CDS Position',\n",
    "                                  labelFontSize = 16,\n",
    "                                  titleFontSize = 20\n",
    "                                 ),\n",
    "                  scale = alt.Scale(domain = plot_domain\n",
    "                                   )\n",
    "                 ),\n",
    "        y = alt.Y('median_depth', \n",
    "                  axis = alt.Axis(title = 'Median Normalized Depth',\n",
    "                                  values = [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0],\n",
    "                                 labelFontSize = 16,\n",
    "                                 titleFontSize = 20\n",
    "                                 ),\n",
    "                  scale = alt.Scale(domain = [0.05, 1.15]\n",
    "                                   )\n",
    "                 ),\n",
    "        color = alt.Color('target', \n",
    "                          sort = sorted, \n",
    "                          scale = alt.Scale(range = combined_colors[:34]),\n",
    "                          legend = None\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('CDSpos', title = 'CDS Position: '),\n",
    "                   alt.Tooltip('pos', title = 'Genomic Coordinate: '),\n",
    "                   alt.Tooltip('amino_acid', title = 'Amino Acid: '),\n",
    "                   alt.Tooltip('median_depth', title = 'Median Normalized Depth: ')]\n",
    "    ).properties(\n",
    "        width = 1700, \n",
    "        height = 300,\n",
    "        title = alt.TitleParams(text = 'Median Read Depth by Target',\n",
    "                                fontSize = 22\n",
    "                               )\n",
    "    ).interactive()\n",
    "\n",
    "    #Adds ticks to mark cutsites\n",
    "    ticks = alt.Chart(cds_cut_df).mark_tick(\n",
    "        color='red',\n",
    "        thickness = 2,\n",
    "        size = 20,\n",
    "        dy = 10\n",
    "    ).encode(\n",
    "        x= alt.X('CDSpos:Q',\n",
    "                 scale = alt.Scale(domain = plot_domain\n",
    "                                  )\n",
    "                ),# Your column name\n",
    "        y=alt.datum(0)\n",
    "    )\n",
    "\n",
    "    #Adds rectangles to specify BARD1 domains\n",
    "    y_max = 1\n",
    "    rectangles = alt.Chart(pd.DataFrame({\n",
    "        'x': [76, 1273, 1696],      # Start x-position of each rectangle\n",
    "        'x2': [363, 1636, 2331],     # End x-position of each rectangle  \n",
    "        'y': [y_max * 1.05, y_max * 1.05, y_max * 1.05],    # Bottom of rectangles (5% above max)\n",
    "        'y2': [y_max * 1.15, y_max * 1.15, y_max * 1.15],# Top of rectangles (15% above max)\n",
    "        'color': ['#B9DBF4','#C8DBC8', '#F6BF93'] \n",
    "    })).mark_rect(\n",
    "        opacity=0.7,\n",
    "        stroke = 'black', \n",
    "        strokeWidth = 2\n",
    "    ).encode(\n",
    "        x= 'x:Q',\n",
    "        x2='x2:Q',\n",
    "        y='y:Q',\n",
    "        y2='y2:Q',\n",
    "        color = alt.Color('color:N', scale = None)\n",
    "    )\n",
    "\n",
    "    plot = alt.layer(plot, ticks, rectangles).encode(\n",
    "        x = alt.X(scale = alt.Scale(domain = plot_domain)\n",
    "                 )\n",
    "    ).configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "\n",
    "    #plot.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_2a_dels.png', ppi = 500)\n",
    "    plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def graph_aa_pos_median_depth(df, cut_coords): #Generates final line plot based on median collapsed depth\n",
    "\n",
    "    df = df.dropna(subset = ['AApos'])\n",
    "    df = df.copy()\n",
    "    df['pos'] = df['pos'].astype(int)\n",
    "    df['AApos'] = df['AApos'].astype(int)\n",
    "    \n",
    "    annotate_cutsite_df = df[['pos', 'AApos']]\n",
    "\n",
    "    cds_cut_df = pd.merge(cut_coords, annotate_cutsite_df, on = 'pos', how = 'inner')\n",
    "    cds_cut_df = cds_cut_df.drop_duplicates()\n",
    "\n",
    "    sorted = natsorted(set(df['target'].tolist()))\n",
    "\n",
    "    tableau20 = ['#4E79A7', '#F28E2C', '#E15759', '#76B7B2', '#59A14F', \n",
    "             '#EDC949', '#AF7AA1', '#FF9DA7', '#9C755F', '#BAB0AB',\n",
    "             '#4E79A7', '#F28E2C', '#E15759', '#76B7B2', '#59A14F',\n",
    "             '#EDC949', '#AF7AA1', '#FF9DA7', '#9C755F', '#BAB0AB']\n",
    "\n",
    "    category20 = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd',\n",
    "                  '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf',\n",
    "                  '#aec7e8', '#ffbb78', '#98df8a', '#ff9896', '#c5b0d5',\n",
    "                  '#c49c94', '#f7b6d2', '#c7c7c7', '#dbdb8d', '#9edae5']\n",
    "    \n",
    "    # Combine them\n",
    "    combined_colors = tableau20 + category20  # 40 colors total\n",
    "    \n",
    "    # Or remove similar colors\n",
    "    plot_domain = [-10, 787]\n",
    "    combined_colors = list(dict.fromkeys(tableau20 + category20))\n",
    "\n",
    "    #Builds median read depth plot\n",
    "    plot =  alt.Chart(df).mark_line(point = True).encode(\n",
    "        x = alt.X('AApos',\n",
    "                  axis = alt.Axis(title = 'Amino Acid Position',\n",
    "                                  labelFontSize = 16,\n",
    "                                  titleFontSize = 20\n",
    "                                 ),\n",
    "                  scale = alt.Scale(domain = plot_domain\n",
    "                                   )\n",
    "                 ),\n",
    "        y = alt.Y('median_depth', \n",
    "                  axis = alt.Axis(title = 'Median Normalized Depth',\n",
    "                                  values = [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0],\n",
    "                                 labelFontSize = 16,\n",
    "                                 titleFontSize = 20\n",
    "                                 ),\n",
    "                  scale = alt.Scale(domain = [0.05, 1.15]\n",
    "                                   )\n",
    "                 ),\n",
    "        color = alt.Color('target', \n",
    "                          sort = sorted, \n",
    "                          scale = alt.Scale(range = combined_colors[:34]),\n",
    "                          legend = None\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('CDSpos', title = 'CDS Position: '),\n",
    "                   alt.Tooltip('pos', title = 'Genomic Coordinate: '),\n",
    "                   alt.Tooltip('amino_acid', title = 'Amino Acid: '),\n",
    "                   alt.Tooltip('median_depth', title = 'Median Normalized Depth: ')]\n",
    "    ).properties(\n",
    "        width = 1700, \n",
    "        height = 300,\n",
    "        title = alt.TitleParams(text = 'Median Read Depth by Target',\n",
    "                                fontSize = 22\n",
    "                               )\n",
    "    ).interactive()\n",
    "\n",
    "\n",
    "\n",
    "    #Adds ticks to mark cutsites\n",
    "    ticks = alt.Chart(cds_cut_df).mark_tick(\n",
    "        color='red',\n",
    "        thickness = 2,\n",
    "        size = 20,\n",
    "        dy = 10\n",
    "    ).encode(\n",
    "        x= alt.X('AApos:Q',\n",
    "                 scale = alt.Scale(domain = plot_domain\n",
    "                                  )\n",
    "                ),# Your column name\n",
    "        y=alt.datum(0)\n",
    "    )\n",
    "\n",
    "    #Adds rectangles to specify BARD1 domains\n",
    "    y_max = 1\n",
    "    rectangles = alt.Chart(pd.DataFrame({\n",
    "        'x': [26, 425, 568],      # Start x-position of each rectangle\n",
    "        'x2': [122, 545, 777],     # End x-position of each rectangle  \n",
    "        'y': [y_max * 1.05, y_max * 1.05, y_max * 1.05],    # Bottom of rectangles (5% above max)\n",
    "        'y2': [y_max * 1.15, y_max * 1.15, y_max * 1.15],# Top of rectangles (15% above max)\n",
    "        'color': ['#B9DBF4','#C8DBC8', '#F6BF93'] \n",
    "    })).mark_rect(\n",
    "        opacity=0.7,\n",
    "        stroke = 'black', \n",
    "        strokeWidth = 2\n",
    "    ).encode(\n",
    "        x= 'x:Q',\n",
    "        x2='x2:Q',\n",
    "        y='y:Q',\n",
    "        y2='y2:Q',\n",
    "        color = alt.Color('color:N', scale = None)\n",
    "    )\n",
    "\n",
    "    plot = alt.layer(plot, ticks, rectangles).encode(\n",
    "        x = alt.X(scale = alt.Scale(domain = plot_domain)\n",
    "                 )\n",
    "    ).configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "\n",
    "\n",
    "    \n",
    "    #plot.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_2a_dels.png', ppi = 500)\n",
    "    plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def graph_aapos_alphafold_rects(df, cut_coords): #Generates final line plot based on median collapsed depth\n",
    "\n",
    "    df = df.dropna(subset = ['AApos'])\n",
    "    df = df.copy()\n",
    "    df['pos'] = df['pos'].astype(int)\n",
    "    df['AApos'] = df['AApos'].astype(int)\n",
    "    \n",
    "    annotate_cutsite_df = df[['pos', 'AApos']]\n",
    "\n",
    "    #cds_cut_df = pd.merge(cut_coords, annotate_cutsite_df, on = 'pos', how = 'left')\n",
    "    cds_cut_df = cut_coords\n",
    "    cds_cut_df = cds_cut_df.drop_duplicates()\n",
    "\n",
    "    sorted = natsorted(set(df['target'].tolist()))\n",
    "\n",
    "    tableau20 = ['#4E79A7', '#F28E2C', '#E15759', '#76B7B2', '#59A14F', \n",
    "             '#EDC949', '#AF7AA1', '#FF9DA7', '#9C755F', '#BAB0AB',\n",
    "             '#4E79A7', '#F28E2C', '#E15759', '#76B7B2', '#59A14F',\n",
    "             '#EDC949', '#AF7AA1', '#FF9DA7', '#9C755F', '#BAB0AB']\n",
    "\n",
    "    category20 = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd',\n",
    "                  '#8c564b', '#e377c2', '#7f7f7f', '#bcbd22', '#17becf',\n",
    "                  '#aec7e8', '#ffbb78', '#98df8a', '#ff9896', '#c5b0d5',\n",
    "                  '#c49c94', '#f7b6d2', '#c7c7c7', '#dbdb8d', '#9edae5']\n",
    "    \n",
    "    # Combine them\n",
    "    combined_colors = tableau20 + category20  # 40 colors total\n",
    "    \n",
    "    # Or remove similar colors\n",
    "    plot_domain = [-10, 787]\n",
    "    combined_colors = list(dict.fromkeys(tableau20 + category20))\n",
    "\n",
    "    #Builds median read depth plot\n",
    "    plot =  alt.Chart(df).mark_line(point = True).encode(\n",
    "        x = alt.X('AApos',\n",
    "                  axis = alt.Axis(title = 'Amino Acid Position',\n",
    "                                  labelFontSize = 16,\n",
    "                                  titleFontSize = 20\n",
    "                                 ),\n",
    "                  scale = alt.Scale(domain = plot_domain\n",
    "                                   )\n",
    "                 ),\n",
    "        y = alt.Y('median_depth', \n",
    "                  axis = alt.Axis(title = 'Median Normalized Depth',\n",
    "                                  values = [0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0],\n",
    "                                 labelFontSize = 16,\n",
    "                                 titleFontSize = 20\n",
    "                                 ),\n",
    "                  scale = alt.Scale(domain = [0.05, 1.4]\n",
    "                                   )\n",
    "                 ),\n",
    "        color = alt.Color('target', \n",
    "                          sort = sorted, \n",
    "                          scale = alt.Scale(range = combined_colors[:34]),\n",
    "                          legend = None\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('CDSpos', title = 'CDS Position: '),\n",
    "                   alt.Tooltip('pos', title = 'Genomic Coordinate: '),\n",
    "                   alt.Tooltip('amino_acid', title = 'Amino Acid: '),\n",
    "                   alt.Tooltip('median_depth', title = 'Median Normalized Depth: ')]\n",
    "    ).properties(\n",
    "        width = 1700, \n",
    "        height = 300,\n",
    "        title = alt.TitleParams(text = 'Median Read Depth by Target',\n",
    "                                fontSize = 22\n",
    "                               )\n",
    "    ).interactive()\n",
    "\n",
    "\n",
    "\n",
    "    #Adds ticks to mark cutsites\n",
    "    ticks = alt.Chart(cds_cut_df).mark_tick(\n",
    "        color='red',\n",
    "        thickness = 2,\n",
    "        size = 20,\n",
    "        dy = 10\n",
    "    ).encode(\n",
    "        x= alt.X('AApos:Q',\n",
    "                 scale = alt.Scale(domain = plot_domain\n",
    "                                  )\n",
    "                ),# Your column name\n",
    "        y=alt.datum(0)\n",
    "    )\n",
    "\n",
    "    #Adds rectangles to specify BARD1 domains\n",
    "    y_max = 1\n",
    "\n",
    "    rect_data = pd.DataFrame({\n",
    "        'x': [1, 26, 120,141, 165, 210,220, 421, 548, 565, 26, 425, 568],      # Start x-position of each rectangle\n",
    "        'x2':[26, 120, 141,165, 210, 220,421, 548, 565, 777, 120, 545, 777],     # End x-position of each rectangle  \n",
    "        'y': [y_max * 1.05, y_max * 1.05, y_max * 1.05, y_max * 1.05, y_max * 1.05, y_max * 1.05, y_max * 1.05, y_max * 1.05, y_max * 1.05, y_max * 1.05, y_max * 1.2, y_max * 1.2, y_max * 1.2],    # Bottom of rectangles (5% above max)\n",
    "        'y2': [y_max * 1.15, y_max * 1.15, y_max * 1.15, y_max * 1.15, y_max * 1.15, y_max * 1.15, y_max * 1.15, y_max * 1.15, y_max * 1.15, y_max * 1.15, y_max * 1.35, y_max * 1.35, y_max * 1.35],# Top of rectangles (15% above max)\n",
    "        'color': ['#b1b1b1', '#DEDBEE', '#b1b1b1', '#DEDBEE', '#b1b1b1', '#DEDBEE', '#b1b1b1', '#DEDBEE', '#b1b1b1', '#DEDBEE', '#B9DBF4', '#C8DBC8', '#F6BF93'],\n",
    "        'label': ['', 'Structured', '', '', '', '', 'Disordered', '', '', '', 'RING', 'ARD', 'BRCT']\n",
    "    })\n",
    "    \n",
    "    rectangles = alt.Chart(rect_data).mark_rect(\n",
    "        opacity=0.7,\n",
    "        stroke = 'black', \n",
    "        strokeWidth = 2\n",
    "    ).encode(\n",
    "        x= 'x:Q',\n",
    "        x2='x2:Q',\n",
    "        y='y:Q',\n",
    "        y2='y2:Q',\n",
    "        color = alt.Color('color:N', scale = None)\n",
    "    )\n",
    "\n",
    "    text = alt.Chart(rect_data).mark_text(\n",
    "        align='center',\n",
    "        baseline='middle',\n",
    "        fontWeight = 'bold',\n",
    "        fontSize=24,\n",
    "        angle=0,  # or 90/-90 for vertical text\n",
    "        color='black',\n",
    "        limit=1000  # truncate long text\n",
    "    ).encode(\n",
    "        x=alt.X('x_center:Q'),\n",
    "        y=alt.Y('y_center:Q'),\n",
    "        text='label:N'\n",
    "    ).transform_calculate(\n",
    "        x_center='(datum.x + datum.x2) / 2',\n",
    "        y_center='(datum.y + datum.y2) / 2'\n",
    "    )\n",
    "\n",
    "    plot = alt.layer(plot, ticks, rectangles,text).encode(\n",
    "        x = alt.X(scale = alt.Scale(domain = plot_domain)\n",
    "                 )\n",
    "    ).configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "\n",
    "\n",
    "    \n",
    "    #plot.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_2a_dels.png', ppi = 500)\n",
    "    plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def graph_cutsite_depth(df):\n",
    "\n",
    "    sorted = natsorted(set(df['target'].tolist()))\n",
    "    plot =  alt.Chart(df).mark_line(point = True).encode(\n",
    "        x = alt.X('cut_site_distance',\n",
    "                  axis = alt.Axis(title = 'Distance to Cut Site')\n",
    "                 ),\n",
    "        y = alt.Y('median_depth', \n",
    "                  axis = alt.Axis(title = 'Median Normalized Depth'),\n",
    "                  scale = alt.Scale(domain = [0, 1.03]\n",
    "                                   )\n",
    "                 ),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('CDSpos', title = 'CDS Position: '),\n",
    "                   alt.Tooltip('pos', title = 'Genomic Coordinate: '),\n",
    "                   alt.Tooltip('amino_acid', title = 'Amino Acid: '),\n",
    "                   alt.Tooltip('median_depth', title = 'Median Normalized Depth: ')]\n",
    "    ).properties(\n",
    "        width = 600, \n",
    "        height = 400,\n",
    "        title = 'Median Read Depth by Target'\n",
    "    ).interactive()\n",
    "\n",
    "    plot = plot.facet(\n",
    "        facet = alt.Facet('target:N',\n",
    "                          sort = sorted\n",
    "                         ),\n",
    "        columns = 2\n",
    "    ).resolve_scale(\n",
    "        x = 'independent',\n",
    "        y = 'independent'\n",
    "    )\n",
    "    plot.display()\n",
    "\n",
    "    '''\n",
    "    cut_site_collapsed = df.groupby('cut_site_distance').agg({\n",
    "        'median_depth': 'median',\n",
    "        'cut_site_distance': 'first'\n",
    "    }\n",
    "                                                            )\n",
    "    print(cut_site_collapsed)\n",
    "    collapsed_plot =  alt.Chart(df).mark_line(point = True).encode(\n",
    "        x = alt.X('cut_site_distance',\n",
    "                  axis = alt.Axis(title = 'Distance to Cut Site')\n",
    "                 ),\n",
    "        y = alt.Y('median_depth', \n",
    "                  axis = alt.Axis(title = 'Median Normalized Depth'),\n",
    "                  scale = alt.Scale(domain = [0, 1.03]\n",
    "                                   )\n",
    "                 )\n",
    "    ).properties(\n",
    "        width = 600, \n",
    "        height = 400,\n",
    "        title = 'Median Read Depth by Target'\n",
    "    ).interactive()\n",
    "\n",
    "    collapsed_plot.display()\n",
    "    '''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    coding_coords, cut_coords = read_input(gsp_input_file, cut_sites)\n",
    "    all_reps_depth = process_depth(read_depth_path, sge_file, coding_coords, target_coords, cut_coords)\n",
    "    collapsed_depth = process_read_depth(all_reps_depth)\n",
    "    graph_rep_depth(all_reps_depth, collapsed_depth)\n",
    "    graph_median_depth(collapsed_depth, cut_coords)\n",
    "    graph_aa_pos_median_depth(collapsed_depth, cut_coords)\n",
    "    graph_aapos_alphafold_rects(collapsed_depth, cut_coords)\n",
    "    #collapsed_depth.to_excel('/Users/ivan/Desktop/test_excel_outputs/20250717_BARD1_CollapsedMedianDepth.xlsx', index = False)\n",
    "    graph_cutsite_depth(collapsed_depth)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
