{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import numpy as np\n",
    "import statistics\n",
    "from scipy import stats\n",
    "from pathlib import Path\n",
    "from natsort import natsorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#For all regions excluding X1A ATG double mutant library\n",
    "rna_count_folder = '../Data/RNAscoring_dev_data/RNA_counts/' #path to RNA counts folder\n",
    "dna_count_folder = '../Data/RNAscoring_dev_data/DNA_counts/' #paths to DNA counts folder\n",
    "sge_score = '../Data/20250813_BARD1scores_final_FILTERED.xlsx' #most up to date scores\n",
    "sge_threshold_file = '../Data/20250813_BARD1_thresholds.tsv' #SGE Thresholds file\n",
    "cds_pos = '../Data/BARD1_GenomicCoords_wCDS.xlsx' #CDS annotated BARD1 coordinates\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_counts(rna_counts, dna_counts, sge_file, hreshold_file): #reads RNA/DNA count files and merges them. Also reads SGE scores\n",
    "\n",
    "    dna_scores = pd.read_excel(sge_file) #Reads DNA SGE Scores\n",
    "\n",
    "    threshold_df = pd.read_csv(threshold_file, sep = '\\t')\n",
    "\n",
    "    lwr = threshold_df['lthresh'][0] * threshold_df['std_neut'][0] + threshold_df['#mu_neut'][0]\n",
    "    uppr = threshold_df['uthresh'][0] * threshold_df['std_neut'][0] + threshold_df['#mu_neut'][0]\n",
    "\n",
    "    sge_thresholds = [lwr, uppr]\n",
    "    \n",
    "    #dna_scores = pd.read_csv(sge_file, sep = '\\t')\n",
    "    dna_scores = dna_scores.rename(columns = {'consequence': 'Consequence',   'score': 'snv_score', 'amino_acid_change': 'AAsub'})\n",
    "    dna_scores['pos'] = dna_scores['pos'].astype(str)\n",
    "    #dna_scores['pos_id'] = dna_scores['pos'] + ':' + dna_scores['alt']\n",
    "\n",
    "    \n",
    "    rna_path = Path(rna_counts) #Creates path object to RNA counts\n",
    "    dna_path = Path(dna_counts) #Creates path object to DNA counts\n",
    "\n",
    "    rna_outputs = sorted(list(rna_path.glob('*.tsv'))) #Gets list of RNA count files. Sort function added to order replicates\n",
    "    dna_outputs = sorted(list(dna_path.glob('*.tsv'))) #Gets list of DNA count files. Sort function added to order replicates\n",
    "\n",
    "    all_outputs = [] #list to hold tuples of DNA/RNA count files paired by replicate\n",
    "\n",
    "    i = 0\n",
    "    while i < len(rna_outputs): #Iterates through RNA count files and constructs the paired DNA/RNA tuples\n",
    "        \n",
    "        paired_output = (rna_outputs[i], dna_outputs[i]) #Paired RNA/DNA files\n",
    "        all_outputs.append(paired_output) #Appends ot list\n",
    "\n",
    "        i += 1\n",
    "\n",
    "    all_rep_counts = {} #Dictionary to hold merged RNA/DNA count dataframe\n",
    "    \n",
    "    for elem in all_outputs: #Iterates through list of paired RNA/DNA count files\n",
    "        rna, dna = elem #Breaks up tuple into constituent RNA/DNA count file paths\n",
    "        rna_counts = pd.read_csv(rna, sep = '\\t') #Reads RNA count file\n",
    "        sampleid = rna_counts['sampleid'][0] #Gets sample ID from first row\n",
    "\n",
    "        #Booleans to get replicate number from sample ID\n",
    "        if 'R3R6' in sampleid or 'R7R8R9' in sampleid:\n",
    "            rep = 'rep_3'\n",
    "        elif 'R2R5'in sampleid or 'R4R5R6' in sampleid:\n",
    "            rep = 'rep_2'\n",
    "        elif 'R1R4' in sampleid or 'R1R2R3' in sampleid:\n",
    "            rep = 'rep_1'\n",
    "\n",
    "        #Creates a column that has target and replicate number\n",
    "        target_rep_name = rna_counts['target'] + rep \n",
    "        rna_counts['target_rep'] = target_rep_name \n",
    "        rna_counts = rna_counts.rename(columns = {'count': 'RNAcount'}) #Renames count column in RNA file\n",
    "        rna_counts['pos'] = rna_counts['pos'].astype(str) #Sets 'pos' column to string data type for merging\n",
    "        rna_counts['pos_id'] = rna_counts['pos'] + ':' + rna_counts['allele'] #Creates position ID that is genomic coordinate : allele for merging\n",
    "\n",
    "        \n",
    "        dna_counts = pd.read_csv(dna, sep = '\\t') #Reads DNA file\n",
    "        dna_counts = dna_counts.rename(columns = {'count': 'DNAcount'}) #Renames count column in DNA file\n",
    "        dna_counts['pos'] = dna_counts['pos'].astype(str) #Sets 'pos' column to string data type for merging\n",
    "        dna_counts['pos_id'] = dna_counts['pos'] + ':' + dna_counts['allele'] #Creates position ID column for merging\n",
    "         \n",
    "        merged = pd.merge(rna_counts, dna_counts, how = 'inner', on = 'pos_id') #Merges RNA and DNA dataframes on intersection of pos_id column \n",
    "        merged = merged.drop(columns = ['sampleid_x', 'sampleid_y', 'target_x','chrom_x', 'chrom_y', 'pos_x', 'pos_y', 'allele_x', 'allele_y']) #drops duplicate columns \n",
    "        merged = merged.rename(columns  = {'target_y' : 'target'}) #Renames target column to keep exon SGE target name\n",
    "        merged = merged.loc[:, ['target_rep','target', 'pos_id', 'RNAcount', 'DNAcount']] #Columns reordered\n",
    "        \n",
    "        all_rep_counts[target_rep_name[0]] = merged #merged dataframe added to dictionary\n",
    "\n",
    "    return all_rep_counts, dna_scores, sge_thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def rna_v_dna_plot(all_counts): #Generates scatter plots of RNA count vs. DNA count\n",
    "    \n",
    "    rep_regions = list(all_counts.keys()) #Gets list of all replicates and regions\n",
    "\n",
    "    df_list = [] #List to hold data frames\n",
    "    for elem in rep_regions: #Loop gets all dataframes\n",
    "        df_list.append(all_counts[elem])\n",
    "\n",
    "\n",
    "    all_df = pd.concat(df_list) #All dataframes concatenated \n",
    "\n",
    "    all_df = all_df[~((all_df['RNAcount'] < 50) | (all_df['DNAcount'] < 50))] #Data filtered against very low read variants\n",
    "\n",
    "    #Builds RNA count vs. DNA count scatterplot\n",
    "    scatter = alt.Chart(all_df).mark_circle().encode(\n",
    "        x = alt.X('RNAcount', title = 'RNA Count'),\n",
    "        y = alt.Y('DNAcount', title = 'DNA Count'),\n",
    "        tooltip = [alt.Tooltip('RNAcount', title = 'RNA Count: '),\n",
    "                   alt.Tooltip('DNAcount', title = 'DNA Count: '),\n",
    "                   alt.Tooltip('pos_id', title = 'Var: ')\n",
    "                  ]\n",
    "    )\n",
    "\n",
    "    #Builds trendline\n",
    "    trend_line = alt.Chart(all_df).transform_regression(\n",
    "        'RNAcount', 'DNAcount',\n",
    "        groupby=['target_rep']\n",
    "    ).mark_line(color='blue').encode(\n",
    "        x='RNAcount',\n",
    "        y='DNAcount'\n",
    "    )\n",
    "\n",
    "\n",
    "    #Combines scatterplot and trendline\n",
    "    combined = (scatter + trend_line).facet(\n",
    "        facet=alt.Facet('target_rep', title = 'DNA vs. RNA Counts')  # Set maximum of 4 columns per row\n",
    "        ).resolve_scale(\n",
    "            x = 'independent',\n",
    "            y = 'independent'\n",
    "        ).properties(\n",
    "            columns = 6\n",
    "        ).interactive()\n",
    "\n",
    "    combined.display()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rna_scores(rep_dict): #Handles RNA scoring of variants (note, as of 4/25/2025, RNA scores generated are no longer used downstream, but RNA/DNA counts/frequencies are)\n",
    "\n",
    "    target_reps = list(rep_dict.keys()) #Gets list of all targets and replicates present in dictionary\n",
    "    rna_scored = {} #Diciotnary to hold RNA-scored dataframes\n",
    "    targets = [] #list of SGE targets present\n",
    "    \n",
    "    #Does RNA scoring\n",
    "    for elem in target_reps:\n",
    "        df = rep_dict[elem] #Gets dataframe\n",
    "        df = df[~((df['RNAcount'] == 0) | (df['DNAcount'] == 0))].copy() #Drops rows with RNA or DNA count of 0\n",
    "        df = df[~((df['RNAcount'] < 25) | (df['DNAcount'] < 25))].copy() #Drops rows with RNA or DNA count less than 25\n",
    "\n",
    "        df = df.reset_index(drop = True) #Resets index\n",
    "         \n",
    "        DNAcount = df['DNAcount'].sum() #Gets total DNA counts\n",
    "        RNAcount = df['RNAcount'].sum() #Gets total RNA counts\n",
    "\n",
    "        df['DNAfreq'] = df['DNAcount'] / DNAcount #Calculates DNA frequency\n",
    "        df['RNAfreq'] = df['RNAcount'] / RNAcount #Calculates RNA frequency\n",
    "\n",
    "        slope, intercept, r_value, p_value, std_err = stats.linregress(df['RNAfreq'], df['DNAfreq'])\n",
    "        \n",
    "        df['RNA/DNA'] = df['RNAfreq'] / df['DNAfreq'] #Calculates ratio of RNA to DNA frequency for each replicate\n",
    "\n",
    "        rna_scored[df['target_rep'][0]] = df #RNA-scored dataframe is added to rna_scored dataframe\n",
    "\n",
    "        if df['target'][0] not in targets: #Appends SGE target names\n",
    "            targets.append(df['target'][0])\n",
    "        \n",
    "\n",
    "    return rna_scored, targets\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rna_dna_freq_corr_plots(dict, sge_scores): #Builds scatterplots of RNA frequency vs. DNA frequency for all 3 replicates for an SGE target\n",
    "    \n",
    "    all_rna = list(dict.values()) #Gets all dataframes\n",
    "    df = pd.concat(all_rna) #Complete dataframe\n",
    "\n",
    "    df = pd.merge(df, sge_scores, on = 'pos_id', how = 'left') #Merges with SGE scores\n",
    "    df = df.rename(columns = {'target_x': 'target'})#Columns renamed\n",
    "    df = df.rename(columns = {'simplified_consequence': 'Consequence'}) #Columns renamed - may be removed in future input data release\n",
    "    df = df[['target_rep', 'target', 'pos_id', 'RNAfreq', 'DNAfreq', 'RNA/DNA', 'AAsub', 'Consequence', 'snv_score', 'functional_consequence']] #Columsn pulled for downstream functions\n",
    "    df = df.loc[df['Consequence'].isin(['missense_variant', 'synonymous_variant', 'stop_gained', 'splicing_variant', 'stop_lost', 'start_lost', 'UTR_vairant'])] #Pulls out these variant types only\n",
    "\n",
    "    grouped = df.groupby('target') #Groups by target\n",
    "    charts = [] #List to hold plots\n",
    "    rna_dna_corr = {}\n",
    "    \n",
    "    #Iterates through SGE targets and builds plots\n",
    "    for target, data in grouped:\n",
    "        target_charts = [] #Charts for single SGE target\n",
    "        \n",
    "        #Gets replicate dataframes\n",
    "        rep1 = data.loc[data['target_rep'].str.contains('rep_1')].reset_index(drop = True)\n",
    "        rep2 = data.loc[data['target_rep'].str.contains('rep_2')].reset_index(drop = True)\n",
    "        rep3 = data.loc[data['target_rep'].str.contains('rep_3')].reset_index(drop = True)\n",
    "\n",
    "        pairs = [\n",
    "            ('RNAfreq', 'DNAfreq', rep1, 'DNAfreq', 'RNAfreq', 'Rep 1'),\n",
    "            ('RNAfreq', 'DNAfreq', rep2, 'DNAfreq', 'RNAfreq', 'Rep 2'),\n",
    "            ('RNAfreq', 'DNAfreq', rep3, 'DNAfreq', 'RNAfreq', 'Rep 3')\n",
    "        ] #Pairs for replicate testing\n",
    "\n",
    "        target_corr = []\n",
    "        #iterates through pairs and creates plots\n",
    "        for rna, dna, rep, x_col, y_col, replicate in pairs:\n",
    "            corr_pearson = np.corrcoef(rep[x_col], rep[y_col])[0,1]\n",
    "            target_corr.append(corr_pearson)\n",
    "            \n",
    "            # Calculate min/max for diagonal line (y = x)\n",
    "            min_val = min(rep[x_col].min(), rep[y_col].min())\n",
    "            max_val = max(rep[x_col].max(), rep[y_col].max())\n",
    "            \n",
    "            line_df = pd.DataFrame({\n",
    "                'x': [min_val, max_val],\n",
    "                'y': [min_val, max_val]\n",
    "            })\n",
    "\n",
    "            # Create scatter plot\n",
    "            scatter = alt.Chart(rep).mark_circle(opacity=0.7).encode(\n",
    "                x=alt.X(x_col, title=f'{x_col} '),\n",
    "                y=alt.Y(y_col, title=f'{y_col} '),\n",
    "                color = alt.Color('Consequence', sort = ['synonymous_variant', 'missense_variant', 'stop_gained']),\n",
    "                tooltip = [alt.Tooltip('pos_id', title = 'Variant: '),\n",
    "                           alt.Tooltip(x_col),\n",
    "                           alt.Tooltip(y_col),\n",
    "                           alt.Tooltip('AAsub', title = 'Substitution: '), \n",
    "                           alt.Tooltip('snv_score', title = 'Score: ')\n",
    "                          ]\n",
    "            )\n",
    "            \n",
    "            # Create reference line\n",
    "            line = alt.Chart(line_df).mark_line(\n",
    "                color='red', \n",
    "                strokeDash=[4, 4]\n",
    "            ).encode(\n",
    "                x='x:Q',\n",
    "                y='y:Q'\n",
    "            )\n",
    "\n",
    "\n",
    "            #Code for trendline if desired\n",
    "            '''\n",
    "            trend_line = alt.Chart(rep).transform_regression(\n",
    "                'DNAfreq', 'RNAfreq',\n",
    "                groupby=['target_rep']\n",
    "            ).mark_line(color='blue').encode(\n",
    "                x='DNAfreq',\n",
    "                y='RNAfreq'\n",
    "            )\n",
    "\n",
    "            '''\n",
    "            \n",
    "            # Combine and add title with correlation\n",
    "            combined = (scatter + line).properties(\n",
    "                width=250,\n",
    "                height=250,\n",
    "                title= target + ' ' + replicate + ' ' + f'{rna} vs {dna} (r = {corr_pearson:.3f})' \n",
    "            ).interactive()\n",
    "\n",
    "            target_charts.append(combined)\n",
    "\n",
    "        target_chart = alt.hconcat(target_charts[0], target_charts[1], target_charts[2]) #All charts concatenated\n",
    "        median_corr = statistics.median(target_corr)\n",
    "        rna_dna_corr[target] = median_corr\n",
    "        charts.append(target_chart)\n",
    "        \n",
    "    # Concatenate all charts horizontally\n",
    "    final_chart = alt.vconcat(*charts).properties(\n",
    "        title=''\n",
    "    )\n",
    "    \n",
    "    # Display the chart\n",
    "    #final_chart.display()\n",
    "    \n",
    "    return df, rna_dna_corr        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def agg_rnafreq(df): #Aggregates replicate dataframes into one target-level dataframe for each target\n",
    "    df = df[['target', 'pos_id', 'RNAfreq', 'DNAfreq', 'RNA/DNA', 'AAsub', 'Consequence', 'snv_score', 'functional_consequence']] #Gets these columns\n",
    "    \n",
    "    df['target_id'] = df['target'] + ':' + df['pos_id'] #Creates unique pos_id that also has target identifier for aggregation\n",
    "\n",
    "\n",
    "    #Aggregates information \n",
    "    summary_df = df.groupby('target_id').agg({\n",
    "    'target': 'first',\n",
    "    'pos_id': 'first',\n",
    "    'RNAfreq': 'median',\n",
    "    'DNAfreq': 'median',\n",
    "    'RNA/DNA': ['median', 'sem'],  # Changed to include both median and SEM\n",
    "    'AAsub': 'first',\n",
    "    'Consequence': 'first',\n",
    "    'snv_score': 'first',\n",
    "    'functional_consequence': 'first'\n",
    "    }).reset_index()\n",
    "\n",
    "    # Flatten the multi-level column names for SEM column\n",
    "    summary_df.columns = [\n",
    "    'RNA/DNA' if col == ('RNA/DNA', 'median') else\n",
    "    'RNA/DNA_sem' if col == ('RNA/DNA', 'sem') else\n",
    "    col[0] if isinstance(col, tuple) else col\n",
    "    for col in summary_df.columns\n",
    "    ]\n",
    "    \n",
    "    # Rename to cleaner names\n",
    "    summary_df.rename(columns={\n",
    "        'RNA/DNA_median': 'RNA/DNA',\n",
    "        'RNA/DNA_sem': 'RNA/DNA_sem'\n",
    "    }, inplace=True)\n",
    "\n",
    "    summary_df['L2RNA/DNA'] = np.log2(summary_df['RNA/DNA']) #Does Log2 of RNA/DNA ratio (not in use)\n",
    "\n",
    "    return summary_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_scores_mean(df): #Beta normalization method based on distribution of synonymous variants in each SGE target rather than distribution of all synonymous variants\n",
    "\n",
    "    grouped = df.groupby('target')\n",
    "\n",
    "    normalized_dfs = []\n",
    "    for target, data in grouped:\n",
    "        threshold_df = df.loc[df['Consequence'].isin(['synonymous_variant'])]\n",
    "        percentile_99 = threshold_df['RNA/DNA'].quantile(0.99) #Gets 99th percentile RNA/DNA \n",
    "        percentile_1 = threshold_df['RNA/DNA'].quantile(0.01) #Gets 1st percentile RNA/DNA\n",
    "\n",
    "        threshold_df = data[(data['RNA/DNA'] >= percentile_1) & \n",
    "            (data['RNA/DNA'] <= percentile_99)]\n",
    "\n",
    "        mean_rnadna = threshold_df['RNA/DNA'].mean() #Gets mean RNA/DNA ratio\n",
    "        rnadna_std = threshold_df['RNA/DNA'].std()\n",
    "\n",
    "        upper_thresh = mean_rnadna + 2 * rnadna_std #Gets upper threshold by calculating 2 standard deviations above mean\n",
    "        lower_thresh = mean_rnadna - 2 * rnadna_std\n",
    "\n",
    "        data['RNA/DNA'] = (data['RNA/DNA'] - lower_thresh)/(upper_thresh - lower_thresh) #min-max normalization\n",
    "\n",
    "        normalized_dfs.append(data)\n",
    "\n",
    "    final_df = pd.concat(normalized_dfs)\n",
    "    final_df = final_df.reset_index(drop = True)\n",
    "\n",
    "\n",
    "    return final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rnafreq_vs_sge(df, sge_threshold): #Gets thresholds for classifying RNA performance of a variant and various visualizations\n",
    "\n",
    "    #Threshold setting based on distribution of all scored synonymous variants in data set\n",
    "    threshold_df = df.loc[df['Consequence'].isin(['synonymous_variant'])]\n",
    "    percentile_99 = threshold_df['RNA/DNA'].quantile(0.99) #Gets 99th percentile RNA/DNA \n",
    "    percentile_1 = threshold_df['RNA/DNA'].quantile(0.01) #Gets 1st percentile RNA/DNA\n",
    "\n",
    "    threshold_df = threshold_df[(threshold_df['RNA/DNA'] >= percentile_1) & \n",
    "            (threshold_df['RNA/DNA'] <= percentile_99)] #Gets threshold setting dataframe by removing outliers\n",
    "    '''\n",
    "    percentile_99 = df['RNA/DNA'].quantile(0.99) #Gets 99th percentile RNA/DNA \n",
    "    percentile_1 = df['RNA/DNA'].quantile(0.01) #Gets 1st percentile RNA/DNA\n",
    "\n",
    "    threshold_df = df[(df['RNA/DNA'] >= percentile_1) & \n",
    "            (df['RNA/DNA'] <= percentile_99)] #Gets threshold setting dataframe by removing outliers\n",
    "    '''\n",
    "    \n",
    "    mean_rnadna = threshold_df['RNA/DNA'].mean() #Gets mean RNA/DNA ratio\n",
    "    rnadna_std = threshold_df['RNA/DNA'].std() #Gets standard deviation of RNA/DNA ratio\n",
    "\n",
    "    upper_thresh = mean_rnadna + 2 * rnadna_std #Gets upper threshold by calculating 2 standard deviations above mean\n",
    "    lower_thresh = mean_rnadna - 2 * rnadna_std #Gets lower threshold by calculating 3 standard deviations below mean\n",
    "\n",
    "    thresholds = (lower_thresh, upper_thresh) #Tuple to be passed onto next functions\n",
    "\n",
    "    #Vertical lines for upper and lower thresholds\n",
    "    minus_std = alt.Chart(pd.DataFrame({'x': [lower_thresh]})).mark_rule(color = 'red').encode(\n",
    "        x = 'x')\n",
    "    plus_std = alt.Chart(pd.DataFrame({'x': [upper_thresh]})).mark_rule(color = 'red').encode(\n",
    "        x = 'x')\n",
    "\n",
    "    #Horizontal lines for upper and lower thresholds\n",
    "    minus_std_horz = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        y = alt.Y(datum = (mean_rnadna - 2 *rnadna_std))\n",
    "    )\n",
    "\n",
    "    plus_std_horz = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        y = alt.Y(datum = (mean_rnadna + 2 *rnadna_std))\n",
    "    )\n",
    "\n",
    "    #Vertical Lines for SGE Score Thresholds\n",
    "    LFClower = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        x = alt.X(datum = (sge_threshold[0]))\n",
    "    )\n",
    "\n",
    "    LFCupper = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        x = alt.X(datum = (sge_threshold[1]))\n",
    "    )\n",
    "    #Base scatter plot of RNA/DNA ratio vs. SGE score\n",
    "    chart = alt.Chart(df).mark_circle().encode(\n",
    "        x = alt.X('snv_score',\n",
    "                  scale = alt.Scale(domain = [-0.4, 0.1]\n",
    "                                   )\n",
    "                 ),\n",
    "        y = alt.Y('RNA/DNA',\n",
    "                  scale = alt.Scale(domain = [-0.5, 2.5])\n",
    "                 ),\n",
    "        color = alt.Color('Consequence', \n",
    "                          sort = ['synonymous_variant', 'missense_variant', 'stop_gained']\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('snv_score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('RNAfreq', title = 'D05 RNA Frequency: '),\n",
    "                   alt.Tooltip('Consequence', title = 'Conseuqence: '),\n",
    "                   alt.Tooltip('AAsub', title = 'Substitution: '),\n",
    "                   alt.Tooltip('pos_id', title = 'Variant: ')\n",
    "                  ]\n",
    "    )\n",
    "\n",
    "\n",
    "    #Scatter of median collapsed RNA freq vs. median collapsed DNA freq\n",
    "    scatter = alt.Chart(df).mark_circle().encode(\n",
    "        x = 'DNAfreq', \n",
    "        y = 'RNAfreq',\n",
    "        color = alt.Color('Consequence',\n",
    "                          sort = ['syonymous_variant',\n",
    "                                  'missense_variant', \n",
    "                                  'stop_gained'\n",
    "                                 ]\n",
    "                         )\n",
    "    ).facet(\n",
    "        facet = alt.Facet('target')\n",
    "    ).resolve_scale(\n",
    "        x = 'independent',\n",
    "        y = 'independent'\n",
    "    ).interactive()\n",
    "\n",
    "    #Base histogram for visualizing distribution of RNA/DNA ratios\n",
    "    base_histogram = alt.Chart(df).mark_bar().encode(\n",
    "        alt.X('RNA/DNA:Q',\n",
    "                  bin = alt.Bin(maxbins = 100, step = 0.1),\n",
    "                  scale = alt.Scale(\n",
    "                      domain = [0,3]\n",
    "                  )\n",
    "                 ),\n",
    "        y = 'count()',\n",
    "        color = alt.Color('Consequence', \n",
    "                          sort = ['synonymous_variant', 'missense_variant', 'stop_gained']\n",
    "                         )\n",
    "    ).interactive()\n",
    "    \n",
    "    \n",
    "\n",
    "    histogram = base_histogram + minus_std + plus_std #Threshold lines added to histogram\n",
    "\n",
    "    #Base scatter plot faceted\n",
    "    targets = natsorted(list(set(df['target'].tolist())))\n",
    "    chart = chart + minus_std_horz + plus_std_horz + LFCupper + LFClower\n",
    "    chart = chart.facet(facet = alt.Facet(\n",
    "        'target',\n",
    "        sort = targets\n",
    "        ),columns = 5\n",
    "    ).interactive()\n",
    "\n",
    "    chart.display()\n",
    "    #scatter.display()\n",
    "    histogram.display()\n",
    "    \n",
    "\n",
    "    return thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rnafreq_vs_sge_beta(df, sge_threshold):\n",
    "\n",
    "    upper_thresh = 1 #Gets upper threshold by calculating 2 standard deviations above mean\n",
    "    lower_thresh = 0 #Gets lower threshold by calculating 3 standard deviations below mean\n",
    "\n",
    "    thresholds = (lower_thresh, upper_thresh) #Tuple to be passed onto next functions\n",
    "    print(thresholds)\n",
    "    #Vertical lines for upper and lower thresholds\n",
    "    minus_std = alt.Chart(pd.DataFrame({'x': [lower_thresh]})).mark_rule(color = 'red').encode(\n",
    "        x = 'x')\n",
    "    plus_std = alt.Chart(pd.DataFrame({'x': [upper_thresh]})).mark_rule(color = 'red').encode(\n",
    "        x = 'x')\n",
    "\n",
    "    #Horizontal lines for upper and lower thresholds\n",
    "    minus_std_horz = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        y = alt.Y(datum = lower_thresh)\n",
    "    )\n",
    "\n",
    "    plus_std_horz = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        y = alt.Y(datum = upper_thresh)\n",
    "    )\n",
    "\n",
    "    #Vertical Lines for SGE Score Thresholds\n",
    "    LFClower = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        x = alt.X(datum = (sge_threshold[0]))\n",
    "    )\n",
    "\n",
    "    LFCupper = alt.Chart(df).mark_rule(color = 'red').encode(\n",
    "        x = alt.X(datum = (sge_threshold[1]))\n",
    "    )\n",
    "\n",
    "    #Base scatter plot of RNA/DNA ratio vs. SGE score\n",
    "    chart = alt.Chart(df).mark_circle().encode(\n",
    "        x = alt.X('snv_score',\n",
    "                  scale = alt.Scale(domain = [-0.4, 0.1])\n",
    "                ),\n",
    "        y = alt.Y('RNA/DNA',\n",
    "                  scale = alt.Scale(domain = [-0.5, 2.5])\n",
    "                 ),\n",
    "        color = alt.Color('Consequence', \n",
    "                          sort = ['synonymous_variant', 'missense_variant', 'stop_gained']\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '),\n",
    "                   alt.Tooltip('snv_score', title = 'SGE Score: '),\n",
    "                   alt.Tooltip('RNAfreq', title = 'D05 RNA Frequency: '),\n",
    "                   alt.Tooltip('Consequence', title = 'Conseuqence: '),\n",
    "                   alt.Tooltip('AAsub', title = 'Substitution: '),\n",
    "                   alt.Tooltip('pos_id', title = 'Variant: ')\n",
    "                  ]\n",
    "    )\n",
    "\n",
    "\n",
    "    #Scatter of median collapsed RNA freq vs. median collapsed DNA freq\n",
    "    scatter = alt.Chart(df).mark_circle().encode(\n",
    "        x = 'DNAfreq', \n",
    "        y = 'RNAfreq',\n",
    "        color = alt.Color('Consequence',\n",
    "                          sort = ['syonymous_variant',\n",
    "                                  'missense_variant', \n",
    "                                  'stop_gained'\n",
    "                                 ]\n",
    "                         )\n",
    "    ).facet(\n",
    "        facet = alt.Facet('target')\n",
    "    ).resolve_scale(\n",
    "        x = 'independent',\n",
    "        y = 'independent'\n",
    "    ).interactive()\n",
    "\n",
    "    #Base histogram for visualizing distribution of RNA/DNA ratios\n",
    "    base_histogram = alt.Chart(df).mark_bar().encode(\n",
    "        alt.X('RNA/DNA:Q',\n",
    "                  bin = alt.Bin(maxbins = 100, step = 0.1),\n",
    "                  scale = alt.Scale(\n",
    "                      domain = [0,3]\n",
    "                  )\n",
    "                 ),\n",
    "        y = 'count()',\n",
    "        color = alt.Color('Consequence', \n",
    "                          sort = ['synonymous_variant', 'missense_variant', 'stop_gained']\n",
    "                         )\n",
    "    ).interactive()\n",
    "    \n",
    "    \n",
    "\n",
    "    histogram = base_histogram + minus_std + plus_std #Threshold lines added to histogram\n",
    "\n",
    "    #Base scatter plot faceted\n",
    "    targets = natsorted(list(set(df['target'].tolist())))\n",
    "    chart = chart + minus_std_horz + plus_std_horz + LFCupper + LFClower\n",
    "    chart = chart.facet(facet = alt.Facet(\n",
    "        'target',\n",
    "        sort = targets\n",
    "        ),columns = 5\n",
    "    ).interactive()\n",
    "\n",
    "    chart.display()\n",
    "    #scatter.display()\n",
    "    histogram.display()\n",
    "    \n",
    "\n",
    "    return thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_abnormal_rna(df, cds, corr_dict, thresholds): #Gets final dataframe of variants with abnormal RNA performance\n",
    "    \n",
    "    min_threshold, max_threshold = thresholds #Gets thresholds\n",
    "    \n",
    "    grouped = df.groupby('pos_id') #Groups by position ID\n",
    "\n",
    "    #Iterates through each variant. For double-covered variants, data from library with higher RNA to DNA correalation kept\n",
    "    for position, data in grouped:\n",
    "        target_list = list(set(data['target'].tolist()))\n",
    "\n",
    "        if len(target_list) == 1:\n",
    "            continue\n",
    "        else:\n",
    "            lib1 = target_list[0]\n",
    "            lib2 = target_list[1]\n",
    "\n",
    "            lib1_corr = corr_dict[lib1]\n",
    "            lib2_corr = corr_dict[lib2]\n",
    "\n",
    "            if lib1_corr > lib2_corr:\n",
    "                to_remove = lib2 + ':' + position\n",
    "\n",
    "                df = df.loc[~(df['target_id'].isin([to_remove]))]\n",
    "                df = df.copy()\n",
    "\n",
    "            elif lib2_corr > lib1_corr: \n",
    "                to_remove = lib1 + ':' + position\n",
    "\n",
    "                df = df.loc[~(df['target_id'].isin([to_remove]))]\n",
    "                df = df.copy()\n",
    "                \n",
    "    cds_pos = pd.read_excel(cds)\n",
    "        \n",
    "    \n",
    "    df['RNA_classification'] = 'normal'\n",
    "\n",
    "    #Classifies variant RNA performance\n",
    "    df.loc[df['RNA/DNA'] > max_threshold, 'RNA_classification'] = 'high'\n",
    "    df.loc[df['RNA/DNA'] < min_threshold, 'RNA_classification'] = 'low'\n",
    "\n",
    "    lowRNA_funcNormal = len(df.loc[(df['RNA_classification'] == 'low') & (df['functional_consequence'] == 'functionally_normal')])\n",
    "    perc_misclassified = (lowRNA_funcNormal / len(df)) * 100\n",
    "\n",
    "    df['pos'] = df['pos_id'].transform(lambda x: x.split(':')[0])\n",
    "    df['pos'] = df['pos'].astype(int)\n",
    "\n",
    "    df = pd.merge(df, cds_pos, on = 'pos', how = 'inner')\n",
    "\n",
    "    df = df[['target', 'pos_id', 'AAsub', 'Consequence', 'snv_score', 'functional_consequence', 'RNA/DNA', 'RNA/DNA_sem', 'RNA_classification', 'pos', 'CDSpos']]\n",
    "    \n",
    "    print('% Low RNA & Functionally Normal: ', str(perc_misclassified),'\\n',\n",
    "         '# Low RNA & Functionally Normal: ', str(lowRNA_funcNormal))\n",
    "    df = df.reset_index(drop = True)\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_dna(dict, gdna_scores): #This function merges the DNA SGE scores the median RNA scores for each variant\n",
    "\n",
    "    dna_scores = dna_scores.drop(columns = ['chrom', 'pos', 'allele', 'R1_score', 'R2_score', 'R3_score', 'target']) #Drops these columns\n",
    "    all_rna = list(dict.values()) #Gets all dataframes stored in the RNAscore dictionary\n",
    "\n",
    "    df = pd.concat(all_rna) #Concatenates all RNAscore dataframes\n",
    "\n",
    "    merged = pd.merge(df, dna_scores, how = 'inner', on = 'pos_id') #Merges based on position ID\n",
    "\n",
    "\n",
    "    multi_target_vars = merged.groupby('pos_id').agg({'RNAscore_med': 'median'}).reset_index()\n",
    "\n",
    "    merged = pd.merge(multi_target_vars, merged, how = 'left', on = 'pos_id')\n",
    "    merged = merged.rename(columns = {'RNAscore_med_x': 'RNAscore', 'RNAscore_med_y': 'target_RNAscore'})\n",
    "    final_df = merged.loc[:, ['target', 'pos_id', 'Consequence', 'AAsub', 'snv_score', 'target_RNAscore', 'RNAscore']]\n",
    "\n",
    "    \n",
    "    return final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    all_counts, sge_scores, sge_thresholds = read_counts(rna_count_folder, dna_count_folder, sge_score, sge_threshold_file)\n",
    "\n",
    "    #rna_v_dna_plot(all_counts)\n",
    "    \n",
    "    rna_scored, sge_targets = get_rna_scores(all_counts)\n",
    "    merged_df, rna_dna_corr = rna_dna_freq_corr_plots(rna_scored, sge_scores)\n",
    "\n",
    "    #Code for Whole-Gene Synonymous Normalization\n",
    "    final_df = agg_rnafreq(merged_df)\n",
    "    print('Whole gene, syn. distribution: ')\n",
    "    threshold = rnafreq_vs_sge(final_df)\n",
    "\n",
    "    #Code for Target-by-Target Synonymous Normalization\n",
    "    final_df_mean_norm = normalize_scores_mean(final_df)\n",
    "    print('Target by target, syn. distribution: ')\n",
    "    target_threshold = rnafreq_vs_sge_beta(final_df_mean_norm)\n",
    "    \n",
    "    print('Whole Gene QC: ')\n",
    "    whole_gene_output = get_abnormal_rna(final_df,cds_pos, rna_dna_corr, threshold, sge_thresholds)\n",
    "    print('Target-by-Target QC: ')\n",
    "    target_by_target_output = get_abnormal_rna(final_df_mean_norm, cds_pos, rna_dna_corr, target_threshold, sge_thresholds)\n",
    "\n",
    "    output = 'Whole'\n",
    "\n",
    "    if output == 'Whole':\n",
    "        final_df = whole_gene_output\n",
    "        final_threshold = threshold\n",
    "        output_path_str = '../Data/20250813/20250813_BARD1_finaldata_RNA_wholeGene.xlsx'\n",
    "    elif output == 'Target':\n",
    "        final_df == target_by_target_output\n",
    "        final_threshold = target_threshold\n",
    "        output_path_str = '../Data/20250813_BARD1_finaldata_RNA_Targetbased.xlsx'\n",
    "    \n",
    "    print(final_df)\n",
    "    min_threshold, max_threshold = final_threshold\n",
    "    \n",
    "    threshold_df = pd.DataFrame({'min': [min_threshold], 'max': [max_threshold]})\n",
    "\n",
    "    dfs = {'data': final_df,\n",
    "           'thresholds': threshold_df\n",
    "          }\n",
    "\n",
    "\n",
    "    '''\n",
    "    with pd.ExcelWriter(output_path_str) as writer:\n",
    "        for sheet_name, df in dfs.items():\n",
    "            df.to_excel(writer, sheet_name=sheet_name, index=False)\n",
    "    '''\n",
    "\n",
    "    #whole_gene_output.to_excel('../Data/20250813/20250813_BARD1_finaldata_RNA_wholeGene.xlsx', index = False)\n",
    "    #target_by_target_output.to_excel('../Data/20250813_BARD1_finaldata_RNA_Targetbased.xlsx', index = False)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
