{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0",
   "metadata": {},
   "source": [
    "This notebook generates figures 5A (RNA score vs. SGE score scatterplot), 5B (stem plots for splice-disrupting syn. variants), and S8 (RNA score vs. SGE score stem plots). Figure 5B is generated through figure S8 (bottom), and a special stem plot pulling only targets X4G - X4L (manually set in the code). Plots not in the publication are also coded and can be shown"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import numpy as np\n",
    "from natsort import natsorted\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = '../Data/final_tables/BARD1_SGE_final_table.xlsx'\n",
    "alt.data_transformers.disable_max_rows()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_output(file):\n",
    "    df = pd.read_excel(file, sheet_name = 'scores')\n",
    "    df = df.loc[df['var_type'].isin(['snv'])]\n",
    "    df = df.loc[~df['variant_qc_flag'].isin(['WARN'])]\n",
    "    df = df.dropna(subset = ['RNAscore'])\n",
    "    df = df.rename(columns = {'consequence': 'Consequence', 'score': 'snv_score', 'RNAscore': 'RNA/DNA', 'CDS_position': 'CDSpos',  'amino_acid_change': 'AAsub'})\n",
    "    \n",
    "    df.loc[df['Consequence'].str.contains('missense'), 'Consequence'] = 'Missense'\n",
    "    df.loc[df['Consequence'] == 'synonymous_variant', 'Consequence'] = 'Synonymous'\n",
    "    df.loc[df['Consequence'] == 'intron_variant', 'Consequence'] = 'Intron'\n",
    "    df.loc[df['Consequence'] == 'stop_gained', 'Consequence'] = 'Stop Gained'\n",
    "    df.loc[df['Consequence'] == 'stop_lost', 'Consequence'] = 'Stop Lost'\n",
    "    df.loc[df['Consequence'].str.contains('site'), 'Consequence'] = 'Canonical Splice'\n",
    "    df.loc[df['Consequence'].str.contains('ing_var'), 'Consequence'] = 'Splice Region'\n",
    "    df.loc[df['Consequence'].str.contains('UTR'), 'Consequence'] = 'UTR Variant'\n",
    "    df.loc[df['Consequence'] == 'start_lost', 'Consequence'] = 'Start Lost'\n",
    "    \n",
    "    threshold_df = pd.read_excel(file, sheet_name = 'thresholds')\n",
    "\n",
    "    min_threshold = threshold_df['min'][0]\n",
    "    max_threshold = threshold_df['max'][0]\n",
    "    rna_threshold = threshold_df['rna'][0]\n",
    "    thresholds = (rna_threshold)\n",
    "              \n",
    "    \n",
    "    df['AApos'] = df['AAsub'].transform(lambda x: x[1: -1])\n",
    "    df = df.loc[~(df['AApos'].isin(['-']))]\n",
    "    df['AApos'] = df['AApos'].astype(int)\n",
    "    df['exon'] = df['target'].str.extract(r'(BARD1_X\\d+)')\n",
    "    \n",
    "    abnormal_df = df[df['RNA_consequence'].isin(['low'])]\n",
    "    normal_df = df[df['RNA_consequence'].isin(['normal'])]\n",
    "\n",
    "    sge_thresholds = [min_threshold, max_threshold]\n",
    "\n",
    "    return abnormal_df, normal_df, df, thresholds, sge_thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_rna_performance(df, thresholds, sge_threshold):\n",
    "\n",
    "    min_threshold = thresholds\n",
    "    \n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a', # yellow\n",
    "    '#93C47D', # light green\n",
    "    '#888888', # med gray\n",
    "    '#000000', # black\n",
    "    '#CFCFCF' # light gray\n",
    "        \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained',\n",
    "        'UTR Variant',\n",
    "        'Stop Lost',\n",
    "        'Start Lost',\n",
    "        'Splice Region',\n",
    "    ]\n",
    "\n",
    "    plot = alt.Chart(df).mark_point().encode(\n",
    "        x = 'AApos',\n",
    "        y = alt.Y('RNA/DNA',\n",
    "                  scale = alt.Scale(domain = [-8, 3]\n",
    "                                   )\n",
    "                 ),\n",
    "        color = alt.Color('functional_consequence', \n",
    "                          sort = ['functionally_normal', 'indeterminate', 'functionally_abnormal']\n",
    "                         ),\n",
    "        shape = 'RNA_consequence',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Region: '),\n",
    "                   alt.Tooltip('AAsub', title = 'Amino Acid Substituion: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 1200,\n",
    "        height = 400, \n",
    "        title = 'Distribution of RNA Scores Across BARD1'\n",
    "    ).interactive()\n",
    "\n",
    "\n",
    "    chart = alt.Chart(df).mark_point().encode(\n",
    "        x = alt.X('RNA/DNA',\n",
    "                  title = 'RNA Score',\n",
    "                  axis = alt.Axis(labelFontSize = 16,\n",
    "                                  titleFontSize = 20,\n",
    "                                  values = [-8, -6, -4, -2, 0, 2]\n",
    "                                 ),\n",
    "                  scale = alt.Scale(domain = [-8, 3]\n",
    "                                   )\n",
    "                 ),\n",
    "        y = alt.Y('snv_score',\n",
    "                 title = 'SGE Score',\n",
    "                 axis = alt.Axis(labelFontSize =16,\n",
    "                                titleFontSize = 20,\n",
    "                                values = [-0.6,-0.5, -0.4, -0.3, -0.2, -0.1, 0, 0.1, 0.2, 0.3]\n",
    "                                ),\n",
    "                  scale = alt.Scale(domain = [-0.6, 0.3]\n",
    "                                   )\n",
    "                 ),\n",
    "        color = alt.Color('Consequence',\n",
    "                      scale = alt.Scale(range = palette,\n",
    "                                        domain = variant_types\n",
    "                                       ),\n",
    "                      legend = alt.Legend(titleFontSize = 16, \n",
    "                                          labelFontSize = 14\n",
    "                                         )\n",
    "                     ),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Region: '),\n",
    "                   alt.Tooltip('AAsub', title = 'Amino Acid Substituion: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 600,\n",
    "        height = 400, \n",
    "        title = alt.TitleParams(text = 'SGE Score vs. RNA Score',\n",
    "                                fontSize = 22\n",
    "                               )               \n",
    "    ).interactive()\n",
    "\n",
    "    nf_line = alt.Chart(pd.DataFrame({'snv_score': [sge_threshold[0]]})).mark_rule(color = 'red').encode(\n",
    "        y = 'snv_score')\n",
    "\n",
    "    func_line = alt.Chart(pd.DataFrame({'snv_score': [sge_threshold[1]]})).mark_rule(color = 'blue').encode(\n",
    "        y = 'snv_score')\n",
    "    \n",
    "    rna_nf_line = alt.Chart(pd.DataFrame({'RNA/DNA': [min_threshold]})).mark_rule(color = 'red').encode(\n",
    "        x = 'RNA/DNA')\n",
    "\n",
    "    \n",
    "    chart = (chart + nf_line + func_line + rna_nf_line).configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    ).interactive()\n",
    "    \n",
    "    #plot.display()\n",
    "    chart.display()\n",
    "\n",
    "    return chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def stem_plot(df, thresholds):\n",
    "\n",
    "    df = df.loc[df['RNA/DNA'] < 10]\n",
    "    df = df.copy()\n",
    "\n",
    "    min_threshold = thresholds\n",
    "    \n",
    "    abnormal_df = df.loc[(df['functional_consequence'].isin(['functionally_abnormal'])) & (df['RNA_consequence'].isin(['low']))]\n",
    "\n",
    "    df['rule_value'] = np.nan\n",
    "    df.loc[(df['functional_consequence'] == 'functionally_abnormal') & (df['RNA_consequence'] == 'low'), 'rule_value'] = df.loc[(df['functional_consequence'] == 'functionally_abnormal') & (df['RNA_consequence'] == 'low'), 'RNA/DNA']\n",
    "\n",
    "    df.loc[df['functional_consequence'] == 'functionally_abnormal', 'functional_consequence'] = 'Functionally Abnormal'\n",
    "    df.loc[df['functional_consequence'] == 'functionally_normal', 'functional_consequence'] = 'Functionally Normal'\n",
    "    df.loc[df['functional_consequence'] == 'indeterminate', 'functional_consequence'] = 'Indeterminate'\n",
    "\n",
    "    #df = df.loc[df['target'].isin(['BARD1_X4G', 'BARD1_X4H', 'BARD1_X4I', 'BARD1_X4J', 'BARD1_X4K', 'BARD1_X4L'])] #use to make zoomed in subset for X4 focusing on the splice disrupting synonymous variants\n",
    "    consequence_sort = ['Functionally Normal', 'Fucntionally Abnormal', 'Indeterminate']\n",
    "    \n",
    "    exons = list(set(df['exon'].tolist()))\n",
    "    exons = natsorted(exons)\n",
    "\n",
    "    palette = [\n",
    "    '#006616', # dark green,\n",
    "    '#81B4C7', # dusty blue\n",
    "    '#ffcd3a', # yellow\n",
    "    '#93C47D', # light green\n",
    "    '#888888', # med gray\n",
    "    '#000000', # black\n",
    "    '#CFCFCF' # light gray\n",
    "        \n",
    "    ]\n",
    "    \n",
    "    \n",
    "    variant_types = [\n",
    "        'Synonymous',\n",
    "        'Missense',  \n",
    "        'Stop Gained',\n",
    "        'UTR Variant',\n",
    "        'Stop Lost',\n",
    "        'Start Lost',\n",
    "        'Splice Region',\n",
    "    ]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    df['min_threshold'] = min_threshold\n",
    "    base = alt.Chart(df)\n",
    "    base_dot = base.mark_point(filled = True, size = 75).encode(\n",
    "        x = alt.X('CDSpos:Q', \n",
    "                  title = 'CDS Position',\n",
    "                  axis = alt.Axis(labelFontSize = 16,\n",
    "                                  titleFontSize = 20\n",
    "                                 ),\n",
    "                  scale = alt.Scale(zero = False, \n",
    "                                    padding = 5\n",
    "                                   )\n",
    "                 ),\n",
    "        y = alt.Y('RNA/DNA:Q', \n",
    "                  title = 'RNA Score',\n",
    "                  axis = alt.Axis(labelFontSize = 16,\n",
    "                                  titleFontSize = 20,\n",
    "                                  values = [-8, -6, -4, -2, 0, 2]\n",
    "                                 ),\n",
    "                  scale = alt.Scale(domain = [-8, 3],\n",
    "                                    padding = 5)\n",
    "                 ),\n",
    "        color = alt.Color('Consequence',\n",
    "                          scale = alt.Scale(\n",
    "                              range = palette, \n",
    "                              domain = variant_types\n",
    "                          ),\n",
    "                          legend = alt.Legend(\n",
    "                              titleFontSize = 16,\n",
    "                              labelFontSize =14\n",
    "                          )\n",
    "                         ),\n",
    "        shape = alt.Shape('functional_consequence',\n",
    "                          sort = consequence_sort,\n",
    "                          legend = alt.Legend(\n",
    "                              titleLimit  = 500,\n",
    "                              title = 'Functional Consequence',\n",
    "                              titleFontSize = 16,\n",
    "                              labelFontSize = 14\n",
    "                          )\n",
    "                         ),\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Target: '), \n",
    "                   alt.Tooltip('AAsub', title = 'Amino Acid Sub: '),\n",
    "                   alt.Tooltip('Consequence', title = 'Consequence: '),\n",
    "                   alt.Tooltip('snv_score', title = 'SGE Score: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 800, \n",
    "        height = 400\n",
    "    )\n",
    "\n",
    "    rule = base.mark_rule().encode(\n",
    "        x=alt.X('CDSpos:Q',\n",
    "                scale=alt.Scale(zero=False, padding=5)\n",
    "        ),\n",
    "        y=alt.Y('min_threshold:Q'),  # Reference as a column\n",
    "        y2=alt.Y2('rule_value:Q')\n",
    "    )\n",
    "\n",
    "    rna_nf_line = alt.Chart(pd.DataFrame({'y': [min_threshold]})).mark_rule(color = 'red').encode(\n",
    "    y = 'y')\n",
    "    \n",
    "    chart = (rule + base_dot + rna_nf_line)\n",
    "    \n",
    "\n",
    "    chart = chart.facet(facet = alt.Facet('exon:N',\n",
    "                                         sort = exons\n",
    "                                         ),\n",
    "                        columns = 2\n",
    "                       ).resolve_scale(\n",
    "        x = 'independent', \n",
    "        y = 'independent'\n",
    "    ).resolve_axis(\n",
    "    x='independent',\n",
    "    y='independent'\n",
    "    )\n",
    "\n",
    "    chart = chart.configure_axis(\n",
    "        grid = False\n",
    "    ).configure_view(\n",
    "        stroke = None\n",
    "    )\n",
    "    \n",
    "    chart.display()\n",
    "\n",
    "    return chart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nmd_across_gene(df):\n",
    "    \n",
    "    df = df.loc[df['Consequence'].isin(['Stop Gained'])]\n",
    "\n",
    "    color_scale = alt.Scale(\n",
    "        domain =['functionally_normal', 'functionally_abnormal'],\n",
    "        range = ['#1f77b4', '#d62728']\n",
    "    )\n",
    "    \n",
    "    plot = alt.Chart(df).mark_point().encode(\n",
    "        x = 'AApos',\n",
    "        y = alt.Y('RNA/DNA',\n",
    "                  title = 'RNA Score',\n",
    "                  scale = alt.Scale(domain = [-8, 3]),\n",
    "                  axis = alt.Axis(values = list(range(-8,3, 2))\n",
    "                                 )\n",
    "                 ),\n",
    "        color = alt.Color('functional_consequence', \n",
    "                          scale = color_scale,\n",
    "                         ),\n",
    "        shape = 'RNA_consequence',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Region: '),\n",
    "                   alt.Tooltip('AApos', title = 'Amino Acid Position: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 1200,\n",
    "        height = 400, \n",
    "        title = 'NMD Across BARD1'\n",
    "    ).interactive()\n",
    "\n",
    "    plot.display()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def missense_rna_performance(df):\n",
    "\n",
    "    df = df.loc[df['Consequence'].isin(['Missense'])]\n",
    "\n",
    "    color_scale = alt.Scale(\n",
    "        domain =['functionally_normal', 'functionally_abnormal'],\n",
    "        range = ['#1f77b4', '#d62728']\n",
    "    )\n",
    "    \n",
    "    plot = alt.Chart(df).mark_point().encode(\n",
    "        x = 'AApos',\n",
    "        y = alt.Y('RNA/DNA',\n",
    "                  title = 'RNA Score',\n",
    "                  scale = alt.Scale(domain = [-8, 3]),\n",
    "                  axis = alt.Axis(values = list(range(-8,3, 2))\n",
    "                                 )\n",
    "                 ),\n",
    "        color = alt.Color('functional_consequence', \n",
    "                          scale = color_scale,\n",
    "                         ),\n",
    "        shape = 'RNA_consequence',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Region: '),\n",
    "                   alt.Tooltip('AApos', title = 'Amino Acid Position: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 1200,\n",
    "        height = 400, \n",
    "        title = 'Missense Vars. RNA Scores'\n",
    "    ).interactive()\n",
    "\n",
    "    plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def synonymous_rna_performance(df):\n",
    "    \n",
    "    df = df.loc[df['Consequence'].isin(['Synonymous'])]\n",
    "\n",
    "    color_scale = alt.Scale(\n",
    "        domain =['functionally_normal', 'functionally_abnormal'],\n",
    "        range = ['#1f77b4', '#d62728']\n",
    "    )\n",
    "    \n",
    "    plot = alt.Chart(df).mark_point().encode(\n",
    "        x = 'AApos',\n",
    "        y = alt.Y('RNA/DNA',\n",
    "                  title = 'RNA Score',\n",
    "                  scale = alt.Scale(domain = [-8, 3]),\n",
    "                  axis = alt.Axis(values = list(range(-8,3, 2))\n",
    "                                 )\n",
    "                 ),\n",
    "        color = alt.Color('functional_consequence', \n",
    "                          scale = color_scale,\n",
    "                         ),\n",
    "        shape = 'RNA_consequence',\n",
    "        tooltip = [alt.Tooltip('target', title = 'SGE Region: '),\n",
    "                   alt.Tooltip('AApos', title = 'Amino Acid Position: ')\n",
    "                  ]\n",
    "    ).properties(\n",
    "        width = 1200,\n",
    "        height = 400, \n",
    "        title = 'Synonymous Vars. RNA Scores'\n",
    "    ).interactive()\n",
    "\n",
    "    plot.display()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    ab_data, norm_data, all_data, rna_thresholds, sge_thresholds = read_output(file)\n",
    "\n",
    "\n",
    "    sge_vs_rna = all_rna_performance(all_data, rna_thresholds, sge_thresholds)\n",
    "    rna_stem_plot = stem_plot(all_data, rna_thresholds)\n",
    "\n",
    "    '''\n",
    "    nmd_across_gene(all_data)\n",
    "    missense_rna_performance(all_data)\n",
    "    synonymous_rna_performance(all_data)\n",
    "    '''\n",
    "    #sge_vs_rna.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_3a.png', ppi = 500)\n",
    "    #rna_stem_plot.save('/Users/ivan/Desktop/BARD1_draft_figs/supp_figs/suppfig_RNA_Stemplot.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
