{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import altair as alt\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "from scipy import stats\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "sge = '../Data/20250825_BARD1snvscores_filtered.xlsx'\n",
    "gnomad = '../Data/20240905_BARD1_gnomADv4.1.0_SNVs.xlsx' #for Ambry\n",
    "carriers = '/Users/ivan/Downloads/7_genes_for_Starita_02282025_hg38.xlsx' #carriers data\n",
    "ambry = '/Users/ivan/Downloads/Ambry_BARD1.xlsx' #ambry set\n",
    "bridges_all = '../Data/BRIDGES_data/20250815_BRIDGES_missense_all.xlsx' #bridges all missense variants\n",
    "bridges_population = '../Data/BRIDGES_data/20250815_BRIDGES_missense_population.xlsx' #bridges population missense variants\n",
    "bridges_all_ptv = '../Data/BRIDGES_data/20250815_BRIDGES_PTVs_all.xlsx' #bridges all PTVs\n",
    "bridges_pop_ptv = '../Data/BRIDGES_data/20250815_BRIDGES_PTVs_pop.xlsx' #bridges population PTVs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_sge(sge):\n",
    "    sge = pd.read_excel(sge) #Reads SGE data\n",
    "    sge = sge.loc[~(sge['functional_consequence'].isin(['indeterminate']))]\n",
    "\n",
    "    sge = sge.rename(columns = {'functional_consequence': 'Classification', \n",
    "                               'consequence': 'Consequence'})\n",
    "\n",
    "    sge.loc[sge['Classification'] == 'functionally_normal', 'Classification'] = 'F'\n",
    "    sge.loc[sge['Classification'] == 'functionally_abnormal', 'Classification'] = 'NF'\n",
    "    sge.loc[sge['Classification'] == 'indeterminate', 'Classification'] = 'I'\n",
    "\n",
    "    return sge"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_ambry(ambry, gnomad):\n",
    "    ambry = pd.read_excel(ambry)\n",
    "    ambry = ambry[['Gene', 'c_variant', 'p_variant', 'Classification']]\n",
    "    ambry['amino_acid_change'] = ambry['p_variant'].str[2:]\n",
    "\n",
    "    ambry = ambry.dropna(subset = ['amino_acid_change']).copy()\n",
    "    ambry = ambry.reset_index(drop = True)\n",
    "\n",
    "    \n",
    "    gnomad = pd.read_excel(gnomad)\n",
    "    gnomad = gnomad[['Chromosome', 'Position', 'Reference', 'Alternate', 'Allele Frequency', 'Protein Consequence']]\n",
    "    gnomad['Position'] = gnomad['Position'].astype(str)\n",
    "    gnomad['pos_id'] = gnomad['Position'] + ':' + gnomad['Alternate']\n",
    "\n",
    "    return ambry, gnomad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_carriers_data(cc): #Reads all data\n",
    "    cc = pd.read_excel(cc) #Reads case-control data\n",
    "    cc = cc[cc['CAVA_GENE'].isin(['BARD1'])] #Filters only for BARD1\n",
    "    cc = cc[['Classification', '#CHROM', 'REF', 'ALT', 'CAVA_GENE', 'CAVA_CSN', 'CAVA_SO', 'Sample_AAF', 'Sample_ID', 'CaseControl','ER_status1', 'hg38_start']].copy() #Keeps necessary columns\n",
    "    \n",
    "    cc['pos_id'] = None #Creates emtpy pos_id column\n",
    "    cc = cc[cc['ALT'].str.len() == 1].copy()\n",
    "    cc['hg38_start'] = cc['hg38_start'].astype(str) #Sets hg38 coordinates as str data type\n",
    "    cc['pos_id'] = cc['hg38_start'] + ':' + cc['ALT'] #Creates position ID\n",
    "\n",
    "    cc_er = cc[cc['ER_status1'].isin([0, 777])]\n",
    "\n",
    "    return cc, cc_er"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_bridges(all, population, all_ptv, pop_ptv):\n",
    "    bridges_all = pd.read_excel(all, sheet_name = 'BARD1')\n",
    "    bridges_pop = pd.read_excel(population, sheet_name = 'BARD1')\n",
    "    bridges_all_ptv = pd.read_excel(all_ptv, sheet_name = 'BARD1')\n",
    "    bridges_pop_ptv = pd.read_excel(pop_ptv, sheet_name = 'BARD1')\n",
    "\n",
    "    raw_dfs = [bridges_all, bridges_pop, bridges_all_ptv, bridges_pop_ptv]\n",
    "    cleaned_dfs = []\n",
    "    \n",
    "    for df in raw_dfs:\n",
    "        df = df[['Cases', 'Controls', 'chr', 'ref', 'alt', 'hg38_pos']]\n",
    "        df = df.loc[(df['ref'].str.len() == 1) & (df['alt'].str.len() == 1)]\n",
    "        \n",
    "        df = df.rename(columns = {'hg38_pos': 'pos'})\n",
    "\n",
    "        df['pos_id'] = df['pos'].astype(str) + ':' + df['alt']\n",
    "\n",
    "        df = df[['Cases', 'Controls', 'pos_id']]\n",
    "        cleaned_dfs.append(df)\n",
    "\n",
    "\n",
    "    final_dfs = []\n",
    "    bridges_all = pd.concat([cleaned_dfs[0], cleaned_dfs[2]])\n",
    "    bridges_pop = pd.concat([cleaned_dfs[1], cleaned_dfs[3]])\n",
    "\n",
    "    dfs = [bridges_all, bridges_pop]\n",
    "    \n",
    "    for elem in dfs:\n",
    "        final_dfs.append(elem)\n",
    "\n",
    "    return bridges_all, bridges_pop   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_vars(sge):\n",
    "\n",
    "    #under development - trying to figure out which parameters are best\n",
    "    sge['AApos'] = sge['amino_acid_change'].str.extract(r'([0-9]+)', expand=False).astype(float)\n",
    "    sge_all = sge #All variants\n",
    "    \n",
    "    sge_miss = sge[sge['Consequence'].isin(['missense_variant'])] #missense variants only \n",
    "\n",
    "\n",
    "    not_domain_coords = list(range(1, 26)) + list(range(123, 425)) + list(range(546, 568))\n",
    "    sge_ring = sge_miss.loc[sge_miss['AApos'].isin(list(range(26,123)))] #RING missense variants only\n",
    "    sge_ard = sge_miss.loc[sge_miss['AApos'].isin(list(range(425, 546)))] #ARD missense variants only\n",
    "    sge_brct = sge_miss.loc[sge_miss['AApos'].isin(list(range(568, 778)))] #BRCT missense variants only \n",
    "    sge_structured = sge_miss.loc[~(sge_miss['AApos'].isin(not_domain_coords))] #All structured missense variants only\n",
    "        \n",
    "    sge_dict = {'miss': sge_miss, 'all': sge_all, 'ring': sge_ring,\n",
    "                'ard': sge_ard, 'brct': sge_brct, 'structured': sge_structured\n",
    "               }\n",
    "\n",
    "    sge_keys = list(sge_dict.keys())\n",
    "    \n",
    "    return sge_dict, sge_keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_ambry(sge, gnomad, ambry):\n",
    "    \n",
    "    sge_gnomad = pd.merge(sge, gnomad, on = 'pos_id', how = 'inner')\n",
    "\n",
    "    gnomad_func = sge_gnomad['Classification'].value_counts()['F']\n",
    "    gnomad_nf = sge_gnomad['Classification'].value_counts()['NF']\n",
    "    \n",
    "    sge_ambry = pd.merge(sge,ambry, on = 'amino_acid_change', how = 'inner')\n",
    "\n",
    "    ambry_func = sge_ambry['Classification_x'].value_counts()['F']\n",
    "    ambry_nf = sge_ambry['Classification_x'].value_counts()['NF'] \n",
    "\n",
    "    contingency_table = np.array([[ambry_nf, ambry_func],\n",
    "                                    [gnomad_nf, gnomad_func]])\n",
    "\n",
    "    df = pd.DataFrame(contingency_table, columns = ['NF', 'F'], index = ['Ambry', 'gnomAD'])\n",
    "\n",
    "\n",
    "    oddsratio, p_value = stats.fisher_exact(contingency_table) #Tabulates odds-ratio and p-value from Fischer's exact test\n",
    "    table = sm.stats.Table2x2(contingency_table) #Generates confidence intervals\n",
    "    \n",
    "    #result = table.summary() #Summary table with Odds and Risk ratios\n",
    "    #display(result) #Displays summary stats\n",
    "    \n",
    "    '''\n",
    "    #Prints out summary stats\n",
    "    print('Analysis for Ambry: ')\n",
    "    display(result) #Displays summary stats\n",
    "    print(f\"Odds Ratio: {table.oddsratio:.2f}\")\n",
    "    print(f\"95% CI: ({table.oddsratio_confint()[0]:.2f}, {table.oddsratio_confint()[1]:.2f})\")\n",
    "    print(f\"P-value (Fisher's exact): {p_value:.4f}\")\n",
    "    oddsratio, p_value = stats.fisher_exact(contingency_table)\n",
    "    '''\n",
    "\n",
    "    to_return = (table.oddsratio, [table.oddsratio_confint()[0], table.oddsratio_confint()[1]])\n",
    "    \n",
    "    return to_return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_carriers(cc, sge):\n",
    "    merged = pd.merge(cc, sge, on = 'pos_id', how = 'inner') #Merges case-control and SGE data\n",
    "    merged = merged.dropna(subset = ['Classification_y']) #drops any columsn without a classification\n",
    "    \n",
    "    contingency_tab = merged[['CaseControl', 'Classification_y']] #Creates dataframe for contingency table\n",
    "\n",
    "    contingency_tab = pd.crosstab(merged['CaseControl'], merged['Classification_y']) #Creates contingency table\n",
    "    contingency_tab = contingency_tab[contingency_tab.columns[::-1]] #Swaps NF and F columns in df\n",
    "    table_array = contingency_tab.values #Gets values from contingency table\n",
    "\n",
    "    oddsratio, p_value = stats.fisher_exact(table_array) #Tabulates odds-ratio and p-value from Fischer's exact test\n",
    "    table = sm.stats.Table2x2(table_array) #Generates confidence intervals\n",
    "    \n",
    "    #result = table.summary() #Summary table with Odds and Risk ratios\n",
    "    \n",
    "    \n",
    "\n",
    "    '''\n",
    "    #Prints out summary stats\n",
    "    print('Analysis for CARRIERS Data Set: ')\n",
    "    display(contingency_tab) #Displays contingency table \n",
    "    display(result) #Displays summary stats\n",
    "    print(f\"Odds Ratio: {table.oddsratio:.2f}\")\n",
    "    print(f\"95% CI: ({table.oddsratio_confint()[0]:.2f}, {table.oddsratio_confint()[1]:.2f})\")\n",
    "    print(f\"P-value (Fisher's exact): {p_value:.4f}\")\n",
    "    '''\n",
    "    \n",
    "    to_return = (table.oddsratio, [table.oddsratio_confint()[0], table.oddsratio_confint()[1]])\n",
    "\n",
    "\n",
    "    return to_return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_bridges(sge, df):\n",
    "\n",
    "    merged = pd.merge(df, sge, on = 'pos_id', how = 'inner')\n",
    "\n",
    "    summary = merged.pivot_table(\n",
    "        values = ['Cases', 'Controls'],\n",
    "        index = 'Classification',\n",
    "        aggfunc = 'sum'\n",
    "    )\n",
    "\n",
    "    #print(summary.transpose())\n",
    "    table_array = np.array([[summary['Cases']['NF'], summary['Cases']['F']],\n",
    "                             [summary['Controls']['NF'], summary['Controls']['F']]])\n",
    "\n",
    "    \n",
    "    oddsratio, p_value = stats.fisher_exact(table_array) #Tabulates odds-ratio and p-value from Fischer's exact test\n",
    "    table = sm.stats.Table2x2(table_array) #Generates confidence intervals\n",
    "    \n",
    "    #result = table.summary() #Summary table with Odds and Risk ratios\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "    #Prints out summary stats\n",
    "    '''\n",
    "    print('Analysis for ', 'BRIDGES', ' Data Set: ')\n",
    "    display(result) #Displays summary stats\n",
    "    print(f\"Odds Ratio: {table.oddsratio:.2f}\")\n",
    "    print(f\"95% CI: ({table.oddsratio_confint()[0]:.2f}, {table.oddsratio_confint()[1]:.2f})\")\n",
    "    print(f\"P-value (Fisher's exact): {p_value:.4f}\")        \n",
    "    '''\n",
    "    \n",
    "    to_return = (table.oddsratio, [table.oddsratio_confint()[0], table.oddsratio_confint()[1]])\n",
    "\n",
    "    return to_return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def process_data_dict(dict, key_dict):\n",
    "    all_keys = list(dict.keys())\n",
    "\n",
    "\n",
    "    odds_ratios = []\n",
    "\n",
    "    lwr_ci = []\n",
    "    uppr_ci = []\n",
    "    for key in all_keys:\n",
    "        odds, ci_list = dict[key]\n",
    "        odds_ratios.append(odds)\n",
    "        lwr_ci.append(ci_list[0])\n",
    "        uppr_ci.append(ci_list[1])\n",
    "\n",
    "\n",
    "    datasets = ['Ambry', 'CARRIERS (All)', 'CARRIERS (ER-)', 'BRIDGES (All)', 'BRIDGES (Pop.)',\n",
    "                'Ambry', 'CARRIERS (All)', 'CARRIERS (ER-)', 'BRIDGES (All)', 'BRIDGES (Pop.)',\n",
    "                'Ambry', 'CARRIERS (All)', 'CARRIERS (ER-)', 'BRIDGES (All)', 'BRIDGES (Pop.)',\n",
    "                'Ambry', 'CARRIERS (All)', 'CARRIERS (ER-)', 'BRIDGES (All)', 'BRIDGES (Pop.)',\n",
    "                'Ambry', 'CARRIERS (All)', 'CARRIERS (ER-)', 'BRIDGES (All)', 'BRIDGES (Pop.)',\n",
    "                'Ambry', 'CARRIERS (All)', 'CARRIERS (ER-)', 'BRIDGES (All)', 'BRIDGES (Pop.)'\n",
    "               ]\n",
    "\n",
    "    vars_incl = ['All Mis.', 'All Mis.', 'All Mis.', 'All Mis.', 'All Mis.', \n",
    "                 'All Vars.', 'All Vars.', 'All Vars.', 'All Vars.', 'All Vars.', \n",
    "                 'RING Mis.', 'RING Mis.', 'RING Mis.', 'RING Mis.', 'RING Mis.', \n",
    "                 'ARD Mis.', 'ARD Mis.','ARD Mis.','ARD Mis.','ARD Mis.',\n",
    "                 'BRCT Mis.', 'BRCT Mis.','BRCT Mis.','BRCT Mis.','BRCT Mis.',\n",
    "                 'Structured Mis.', 'Structured Mis.','Structured Mis.','Structured Mis.','Structured Mis.'\n",
    "                ]\n",
    "\n",
    "    \n",
    "    df = pd.DataFrame({'analysis': all_keys, 'OR': odds_ratios, 'lwr_ci': lwr_ci, 'uppr_ci': uppr_ci,\n",
    "                      'dataset': datasets, 'incl_vars': vars_incl})\n",
    "\n",
    "    df['full_type'] =  df['dataset'] + ' ' + df['incl_vars']\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cc_fig(df):\n",
    "\n",
    "    base = alt.Chart(df)\n",
    "\n",
    "    sort_order = ['Ambry All Vars.', 'BRIDGES (All) All Vars.', 'BRIDGES (Pop.) All Vars.', 'CARRIERS (All) All Vars.', 'CARRIERS (ER-) All Vars.',\n",
    "                  'Ambry All Mis.', 'BRIDGES (All) All Mis.', 'BRIDGES (Pop.) All Mis.', 'CARRIERS (All) All Mis.', 'CARRIERS (ER-) All Mis.',\n",
    "                  'Ambry RING Mis.', 'BRIDGES (All) RING Mis.', 'BRIDGES (Pop.) RING Mis.', 'CARRIERS (All) RING Mis.', 'CARRIERS (ER-) RING Mis.',\n",
    "                  'Ambry ARD Mis.', 'BRIDGES (All) ARD Mis.', 'BRIDGES (Pop.) ARD Mis.', 'CARRIERS (All) ARD Mis.', 'CARRIERS (ER-) ARD Mis.',\n",
    "                  'Ambry BRCT Mis.', 'BRIDGES (All) BRCT Mis.', 'BRIDGES (Pop.) BRCT Mis.', 'CARRIERS (All) BRCT Mis.', 'CARRIERS (ER-) BRCT Mis.',\n",
    "                  'Ambry Structured Mis.', 'BRIDGES (All) Structured Mis.', 'BRIDGES (Pop.) Structured Mis.', 'CARRIERS (All) Structured Mis.', 'CARRIERS (ER-) Structured Mis.']\n",
    "    \n",
    "    points = base.mark_point(\n",
    "        filled = True,\n",
    "        size = 50, \n",
    "        color = 'black'\n",
    "        ).encode(\n",
    "        y = alt.Y('full_type:O',\n",
    "                  scale = alt.Scale(domain = sort_order),\n",
    "                 axis = alt.Axis(title = '',\n",
    "                                 labelFontSize = 16, \n",
    "                                 labelLimit = 1000\n",
    "                                )\n",
    "                 ),\n",
    "        x = alt.X('OR',\n",
    "                 axis = alt.Axis(\n",
    "                     title = 'Odds Ratio',\n",
    "                     labelFontSize = 16,\n",
    "                     titleFontSize = 18,\n",
    "                     values = list(range(0, 9))\n",
    "                                 ),\n",
    "                  scale = alt.Scale(domain = [0, 8]\n",
    "                                   )\n",
    "                 ),\n",
    "        tooltip = ['OR']\n",
    "        )\n",
    "    \n",
    "    ci_bars = base.mark_errorbar().encode(\n",
    "        y = 'full_type',\n",
    "        x = alt.Y('lwr_ci:Q', axis = alt.Axis(title = '')),\n",
    "        x2 = 'uppr_ci:Q'\n",
    "        )\n",
    "\n",
    "    line = alt.Chart(pd.DataFrame({'Odds Ratio': [1]})).mark_rule(color = 'red').encode(\n",
    "        x = 'Odds Ratio')\n",
    "\n",
    "\n",
    "    plot = (points + ci_bars + line).configure_view(\n",
    "        stroke = None\n",
    "        ).configure_axis(\n",
    "        grid = False\n",
    "        ).interactive()\n",
    "\n",
    "    \n",
    "    plot.display()\n",
    "\n",
    "    return plot \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    raw_sge_df = read_sge(sge)\n",
    "    ambry_df, gnomad_df = read_ambry(ambry, gnomad)\n",
    "    \n",
    "    cc_df, cc_er_df = read_carriers_data(carriers)\n",
    "    bridges_all_df, bridges_pop_df = read_bridges(bridges_all, bridges_population, bridges_all_ptv, bridges_pop_ptv)\n",
    "    \n",
    "    sge_dfs, sge_keys = classify_vars(raw_sge_df)\n",
    "\n",
    "    data_keys = {'miss': ['ambry_miss', 'cc_miss', 'cc_miss_er', 'bridges_all_miss', 'bridges_pop_miss'], \n",
    "                 'all': ['ambry_all', 'cc_all', 'cc_all_er', 'bridges_all_all', 'bridges_pop_all'],\n",
    "                 'ring': ['ambry_ring', 'cc_ring', 'cc_ring_er', 'bridges_all_ring', 'bridges_pop_ring'],\n",
    "                 'ard': ['ambry_ard', 'cc_ard', 'cc_ard_er', 'bridges_all_ard', 'bridges_pop_ard'],\n",
    "                 'brct': ['ambry_brct', 'cc_brct', 'cc_brct_er', 'bridges_all_brct', 'bridges_pop_brct'],\n",
    "                 'structured': ['ambry_structured', 'cc_structured', 'cc_structured_er', 'bridges_all_structured', 'bridges_pop_structured']\n",
    "                }\n",
    "\n",
    "    data_dict = {}\n",
    "\n",
    "\n",
    "\n",
    "    for key in sge_keys:\n",
    "        sge_df = sge_dfs[key]\n",
    "        output_keys = data_keys[key]\n",
    "\n",
    "        data_dict[output_keys[0]] = analyze_ambry(sge_df, gnomad_df, ambry_df)\n",
    "        data_dict[output_keys[1]] = analyze_carriers(cc_df, sge_df)\n",
    "        data_dict[output_keys[2]] = analyze_carriers(cc_er_df, sge_df)\n",
    "        data_dict[output_keys[3]] = analyze_bridges(sge_df, bridges_all_df)\n",
    "        data_dict[output_keys[4]] = analyze_bridges(sge_df, bridges_pop_df)\n",
    "\n",
    "\n",
    "    final_df = process_data_dict(data_dict, data_keys)\n",
    "    \n",
    "\n",
    "    '''\n",
    "    carriers_all, carriers_controls, carriers_cases = process_carriers(cc_df)\n",
    "    analyze_ambry(raw_sge_df, gnomad_df, ambry_df)\n",
    "    cc_df = analyze_carriers(carriers_all, sge_all)\n",
    "    analyze_bridges(sge_all, bridges_dfs)\n",
    "    '''\n",
    "    \n",
    "    final_plot = cc_fig(final_df)\n",
    "    #final_plot.save('/Users/ivan/Desktop/BARD1_draft_figs/fig_4d_ORplot.png', ppi = 500)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
