{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import necessary modules\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Path to the .fa file\n",
    "fasta_path = '../Data/SNV_filtering_inputs/PillarProject_filtering/PALB2/20250822_PALB2.fasta'\n",
    "entry_file = '../Data/SNV_filtering_inputs/PillarProject_filtering/PALB2/20250822_PALB2_filter_entry.xlsx' #Entry file contains 3 sheets: SGE targets with start/end coordinates, start/end coordinates for exons of the gene, start/end coordinates and sense (0 - anti, 1 - sense) of provided fasta file\n",
    "final_name = '../Data/SNV_filtering_inputs/PillarProject_filtering/PALB2/20250822_PALB2_PillarProjectRef' + '.xlsx' #Final path for reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_inputs(file): #Reads input files into list of tuples\n",
    "    \n",
    "    lib_df = pd.read_excel(file, sheet_name = 'targets') #Reads targets sheets with information about each SGE target\n",
    "    exon_df = pd.read_excel(file, sheet_name = 'exons') #Reads exons sheet with information about each exon of the gene\n",
    "    start_end = pd.read_excel(file, sheet_name = 'gene') #Reads gene sheet with information about provided fasta\n",
    "\n",
    "    libs =  [] #Empty list to hold tuples that are (SGE target, target start coordinate, target end coordinate)\n",
    "    i = 0\n",
    "    while i < len(lib_df):\n",
    "        target = lib_df['target'][i]\n",
    "        target_start = lib_df['start'][i]\n",
    "        target_end = lib_df['end'][i]\n",
    "\n",
    "        handoff_tuple = (target, target_start, target_end)\n",
    "        libs.append(handoff_tuple)\n",
    "\n",
    "        i += 1\n",
    "\n",
    "    exons = [] #Empty list to tuples that are (exon start coordinate, exon end coordinate)\n",
    "    j = 0\n",
    "    while j < len(exon_df):\n",
    "        exon_start = exon_df['start'][j]\n",
    "        exon_end = exon_df['end'][j]\n",
    "\n",
    "        exon_tuple = (exon_start, exon_end)\n",
    "\n",
    "        exons.append(exon_tuple)\n",
    "        j += 1\n",
    "\n",
    "    start = start_end['start'][0] #Gets fasta start\n",
    "    end = start_end['end'][0] #Gets fasta end\n",
    "    sense = start_end['sense'][0] #Gets gene sense\n",
    "    \n",
    "    return libs, exons, start, end, sense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse the .fa file\n",
    "def parse(fasta_path, libs, start, end, sense):\n",
    "    start = start\n",
    "    end = end\n",
    "\n",
    "    if sense == 0:\n",
    "        #creates lists of coordinates (adj is to pull from fasta, not adj is genomic) and region names\n",
    "        lib_coords_adj = []\n",
    "        lib_coords = []\n",
    "        regions = []\n",
    "\n",
    "        \n",
    "        for elem in libs: \n",
    "            region, lib_start, lib_end = elem\n",
    "            lib_coords.append(lib_start)\n",
    "            lib_coords.append(lib_end)\n",
    "            \n",
    "            pos_adjusted_start = start - lib_start\n",
    "            pos_adjusted_end = start - lib_end\n",
    "            lib_coords_adj.append(pos_adjusted_start)\n",
    "            lib_coords_adj.append(pos_adjusted_end)\n",
    "            regions.append(region)\n",
    "\n",
    "        #creates list of library sequences from fasta\n",
    "        list_lib_seq = []\n",
    "        for record in SeqIO.parse(fasta_path, \"fasta\"):\n",
    "            i = 0\n",
    "            while i < len(lib_coords):\n",
    "                lib_seq = record.seq[lib_coords_adj[i] - 1:lib_coords_adj[i+1]] #indexing of lib_coords_adj allows for reading every pair of coordinates\n",
    "                lib_seq = str(lib_seq.complement())[::-1] #reverse complements for antisense sequence\n",
    "                list_lib_seq.append(str(lib_seq)) #appends sequence to list\n",
    "                print(lib_seq)\n",
    "                i += 2 #increment of 2 to account for reading each adjacent pair of coordinates\n",
    "        \n",
    "        #creates list of data frames where each dataframe contains a region with its reference and genomic coordinates at each bp\n",
    "        df_list = []\n",
    "        j = 0\n",
    "        z = 0\n",
    "        while j < len(regions):\n",
    "            name = regions[j] + '_df' #name of df\n",
    "            region_name = regions[j] #region name\n",
    "            bp = [] #list where each element is a base pair\n",
    "            coords = [] #list where each element is a coordinate\n",
    "            \n",
    "            seq = list_lib_seq[j] #pulls out sequence from list of sequences created by previous loop\n",
    "            for char in seq: #iterates through sequence and appends each character to bp list\n",
    "                bp.append(char)\n",
    "                \n",
    "            start = lib_coords[z] #gets start coordinate from list of coordinates\n",
    "            end = lib_coords[z+1] #gets end coordinate from list of coordinates\n",
    "            for k in range(end, start + 1): #Creates coordinates (end and start flipped for gene on antisense strand after reverse compliment)\n",
    "                coords.append(k)\n",
    "    \n",
    "            #creates and appends dataframe to list\n",
    "            print(region_name, len(bp),len(coords))\n",
    "            name = pd.DataFrame({'target': region_name, 'Reference': bp, 'pos': coords})\n",
    "            df_list.append(name)\n",
    "            \n",
    "            j += 1\n",
    "            z += 2\n",
    "            \n",
    "        final_ref = pd.concat(df_list)\n",
    "\n",
    "    elif sense == 1:\n",
    "        \n",
    "        #creates lists of coordinates (adj is to pull from fasta, not adj is genomic) and region names\n",
    "        lib_coords_adj = []\n",
    "        lib_coords = []\n",
    "        regions = []\n",
    "        for elem in libs: \n",
    "            region, lib_start, lib_end = elem\n",
    "            lib_coords.append(lib_start)\n",
    "            lib_coords.append(lib_end)\n",
    "            \n",
    "            pos_adjusted_start = -(start - lib_start)\n",
    "            pos_adjusted_end = -(start - lib_end)\n",
    "            lib_coords_adj.append(pos_adjusted_start)\n",
    "            lib_coords_adj.append(pos_adjusted_end)\n",
    "            regions.append(region)\n",
    "\n",
    "        #creates list of library sequences from fasta\n",
    "        list_lib_seq = []\n",
    "        for record in SeqIO.parse(fasta_path, \"fasta\"):\n",
    "            i = 0\n",
    "            while i < len(lib_coords):\n",
    "                lib_seq = record.seq[lib_coords_adj[i]:lib_coords_adj[i+1] + 1] #indexing of lib_coords_adj allows for reading every pair of coordinates\n",
    "                #lib_seq = str(lib_seq.complement())[::-1] #reverse complements for antisense sequence\n",
    "                list_lib_seq.append(str(lib_seq)) #appends sequence to list\n",
    "                print(region, lib_seq)\n",
    "                i += 2 #increment of 2 to account for reading each adjacent pair of coordinates\n",
    "                \n",
    "        df_list = []\n",
    "        j = 0\n",
    "        z = 0\n",
    "        while j < len(regions):\n",
    "            name = regions[j] + '_df' #name of df\n",
    "            region_name = regions[j] #region name\n",
    "            bp = [] #list where each element is a base pair\n",
    "            coords = [] #list where each element is a coordinate\n",
    "            \n",
    "            seq = list_lib_seq[j] #pulls out sequence from list of sequences created by previous loop\n",
    "            for char in seq: #iterates through sequence and appends each character to bp list\n",
    "                bp.append(char)\n",
    "                \n",
    "            start = lib_coords[z] #gets start coordinate from list of coordinates\n",
    "            end = lib_coords[z+1] #gets end coordinate from list of coordinates\n",
    "\n",
    "            for k in range(start, end + 1): #Creates coordinates for each region\n",
    "                coords.append(k)\n",
    "    \n",
    "            #creates and appends dataframe to list\n",
    "            name = pd.DataFrame({'target': region_name, 'Reference': bp, 'pos': coords})\n",
    "            df_list.append(name)\n",
    "            \n",
    "            j += 1\n",
    "            z += 2\n",
    "            \n",
    "        final_ref = pd.concat(df_list)\n",
    "\n",
    "    return final_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_introns(df, exons):\n",
    "    df['Intron/Exon'] = None\n",
    "    \n",
    "    exon_coords = []\n",
    "    for elem in exons:\n",
    "        start, end = elem\n",
    "        for i in range(start, end + 1):\n",
    "            exon_coords.append(i)\n",
    "\n",
    "    df.loc[df['pos'].isin(exon_coords),'Intron/Exon'] = 'Exon'\n",
    "    df.loc[~df['pos'].isin(exon_coords),'Intron/Exon'] = 'Intron'\n",
    "\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#saves to excel\n",
    "def save_to_xlsx(df):\n",
    "    df.to_excel(final_name, index = False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    libs, exons, start, end, sense = read_inputs(entry_file)\n",
    "    ref_seqs = parse(fasta_path, libs, start, end, sense)\n",
    "    annotated_ref = annotate_introns(ref_seqs, exons)\n",
    "    save_to_xlsx(ref_seqs)\n",
    "    print(annotated_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
