{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import necessary modules\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "\n",
    "# Path to the .fa file\n",
    "file_path = '/Users/ivan/Downloads/bard1.fasta'\n",
    "\n",
    "#list of Libraries and coordinates\n",
    "#(\"Region Name\", start, end) - start is higher than end for genes on antisense strand and/or if fasta is sequence on antisense strand\n",
    "libs = [('BARD1_X1A',214809584,214809476),('BARD1_X1B',214809500,214809379),\n",
    "        ('BARD1_X2',214797156,214797050),('BARD1_X3A',214792474, 214792354),\n",
    "        ('BARD1_X3B',214792400,214792286),('BARD1_X4A',214781543,214781431),\n",
    "        ('BARD1_X4B',214781448,214781334),('BARD1_X4C',214781383,214781206),\n",
    "        ('BARD1_X4D',214781326,214781202),('BARD1_X4E',214781206,214781087),\n",
    "        ('BARD1_X4F',214781131,214781012),('BARD1_X4G',214781020,214780906),\n",
    "        ('BARD1_X4H',214780958,214780857),('BARD1_X4I',214780866,214780767),\n",
    "        ('BARD1_X4J',214780821,214780697),('BARD1_X4K',214780740,214780627),\n",
    "        ('BARD1_X4L',214780646,214780532),('BARD1_X5A',214769383,214769267),\n",
    "        ('BARD1_X5B',214769286,214769166),('BARD1_X6A',214767689,214767579),\n",
    "        ('BARD1_X6B',214767615,214767507),('BARD1_X6C',214767549,214767430),\n",
    "        ('BARD1_X7A',214752680,214752442),('BARD1_X7B',214752491,214752370),\n",
    "        ('BARD1_X8A',214745888,214745778),('BARD1_X8B',214745791,214745675),\n",
    "        ('BARD1_X9A',214745199,214745103),('BARD1_X9B',214745123,214745004),\n",
    "        ('BARD1_X10A',214730531,214730428),('BARD1_X10B',214730469,214730353),\n",
    "        ('BARD1_X11A',214729018,214728899),('BARD1_X11B',214728917,214728818),\n",
    "        ('BARD1_X11C',214728835,214728716),('BARD1_X11D',214728735,214728632)]\n",
    "\n",
    "\n",
    "exons = [(214809412,214809569),(214797061,214797117),(214792297,214792445),\n",
    "         (214780560,214781509),(214769232,214769312),(214767482,214767654),\n",
    "         (214752447,214752555),(214745722,214745854),(214745067,214745159),\n",
    "         (214730411,214730508),(214728676,214729008)\n",
    "        ]\n",
    "\n",
    "final_name = '20240809_BARD1_SNVlib_ref_seqs_intron_annotated' + '.xlsx'\n",
    "start = 214809683\n",
    "end = 214725646"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parse the .fa file\n",
    "def parse():\n",
    "    start = 214809683\n",
    "    end = 214725646\n",
    "\n",
    "    #creates lists of coordinates (adj is to pull from fasta, not adj is genomic) and region names\n",
    "    lib_coords_adj = []\n",
    "    lib_coords = []\n",
    "    regions = []\n",
    "    for elem in libs: \n",
    "        region, lib_start, lib_end = elem\n",
    "        lib_coords.append(lib_start)\n",
    "        lib_coords.append(lib_end)\n",
    "        \n",
    "        pos_adjusted_start = start - lib_start\n",
    "        pos_adjusted_end = start - lib_end\n",
    "        lib_coords_adj.append(pos_adjusted_start)\n",
    "        lib_coords_adj.append(pos_adjusted_end)\n",
    "        regions.append(region)\n",
    "\n",
    "    #creates list of library sequences from fasta\n",
    "    list_lib_seq = []\n",
    "    for record in SeqIO.parse(file_path, \"fasta\"):\n",
    "        i = 0\n",
    "        while i < len(lib_coords):\n",
    "            lib_seq = record.seq[lib_coords_adj[i]:lib_coords_adj[i+1] + 1] #indexing of lib_coords_adj allows for reading every pair of coordinates\n",
    "            lib_seq = str(lib_seq.complement())[::-1] #reverse complements for antisense sequence\n",
    "            list_lib_seq.append(str(lib_seq)) #appends sequence to list\n",
    "            i += 2 #increment of 2 to account for reading each adjacent pair of coordinates\n",
    "    \n",
    "    #creates list of data frames where each dataframe contains a region with its reference and genomic coordinates at each bp\n",
    "    df_list = []\n",
    "    j = 0\n",
    "    z = 0\n",
    "    while j < len(regions):\n",
    "        name = regions[j] + '_df' #name of df\n",
    "        region_name = regions[j] #region name\n",
    "        bp = [] #list where each element is a base pair\n",
    "        coords = [] #list where each element is a coordinate\n",
    "        \n",
    "        seq = list_lib_seq[j] #pulls out sequence from list of sequences created by previous loop\n",
    "        for char in seq: #iterates through sequence and appends each character to bp list\n",
    "            bp.append(char)\n",
    "            \n",
    "        start = lib_coords[z] #gets start coordinate from list of coordinates\n",
    "        end = lib_coords[z+1] #gets end coordinate from list of coordinates\n",
    "        for k in range(end, start + 1): #Creates coordinates (end and start flipped for gene on antisense strand after reverse compliment)\n",
    "            coords.append(k)\n",
    "\n",
    "        #creates and appends dataframe to list\n",
    "        name = pd.DataFrame({'target': region_name, 'Reference': bp, 'pos': coords})\n",
    "        df_list.append(name)\n",
    "        \n",
    "        j += 1\n",
    "        z += 2\n",
    "        \n",
    "    final_ref = pd.concat(df_list)   \n",
    "\n",
    "    return final_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_introns(df):\n",
    "    df['Intron/Exon'] = None\n",
    "    \n",
    "    exon_coords = []\n",
    "    for elem in exons:\n",
    "        start, end = elem\n",
    "        for i in range(start, end + 1):\n",
    "            exon_coords.append(i)\n",
    "\n",
    "    df.loc[df['pos'].isin(exon_coords),'Intron/Exon'] = 'Exon'\n",
    "    df.loc[~df['pos'].isin(exon_coords),'Intron/Exon'] = 'Intron'\n",
    "\n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#saves to excel\n",
    "def save_to_xlsx(df):\n",
    "    df.to_excel(final_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    ref_seqs = parse()\n",
    "    annotated_ref = annotate_introns(ref_seqs)\n",
    "    save_to_xlsx(ref_seqs)\n",
    "    print(ref_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
