{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import statistics\n",
    "import warnings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_path = '../Data/SNV_filtering_inputs/PillarProject_filtering/XRCC2/20250822_XRCC2_PillarProjectRef.xlsx' #path to annotated (non-coding/coding) reference for each library\n",
    "sge_scores = '../Data/SNV_filtering_inputs/PillarProject_filtering/XRCC2/20250825.XRCC2.snvscores.tsv'  #path to SGE datafile\n",
    "gene = 'XRCC2' #name of your gene\n",
    "ref_sense = 0 #Sense of the reference file you provide\n",
    "filtered_file_name = '../Data/SNV_filtering_inputs/PillarProject_filtering/XRCC2/20250825_XRCC2snvscores_filtered.xlsx' #name of saved file\n",
    "targets_input = '../Data/SNV_filtering_inputs/PillarProject_filtering/XRCC2/202507_XRCC2_targets.tsv' #Targets input contains positions and base changed for all fixed edits on the positive sense strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def reverse_complement_string(seq_string): #Reverse complement and returns string\n",
    "    reverse_seq = seq_string[::-1]\n",
    "    reverse_comp_list = []\n",
    "    for char in reverse_seq:\n",
    "        if char == \"A\":\n",
    "            reverse_comp_list.append(\"T\")\n",
    "        elif char == \"G\":\n",
    "            reverse_comp_list.append(\"C\")\n",
    "        elif char == \"C\":\n",
    "            reverse_comp_list.append(\"G\")\n",
    "        else:\n",
    "            reverse_comp_list.append(\"A\")\n",
    "    reverse_compliment_str = \"\".join(reverse_comp_list)\n",
    "    return reverse_compliment_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def mutate_snvs(dna_sequence): #Mutates all possible SNVs of provided DNA sequence\n",
    "    snvs = []\n",
    "    i = 0\n",
    "    while i < len(dna_sequence):\n",
    "        if dna_sequence[i] == \"A\":\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        elif dna_sequence[i] == \"T\":\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        elif dna_sequence[i] == \"C\":\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"G\" + dna_sequence[i + 1 :])\n",
    "        else:\n",
    "            snvs.append(dna_sequence[:i] + \"A\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"T\" + dna_sequence[i + 1 :])\n",
    "            snvs.append(dna_sequence[:i] + \"C\" + dna_sequence[i + 1 :])\n",
    "        i += 1\n",
    "    return snvs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def string_compare(string1, string2): #Function to compare strings and return the position and character that is different\n",
    "    \n",
    "    if len(string1) != len(string2): #Function expects strings of same length\n",
    "        print(len(string1), len(string2))\n",
    "        print('String1: ', string1, ' ', 'String2: ', string2)\n",
    "        raise ValueError('Different Length Strings')\n",
    "\n",
    "    else: #Checking all string indices\n",
    "        i = 0\n",
    "        while i < len(string1):\n",
    "            char = string1[i]\n",
    "            if char == string2[i]: #Next interation if characters are the same\n",
    "                i += 1\n",
    "            else: #Returns index and character if not the same\n",
    "                return i, char\n",
    "                break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_same_codon(numbers): #Filtering function to identify if any fixed edits are within the same codon\n",
    "    \n",
    "    pairs = [] #List to hold pairs of positions\n",
    "    \n",
    "    for i in range(len(numbers)): #Iterates through positions of all fixed edits\n",
    "        for j in range(i + 1, len(numbers)): #Loop to extract positions from list\n",
    "            \n",
    "            if abs(numbers[i] - numbers[j]) == 2: #Fixed edits within the same codon will have distance of 2\n",
    "                pairs.append([numbers[i], numbers[j]]) #Appended to list to be returned\n",
    "\n",
    "    return pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_inputs(reference, sge, targets): #Reads all inputs\n",
    "    \n",
    "    reference = pd.read_excel(reference) #Reads provided reference file\n",
    "    reference['Reference'] = reference['Reference'].transform(lambda x: x.upper()) #Turns all bases in reference to upper case\n",
    "    raw_data = pd.read_csv(sge, sep = '\\t') #Reads raw data\n",
    "    targets = pd.read_csv(targets, sep = '\\t') #Reads targets file\n",
    "\n",
    "    sge_targets = list(set(raw_data['target'].tolist())) #Gets all SGE targets\n",
    "    print(sge_targets)\n",
    "\n",
    "    targets = targets.loc[targets['target'].isin(sge_targets)] #Filters targets input to include SGE targets in data only\n",
    "    targets = targets.reset_index(drop = True) #Resets index\n",
    "    \n",
    "    print(targets)\n",
    "    return reference, raw_data, targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_edits(df): #Gets all fixed edits from targets.tsv\n",
    "    \n",
    "    i = 0\n",
    "    edits_dict = {} #Dictionary to hold edits\n",
    "    while i < len(df): #iterates through each row of targets dataframe\n",
    "        target = df['target'][i] #Gets target\n",
    "        all_edits = df['required_edits'][i] #Gets all edit\n",
    "        edits_split = all_edits.split(',') #Splits all edits\n",
    "        \n",
    "        edits_dict[target] = edits_split #Appended to dictionary\n",
    "\n",
    "        i += 1\n",
    "\n",
    "    return edits_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter(edit_dict, data, ref, sense): #Filters out amino acid changes possible only in context of required fixed edits\n",
    "\n",
    "    grouped = data.groupby('target') #Groups by target\n",
    "\n",
    "    to_filter = [] #List to hold variants to remove\n",
    "    \n",
    "    for target, df in grouped: #Iterates through data for each target\n",
    "        target_ref = ref.loc[ref['target'].isin([target])] #Gets reference oligo for target\n",
    "        edits = edit_dict[target] #Gets all edits\n",
    "        edits_split = {int(s[:-1]): s[-1] for s in edits} #Splits edits into dictionary that is {position: edit}\n",
    "\n",
    "        edit_pos = list(edits_split.keys()) #Gets all edit positions\n",
    "        same_codon = identify_same_codon(edit_pos) #Runs function to identify fixed edits in the same codon\n",
    "        \n",
    "        if len(same_codon) > 0: #Tests if any same codon edits are detected\n",
    "            same_codon = same_codon[0] #Gets positions that have fixed edits in same codon\n",
    "            pos_to_remove = int(statistics.median(same_codon)) #Gets basepair in middle that is not fixed edit\n",
    "            \n",
    "            pos_to_remove = str(pos_to_remove) #Sets coordinate to remove to string\n",
    "            \n",
    "            to_remove = [pos_to_remove + ':' + 'A',\n",
    "                         pos_to_remove + ':' + 'T',\n",
    "                         pos_to_remove + ':' + 'G',\n",
    "                         pos_to_remove + ':' + 'C'\n",
    "                        ] #List of variants to remove in the middle basepair\n",
    "\n",
    "            for var in to_remove: #Appends all variants to filtering list\n",
    "                target_pos_id = target + ':' + var\n",
    "                to_filter.append(target_pos_id)\n",
    "                \n",
    "            print('Doubles: ', pos_to_remove) #Prints position removed due to two fixed edits in same codon\n",
    "            \n",
    "        for pos in edit_pos: #Iterates through each fixed edit and determines variants to remove\n",
    "            if sense == 0: #For antisense genes\n",
    "                edited_codon_coords = list(range(pos, pos + 3)) #Gets coordinates of the edited codon \n",
    "                \n",
    "                print('1 - Edited Codon ', edited_codon_coords) #Prints edited codon coords (should decrease to fixed edit for antisense gene)\n",
    "                \n",
    "                wt_codon_df = target_ref.loc[target_ref['pos'].isin(edited_codon_coords)] #Gets WT codon from reference in df form\n",
    "                wt_codon = wt_codon_df['Reference'].tolist() #WT reference in list form\n",
    "                wt_codon = ''.join(wt_codon) #WT reference in string form\n",
    "                wt_codon = reverse_complement_string(wt_codon) #Reverse complementation for antisense gene\n",
    "                print(\"2 - WT \", wt_codon) #Prints WT codon\n",
    "\n",
    "                if len(wt_codon) != 3: #Error checking for fixed edits that are at very edges of libraries\n",
    "                    warnings.warn('Incomplete Codon')\n",
    "                    error_string = 'Incomplete Codon at pos ' + str(pos)\n",
    "                    print(error_string)\n",
    "                    continue\n",
    "                    \n",
    "                canonical_snvs = mutate_snvs(wt_codon) #Gets all possible SNVs with the WT codon\n",
    "                possible_aa = [] #List to hold possible alternate amino acids with the canonical SNVs\n",
    "                \n",
    "                for elem in canonical_snvs: #Generates the possible alternate amino acids based on WT codon\n",
    "                    var = Seq(elem)\n",
    "                    aa = var.translate()\n",
    "                    possible_aa.append(str(aa))\n",
    "\n",
    "                mut_codon = wt_codon[0:2] + reverse_complement_string(edits_split[pos]) #Gets codon with fixed edit (first two positions are same, 3rd position is fixed edit)\n",
    "                fixed_edit = reverse_complement_string(edits_split[pos]) #Reverse complementation for antisense gene\n",
    "                mut_codon = wt_codon[0:2] #Gets bases not including fixed edit\n",
    "                mut_snvs = mutate_snvs(mut_codon) #mutates all possible SNVs for those 2 bases\n",
    "\n",
    "                fixed_edit_aa = [] #List to hold amino acid changes with the fixed edit\n",
    "                impossible = [] #List to hold missense changes possible only in context of the fixed edit\n",
    "                \n",
    "                for elem in mut_snvs: #Iterates and translates SNVs with fixed edit\n",
    "                    full_mut_codon = elem + fixed_edit #Buils full codon with fixed edit\n",
    "                    \n",
    "                    normalized_pos, mut_codon_snv = string_compare(full_mut_codon[0:2], wt_codon[0:2]) #String compare to get position that is different from WT (only first 2 bases checked)\n",
    "                    if normalized_pos == 0: #Gets position if first bp is different (2bp upstream of edit)\n",
    "                        mut_codon_pos = pos + 2 \n",
    "                    elif normalized_pos == 1: #Gets position if 2nd bp is different (1bp upstream of edit)\n",
    "                        mut_codon_pos = pos + 1\n",
    "\n",
    "                    print('mut codon', full_mut_codon, mut_codon_pos, normalized_pos) #Prints full codon with fixed edit, and positions\n",
    "                    mut_codon_pos_id = target + ':' + str(mut_codon_pos) + ':' + reverse_complement_string(mut_codon_snv) #Builds unique identifier for filtering\n",
    "\n",
    "                    #Variants translated\n",
    "                    var = Seq(full_mut_codon)\n",
    "                    aa = str(var.translate())\n",
    "                    fixed_edit_aa.append(aa)\n",
    "                    \n",
    "                    if aa in possible_aa: #Checks if amino acid change is possible with WT SNVs\n",
    "                        continue\n",
    "                    else:\n",
    "                        impossible.append(aa) \n",
    "                        to_filter.append(mut_codon_pos_id) #Identifier appended to filter list if not in list of amino acid changes possible with WT SNVs\n",
    "                                         \n",
    "                print('3 - ', target, pos, possible_aa, fixed_edit_aa, impossible) #Prints the changes\n",
    "                print('4 - ', len(impossible), len(to_filter)) #Running counter for variants to filter\n",
    "\n",
    "            elif sense == 1: #Analogous code for genes where coding strand is sense strand. Reverse complementation removed. Math to determine positions of codons changed\n",
    "                edited_codon_coords = list(range(pos - 2, pos + 1))\n",
    "                \n",
    "                print('1 - Edited Codon ', edited_codon_coords)\n",
    "                \n",
    "                wt_codon_df = target_ref.loc[target_ref['pos'].isin(edited_codon_coords)]\n",
    "                wt_codon = wt_codon_df['Reference'].tolist()\n",
    "                wt_codon = ''.join(wt_codon)\n",
    "                print('2 - WT ', wt_codon)\n",
    "                if len(wt_codon) != 3:\n",
    "                    warnings.warn('Incomplete Codon')\n",
    "                    error_string = 'Incomplete Codon at pos ' + str(pos)\n",
    "                    print(error_string)\n",
    "                    continue\n",
    "\n",
    "                canonical_snvs = mutate_snvs(wt_codon)\n",
    "                possible_aa = []\n",
    "                \n",
    "                for elem in canonical_snvs:\n",
    "                    var = Seq(elem)\n",
    "                    aa = var.translate()\n",
    "                    possible_aa.append(str(aa))\n",
    "\n",
    "                mut_codon = wt_codon[0:2] + (edits_split[pos])\n",
    "                fixed_edit = edits_split[pos]\n",
    "                mut_codon = wt_codon[0:2]\n",
    "                mut_snvs = mutate_snvs(mut_codon)\n",
    "\n",
    "                fixed_edit_aa = []\n",
    "                impossible = []\n",
    "                \n",
    "                for elem in mut_snvs:\n",
    "                    full_mut_codon = elem + fixed_edit\n",
    "                    normalized_pos, mut_codon_snv = string_compare(full_mut_codon[0:2], wt_codon[0:2])\n",
    "                    if normalized_pos == 0:\n",
    "                        mut_codon_pos = pos - 2\n",
    "                    elif normalized_pos == 1:\n",
    "                        mut_codon_pos = pos -1\n",
    "                        \n",
    "                    print('mut codon', full_mut_codon, mut_codon_pos, normalized_pos)\n",
    "                    mut_codon_pos_id = target + ':' + str(mut_codon_pos) + ':' + mut_codon_snv\n",
    "\n",
    "                    var = Seq(full_mut_codon)\n",
    "                    aa = str(var.translate())\n",
    "                    fixed_edit_aa.append(aa)\n",
    "                    if aa in possible_aa:\n",
    "                        continue\n",
    "                    else:\n",
    "                        impossible.append(aa)\n",
    "                        to_filter.append(mut_codon_pos_id)\n",
    "                                         \n",
    "                print('3 - ', target, pos, possible_aa, fixed_edit_aa, impossible)\n",
    "                print('4 - ', len(impossible), len(to_filter))\n",
    "    \n",
    "\n",
    "    return to_filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_data(sge, to_filter): #Filters out variants\n",
    "\n",
    "    #Builds unique identifier into raw SGE data\n",
    "    sge['pos'] = sge['pos'].astype(str) #Sets 'pos' to string\n",
    "    sge['pos_id'] = sge['pos'] + ':' + sge['alt'] #Builds position ID\n",
    "    sge['target_pos_id'] = sge['target']+ ':' +  sge['pos_id'] #Builds target into position ID for filtering (needed to filter/skip variants covered in overlapping targets)\n",
    "    sge['pos'] = sge['pos'].astype(int) #Resets 'pos' to integer datatype   \n",
    "\n",
    "    filtered = sge.loc[~((sge['target_pos_id'].isin(to_filter)) & (sge['amino_acid_change'] != '---'))] #Variants with correct target position ID and in coding sequence filtered\n",
    "\n",
    "    filtered = filtered.copy() #Copy made\n",
    "\n",
    "    #Adds columns for functional classification based on 0.95 and 0.99 GMM density cutoffs\n",
    "    filtered['gmm_consequence_0.95'] = 'indeterminate'\n",
    "    filtered.loc[filtered['gmm_density_abnormal'] >= 0.95, 'gmm_consequence_0.95'] = 'functionally_abnormal'\n",
    "    filtered.loc[filtered['gmm_density_normal'] >= 0.95, 'gmm_consequence_0.95'] = 'functionally_normal'\n",
    "\n",
    "    filtered['gmm_consequence_0.99'] = 'indeterminate'\n",
    "    filtered.loc[filtered['gmm_density_abnormal'] >= 0.99, 'gmm_consequence_0.99'] = 'functionally_abnormal'\n",
    "    filtered.loc[filtered['gmm_density_normal'] >= 0.99, 'gmm_consequence_0.99'] = 'functionally_normal'\n",
    "    \n",
    "    return filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10",
   "metadata": {},
   "outputs": [],
   "source": [
    "def main():\n",
    "    ref, raw_data, targets = read_inputs(ref_path, sge_scores, targets_input) #Reads inputs\n",
    "    all_edits = get_edits(targets) #Gets edits\n",
    "    to_filter = filter(all_edits, raw_data, ref, ref_sense) #Gets variants to filter\n",
    "    filtered_data = filter_data(raw_data, to_filter) #Variants filtered\n",
    "\n",
    "\n",
    "    filtered_data.to_excel(filtered_file_name, index = False) #Filtered data saved\n",
    "    print(to_filter) #Variants to filter printed\n",
    "\n",
    "    #Prints filtering statistics\n",
    "    print('Filtering statistics: ', '\\n',\n",
    "            'RAW DATA: ', len(raw_data), '\\n',\n",
    "           'VARS. TO FILTER: ', len(to_filter), '\\n',\n",
    "            'POST FILTER: ', len(filtered_data), '\\n'\n",
    "         )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11",
   "metadata": {},
   "outputs": [],
   "source": [
    "main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
